Before filtering overlapping supplementary alignments and before filtering
partially-mapped reads, the following supplementary alignment statistics held.

In the CAMP genome, 8,645 / 503,385 (1.72\%) unique reads aligned to within the genome had supplementary alignments within CAMP.
On average, these 8,645 reads had 2.07 alignments within CAMP.
Furthermore, 2,180 / 503,385 (0.43\%) unique reads aligned to within CAMP had supplementary alignments within CAMP \emph{and} had reference overlap between at least one pair of their alignments on CAMP.
The average length of these overlaps (considering all pairs of overlapping alignments in CAMP from the same read) was 580.12 bp.

In the BACT1 genome, 6,067 / 268,075 (2.26\%) unique reads aligned to within the genome had supplementary alignments within BACT1.
On average, these 6,067 reads had 2.04 alignments within BACT1.
Furthermore, 470 / 268,075 (0.18\%) unique reads aligned to within BACT1 had supplementary alignments within BACT1 \emph{and} had reference overlap between at least one pair of their alignments on BACT1.
The average length of these overlaps (considering all pairs of overlapping alignments in BACT1 from the same read) was 1,417.80 bp.

In the BACT2 genome, 5,497 / 745,461 (0.74\%) unique reads aligned to within the genome had supplementary alignments within BACT2.
On average, these 5,497 reads had 2.04 alignments within BACT2.
Furthermore, 728 / 745,461 (0.10\%) unique reads aligned to within BACT2 had supplementary alignments within BACT2 \emph{and} had reference overlap between at least one pair of their alignments on BACT2.
The average length of these overlaps (considering all pairs of overlapping alignments in BACT2 from the same read) was 2,561.42 bp.
\endinput