Before filtering overlapping supplementary alignments and before filtering
partially-mapped reads, the following supplementary alignment statistics held.

In the CAMP genome, 8,645 / 503,385 (1.72\%) unique reads aligned to within the genome had supplementary alignments within the genome.
On average, these 8,645 reads had 2.07 alignments.
Furthermore, 2,180 / 503,385 (0.43\%) unique reads aligned to within the genome had supplementary alignments within the genome \emph{and} had overlap between at least one pair of their alignments on the reference genome.
The average length of these overlaps (considering all pairs of overlapping alignments from the same read) was 580.12 bp.

In the BACTERIA genome, 6,067 / 268,075 (2.26\%) unique reads aligned to within the genome had supplementary alignments within the genome.
On average, these 6,067 reads had 2.04 alignments.
Furthermore, 470 / 268,075 (0.18\%) unique reads aligned to within the genome had supplementary alignments within the genome \emph{and} had overlap between at least one pair of their alignments on the reference genome.
The average length of these overlaps (considering all pairs of overlapping alignments from the same read) was 1,417.80 bp.

In the BACTEROIDALES genome, 5,497 / 745,461 (0.74\%) unique reads aligned to within the genome had supplementary alignments within the genome.
On average, these 5,497 reads had 2.04 alignments.
Furthermore, 728 / 745,461 (0.10\%) unique reads aligned to within the genome had supplementary alignments within the genome \emph{and} had overlap between at least one pair of their alignments on the reference genome.
The average length of these overlaps (considering all pairs of overlapping alignments from the same read) was 2,561.42 bp.
\endinput