{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate \"mutation matrices\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run \"Header.ipynb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import copy\n",
    "import time\n",
    "import pysam\n",
    "import skbio\n",
    "from collections import defaultdict, Counter\n",
    "from parse_sco import parse_sco"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialize data structures that we'll store frequency data in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 64x63 dict: each key is a triplet of {A, C, G, T}, and each value is another dict with all the other codons\n",
    "codon2codon2freq = {}\n",
    "\n",
    "# 21x20 dict: each key is a proteinogenic amino acid (A, C, D, E, F, ...), limited to just\n",
    "# stuff in the standard genetic code (i.e. ignoring selenocystine and pyrrolsine) but including\n",
    "# \"*\", representing a stop codon.\n",
    "aa2aa2freq = {}\n",
    "\n",
    "# 64-key dict: maps each triplet to an integer indicating how frequently this triplet occurs in all genes\n",
    "# in the genomes (i.e. not counting mutations into this triplet).\n",
    "codon2freq = {}\n",
    "\n",
    "# 21-key dict: maps amino acid/stop codon to integer indicating frequency across all genes.\n",
    "aa2freq = {}\n",
    "\n",
    "# There's probably a fancier way of generating this list, but this is fine.\n",
    "codons = []\n",
    "# Also, we figure out the reverse complements of each of the 64 3-mers in advance -- this avoids\n",
    "# us having to call str(skbio.DNA(c).reverse_complement()) every time we see a codon, and saves a tiny\n",
    "# amount of time per read (the skbio approach took ~9e-5 seconds every time; the new approach takes ~9e-7\n",
    "# seconds every time). Considering we're going through well over a million reads, the time savings comes out\n",
    "# to ... 130.977 seconds, aka 2 minutes 10 seconds, if I'm computing this correctly. So, not much, but it's\n",
    "# something!\n",
    "codon2revcomp = {}\n",
    "nts = \"ACGT\"\n",
    "for i in nts:\n",
    "    for j in nts:\n",
    "        for k in nts:\n",
    "            c = \"{}{}{}\".format(i, j, k)\n",
    "            codons.append(c)\n",
    "            codon2revcomp[c] = str(skbio.DNA(c).reverse_complement())\n",
    "\n",
    "aas = set([])\n",
    "for c in codons:\n",
    "    aas.add(str(skbio.DNA(c).translate()))\n",
    "    \n",
    "# Initialize dicts to 0s\n",
    "for c1 in codons:\n",
    "    codon2codon2freq[c1] = {c2: 0 for c2 in set(codons) - set([c1])}\n",
    "    codon2freq[c1] = 0\n",
    "    \n",
    "for aa1 in aas:\n",
    "    aa2aa2freq[aa1] = {aa2: 0 for aa2 in set(aas) - set([aa1])}\n",
    "    aa2freq[aa1] = 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Go through all reads aligned to each genome\n",
    "\n",
    "Define a dict which we'll use to keep track of aligned codon frequencies for each codon, for each gene, for each genome.\n",
    "\n",
    "- For each read, see which predicted genes (if any) this read intersects within the genome. Note that \"intersects\" doesn't mean \"fully covers\".\n",
    "\n",
    "- For each of these genes, see which codons (if any) this read fully covers within the gene.\n",
    "\n",
    "- Increment aligned codon frequencies for all codons accordingly.\n",
    "\n",
    "The reason we do things this way, as opposed to iterating over just the reads overlapping each codon in each gene, is that doing things that way is really slow! I'm pretty sure it's because \"find out which reads overlap this region\" is a pretty slow operation when working with large datasets -- and also since these are long reads, doing this on the level of each codon means we're effectively doing a lot of redundant work (you can imagine that, for a given codon, the odds are pretty good that most reads overlapping it will also overlap adjacent codon(s))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Maps sequence IDs to genes (keyed by their Index in the .sco file) to codons (keyed by (0-indexed!)\n",
    "# left end, i.e. the lower of the two positional boundaries of the codon, regardless of if its gene\n",
    "# is on the + or - strand) to observed aligned codon frequencies (keyed by just the triplet, e.g. \"AAA\").\n",
    "#\n",
    "# Example:\n",
    "# {\"edge_6104\":                                Sequence\n",
    "#     {1:                                      Gene index in the .sco file\n",
    "#         {265:                                Left codon position\n",
    "#             {\"TTA\": 1000, \"TTT\": 1, ... }    Aligned codon frequencies for this particular codon\n",
    "#         }\n",
    "#     }\n",
    "# }\n",
    "seq2gene2codon2alignedcodons = {}\n",
    "\n",
    "# This mapping has the same structure as the one above for the first three levels; however, the bottom level\n",
    "# instead maps codons to the read names we've seen aligned to this codon.\n",
    "#\n",
    "# The reason for this is so we can check to see if there are any reads that happen to cover the same codon\n",
    "# more than once -- if so we just ignore these additional alignments, I guess? Or we could include them, but\n",
    "# in either case we should mention this in the report because this is a weird thing.\n",
    "#\n",
    "# (Since we've filtered out secondary alignments, the only case in which reads could align to a codon > 1\n",
    "# time -- at least as far as I can tell -- is if the read has supplementary alignment(s) that overlap. This is\n",
    "# possible but probably not a very common scenario, since it's expected per the SAM specification that\n",
    "# supplementary alignments have only small overlaps.)\n",
    "seq2gene2codon2seenreads = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-2-38182d14e5c3>, line 35)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-2-38182d14e5c3>\"\u001b[0;36m, line \u001b[0;32m35\u001b[0m\n\u001b[0;31m    for gene_data.in df.itertuples():\u001b[0m\n\u001b[0m                   ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "bf = pysam.AlignmentFile(\"../main-workflow/output/fully-filtered-and-sorted-aln.bam\", \"rb\")\n",
    "\n",
    "for seq in SEQS:\n",
    "    fasta = skbio.DNA.read(\"../seqs/{}.fasta\".format(seq))\n",
    "    df = parse_sco(\"../seqs/genes/{}.sco\".format(seq))\n",
    "    \n",
    "    # Initialize some of the data structures\n",
    "    # NOTE: this is kind of slow. However, it still finishes within a few seconds, so not the most\n",
    "    # important thing to optimize\n",
    "    seq2gene2codon2alignedcodons[seq] = {}\n",
    "    seq2gene2codon2seenreads[seq] = {}\n",
    "    for gene_data in df.itertuples():\n",
    "        \n",
    "        # Should never happen, but check this so that we can compute overlap easily and with peace of mind later\n",
    "        if gene_data.LeftEnd >= gene_data.RightEnd:\n",
    "            raise ValueError(\"Gene {}'s coordinates seem messed up: left = {}, right = {}\".format(\n",
    "                gene_data.Index, gene_data.LeftEnd, gene_data.RightEnd\n",
    "            ))\n",
    "        \n",
    "        seq2gene2codon2alignedcodons[seq][gene_data.Index] = {}\n",
    "        seq2gene2codon2seenreads[seq][gene_data.Index] = {}\n",
    "        \n",
    "        codon_positions = [\n",
    "            i for i in range(gene_data.LeftEnd, gene_data.RightEnd + 1, 3)\n",
    "        ]\n",
    "\n",
    "        # For each codon in this gene...\n",
    "        for cpleft in codon_positions:\n",
    "            seq2gene2codon2alignedcodons[seq][gene_data.Index][cpleft] = defaultdict(int)\n",
    "            seq2gene2codon2seenreads[seq][gene_data.Index][cpleft] = []\n",
    "            \n",
    "    print(\"Finished initialization for seq = {}\".format(seq))\n",
    "    readtimes = []\n",
    "    \n",
    "    # Note that this isn't really a \"read\" so much as it is an aligned linear segment (and a read\n",
    "    # can have multiple such segments derived from it, as discussed above).\n",
    "    for ri, read in enumerate(bf.fetch(seq)):\n",
    "        \n",
    "        t1 = time.time()\n",
    "        \n",
    "        # Find all genes that this read intersects in this genome\n",
    "        \n",
    "        # These are 0-indexed coordinates (and segright is offset to the right by one; see\n",
    "        # https://pysam.readthedocs.io/en/latest/api.html#pysam.AlignedSegment.reference_end)\n",
    "        segleft = read.reference_start\n",
    "        segright = read.reference_end\n",
    "        \n",
    "        if segleft is None or segright is None:\n",
    "            raise ValueError(\"Read {} is unmapped? This shouldn't happen!\".format(seg.query_name))\n",
    "        \n",
    "        if segleft >= segright:\n",
    "            raise ValueError(\"Read {}'s coordinates in pysam seem messed up: left = {}, right = {}\".format(\n",
    "                seg.query_name, segleft, segright\n",
    "            ))\n",
    "        \n",
    "        # Convert aligned segment boundaries to 1-indexed coordinates to make comparing with gene\n",
    "        # coordinates from the .sco file easier.\n",
    "        # Since segright was already offset to the right by 1, we don't need to do anything for it\n",
    "        # (the gene coordinates are exact: a gene from [266, 712] starts at base 266 and ends at base 712,\n",
    "        # using 1-indexing. So in order to make the read boundaries match, we'd add 1 for segright and then\n",
    "        # subract 1 since segright was already 1 base off -- and n + 1 - 1 = n. (...math is hard)\n",
    "        segleft += 1\n",
    "        \n",
    "        # This could probably be vectorized to make it faster\n",
    "        genes_overlapping_read = []\n",
    "        for gene_data in df.itertuples():\n",
    "            # Quick test for overlap between two ranges where left is guaranteed to be < right.\n",
    "            # See https://stackoverflow.com/a/325964.\n",
    "            if segleft <= gene_data.RightEnd and gene_data.LeftEnd <= segright:\n",
    "                genes_overlapping_read.append(gene_data)\n",
    "\n",
    "        # (Debugging code)\n",
    "        # print(\"{} genes overlap read {}\".format(len(genes_overlapping_read), ri))\n",
    "        # print(\"Read {}, which ranges from {} to {}, overlaps these genes:\".format(ri, segleft, segright))\n",
    "        # print(genes_overlapping_read)\n",
    "                \n",
    "        # If no genes overlap this read, we are free to move on to the next read\n",
    "        if len(genes_overlapping_read) > 0:\n",
    "            \n",
    "            # Computing this is relatively slow, which is why we jump through so many hoops before we do this.\n",
    "            # Each entry in get_aligned_pairs() is a tuple with 2 elements:\n",
    "            # the first is the read pos and the second is the reference pos.\n",
    "            ap = read.get_aligned_pairs(matches_only=True)\n",
    "            \n",
    "            # We only consider the leftmost position of each codon, so we don't need to bother checking the last\n",
    "            # two pairs of positions (since neither could be the leftmost position of a codon that this read\n",
    "            # fully covers).\n",
    "            for api, pair1 in enumerate(ap[:-2]):\n",
    "\n",
    "                # Convert to 1-indexed position for ease of comparison with gene coordinates\n",
    "                pair1_refpos = pair1[1] + 1\n",
    "                for gene_data in genes_overlapping_read:\n",
    "                    \n",
    "                    gi = gene_data.Index\n",
    "                    gl = gene_data.LeftEnd\n",
    "                    gr = gene_data.RightEnd\n",
    "                    \n",
    "                    # If this pair is located within a gene:\n",
    "                    if pair1_refpos >= gl and pair1_refpos <= gr:\n",
    "                        \n",
    "                        # If this pair is located at the leftmost position of a codon in this gene:\n",
    "                        # (Note that check works for both + or - strand genes. whether the leftmost position\n",
    "                        # is the \"start\" [i.e. CP 1] or \"end\" [i.e. CP 3] of the gene changes with the strand\n",
    "                        # of the gene, but underlying math is the same.)\n",
    "                        if (pair1_refpos - gl) % 3 == 0:\n",
    "                            \n",
    "                            # Check that the pairs are all consecutive (i.e. no \"jumps\" in the read,\n",
    "                            # and no \"jumps\" in the reference)\n",
    "                            # Since we don't consider the last two pairs in ap, pair2 and pair3 should\n",
    "                            # always be available.\n",
    "                            pair2 = ap[api + 1]\n",
    "                            pair3 = ap[api + 2]\n",
    "                            readpos_consecutive = pair2[0] == pair1[0] + 1 and pair3[0] == pair2[0] + 1\n",
    "                            refpos_consecutive =  pair2[1] == pair1[1] + 1 and pair3[1] == pair2[1] + 1\n",
    "                            \n",
    "                            if refpos_consecutive and readpos_consecutive:\n",
    "                                # Nice! This read fully covers this codon.\n",
    "                                # Figure out what the read actually *says* in the alignment to the codon.\n",
    "                                # (It'll probably be a complete match most of the time, but there will\n",
    "                                # be some occasional mismatches -- and seeing those is ... the whole point\n",
    "                                # of this notebook.)\n",
    "\n",
    "                                # We make sure to index the read by read coords, not reference coords!\n",
    "                                aligned_codon = read.query_sequence[pair1[0]: pair1[0] + 3]\n",
    "\n",
    "                                if gene_data.Strand == \"-\":\n",
    "                                    aligned_codon = codon2revcomp[aligned_codon]\n",
    "                                    \n",
    "                                if len(aligned_codon) != 3:\n",
    "                                    # This actually happened when I indexed the reads incorrectly and Python\n",
    "                                    # just silently returned an empty string lol\n",
    "                                    print(read, read.query_sequence, gene_data, pair, aligned_codon)\n",
    "                                    raise ValueError(\"^^^Something went very wrong with read coordinates!\")\n",
    "\n",
    "                                # Finally, update information about codon frequencies.\n",
    "                                seq2gene2codon2alignedcodons[seq][gi][pair1_refpos][aligned_codon] += 1\n",
    "                                if read.query_name not in seq2gene2codon2seenreads[seq][gi][pair1_refpos]:\n",
    "                                    seq2gene2codon2seenreads[seq][gi][pair1_refpos].append(read.query_name)\n",
    "                                else:\n",
    "                                    raise ValueError(\n",
    "                                        \"Oh wow, read {} covers codon {} in {} more than once???\".format(\n",
    "                                            read.query_name, pair1_refpos, seq\n",
    "                                        )\n",
    "                                    )\n",
    "        t2 = time.time()\n",
    "        readtimes.append(t2 - t1)\n",
    "        if ri % 100 == 0:\n",
    "            print(\"Seen {} reads so far in {}.\".format(ri, seq))\n",
    "    # At this point, we've seen all the reads aligned to all the codons in this genome.\n",
    "    # We can now \"call\" mutations based on the frequencies we've counted.\n",
    "    # TODO: do that!\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bf = pysam.AlignmentFile(\"../main-workflow/output/fully-filtered-and-sorted-aln.bam\", \"rb\")\n",
    "\n",
    "for seq in SEQS:\n",
    "    fasta = skbio.DNA.read(\"../seqs/{}.fasta\".format(seq))\n",
    "    df = parse_sco(\"../seqs/genes/{}.sco\".format(seq))\n",
    "    # Note that we don't explicitly account for gene overlap, so the same position could be considered multiple\n",
    "    # times if it's present within multiple genes.\n",
    "    for gene_data in df.itertuples():\n",
    "        \n",
    "\n",
    "        \n",
    "        # For each codon in this gene...\n",
    "        for cps in codon_positions:\n",
    "            \n",
    "            # The codon positions we computed are 1-indexed, so in order to work with pysam/other python libs\n",
    "            # (which usually use 0-indexing) we reduce the start position by 1 and don't add 1 to the stop\n",
    "            # position.\n",
    "            start = min(cps) - 1\n",
    "            stop = max(cps)\n",
    "            \n",
    "            # Make note of the codon sequence and amino acid encoded by this codon in the \"reference\" genome.\n",
    "            codon_dna = fasta[start: stop]\n",
    "            if gene_data.Strand == \"-\":\n",
    "                codon_dna = codon_dna.reverse_complement()\n",
    "\n",
    "            codon_seq = str(codon_dna)\n",
    "            aa = str(codon_dna.translate())\n",
    "            \n",
    "            # Update frequencies accordingly.\n",
    "            codon2freq[codon_seq] += 1\n",
    "            aa2freq[aa] += 1\n",
    "\n",
    "            # Find all reads that span this codon without indels/skips,\n",
    "            # and save their sequence here.\n",
    "            # Idea to use defaultdict(int) for this from https://stackoverflow.com/a/722702\n",
    "            aligned_codons = defaultdict(int)\n",
    "            \n",
    "            num_good_reads = 0\n",
    "            seen_read_names = []\n",
    "            for read in bf.fetch(seq, start, stop):\n",
    "                \n",
    "                # Check that the read has no indels intersecting with this codon.\n",
    "                # We do this by going through all aligned pairs. Once we find the position aligned\n",
    "                # to the start of this codon, check that the next two codon positions follow immediately\n",
    "                # after in the aligned pairs (and that there aren't any Nones in these pairs indicating\n",
    "                # indels/skips/etc).\n",
    "                #\n",
    "                # get_aligned_pairs() is kinda slow -- we can speed it up a bit by passing matches_only=True\n",
    "                # since we don't care about skips/indels. For the sake of safety, and because I originally\n",
    "                # wrote this without using matches_only=True, we still check for Nones anyway.\n",
    "                ap = read.get_aligned_pairs(matches_only=True)\n",
    "                for i, pair in enumerate(ap):\n",
    "                    if pair[1] == start:\n",
    "                        try:\n",
    "                            pair2 = ap[i + 1]\n",
    "                            pair3 = ap[i + 2]\n",
    "                        except IndexError:\n",
    "                            # If there just straight up aren't any pairs after this one, then that's fine:\n",
    "                            # it means this read doesn't fully overlap the codon. Just move on to the next\n",
    "                            # read.\n",
    "                            break\n",
    "\n",
    "                        # Check that none of the pairs have skips/indels.\n",
    "                        # For an aligned pair, [0] is the read pos and [1] is the reference pos\n",
    "                        if pair[0] is not None and pair2[0] is not None and pair3[0] is not None:\n",
    "\n",
    "                            # Check that the pairs are all consecutive\n",
    "                            refpos_consecutive =  pair2[1] == start + 1 and pair3[1] == stop - 1\n",
    "                            readpos_consecutive = pair2[0] == pair[0] + 1 and pair3[0] == pair2[0] + 1\n",
    "                            if refpos_consecutive and readpos_consecutive:\n",
    "                                # Make sure to index the read by read coordinates, not reference coords!\n",
    "                                aligned_seq = read.query_sequence[pair[0]: pair3[0] + 1]\n",
    "\n",
    "                                if gene_data.Strand == \"-\":\n",
    "                                    aligned_seq = str(skbio.DNA(aligned_seq).reverse_complement())\n",
    "\n",
    "                                if len(aligned_seq) != 3:\n",
    "                                    # This actually happened when I indexed the reads incorrectly and Python\n",
    "                                    # just silently returned an empty string lol\n",
    "                                    print(read, read.query_sequence, start, stop, aligned_seq)\n",
    "                                    raise ValueError(\"Something went very wrong with read coordinates!\")\n",
    "\n",
    "                                aligned_codons[aligned_seq] += 1\n",
    "                                num_good_reads += 1\n",
    "                                seen_read_names.append(read.query_name)\n",
    "                                \n",
    "                        # Regardless of if this was a \"good\" read or not, this was our once chance at\n",
    "                        # aligning this read to this codon -- so no need to continue going through the aligned\n",
    "                        # pairs. Break out of that loop.\n",
    "                        break\n",
    "                        \n",
    "            if len(set(seen_read_names)) < len(seen_read_names):\n",
    "                print(Counter(seen_read_names).most_common())\n",
    "                raise ValueError(\n",
    "                    \"Duplicate reads covering a single codon: figure out how to handle this situation.\"\n",
    "                )\n",
    "            print(\"{} (good) reads covered codon {}\".format(num_good_reads, codon_seq))\n",
    "            print(\"sum of vals of ac is {}\".format(sum(aligned_codons.values())))\n",
    "            print(\"Codon {} from {} to {} in gene {} in seq {} has mutations: {}\".format(\n",
    "                codon_seq, start, stop, gene_data.Index, seq, aligned_codons\n",
    "            ))\n",
    "            # Now, we can finally compute stats re: number of mismatching and matching codons\n",
    "            num_aligned_codons = sum(aligned_codons.values())\n",
    "            alt_codon_frac = (num_aligned_codons - aligned_codons[codon_seq]) / num_aligned_codons\n",
    "            \n",
    "            # Using minfreq = 0.5%\n",
    "            if alt_codon_frac > 0.005:\n",
    "                \n",
    "                alt_codons = {c: aligned_codons[c] for c in aligned_codons if c != codon_seq}\n",
    "                # Retrieve max-freq alternate codon.\n",
    "                # Based on https://stackoverflow.com/a/280156.\n",
    "                # (Note that if there's a tie, the result is arbitrary. Shouldn't be a big deal. Making note\n",
    "                # of in the paper.)\n",
    "                max_freq_alt_codon = max(alt_codons, key=alt_codons.get)\n",
    "                codon2codon2freq[codon_seq][max_freq_alt_codon] += 1\n",
    "                \n",
    "                print(\"Is mutation! And max freq alt codon is {}\".format(max_freq_alt_codon))\n",
    "                \n",
    "                # NOTE: I guess you could argue that we should do this another way, where we actually compute\n",
    "                # the translations of all the alt codons and then pick the most common AA/stop codon from there?\n",
    "                #\n",
    "                # You could argue this either way: doing it based on just the mutated codon keeps the matrices\n",
    "                # consistent and lessens the impact of small errors, while taking into account all alt codon\n",
    "                # translations could help show weird things where multiple mutations have similar consequences.\n",
    "                # Hmm.\n",
    "                #\n",
    "                # TODO: think about!\n",
    "                alt_codon_aa = str(skbio.DNA(max_freq_alt_codon).translate())\n",
    "                if alt_codon_aa != aa:\n",
    "                    aa2aa2freq[aa][alt_codon_aa] += 1\n",
    "                    print(\"Is nonsyn mutation! Alt {} codes for {}; orig coded for {}\".format(\n",
    "                        max_freq_alt_codon, alt_codon_aa, aa\n",
    "                    ))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basic pseudocode\n",
    "\n",
    "- Set up a 64x64 __codon mutation matrix__ of all zeroes (numpy?). Notably, there is no guarantee of symmetricity here, although we may expect this to be roughly symmetric.\n",
    "  - Define a mapping codon2index for this matrix. This should be in alphabetical order, so AAA -> 0, AAC -> 1, ..., TTT -> 63.\n",
    "\n",
    "\n",
    "- Set up a 21x21 __amino acid mutation matrix__ of all zeroes. Similarly, no symmetric guarantee.\n",
    "  - Define a mapping aa2index for this matrix. Also in alphabetical order: so (see [here](https://en.wikipedia.org/wiki/International_Union_of_Pure_and_Applied_Chemistry#Amino_acid_and_nucleotide_base_codes)), it'll look like A, C, ..., Y, \\*. We add an extra row/column for \\*, a stop codon -- for consistency, since mutations into or from stop codons can happen.\n",
    "  \n",
    "\n",
    "- For every genome\n",
    "  - For every gene\n",
    "    - For every codon (i.e. first look at [0, 1, 2] in this gene, then [3, 4, 5], etc.)\n",
    "      - Filter to reads that completely cover this codon's coordinates using pysam.\n",
    "      - Determine what each of these read spells in this 3-mer.\n",
    "      - Choose the most common codon out of all mutated 3-mers. If its frequency out of all reads covering this codon is at least 0.5%, call it a mutated codon.\n",
    "      - Update the codon mutation matrix and amino acid matrix based on the 3-mer and corresponding amino acid (or stop codon) spelled out by the original and mutated 3-mer sequence.\n",
    "\n",
    "- Download the matrices to my laptop!\n",
    "\n",
    "-----\n",
    "\n",
    "- ALTERNATELY: We could go through just the mutation rates, assuming that all reads intersecting a codon cover the thing entirely (which will probs usually but not always be true). We can find mutated codons by looking at individual positions' mutation rates. However, the method above is closer to what was requested, and at this point makes more sense to me."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (qiime2-2021.2)",
   "language": "python",
   "name": "myenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
