{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Demonstrate the \"target-decoy\" approach, as applied to metagenomic variant calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-19T09:12:46.072627Z",
     "iopub.status.busy": "2021-06-19T09:12:46.071700Z",
     "iopub.status.idle": "2021-06-19T09:12:46.350765Z",
     "shell.execute_reply": "2021-06-19T09:12:46.349866Z"
    }
   },
   "outputs": [],
   "source": [
    "%run \"Header.ipynb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import copy\n",
    "import json\n",
    "import pickle\n",
    "import skbio\n",
    "import pileup\n",
    "from math import ceil\n",
    "from collections import defaultdict\n",
    "from pysam import VariantFile\n",
    "from parse_sco import parse_sco"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq2pos2pileup = pileup.load()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Lower limit of $p$\n",
    "\n",
    "Basically: since we only call a mutation at a given position `pos` using the naive method if `alt(pos) ≥ 2`, it doesn't make sense to use values of $p$ less than `2 / coverage` (coverage varies throughout a MAG, but let's use the rounded average coverage as the denominator here). So,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The mean cov for seq CAMP is 4,158.57x.\n",
      "The mean cov for seq BACT1 is 1,415.07x.\n",
      "The mean cov for seq BACT2 is 2,993.46x.\n",
      "Across the three selected MAGs, the min mean cov is 1,415.07x.\n"
     ]
    }
   ],
   "source": [
    "min_mean_cov = float(\"inf\")\n",
    "# This is copied from the SequenceCoveragePlots.ipynb notebook; easier to just duplicate the code for now, sorry!\n",
    "seq2mean_cov = {}\n",
    "for seq in SEQS:\n",
    "    covs = []\n",
    "    for pos in range(1, seq2len[seq] + 1):\n",
    "        covs.append(pileup.get_cov(seq2pos2pileup[seq][pos]))\n",
    "    mean_cov = mean(covs)\n",
    "    seq2mean_cov[seq] = mean_cov\n",
    "    min_mean_cov = min(mean_cov, min_mean_cov)\n",
    "    print(f\"The mean cov for seq {seq2name[seq]} is {mean_cov:,.2f}x.\")\n",
    "    \n",
    "print(f\"Across the three selected MAGs, the min mean cov is {min_mean_cov:,.2f}x.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "So the lowest value of p it makes sense to use is p = 100 * (2 / 1,415.07) = 0.1413%.\n",
      "Effective min p (taking the ceiling to the nearest 0.01%): 0.1500%.\n",
      "===============================================================================\n",
      "Just wrote out this stuff:\n",
      "===============================================================================\n",
      "The average coverages of the three selected genomes are 4,159x, 1,415x, and 2,993x for the CAMP, BACT1, and BACT2 genomes, respectively. \n",
      "===============================================================================\n",
      "Just wrote out this stuff:\n",
      "===============================================================================\n",
      "We thus set the lower limit of $p$ as $\\frac{2}{1,415.07} \\approx 0.1413\\%$. Rounding this value up to the next largest 0.01\\% gives us a lower limit $p$ of 0.15\\%.\\endinput\n"
     ]
    }
   ],
   "source": [
    "# yeah yeah i know this is technically just 200 / min_mean_cov but clarity > performance imo and also this is\n",
    "# just an analysis notebook\n",
    "min_p = 100 * (2 / min_mean_cov)\n",
    "print(f\"So the lowest value of p it makes sense to use is p = 100 * (2 / {min_mean_cov:,.2f}) = {min_p:.4f}%.\")\n",
    "\n",
    "effective_min_p = ceil(min_p * 100) / 100\n",
    "# I know .4f is unnecessary precision since we just rounded to a multiple of 0.01%,\n",
    "# this is to sanity check that I did this right\n",
    "print(f\"Effective min p (taking the ceiling to the nearest 0.01%): {effective_min_p:.4f}%.\")\n",
    "\n",
    "with open(\"misc-text/fdr-min-p-avg-covs.tex\", \"w\") as mpf:\n",
    "    stuff_to_write = (\n",
    "        f\"The average coverages of the three selected genomes are \"\n",
    "        f\"{seq2mean_cov['edge_6104']:,.0f}x, {seq2mean_cov['edge_1671']:,.0f}x, and \"\n",
    "        f\"{seq2mean_cov['edge_2358']:,.0f}x for the {seq2name['edge_6104']}, {seq2name['edge_1671']}, and \"\n",
    "        f\"{seq2name['edge_2358']} genomes, respectively. \"\n",
    "    )\n",
    "    mpf.write(stuff_to_write)\n",
    "    print(f\"{'=' * 79}\\nJust wrote out this stuff:\\n{'=' * 79}\\n{stuff_to_write}\")\n",
    "    \n",
    "with open(\"misc-text/fdr-min-p-val.tex\", \"w\") as mpf:\n",
    "    stuff_to_write = (\n",
    "        f\"We thus set the lower limit of $p$ as \"\n",
    "        r\"$\\frac{2}{\"\n",
    "        f\"{seq2mean_cov['edge_1671']:,.2f}\"\n",
    "        r\"} \\approx \"\n",
    "        f\"{min_p:.4f}\\%$. Rounding this value up to the next largest \"\n",
    "        f\"0.01\\% gives us a lower limit $p$ of {effective_min_p:.2f}\\%.\"\n",
    "        r\"\\endinput\"\n",
    "    )\n",
    "    mpf.write(stuff_to_write)\n",
    "    print(f\"{'=' * 79}\\nJust wrote out this stuff:\\n{'=' * 79}\\n{stuff_to_write}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## First: naive variant calling\n",
    "\n",
    "We don't limit to \"sufficiently-covered\" positions here -- so we consider all regions throughout a genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-19T09:12:46.900714Z",
     "iopub.status.busy": "2021-06-19T09:12:46.899915Z",
     "iopub.status.idle": "2021-06-19T09:12:46.902209Z",
     "shell.execute_reply": "2021-06-19T09:12:46.901568Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First two percentages: [5.0, 4.99]\n",
      "Last two percentages: [0.16, 0.15]\n",
      "Number of percentages: 486\n"
     ]
    }
   ],
   "source": [
    "# Percentages go from 5%, 4.99%, 4.98%, ..., min_p + 0.01%, min_p%.\n",
    "# Since we draw FDR curves, etc. by varying p by 0.01%, we'll use a lower limit of ceil(min_p * 100) / 100.\n",
    "# We limit the start here to 5% because the whole \"rare mutations\" definition means that we ignore p-mutations\n",
    "# at higher values of p. Since this notebook is pretty inefficient, there's no point unnecessarily computing\n",
    "# stuff for p = 50%, etc.\n",
    "effective_min_p_times_100 = ceil(min_p * 100)\n",
    "percentages = [p / 100 for p in range(effective_min_p_times_100, 501, 1)][::-1]\n",
    "print(f\"First two percentages: {percentages[:2]}\")\n",
    "print(f\"Last two percentages: {percentages[-2:]}\")\n",
    "print(f\"Number of percentages: {len(percentages):,}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_num_mutations_per_mb(num_called_mutations_in_positions, num_positions_considered):\n",
    "    # We have the equation\n",
    "    #\n",
    "    #      # called mutations              f\n",
    "    # ---------------------------- = --------------\n",
    "    #  # of positions to consider     1,000,000 bp\n",
    "    #\n",
    "    # This function just solves for f by multiplying the left side of the equation by 1,000,000.\n",
    "    #\n",
    "    # I guess if you're gonna be calling this thousands of time on the same set of positions\n",
    "    # it might speed this up slightly to precompute the (1,000,000 / # of positions to consider) value,\n",
    "    # but convenience is the most important thing here IMO.\n",
    "    return (num_called_mutations_in_positions / num_positions_considered) * 1000000\n",
    "    \n",
    "\n",
    "def naive_calling(seq, positions_to_consider, verbose=True, superverbose=False):\n",
    "    \"\"\"seq should be in SEQS.\n",
    "    \n",
    "    positions_to_consider should be a collection of 1-indexed positions in the sequence to consider when\n",
    "    calling mutations. This makes it possible to, for example, just consider the CP 2 positions in a sequence.\n",
    "    \n",
    "    Returns a tuple with three elements:\n",
    "    \n",
    "    1. p2called_mutations: A dict mapping values in \"percentages\" (defined above) to a list of 1-indexed\n",
    "       \"called\" p-mutations in the positions to consider in the sequence, using this percentage for p.\n",
    "       \n",
    "    2. p2numpermb: A dict with the same keys as p2called_mutations, but the values are the number of called\n",
    "       p-mutations per megabase (1,000,000 bp = 1 Mbp) in the positions to consider in this sequence.\n",
    "       IT'S DEFINITELY WORTH NOTING that we scale this by the number of positions to consider, not the full\n",
    "       sequence length (although it's possible that those values could be equal if positions_to_consider\n",
    "       is equal to range(1, seq2len[seq] + 1)). So, if you select a subset of positions where most of them\n",
    "       are mutations, this'll result in a really high number of mutations per megabase!\n",
    "       ...I recognize \"numpermb\" doesn't really roll off the tongue, but I couldn't think\n",
    "       of a better name for this :P\n",
    "       \n",
    "    3. poslen: Length of positions_to_consider, for reference.\n",
    "    \n",
    "    NOTE THAT this'll not include \"high-frequency\" mutations in the results, as discussed in the paper.\n",
    "    We will not decrease poslen or anything -- positions with high-frequency mutations will just be\n",
    "    treated as if they were non-mutated positions.\n",
    "    \"\"\"\n",
    "    poslen = len(positions_to_consider)\n",
    "    seqlen = seq2len[seq]\n",
    "    positions_to_consider_pct = 100 * (poslen / seqlen)\n",
    "    if verbose:\n",
    "        print(f\"Naively calling mutations in {seq2name[seq]}.\")\n",
    "        print(f\"\\tConsidering {poslen:,} / {seqlen:,} ({positions_to_consider_pct:.2f}%) positions.\")\n",
    "    p2called_mutations = {p: [] for p in percentages}\n",
    "    p2numpermb = {}\n",
    "    for pi, pos in enumerate(sorted(positions_to_consider), 1):\n",
    "        if verbose and (pi == 1 or pi % 100000 == 0):\n",
    "            print(f\"\\tOn the {pi:,}-th position ({pos:,}) of the specified {poslen:,} positions ({100 * (pi / poslen):.2f}%).\")\n",
    "        pos_pileup = seq2pos2pileup[seq][pos]\n",
    "        for p in percentages:\n",
    "            # Note that this includes the min alt(pos) check made by naively_call_mutation() (see docs there)\n",
    "            # as well as the only-calling-if-rare thing\n",
    "            if pileup.naively_call_mutation(pos_pileup, p, only_call_if_rare=True):\n",
    "                p2called_mutations[p].append(pos)\n",
    "\n",
    "    for p in p2called_mutations:\n",
    "        num_called_mutations = len(p2called_mutations[p])\n",
    "        p2numpermb[p] = compute_num_mutations_per_mb(num_called_mutations, poslen)\n",
    "        # We add an extra layer of verbosity here because printing out 2 lines per value of p gets\n",
    "        # ridiculous when there are 1,000 values of p .____.\n",
    "        if superverbose:\n",
    "            print(f\"\\tp = {p}%: {num_called_mutations:,} called p-mutations in {seq2name[seq]}.\")\n",
    "            print(f\"\\t\\tNumber of called p-mutations per megabase: f = {f:,.2f}.\")\n",
    "        \n",
    "    return (p2called_mutations, p2numpermb, poslen)\n",
    "\n",
    "\n",
    "def naive_calling_fullseq(seq):\n",
    "    \"\"\"Does naive variant calling across all positions in a sequence (should be in SEQS).\"\"\"\n",
    "    return naive_calling(seq, range(1, seq2len[seq] + 1))\n",
    "\n",
    "\n",
    "def get_single_gene_cp2_positions(seq):\n",
    "    cp2_positions = set()\n",
    "    multi_gene_positions = set()\n",
    "    \n",
    "    seqlen = seq2len[seq]\n",
    "    genes_df = parse_sco(f\"../seqs/genes/{seq}.sco\")\n",
    "    \n",
    "    # Code here is adapted from get_parent_gene_info_of_many_positions (in Header.ipynb) a bit\n",
    "    # Faster to compute everything at once, rather than iterate through the genes multiple times\n",
    "    pos_to_genes = defaultdict(list)\n",
    "    \n",
    "    for gene in genes_df.itertuples():\n",
    "        gene_left = int(gene.LeftEnd)\n",
    "        gene_right = int(gene.RightEnd)\n",
    "        gene_num = int(gene.Index)\n",
    "        gene_strand = gene.Strand\n",
    "        \n",
    "        def complainAboutCPs(gn, gs, gcp):\n",
    "            raise ValueError(f\"CP got out of whack: gene {gn}, strand {gs}, cp {gcp}?\")\n",
    "        \n",
    "        if gene_strand == \"+\":\n",
    "            cp = 1\n",
    "        else:\n",
    "            cp = 3\n",
    "        \n",
    "        for pos in range(gene_left, gene_right + 1):\n",
    "            pos_to_genes[pos].append(gene_num)\n",
    "            if len(pos_to_genes[pos]) > 1:\n",
    "                multi_gene_positions.add(pos)\n",
    "                \n",
    "            if cp == 2:\n",
    "                cp2_positions.add(pos)\n",
    "                \n",
    "            # Adjust the CP. I already have some code that does this (in a different context) in the within-\n",
    "            # gene mutation spectrum notebook; ideally this code would be generalized between the notebooks.\n",
    "            if gene_strand == \"+\":\n",
    "                # For + strand genes, this goes 123123123123...\n",
    "                if cp == 1 or cp == 2: cp += 1\n",
    "                elif cp == 3: cp = 1\n",
    "                else: complainAboutCPs(gene_num, gene_strand, cp)\n",
    "            else:\n",
    "                # For - strand genes, this goes 321321321321...\n",
    "                if cp == 3 or cp == 2: cp -= 1\n",
    "                elif cp == 1: cp = 3\n",
    "                else: complainAboutCPs(gene_num, gene_strand, cp)\n",
    "    \n",
    "    single_gene_cp2_positions = cp2_positions - multi_gene_positions\n",
    "    return single_gene_cp2_positions\n",
    "    \n",
    "\n",
    "def naive_calling_cp2seq(seq):\n",
    "    \"\"\"Does naive variant calling across just the CP 2 positions in a sequence (should be in SEQS).\n",
    "    \n",
    "    NOTE that this will filter only to positions that meet the exact criteria:\n",
    "    \n",
    "    - In a single gene (not in a position that is covered by overlapping genes).\n",
    "    - In CP 2 within this single gene.\n",
    "    \n",
    "    Even if a position is in CP 2 of all the multiple genes it's covered by, we'll still ignore it.\n",
    "    I'm pretty sure there should be very few positions that get tossed out as a result; my take is that\n",
    "    it isn't worth the trouble to try to handle these positions.\n",
    "    \"\"\"\n",
    "\n",
    "    print(f\"Identifying CP 2 positions in {seq2name[seq]} so we can use them as a decoy genome...\")\n",
    "    single_gene_cp2_positions = get_single_gene_cp2_positions(seq)\n",
    "    print(f\"In {seq2name[seq]}:\")\n",
    "    #print(f\"\\tThere were {len(cp2_positions):,} CP 2 positions.\")\n",
    "    #print(f\"\\tThere were {len(multi_gene_positions):,} positions in multiple genes.\")\n",
    "    print(f\"\\tThere were {len(single_gene_cp2_positions):,} CP 2 positions in only a single gene.\")\n",
    "    return naive_calling(seq, single_gene_cp2_positions)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Naively call mutations in CAMP and compute $\\mathrm{frac}_{\\mathrm{decoy}}$\n",
    "\n",
    "(We're treating CAMP as a \"decoy\" genome, where we assume that all called mutations within it will be incorrect.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Naively calling mutations in CAMP.\n",
      "\tConsidering 1,289,244 / 1,289,244 (100.00%) positions.\n",
      "\tOn the 1-th position (1) of the specified 1,289,244 positions (0.00%).\n"
     ]
    }
   ],
   "source": [
    "camp_naive_p2called_mutations, camp_naive_p2numpermb, _ = naive_calling_fullseq(\"edge_6104\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "camp_cp2_naive_p2called_mutations, camp_cp2_naive_p2numpermb, num_camp_cp2_pos = naive_calling_cp2seq(\"edge_6104\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For comparison, naively call mutations in BACT1 and compute $\\mathrm{frac}_{\\mathrm{BACT1}}$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bact1_naive_p2called_mutations, bact1_naive_p2numpermb, _ = naive_calling_fullseq(\"edge_1671\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Just so we can update the `misc-text/` file, also do this for  BACT2\n",
    "\n",
    "probs possible to get this info from another notebook but this is the easiest way to handle this imo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bact2_naive_p2called_mutations, bact2_naive_p2numpermb, _ = naive_calling_fullseq(\"edge_2358\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save this info to text files using the json module\n",
    "\n",
    "... Since recomputing this takes, like, an hour"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"misc-output/p2called_mutations.txt\", \"w\") as p2cmf:\n",
    "    for p2cm in (\n",
    "        camp_naive_p2called_mutations, camp_cp2_naive_p2called_mutations,\n",
    "        bact1_naive_p2called_mutations, bact2_naive_p2called_mutations\n",
    "    ):\n",
    "        p2cmf.write(json.dumps(p2cm))\n",
    "        p2cmf.write(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"misc-output/p2numpermb.txt\", \"w\") as p2nf:\n",
    "    for p2n in (\n",
    "        camp_naive_p2numpermb, camp_cp2_naive_p2numpermb,\n",
    "        bact1_naive_p2numpermb, bact2_naive_p2numpermb\n",
    "    ):\n",
    "        p2nf.write(json.dumps(p2n))\n",
    "        p2nf.write(\"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load naive calling info from text files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"misc-output/p2called_mutations.txt\", \"r\") as p2cmf:\n",
    "    camp_naive_p2called_mutations = json.loads(p2cmf.readline().strip())\n",
    "    camp_cp2_naive_p2called_mutations = json.loads(p2cmf.readline().strip())\n",
    "    bact1_naive_p2called_mutations = json.loads(p2cmf.readline().strip())\n",
    "    bact2_naive_p2called_mutations = json.loads(p2cmf.readline().strip())\n",
    "    \n",
    "with open(\"misc-output/p2numpermb.txt\", \"r\") as p2nf:\n",
    "    camp_naive_p2numpermb = json.loads(p2nf.readline().strip())\n",
    "    camp_cp2_naive_p2numpermb = json.loads(p2nf.readline().strip())\n",
    "    bact1_naive_p2numpermb = json.loads(p2nf.readline().strip())\n",
    "    bact2_naive_p2numpermb = json.loads(p2nf.readline().strip())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# If we haven't done naive calling yet, we gotta compute this number also so we can use this in figures, etc.\n",
    "# And knowing the CP 2 positions anyway is needed for the LoFreq stuff.\n",
    "camp_cp2_pos = get_single_gene_cp2_positions(\"edge_6104\")\n",
    "num_camp_cp2_pos = len(camp_cp2_pos)\n",
    "print(f\"There are {num_camp_cp2_pos:,} CP 2 single-gene positions in CAMP.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Output info about FDR estimation for $p=0.5\\%$ to `misc-text/`\n",
    "\n",
    "**NOTE: this only includes \"rare\" (i.e. not high-frequency, as defined above) mutations.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Total numbers of identified p-mutations\n",
    "camp_nump = len(camp_naive_p2called_mutations[\"0.5\"])\n",
    "bact1_nump = len(bact1_naive_p2called_mutations[\"0.5\"])\n",
    "bact2_nump = len(bact2_naive_p2called_mutations[\"0.5\"])\n",
    "# Scaled numbers of identified p-mutations per megabase (comparable across different-length genomes\n",
    "# [at least, if you assume that genome length is the only confounding factor here, which it isn't -- we\n",
    "# should mention this in the paper ofc])\n",
    "camp_numpermb = camp_naive_p2numpermb[\"0.5\"]\n",
    "bact1_numpermb = bact1_naive_p2numpermb[\"0.5\"]\n",
    "bact2_numpermb = bact2_naive_p2numpermb[\"0.5\"]\n",
    "\n",
    "bact1_fdr = camp_numpermb / bact1_numpermb\n",
    "\n",
    "naiveinfo = (\n",
    "    f\"At $p=0.5$\\\\%, we identified {camp_nump:,}, {bact1_nump:,}, and {bact2_nump:,} rare $p$-mutations \"\n",
    "    f\"in the {seq2name['edge_6104']}, {seq2name['edge_1671']}, and {seq2name['edge_2358']} genomes, \"\n",
    "    f\"respectively. This illustrates that there exists a difference of nearly two orders of magnitude \"\n",
    "    f\"in the number of rare $p$-mutations per megabase across these genomes \"\n",
    "    f\"({round(camp_numpermb):,}, {round(bact1_numpermb):,}, and {round(bact2_numpermb):,} for \"\n",
    "    f\"{seq2name['edge_6104']}, {seq2name['edge_1671']}, and {seq2name['edge_2358']}, respectively). \"\n",
    "    f\"If the {seq2name['edge_6104']} genome, which has a relatively low mutation rate, is \"\n",
    "    f\"selected as a decoy, then we estimate the FDR for the {seq2name['edge_1671']} genome at $p=0.5\\\\%$ as \"\n",
    "    \"$\\\\frac{\" + f\"{camp_numpermb:,.2f}\" + \"}\" + \"{\" + f\"{bact1_numpermb:,.2f}\" + \"}\" + f\" \\\\approx {bact1_fdr:.4f}$.\"\n",
    ")\n",
    "with open(\"misc-text/naive-calling-target-decoy.tex\", \"w\") as of:\n",
    "    # see https://tex.stackexchange.com/a/18018\n",
    "    of.write(\"{}\\endinput\".format(naiveinfo))\n",
    "\n",
    "print(\"-\" * 79)\n",
    "print(\"RESULTS that we're gonna output to a TeX file:\")\n",
    "print(\"-\" * 79)\n",
    "print(naiveinfo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bact1_fdr_using_camp = camp_naive_p2numpermb[\"0.5\"] / bact1_naive_p2numpermb[\"0.5\"]\n",
    "bact1_fdr_using_campcp2 = camp_cp2_naive_p2numpermb[\"0.5\"] / bact1_naive_p2numpermb[\"0.5\"]\n",
    "cp2info = (\n",
    "    \"For example, at the frequency threshold $p=0.5$\\%, there are only \"\n",
    "    f\"{len(camp_cp2_naive_p2called_mutations['0.5']):,} rare $p$-mutations in CP2 in the {seq2name['edge_6104']} \"\n",
    "    f\"genome, resulting in {round(camp_cp2_naive_p2numpermb['0.5']):,} mutations per megabase of the CP2 \"\n",
    "    f\"positions (as compared to {round(camp_naive_p2numpermb['0.5']):,} for the entire {seq2name['edge_6104']} genome). \"\n",
    "    f\"Therefore, the estimate of the FDR for called mutations in the {seq2name['edge_1671']} genome \"\n",
    "    f\"can now be reduced from {bact1_fdr_using_camp:,.4f} to {bact1_fdr_using_campcp2:,.4f}.\"\n",
    ")\n",
    "\n",
    "with open(\"misc-text/camp-cp2-fdr.tex\", \"w\") as of:\n",
    "    of.write(\"{}\\endinput\".format(cp2info))\n",
    "    \n",
    "print(\"-\" * 79)\n",
    "print(\"RESULTS that we're gonna output to a TeX file:\")\n",
    "print(\"-\" * 79)\n",
    "print(cp2info)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load LoFreq calls, and output info about FDR estimation for LoFreq to `misc-text/`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load LoFreq VCF info and compute various things\n",
    "\n",
    "**NOTE: this still only includes \"rare\" (i.e. not high-frequency, as defined above) mutations.** LoFreq can call either type of mutations, but we explicitly filter out high-frequency mutations here to make its calls comparable with the naive ones above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load LoFreq info\n",
    "lofreq_calls = VariantFile(\"../seqs/lofreq.vcf\")\n",
    "\n",
    "# Record raw numbers of called mutations from LoFreq\n",
    "camp_lofreq_mutations = set()\n",
    "camp_cp2_lofreq_mutations = set()\n",
    "bact1_lofreq_mutations = set()\n",
    "bact2_lofreq_mutations = set()\n",
    "\n",
    "# (and these things, which we'll use to plot an FDR curve)\n",
    "# NOTE: this code is gross and duplicated because it's 7pm on a friday, should make better\n",
    "bact1_lofreq_p2num_called_mutations_i = {p: 0 for p in percentages}\n",
    "bact1_lofreq_p2numpermb = {}\n",
    "camp_lofreq_p2num_called_mutations_i = {p: 0 for p in percentages}\n",
    "camp_lofreq_p2numpermb = {}\n",
    "camp_cp2_lofreq_p2num_called_mutations_i = {p: 0 for p in percentages}\n",
    "camp_cp2_lofreq_p2numpermb = {}\n",
    "\n",
    "total_bact1_called_variants = 0\n",
    "\n",
    "for c in lofreq_calls.fetch():\n",
    "    \n",
    "    if pileup.is_position_rare(seq2pos2pileup[c.contig][c.pos]):\n",
    "        if c.contig == \"edge_6104\":\n",
    "            camp_lofreq_mutations.add(c.pos)\n",
    "            if c.pos in camp_cp2_pos:\n",
    "                camp_cp2_lofreq_mutations.add(c.pos)\n",
    "\n",
    "        elif c.contig == \"edge_1671\":\n",
    "            bact1_lofreq_mutations.add(c.pos)\n",
    "            total_bact1_called_variants += 1\n",
    "\n",
    "        elif c.contig == \"edge_2358\":\n",
    "            bact2_lofreq_mutations.add(c.pos)\n",
    "\n",
    "print(f\"According to LoFreq, there are {len(camp_lofreq_mutations):,} rare mutations in CAMP.\")\n",
    "print(f\"... and {len(camp_cp2_lofreq_mutations):,} rare mutations in CAMP CP 2.\")\n",
    "print(f\"... and {len(bact1_lofreq_mutations):,} rare mutations in BACT1.\")\n",
    "print(f\"... and {len(bact2_lofreq_mutations):,} rare mutations in BACT2.\")\n",
    "\n",
    "if total_bact1_called_variants != len(bact1_lofreq_mutations):\n",
    "    print(\"-\" * 79)\n",
    "    print(\n",
    "        f\"NOTE: There were actually {total_bact1_called_variants:,} variants total in BACT1 -- due to some\\n\"\n",
    "        \"variants being called at the same position. The code accounts for this, don't worry.\"\n",
    "    )\n",
    "    print(\"-\" * 79)\n",
    "\n",
    "print(\"Being slow and assigning p2num_called_mutations_i for BACT1 and CAMP based on LoFreq...\")\n",
    "# NOTE: iterating lazily over all percentages is a very terrible, slow way to do this.\n",
    "# Possible to make it much faster if we sort the percentages in advance, probs...\n",
    "# if this is a bottleneck for you please go send marcus an angry email, sorry 2021 is a rough year\n",
    "for p in percentages:\n",
    "    # consider each UNIQUE position with a called variant -- positions with > 1 called variant will\n",
    "    # only get one pass in this loop. this is intentional.\n",
    "    #\n",
    "    # so basically, if you're me in a year and you're like \"what is this code even doing\", the idea is\n",
    "    # that a given point in the FDR curve only includes variants with mutation frequency >= that point,\n",
    "    # since we're adjusting \"p\" to create this FDR curve. This is essentially* the same as how the naive\n",
    "    # mutation calling works, so we can reuse that on the LoFreq calls, if this isn't too convoluted.\n",
    "    # (we could also use the reported AF [allele frequency] from LoFreq instead of the pileup thing, but\n",
    "    # i figure we may as well be consistent since we're collapsing things to single positions anyway.)\n",
    "    #\n",
    "    # *we use min_alt_pos = 0 here because that's only used when filtering naive calling, and these\n",
    "    # calls are from LoFreq -- so we're just varying them based on alt(pos) while ignoring the other stuff.\n",
    "    for pos in bact1_lofreq_mutations:\n",
    "        if pileup.naively_call_mutation(seq2pos2pileup[\"edge_1671\"][pos], p, min_alt_pos=0):\n",
    "            bact1_lofreq_p2num_called_mutations_i[p] += 1\n",
    "            \n",
    "    # same as above\n",
    "    for pos in camp_lofreq_mutations:\n",
    "        if pileup.naively_call_mutation(seq2pos2pileup[\"edge_6104\"][pos], p, min_alt_pos=0):\n",
    "            camp_lofreq_p2num_called_mutations_i[p] += 1\n",
    "            \n",
    "    for pos in camp_cp2_lofreq_mutations:\n",
    "        if pileup.naively_call_mutation(seq2pos2pileup[\"edge_6104\"][pos], p, min_alt_pos=0):\n",
    "            camp_cp2_lofreq_p2num_called_mutations_i[p] += 1\n",
    "        \n",
    "print(\"Phew, done with that.\")\n",
    "\n",
    "# Compute numbers of mutations per megabase\n",
    "camp_lofreq_numpermb = compute_num_mutations_per_mb(len(camp_lofreq_mutations), seq2len[\"edge_6104\"])\n",
    "camp_cp2_lofreq_numpermb = compute_num_mutations_per_mb(len(camp_cp2_lofreq_mutations), num_camp_cp2_pos)\n",
    "bact1_lofreq_numpermb = compute_num_mutations_per_mb(len(bact1_lofreq_mutations), seq2len[\"edge_1671\"])\n",
    "bact2_lofreq_numpermb = compute_num_mutations_per_mb(len(bact2_lofreq_mutations), seq2len[\"edge_2358\"])\n",
    "\n",
    "# Create p2num_called_mutations for BACT1 using LoFreq.\n",
    "for p in percentages:\n",
    "    bact1_num_called_mutations = bact1_lofreq_p2num_called_mutations_i[p]\n",
    "    bact1_lofreq_p2numpermb[str(p)] = compute_num_mutations_per_mb(bact1_num_called_mutations, seq2len[\"edge_1671\"])\n",
    "    \n",
    "    camp_num_called_mutations = camp_lofreq_p2num_called_mutations_i[p]\n",
    "    camp_lofreq_p2numpermb[str(p)] = compute_num_mutations_per_mb(camp_num_called_mutations, seq2len[\"edge_6104\"])\n",
    "    \n",
    "    camp_cp2_num_called_mutations = camp_cp2_lofreq_p2num_called_mutations_i[p]\n",
    "    camp_cp2_lofreq_p2numpermb[str(p)] = compute_num_mutations_per_mb(camp_cp2_num_called_mutations, num_camp_cp2_pos)\n",
    "\n",
    "# Compute FDRs!\n",
    "bact1_fdr_lofreq_camp_decoy = camp_lofreq_numpermb / bact1_lofreq_numpermb\n",
    "bact1_fdr_lofreq_camp_cp2_decoy = camp_cp2_lofreq_numpermb / bact1_lofreq_numpermb\n",
    "print(\n",
    "    \"Just using the LoFreq calls, the FDR for BACT1 using CAMP as a decoy is \"\n",
    "    f\"{bact1_fdr_lofreq_camp_decoy:.4f}.\"\n",
    ")\n",
    "print(\n",
    "    \"Just using the LoFreq calls, the FDR for BACT1 using CAMP CP 2 as a decoy is \"\n",
    "    f\"{bact1_fdr_lofreq_camp_cp2_decoy:.4f}.\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compute naïve mutation calling info for $p = 2\\%$ mutations\n",
    "\n",
    "Because it turns out these are pretty similar to what LoFreq calls."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "nc2 = camp_naive_p2called_mutations[\"2.0\"]\n",
    "nb12 = bact1_naive_p2called_mutations[\"2.0\"]\n",
    "nb22 = bact2_naive_p2called_mutations[\"2.0\"]\n",
    "\n",
    "fdr_num_n2 = camp_naive_p2numpermb[\"2.0\"]\n",
    "fdr_den_n2 = bact1_naive_p2numpermb[\"2.0\"]\n",
    "fdr_n2 = fdr_num_n2 / fdr_den_n2\n",
    "print(f\"Estimated FDR at p = 2% is {fdr_num_n2:,.2f} / {fdr_den_n2:,.2f} = {fdr_n2:,.4f}\")\n",
    "\n",
    "seq2naive2ct = {}\n",
    "seq2overlapct = {}\n",
    "for seq, naive_2pct_muts, lf_muts in (\n",
    "    (\"edge_6104\", nc2, camp_lofreq_mutations),\n",
    "    (\"edge_1671\", nb12, bact1_lofreq_mutations),\n",
    "    (\"edge_2358\", nb22, bact2_lofreq_mutations)\n",
    "):\n",
    "    # Print out info to prove to myself that yes, p=2% and LoFreq are pretty similar!\n",
    "    nl = len(naive_2pct_muts)\n",
    "    ll = len(lf_muts)\n",
    "    print(f\"{seq2name[seq]} has {nl:,} p = 2% p-mutations and {ll:,} LoFreq-called mutations.\")\n",
    "    naive_lf_overlap = set(naive_2pct_muts) & set(lf_muts)\n",
    "    ol = len(naive_lf_overlap)\n",
    "    print(f\"\\tNumber of overlapping calls btwn (naive calling at p=2%) and (LoFreq): {ol:,}\")\n",
    "    print(f\"\\tPercentage of (Overlap / LoFreq): {100 * (ol / ll):.2f}%\")\n",
    "    \n",
    "    # ... Save this info to dicts to make it easy to output in the misc text notebook below.\n",
    "    # I guess we could do that here instead but, uh, whatever, no reason to make this more complicated\n",
    "    # than it already is. I don't think we have enough coffee for that.\n",
    "    seq2naive2ct[seq] = nl\n",
    "    seq2overlapct[seq] = ol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output info -- analogous to the naive info output above.\n",
    "# ... BUT WITH MORE STUFF since we're comparing LoFreq to the naive method at p=2%\n",
    "lf_camp_num = len(camp_lofreq_mutations)\n",
    "lf_bact1_num = len(bact1_lofreq_mutations)\n",
    "lf_bact2_num = len(bact2_lofreq_mutations)\n",
    "\n",
    "# Just because we use this a lot...!\n",
    "seqnames = f\"{seq2name['edge_6104']}, {seq2name['edge_1671']}, and {seq2name['edge_2358']}\"\n",
    "\n",
    "lfinfo = (\n",
    "    f\"LoFreq called {lf_camp_num:,}, {lf_bact1_num:,}, and {lf_bact2_num:,} rare mutations \"\n",
    "    f\"in the {seqnames} genomes, \"\n",
    "    f\"respectively. At the frequency threshold $p = 2\\%$, the \"\n",
    "    r'na\\\"ive method called a similar number of rare $p$-mutations '\n",
    "    f\"({seq2naive2ct['edge_6104']:,}, {seq2naive2ct['edge_1671']:,}, and {seq2naive2ct['edge_2358']:,} for \"\n",
    "    f\"{seqnames}, respectively). It turns out that the sets of rare mutations identified by LoFreq and by \"\n",
    "    r'the na\\\"ive method at $p = 2\\%$ are similar: the numbers of overlapping rare mutations between these '\n",
    "    f\"groups are {seq2overlapct['edge_6104']:,}, {seq2overlapct['edge_1671']:,}, and \"\n",
    "    f\"{seq2overlapct['edge_2358']:,} for {seqnames}. This suggests that, at least for this dataset, \"\n",
    "    f\"LoFreq primarily detected rare mutations with frequency of at least 2\\%. Here, we describe an analysis of \"\n",
    "    f\"FDRs which suggests that there exist many more lower-frequency rare mutations.\\n\\n\"\n",
    ")\n",
    "lfinfo += (\n",
    "    f\"Using LoFreq's calls, the numbers of rare mutations per megabase for each genome are \"\n",
    "    f\"{round(camp_lofreq_numpermb):,}, {round(bact1_lofreq_numpermb):,}, and {round(bact2_lofreq_numpermb):,} \"\n",
    "    f\"for {seq2name['edge_6104']}, {seq2name['edge_1671']}, and {seq2name['edge_2358']}, respectively. \"\n",
    "    f\"We can evaluate the FDR of LoFreq's calls in the same way as earlier: using the \"\n",
    "    f\"{seq2name['edge_1671']} genome as a target and the {seq2name['edge_6104']} genome as a decoy, we \"\n",
    "    f\"estimate the FDR for {seq2name['edge_1671']} as \"\n",
    "    \"$\\\\frac{\"\n",
    "    f\"{camp_lofreq_numpermb:,.2f}\"\n",
    "    \"}{\"\n",
    "    f\"{bact1_lofreq_numpermb:,.2f}\"\n",
    "    \"}\"\n",
    "    f\" \\\\approx {bact1_fdr_lofreq_camp_decoy:,.4f}$. \"\n",
    ")\n",
    "lfinfo += (\n",
    "    r'This is similar to the FDR of the na\\\"ive approach at the frequency threshold of $p = 2\\%$, which '\n",
    "    f\"had an estimated FDR (using {seq2name['edge_1671']} as a target and {seq2name['edge_6104']} as a decoy) of \"\n",
    "    \"$\\\\frac{\"\n",
    "    f\"{fdr_num_n2:,.2f}\"\n",
    "    \"}{\"\n",
    "    f\"{fdr_den_n2:,.2f}\"\n",
    "    \"}\"\n",
    "    f\" \\\\approx {fdr_n2:,.4f}$. \"\n",
    "    r'Although both LoFreq and the na\\\"ive approach at $p = 2\\%$ result in the reliable identification of '\n",
    "    \"rare mutations with low FDR, we are still interested in extending the set of identified rare mutations while \"\n",
    "    r'controlling the FDR. For example, lowering the frequency threshold of the na\\\"ive approach to '\n",
    "    f\"$p = 0.5\\%$ results in the identification of {len(bact1_naive_p2called_mutations['0.5']):,} rare mutations in the \"\n",
    "    f\"{seq2name['edge_1671']} genome (an additional \"\n",
    "    f\"{len(bact1_naive_p2called_mutations['0.5']) - seq2naive2ct['edge_1671']:,} rare mutations as compared to $p = 2\\%$) \"\n",
    "    \"with a higher but still relatively low FDR (still using CAMP as a decoy) of \"\n",
    "    f\"{camp_naive_p2numpermb['0.5'] / bact1_naive_p2numpermb['0.5']:,.4f}.\"\n",
    ")\n",
    "\n",
    "with open(\"misc-text/lofreq-target-decoy.tex\", \"w\") as of:\n",
    "    # see https://tex.stackexchange.com/a/18018\n",
    "    of.write(\"{}\\endinput\".format(lfinfo))\n",
    "\n",
    "print(\"-\" * 79)\n",
    "print(\"RESULTS that we're gonna output to a TeX file:\")\n",
    "print(\"-\" * 79)\n",
    "print(lfinfo)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot estimated BACT1 FDR vs. scaled number of identified (rare) mutations\n",
    "\n",
    "Previous versions of this notebook only drew one FDR curve at a time; now, this function accepts multiple decoy genome `p2numpermb` objects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DECOY_CAMP_COLOR = \"#008800\"\n",
    "DECOY_CAMP_CP2_COLOR = cp2color[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_bact1_fdr(\n",
    "    target_p2numpermb, decoy_p2numpermbs, colors, shapes, decoy_labels, fig_basename,\n",
    "    use_log=True, start_p=None, end_p=None, show_p_labels=False,\n",
    "    special_p_markers=[2, 0.22, 0.15],\n",
    "    titleprefix=f\"{seq2name['edge_1671']} FDR curves based on na\\u00efve rare $p$-mutation calling\"\n",
    "):\n",
    "    \"\"\"Plots FDR curves with some fancy annotations.\n",
    "    \n",
    "    target_p2numpermb is analogous to bact1_naive_p2numpermb. This maps values of p (some percentage\n",
    "    that you're using as a cutoff for calling mutations) to the number of called mutations per megabase\n",
    "    for some genome. (This is currently assumed to be BACT1, because of e.g. the title we use for this\n",
    "    figure, but really you could draw these sorta curves for any genome that you have variant calls for.)\n",
    "    \n",
    "    decoy_p2numpermbs, colors, shapes, and decoy_labels should all be collections of identical length\n",
    "    (this lets you pass in and style multiple decoy genomes to be shown on the same plot). So, for example,\n",
    "    you can show how different decoy genomes result in different FDR curves for the same target genome.\n",
    "    \"\"\"\n",
    "    \n",
    "    fig, ax = pyplot.subplots(1)\n",
    "    \n",
    "    for di, decoy_p2numpermb in enumerate(decoy_p2numpermbs):\n",
    "        if start_p is None:\n",
    "            start_p = percentages[0]\n",
    "        \n",
    "        if end_p is None:\n",
    "            end_p = percentages[-1]\n",
    "        \n",
    "        # This is all the percentages we HOPE to use\n",
    "        attempting_to_use_percentages = percentages[percentages.index(start_p) : percentages.index(end_p) + 1]\n",
    "        # This is all the percentages we CAN use (the FDR is d/t, so if t is 0 then we can't show that FDR...)\n",
    "        used_percentages = []\n",
    "        \n",
    "        p2bact1fdr = {}\n",
    "        for p in attempting_to_use_percentages:\n",
    "            d = decoy_p2numpermb[str(p)]\n",
    "            t = target_p2numpermb[str(p)]\n",
    "            if t != 0:\n",
    "                p2bact1fdr[p] = d / t\n",
    "                used_percentages.append(p)\n",
    "            \n",
    "        # FDR\n",
    "        x = []\n",
    "        # number of mutations per megabase\n",
    "        y = []\n",
    "        # list of 2-tuples of (x,y). we'll highlight these points.\n",
    "        special_xys = []\n",
    "        for p in used_percentages:\n",
    "            cx = p2bact1fdr[p]\n",
    "            cy = target_p2numpermb[str(p)]\n",
    "            x.append(cx)\n",
    "            y.append(cy)\n",
    "            if show_p_labels:\n",
    "                # add labels (manually positioned). yeah, i know i know\n",
    "                # we only show these labels once per plot (to avoid the text overlapping itself\n",
    "                # from doing this multiple times)\n",
    "                dy = None\n",
    "                dx = None\n",
    "                if use_log:\n",
    "                    if p == 2: dy = -500; dx = 0.005\n",
    "                    elif p == 0.22: dy = -4000; dx = -0.008\n",
    "                    elif p == 0.15: dy = -6000; dx = -0.008\n",
    "                else:\n",
    "                    if p == 2: dy = 0; dx = 0.005\n",
    "                    elif p == 0.22: dy = -200; dx = 0.005\n",
    "                    elif p == 0.15: dy = 300; dx = -0.008\n",
    "                    \n",
    "                if dy is not None:\n",
    "                    if p >= 1:\n",
    "                        text = f\"$p = {p:.0f}\\%$\"\n",
    "                    elif p >= 0.5:\n",
    "                        text = f\"$p = {p:.1f}\\%$\"\n",
    "                    else:\n",
    "                        text = f\"$p = {p:.2f}\\%$\"\n",
    "                    if di == 0:\n",
    "                        ax.text(cx + dx, cy + dy, text)\n",
    "                        \n",
    "            if p in special_p_markers:\n",
    "                special_xys.append((cx, cy))\n",
    "\n",
    "        ax.plot(x, y, marker=shapes[di], color=colors[di], label=decoy_labels[di])\n",
    "        if len(special_xys) > 0:\n",
    "            ax.scatter([xy[0] for xy in special_xys], [xy[1] for xy in special_xys], color=\"#ffff00\", zorder=2000, s=20)\n",
    "        \n",
    "    ax.set_xlabel(f\"Estimated FDR for called rare $p$-mutations in {seq2name['edge_1671']}\")\n",
    "    ax.set_ylabel(f\"Number of called rare $p$-mutations per megabase in {seq2name['edge_1671']}\")\n",
    "    title = (\n",
    "        f\"{titleprefix},\\nusing {len(used_percentages):,} values of $p$ from {max(used_percentages):.2f}% to {min(used_percentages):.2f}%\"\n",
    "    )\n",
    "    if use_log:\n",
    "        ax.set_yscale(\"symlog\")\n",
    "        title += \" (log scale)\"\n",
    "    else:\n",
    "        title += \" (non-log scale)\"\n",
    "    ax.set_title(title, fontsize=20)\n",
    "    ax.legend()\n",
    "    use_thousands_sep(ax.yaxis)\n",
    "    fig.set_size_inches(15, 8)\n",
    "    fig.savefig(f\"figs/{fig_basename}.png\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_bact1_fdr(\n",
    "    bact1_naive_p2numpermb,\n",
    "    (camp_naive_p2numpermb, camp_cp2_naive_p2numpermb),\n",
    "    (DECOY_CAMP_COLOR, DECOY_CAMP_CP2_COLOR),\n",
    "    (\"s\", \"o\"),\n",
    "    (\n",
    "        f\"Decoy Genome: all {seq2len['edge_6104']:,} positions in {seq2name['edge_6104']}\",\n",
    "        f\"Decoy Genome: only the {num_camp_cp2_pos:,} CP 2 positions in {seq2name['edge_6104']}\",\n",
    "    ),\n",
    "    \"BACT1_FDR_CAMP_decoy\",\n",
    "    show_p_labels=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot_bact1_fdr(\n",
    "#     (camp_naive_p2numpermb, camp_cp2_naive_p2numpermb),\n",
    "#     (\"#005500\", cp2color[2]),\n",
    "#     (\"s\", \"o\"),\n",
    "#     (\n",
    "#         f\"Decoy Genome: all {seq2len['edge_6104']:,} positions in {seq2name['edge_6104']}\",\n",
    "#         f\"Decoy Genome: only the {num_camp_cp2_pos:,} CP 2 positions in {seq2name['edge_6104']}\",\n",
    "#     ),\n",
    "#     \"BACT1_FDR_CAMP_decoy_max2\",\n",
    "#     start_p=2\n",
    "# )\n",
    "plot_bact1_fdr(\n",
    "    bact1_naive_p2numpermb,\n",
    "    (camp_naive_p2numpermb, camp_cp2_naive_p2numpermb),\n",
    "    (DECOY_CAMP_COLOR, DECOY_CAMP_CP2_COLOR),\n",
    "    (\"s\", \"o\"),\n",
    "    (\n",
    "        f\"Decoy Genome: all {seq2len['edge_6104']:,} positions in {seq2name['edge_6104']}\",\n",
    "        f\"Decoy Genome: only the {num_camp_cp2_pos:,} CP 2 positions in {seq2name['edge_6104']}\",\n",
    "    ),\n",
    "    \"BACT1_FDR_CAMP_decoy_nonlog_max2\",\n",
    "    use_log=False,\n",
    "    start_p=2,\n",
    "    show_p_labels=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sanity checking\n",
    "\n",
    "because the lofreq fdr curve looks weird. but it seems legit.\n",
    "\n",
    "causal factors for aforementioned weirdness:\n",
    "\n",
    "- although CAMP CP 2 is strictly smaller (i.e. includes less positions) than all of CAMP, we're not computing the FDR using just raw position counts -- we're using number per mb, so it's possible for CAMP CP 2's FDRs to exceed that of all of CAMP at a fixed value of p.\n",
    "\n",
    "- the \"jumps\" seem to correspond to values of p where we pick up new mutations in the decoy genome. Since there are very few of these identified by LoFreq in CAMP / CAMP CP 2, their impact on the graph is very noticeable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# print(\"According to LoFreq:\")\n",
    "# for pi in range(500, 400, -1):\n",
    "#     try:\n",
    "#         fdr = camp_lofreq_p2numpermb[str(pi / 100)] / bact1_lofreq_p2numpermb[str(pi / 100)]\n",
    "#     except ZeroDivisionError:\n",
    "#         fdr = None\n",
    "#     print(\n",
    "#         f\"At p = {pi / 100}%, CAMP has {camp_lofreq_p2num_called_mutations_i[pi / 100]} rare muts. \"\n",
    "#         f\"BACT1 has {bact1_lofreq_p2num_called_mutations_i[pi / 100]} rare muts. \"\n",
    "#         f\"FDR est is {fdr}.\"\n",
    "#     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_bact1_fdr(\n",
    "    bact1_lofreq_p2numpermb,\n",
    "    (camp_lofreq_p2numpermb, camp_cp2_lofreq_p2numpermb),\n",
    "    (DECOY_CAMP_COLOR, DECOY_CAMP_CP2_COLOR),\n",
    "    (\"s\", \"o\"),\n",
    "    (\n",
    "        f\"Decoy Genome: all {seq2len['edge_6104']:,} positions in {seq2name['edge_6104']}\",\n",
    "        f\"Decoy Genome: only the {num_camp_cp2_pos:,} CP 2 positions in {seq2name['edge_6104']}\",\n",
    "    ),\n",
    "    \"BACT1_FDR_CAMP_decoy_lofreq\",\n",
    "    titleprefix=\"BACT1 FDR curves based on LoFreq calls of rare mutations\",\n",
    "    special_p_markers=[]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Show how adjusting $p$ adjusts the number of $p$-mutations in a genome (and the FDR, using the T/D approach)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# There is the possibility that we can't compute the FDR for certain values of p -- if a genome has 0 mutations\n",
    "# for this value of p. In this case, we label these points with a color of None. In order to force matplotlib to\n",
    "# draw a gray circle for these points, instead of just leaving them blank, we take the default viridis colormap\n",
    "# and apply the set_bad() method to it to use this gray color. This, in conjunction with the \"plotnonfinite\"\n",
    "# argument to scatter(), lets us draw gray circles for values of p with an undefined FDR (see\n",
    "# https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.scatter.html).\n",
    "viridis_cmap = copy.copy(matplotlib.cm.get_cmap(\"viridis\"))\n",
    "viridis_cmap.set_bad(\"#888888\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.ticker import MultipleLocator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq2p2called_mutations = {\n",
    "    \"edge_6104\": camp_naive_p2called_mutations,\n",
    "    \"edge_1671\": bact1_naive_p2called_mutations,\n",
    "    \"edge_2358\": bact2_naive_p2called_mutations,\n",
    "    \"CAMP CP 2\": camp_cp2_naive_p2called_mutations\n",
    "}\n",
    "\n",
    "seq2p2numpermb = {\n",
    "    \"edge_1671\": bact1_naive_p2numpermb,\n",
    "    \"edge_2358\": bact2_naive_p2numpermb,\n",
    "}\n",
    "\n",
    "def p_vs_num_muts_plot(leftmost_p=percentages[0], rightmost_p=percentages[-1],\n",
    "                       fdr_vmin=0, fdr_vmax=1, ylogscale=False, fig_basename=None):\n",
    "    \n",
    "    fig, axes = pyplot.subplots(5, 1, gridspec_kw={\"hspace\": 0.6, \"height_ratios\": [1, 1, 1, 1, 0.2]})\n",
    "    \n",
    "    used_percentages = percentages[percentages.index(leftmost_p):percentages.index(rightmost_p) + 1]\n",
    "    \n",
    "    for si, seq in enumerate([\"edge_6104\", \"edge_1671\", \"edge_2358\", \"CAMP CP 2\"]):\n",
    "        \n",
    "        # Reuse the already-called p2called_mutations dicts to save time.\n",
    "        p2called_mutations = seq2p2called_mutations[seq]\n",
    "        y = [len(p2called_mutations[str(p)]) for p in used_percentages]\n",
    "  \n",
    "        # Two phases: 1) line plot, 2) scatter plot. This lets us color points by FDR.\n",
    "        axes[si].plot(used_percentages, y, c=\"#888888\")\n",
    "\n",
    "        if seq != \"edge_6104\" and seq != \"CAMP CP 2\":\n",
    "            fdrs = []\n",
    "            for p in used_percentages:\n",
    "                if seq2p2numpermb[seq][str(p)] > 0:\n",
    "                    f = camp_naive_p2numpermb[str(p)] / seq2p2numpermb[seq][str(p)]\n",
    "                else:\n",
    "                    # draw a gray dot for this value of p since we can't compute the FDR (due to the\n",
    "                    # denominator, i.e. the number of p-mutations per megabase in the \"target\" genome, being 0)\n",
    "                    f = None\n",
    "                fdrs.append(f)\n",
    "            # https://stackoverflow.com/a/8204981\n",
    "            sc = axes[si].scatter(used_percentages, y, c=fdrs, cmap=viridis_cmap, vmin=fdr_vmin, vmax=fdr_vmax, plotnonfinite=True)\n",
    "        else:\n",
    "            axes[si].scatter(used_percentages, y, c=\"#888888\")\n",
    "\n",
    "        if seq in SEQS:\n",
    "            axes[si].set_title(seq2name[seq])\n",
    "        else:\n",
    "            axes[si].set_title(seq)\n",
    "            \n",
    "        # important to set y scale before calling use_thousands_sep() -- otherwise, the 10^n notation\n",
    "        # will trump the thousands separator stuff\n",
    "        if ylogscale:\n",
    "            axes[si].set_yscale(\"symlog\")\n",
    "            # We use a huge upper bound here of max(y) * 5 because in order to make a visible difference for\n",
    "            # log scales we gotta use really huge numbers (because of, uh, math). Since this is a log10 scale,\n",
    "            # we can think of the 5 coefficient as getting us halfway to the next highest y-axis tick mark. Not\n",
    "            # an ideal way to handle this for sure, but it works.\n",
    "            axes[si].set_ylim(-0.5, max(y) * 5)\n",
    "        else:\n",
    "            # hack to get the y = 0 points to not be only half shown on the plot. in theory margin()\n",
    "            # should also do the trick, but that didn't seem to do anything here??? c'est la vie.\n",
    "            axes[si].set_ylim(-max(y) / 20)\n",
    "            \n",
    "        use_thousands_sep(axes[si].yaxis)\n",
    "        \n",
    "        # alter the amount of padding used based on the number of percentages we have. Use less padding if\n",
    "        # there are only a few percentage points shown.\n",
    "        padding = 1e-4 * len(used_percentages)\n",
    "        axes[si].set_xlim(leftmost_p + padding, rightmost_p - padding)\n",
    "        if leftmost_p <= 1:\n",
    "            axes[si].xaxis.set_major_locator(MultipleLocator(0.1))\n",
    "\n",
    "    fig.text(0.05, 0.33, \"Number of called rare $p$-mutations in a genome\", rotation=90, fontsize=16, color=\"#6b6b6b\")\n",
    "    axes[-2].set_xlabel(\"Value of $p$ used to call $p$-mutations (%)\", fontsize=16)\n",
    "\n",
    "    num_p_vals = len(used_percentages)\n",
    "    title = f\"$p$ vs. number of called rare $p$-mutations: {num_p_vals:,} values of p $\\in$ [{leftmost_p:.2f}%, {rightmost_p:.2f}%]\"\n",
    "    if ylogscale:\n",
    "        title += \"; log scale\"\n",
    "    fig.suptitle(title, x=0.52, y=0.95, fontsize=20)\n",
    "    fig.set_size_inches(15, 13)\n",
    "    \n",
    "    fig.colorbar(sc, cax=axes[-1], orientation=\"horizontal\")\n",
    "    \n",
    "    fdr_desc = \"\"\n",
    "    if fdr_vmin != 0:\n",
    "        fdr_desc = f\"FDRs < {fdr_vmin} are clamped to {fdr_vmin}; \"\n",
    "    fdr_desc += f\"FDRs > {fdr_vmax} are clamped to {fdr_vmax}\"\n",
    "    axes[-1].set_xlabel(f\"Color scaling applied to BACT1 and BACT2 estimated FDRs (using all of CAMP as a decoy)\\n{fdr_desc}\", fontsize=14)\n",
    "    \n",
    "    if fig_basename is not None:\n",
    "        fig.savefig(f\"figs/{fig_basename}.png\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_vs_num_muts_plot(\n",
    "    leftmost_p=4.99,\n",
    "    ylogscale=True, fig_basename=\"p_vs_num_muts_log_clamp\",\n",
    "    fdr_vmin=0, fdr_vmax=0.1,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "p_vs_num_muts_plot(\n",
    "    leftmost_p=1,\n",
    "    ylogscale=False, fig_basename=\"p_vs_num_muts_nonlog_clamp\",\n",
    "    fdr_vmin=0, fdr_vmax=0.1,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FDR curves using nonsynonymous mutation rates and nonsense mutation rates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load info about single-nucleotide codon mutations (syn vs nonsyn and non-nonsense vs nonsense) from files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_pf(basename):\n",
    "    with open(f\"misc-output/{basename}.pickle\", \"rb\") as pf:\n",
    "        return pickle.load(pf)\n",
    "\n",
    "p2seq2obs_si = load_pf(\"p2seq2obs_si\")\n",
    "p2seq2obs_ni = load_pf(\"p2seq2obs_ni\")\n",
    "p2seq2obs_nnsi = load_pf(\"p2seq2obs_nnsi\")\n",
    "p2seq2obs_nsi = load_pf(\"p2seq2obs_nsi\")\n",
    "\n",
    "seq2poss_si = load_pf(\"seq2poss_si\")\n",
    "seq2poss_ni = load_pf(\"seq2poss_ni\")\n",
    "seq2poss_nnsi = load_pf(\"seq2poss_nnsi\")\n",
    "seq2poss_nsi = load_pf(\"seq2poss_nsi\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compute $R_S$, $R_N$, $R_{NNS}$, $R_{NS}$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rates for synonymous and nonsynonymous s.n. mutations\n",
    "p2seq2synvals = {p: {} for p in percentages}\n",
    "# Rates for non-nonsense and nonsense s.n. mutations\n",
    "p2seq2nonsensevals = {p: {} for p in percentages}\n",
    "\n",
    "for p in percentages:\n",
    "    for seq in SEQS:\n",
    "        p2seq2synvals[p][seq] = {}\n",
    "        # R_S\n",
    "        p2seq2synvals[p][seq][\"syn\"] = p2seq2obs_si[p][seq] / seq2poss_si[seq]\n",
    "        # R_N\n",
    "        p2seq2synvals[p][seq][\"nonsyn\"] = p2seq2obs_ni[p][seq] / seq2poss_ni[seq]\n",
    "\n",
    "        p2seq2nonsensevals[p][seq] = {}\n",
    "        # R_NNS\n",
    "        p2seq2nonsensevals[p][seq][\"nonnonsense\"] = p2seq2obs_nnsi[p][seq] / seq2poss_nnsi[seq]\n",
    "        # R_NS\n",
    "        p2seq2nonsensevals[p][seq][\"nonsense\"] = p2seq2obs_nsi[p][seq] / seq2poss_nsi[seq]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot (estimated) FDR curves using nonsynonymous and nonsense mutation rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "fig, ax = pyplot.subplots(1)\n",
    "\n",
    "# Highlight the points at these values of p\n",
    "special_p = [2, 0.15]\n",
    "special_xys = []\n",
    "\n",
    "# x-axis: Estimated FDR (equal to R_N or R_NS divided by 3)\n",
    "fdr_nonsyn = []\n",
    "fdr_nonsense = []\n",
    "\n",
    "# y-axis: Number of called p-mutations per megabase\n",
    "num_called_pmuts_per_mb = []\n",
    "\n",
    "sn_pcts = [p / 100 for p in range(15, 500, 1)][::-1]\n",
    "\n",
    "# limit this to starting at 4.99%, since we're only including rare mutations\n",
    "# but it's weird - not entirely either way. need to decide.\n",
    "for p in sn_pcts:\n",
    "    y = bact1_naive_p2numpermb[str(p)]\n",
    "    x_n = p2seq2synvals[p][\"edge_1671\"][\"nonsyn\"] / 3\n",
    "    x_ns = p2seq2nonsensevals[p][\"edge_1671\"][\"nonsense\"] / 3\n",
    "    \n",
    "    num_called_pmuts_per_mb.append(y)\n",
    "    fdr_nonsyn.append(x_n)\n",
    "    fdr_nonsense.append(x_ns)\n",
    "    \n",
    "    if p in special_p:\n",
    "        print(f\"At p = {p}%, R_N = {x_n} and R_NS = {x_ns}. # p-muts per mb = {y}\")\n",
    "        special_xys.append((x_n, y))\n",
    "        special_xys.append((x_ns, y))\n",
    "        if p == 2: dx = 0.000008; dy = -1800\n",
    "        if p == 0.15: dx = -0.000022; dy = -5000\n",
    "        ax.text(x_n + dx, y + dy, f\"p = {p}%\")\n",
    "              \n",
    "ax.plot(fdr_nonsyn, num_called_pmuts_per_mb, marker=\"o\", color=\"#cc3322\",\n",
    "        label=f\"Decoy Genome: single-nucleotide nonsynonymous positions in {seq2name['edge_1671']}\")\n",
    "\n",
    "# We darken the color representing nonsense (NS) mutations from the NNS / NS bar plot, in order to make it\n",
    "# contrast better with the color representing nonsynonymous (N) mutations\n",
    "ax.plot(fdr_nonsense, num_called_pmuts_per_mb, marker=\"s\", color=\"#772277\",\n",
    "        label=f\"Decoy Genome: single-nucleotide nonsense positions in {seq2name['edge_1671']}\")\n",
    "        \n",
    "ax.scatter([xy[0] for xy in special_xys], [xy[1] for xy in special_xys], color=\"#ffff00\", zorder=2000, s=20)\n",
    "    \n",
    "ax.set_xlabel(f\"Estimated FDR for called rare $p$-mutations in {seq2name['edge_1671']}\" + r\", using $R_N\\ /\\ 3$ and $R_{NS}\\ /\\ 3$\")\n",
    "ax.set_ylabel(\"Number of called rare $p$-mutations per megabase\")\n",
    "title = (\n",
    "    f'BACT1 FDR curves based on na\\u00efve $p$-mutation calling,\\nusing {len(sn_pcts):,} values of $p$ from {max(sn_pcts):.2f}% to {min(sn_pcts):.2f}%'\n",
    ")\n",
    "ax.set_yscale(\"symlog\")\n",
    "title += \" (log scale)\"\n",
    "ax.set_title(title, fontsize=20)\n",
    "ax.legend()\n",
    "use_thousands_sep(ax.yaxis)\n",
    "fig.set_size_inches(15, 8)\n",
    "fig.savefig(f\"figs/bact1_fdr_nonsyn_nonsense.png\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
