{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene analysis utility functions\n",
    "\n",
    "This relies on the `Header` and `LoadMutationJSONData` notebooks already having been run."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "import skbio\n",
    "import pandas as pd\n",
    "from statistics import mean\n",
    "from parse_sco import parse_sco\n",
    "\n",
    "def histogram_of_123(vals, title, axes, xlabel=\"Codon Position\", ylabel=\"Number of Mutated Positions\",\n",
    "                     format_yaxis=\"thousands\", max_val=None, show_ylabel=True, show_xlabel=True):\n",
    "    if len(vals) == 3:\n",
    "        one, two, three = vals\n",
    "        x = [1, 2, 3]\n",
    "        labels = [\"1\", \"2\", \"3\"]\n",
    "    elif len(vals) == 4:\n",
    "        one, two, three, four = vals\n",
    "        x = [1, 2, 3, 4]\n",
    "        labels = [\"1\", \"2\", \"3\", \"Non-Coding\"]\n",
    "    else:\n",
    "        raise ValueError(\"Not 3 or 4 vals passed?\")\n",
    "    axes.bar(\n",
    "        x=x,\n",
    "        height=vals,\n",
    "        color=[cp2color[i + 1] for i in range(len(vals))],\n",
    "        edgecolor=BORDERCOLOR,\n",
    "        tick_label=labels\n",
    "    )\n",
    "    \n",
    "    if max_val is None:\n",
    "        ypad = 0.04 * max(vals)\n",
    "    else:\n",
    "        ypad = 0.04 * max_val\n",
    "    # to make interpreting the plots easier, show text for each bar\n",
    "    # just above it: https://stackoverflow.com/a/30229062\n",
    "    # and https://stackoverflow.com/questions/30228069/how-to-display-the-value-of-the-bar-on-each-bar-with-pyplot-barh/30229062#comment86813015_30229062\n",
    "    for xi, yval in enumerate(vals, 1):\n",
    "        # Used to make decision about normalized vs unnormalized values\n",
    "        if format_yaxis == \"percentages\":\n",
    "            text = f\"{(yval * 100):.2f}%\"\n",
    "        else:\n",
    "            text = f\"{yval:,}\"\n",
    "        axes.text(xi, yval + ypad, text, ha=\"center\")\n",
    "        \n",
    "    print(title, vals)\n",
    "    if three > one and one > two:\n",
    "        axes.set_title(title)\n",
    "    else:\n",
    "        axes.set_title(title, color=\"#cc3322\", fontweight=\"semibold\")\n",
    "            \n",
    "    if show_xlabel:\n",
    "        axes.set_xlabel(xlabel)\n",
    "    if show_ylabel:\n",
    "        axes.set_ylabel(ylabel)\n",
    "        \n",
    "    if format_yaxis == \"thousands\":\n",
    "        use_thousands_sep(axes.get_yaxis())\n",
    "    elif format_yaxis == \"percentages\":\n",
    "        # Make the y-axis show percentages: based on\n",
    "        # https://old.reddit.com/r/learnpython/comments/7adhnk/matplotlib_setting_y_axis_labels_to_percent_yaxis/dp93fwq/\n",
    "        # (This is duplicated within this notebook... TODO, make this a function like use_thousands_sep()?)\n",
    "        axes.get_yaxis().set_major_formatter(matplotlib.ticker.PercentFormatter(xmax=1, decimals=2))\n",
    "    else:\n",
    "        raise ValueError(\"Unrecognized format_yaxis value: {}\".format(format_yaxis))\n",
    "    if max_val is not None:\n",
    "        # Shoddy way to add padding -- see https://stackoverflow.com/a/42804403\n",
    "        axes.set_ylim(0, max_val * 1.3)\n",
    "    else:\n",
    "        axes.set_ylim(0, max(vals) * 1.3)\n",
    "    \n",
    "def get_pos_interval_from_gene(gene):\n",
    "    try:\n",
    "        start = int(gene.LeftEnd)\n",
    "    except ValueError:\n",
    "        # Account for the \"<2\" left end coordinate in one of the genes predicted\n",
    "        start = int(gene.LeftEnd[1:])\n",
    "        \n",
    "    try:\n",
    "        end = int(gene.RightEnd)\n",
    "    except ValueError:\n",
    "        # Account for the \">N\" right end coordinate in one of the genes predicted\n",
    "        end = int(gene.RightEnd[1:])\n",
    "\n",
    "    if gene.Strand == \"-\":\n",
    "        return range(end, start - 1, -1)\n",
    "    elif gene.Strand == \"+\":\n",
    "        return range(start, end + 1)\n",
    "    else:\n",
    "        raise ValueError(\"Unknown strand: {}\".format(gene.Strand))\n",
    "\n",
    "def get_val(seq, pos, p, fasta):\n",
    "    \"\"\"Calls simple p-mutations given a seq, position, value of p, and seq's DNA sequence (\"fasta\").\n",
    "    \n",
    "    Returns 1 if pos is a p-mutation, returns 0 otherwise.\n",
    "    \n",
    "    p should be in the range [0, 1] (i.e. if it's stored in the range [0, 100], you should divide it by\n",
    "    100 first).\n",
    "    \n",
    "    pos should be 1-indexed, to be compatible with the JSONs.\n",
    "    \n",
    "    Note that the DNA sequence str, fasta, uses 0-indexing; that's fine, we'll account for that here.\n",
    "    \n",
    "    Note that this does not (currently) take the idea of minimum sufficient coverage into account. So you\n",
    "    can use p = 0 or whatever without worrying about that causing math problems (although the ideas behind\n",
    "    minimum sufficent coverage still apply, so when this function is used you should be careful to mention\n",
    "    why we are not taking suff coverage into account here). If pos has a coverage of 0 (i.e. no aligned\n",
    "    matching or mismatching reads), this'll always return 0 (i.e. \"not a p-mutation\").\n",
    "    \"\"\" \n",
    "    mismatchct = seq2pos2mismatchct[seq][str(pos)]\n",
    "    matchct = seq2pos2matchct[seq][str(pos)]\n",
    "    \n",
    "    # Note that, as mentioned above, this isn't \"complete\" coverage -- it can be zero\n",
    "    # if e.g. all the reads covering a position are deletions\n",
    "    cov = mismatchct + matchct\n",
    "    if cov > 0:\n",
    "        # skbio sequences use 0-indexed coords, so subtract 1 from pos\n",
    "        ref = fasta[pos - 1]\n",
    "        nonmatches = set(\"ACGT\") - set(ref)\n",
    "        # Instead of aggregating all mismatching nucleotides, only consider the individual mismatching\n",
    "        # nucleotides\n",
    "        alt_freqs = seq2pos2mismatches[seq][str(pos)]\n",
    "        for alt in alt_freqs:\n",
    "            if (alt_freqs[alt] / cov) > p:\n",
    "                return 1\n",
    "        # If we've made it here, none of the alt nucleotide freqs were a p-mutation.\n",
    "        return 0\n",
    "    else:\n",
    "        # 0x coverage (at least, considering matching/mismatching reads).\n",
    "        # We thus do not have the data to detect if this position is a p-mutation.\n",
    "        return 0\n",
    "        \n",
    "def histogram_maker(\n",
    "    p,\n",
    "    title,\n",
    "    axes,\n",
    "    ylabel=\"Number of Mutated Positions\",\n",
    "    normalize=False,\n",
    "    add_noncoding_col=True,\n",
    "    make_yaxes_comparable=False,\n",
    "    output_dict=False,\n",
    "    show_xlabel=True,\n",
    "    show_ylabel=True,\n",
    "):\n",
    "    \"\"\"\n",
    "    Produces histograms of pileup data for the 1st, 2nd, and 3rd positions of predicted genes.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    p: float\n",
    "        In the range [0, 1].\n",
    "        \n",
    "    title: str\n",
    "        Will be included after the seq name in every histogram.\n",
    "        \n",
    "    ylabel: str\n",
    "        Label for the y-axis of the histogram. Only will be used if normalize is False.\n",
    "        \n",
    "    normalize: bool\n",
    "        If True, divides each 1/2/3 value by (number of positions considered). This makes it easier to compare\n",
    "        histograms between different sequences.\n",
    "        This also sets ylabel to \"Number of Mutated Positions / Number of Positions\", ignoring whatever\n",
    "        ylabel's default was (or even ignoring the already-specified ylabel).\n",
    "        \n",
    "    add_noncoding_col: bool\n",
    "        If True, adds a 4th column to each histogram representing all of the positions not contained within\n",
    "        predicted protein-coding genes. If normalize is True, this column's value is divided by the total number\n",
    "        of these positions, so it's also a percentage.\n",
    "        \n",
    "    make_yaxes_comparable: bool\n",
    "        If True, sets the ylim max to just over the max y-value across all seqs' plots (so the y-range is the\n",
    "        same for each row). Could be useful, could be bad if the values wildly differ btwn seqs (in which case\n",
    "        the seqs with lower values could be hard to read)\n",
    "    \"\"\"\n",
    "    if normalize:\n",
    "        ylabel = r\"$\\dfrac{\\mathrm{Number\\ of\\ Mutated\\ Positions}}{\\mathrm{Number\\ of\\ Positions}}$\"\n",
    "    seq2vals = {}\n",
    "    max_val = 0\n",
    "    for seq in SEQS:\n",
    "        df = parse_sco(f\"../seqs/genes/{seq}.sco\")\n",
    "        fasta = str(skbio.DNA.read(f\"../seqs/{seq}.fasta\"))\n",
    "\n",
    "        # Mutation rates at modulo positions 1, 2, 3 in a gene within the genome (goes 1, 2, 3, 1, 2, 3, ...)\n",
    "        m1 = []\n",
    "        m2 = []\n",
    "        m3 = []\n",
    "        total_gene_length = 0\n",
    "        bases_in_genes = set()\n",
    "        for gene in df.itertuples():\n",
    "            i = 1\n",
    "            gm1 = []\n",
    "            gm2 = []\n",
    "            gm3 = []\n",
    "            pos_interval = get_pos_interval_from_gene(gene)\n",
    "            total_gene_length += len(pos_interval)\n",
    "            bases_in_genes |= set(pos_interval)\n",
    "            for pos in pos_interval:\n",
    "                val = get_val(seq, pos, p, fasta)\n",
    "                if i == 1:\n",
    "                    gm1.append(val)\n",
    "                    i = 2\n",
    "                elif i == 2:\n",
    "                    gm2.append(val)\n",
    "                    i = 3\n",
    "                elif i == 3:\n",
    "                    gm3.append(val)\n",
    "                    i = 1\n",
    "                else:\n",
    "                    raise ValueError(\"Marcus forgot how modulos work, go yell at him pls\")\n",
    "\n",
    "            if i != 1:\n",
    "                raise ValueError(\"Gene length not divisible by 3.\")\n",
    "                \n",
    "            m1 += gm1\n",
    "            m2 += gm2\n",
    "            m3 += gm3\n",
    "        \n",
    "        if not (len(m1) == len(m2) == len(m3)):\n",
    "            raise ValueError(\"Imbalance in 1/2/3 positions.\")\n",
    "        \n",
    "        mutation_vals = [m1, m2, m3]\n",
    "        if add_noncoding_col:\n",
    "            mn = []\n",
    "            # Figure out which bases are not in any genes. Computing a set of a range is a pretty inefficient\n",
    "            # way to do this, probably, but it works and is easy to reason about so i'll take it :P\n",
    "            bases_not_in_genes = set(range(1, seq2len[seq] + 1)) - bases_in_genes\n",
    "            for pos in bases_not_in_genes:\n",
    "                val = get_val(seq, pos, p, fasta)\n",
    "                mn.append(val)\n",
    "            mutation_vals.append(mn)\n",
    "\n",
    "        # Set vals to just the number of mutations in each pos in mutation_vals\n",
    "        vals = [sum(m) for m in mutation_vals]\n",
    "        \n",
    "        if normalize:\n",
    "            format_yaxis = \"percentages\"\n",
    "            # Divide, to get (# mutated positions) / (# positions)\n",
    "            # Note that we DON'T divide by just gene length / 3 (which would work for CP 1/2/3). Two reasons\n",
    "            # for this:\n",
    "            # 1. Overlapping genes can mess with this\n",
    "            # 2. For non-coding positions (if we're adding a 4th col for these), this doesn't make sense!\n",
    "            for i in range(len(vals)):\n",
    "                num_total_positions = len(mutation_vals[i])\n",
    "                vals[i] /= num_total_positions\n",
    "        else:\n",
    "            format_yaxis = \"thousands\"\n",
    "            \n",
    "        max_val = max(max_val, max(vals))\n",
    "        seq2vals[seq] = vals\n",
    "\n",
    "    # Delay creating a histogram for a given genome until we've computed values for all of the genomes --\n",
    "    # this lets us be fancy and set all histograms in a row to the same max value on the y-axis if needed\n",
    "    i = 0\n",
    "    for seq in SEQS:\n",
    "        if not make_yaxes_comparable:\n",
    "            max_val = None\n",
    "        histogram_of_123(seq2vals[seq], \"{}: {}\".format(seq2name[seq], title), axes[i],\n",
    "                         ylabel=ylabel, format_yaxis=format_yaxis, max_val=max_val,\n",
    "                         show_ylabel=(i == 0) and show_ylabel, show_xlabel=(i == 1) and show_xlabel)\n",
    "        i += 1\n",
    "        \n",
    "def histogram_matrix_maker(percentage_thresholds, normalize=False, make_yaxes_comparable=False, figfilename=None):\n",
    "    fig, axes = pyplot.subplots(\n",
    "        len(percentage_thresholds),\n",
    "        3,\n",
    "        gridspec_kw={\"hspace\": 0.5, \"wspace\": 0.35}\n",
    "    )\n",
    "    if len(percentage_thresholds) % 2 == 0:\n",
    "        middle_row = len(percentage_thresholds) / 2\n",
    "    else:\n",
    "        # this definitely works when you have 7 items (produces int(7/2) = int(3.5) = 3, which is the\n",
    "        # correct \"middle\" number), but it's 5am and idk if it will hold for every odd number ever\n",
    "        middle_row = int(len(percentage_thresholds) / 2)\n",
    "    row = 0\n",
    "    p2pct = get_p2pct(percentage_thresholds)\n",
    "    for p in percentage_thresholds:\n",
    "        \n",
    "        axes_in_this_row = axes[row, :]\n",
    "        if row == len(percentage_thresholds) - 1:\n",
    "            show_xlabel = True\n",
    "        else:\n",
    "            show_xlabel = False\n",
    "            \n",
    "        show_ylabel = (row == middle_row)\n",
    "        \n",
    "        histogram_maker(\n",
    "            p2pct[p],\n",
    "            \"$p$ = \" + str(p) + \"%\",\n",
    "            axes=axes_in_this_row,\n",
    "            normalize=normalize,\n",
    "            make_yaxes_comparable=make_yaxes_comparable,\n",
    "            show_xlabel=show_xlabel,\n",
    "            show_ylabel=show_ylabel,\n",
    "        )\n",
    "        row += 1\n",
    "    \n",
    "    title = \"$p$-mutation frequencies across coding and non-coding positions\"\n",
    "    titley = 0.93\n",
    "    if normalize:\n",
    "        title += \"\\n(each bar normalized by total number of positions in that category)\"\n",
    "        # a bit extra y space to account for extra line\n",
    "        titley = 0.94\n",
    "    fig.suptitle(title, y=titley, x=0.5, fontsize=20)\n",
    "    fig.set_size_inches(15, 20)\n",
    "    if figfilename is not None:\n",
    "        fig.savefig(\"figs/{}\".format(figfilename), bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
