{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "from matplotlib import pyplot\n",
    "from statistics import median, mean\n",
    "\n",
    "# Make the plots used here look kinda like ggplots\n",
    "pyplot.style.use(\"ggplot\")\n",
    "\n",
    "# Use a consistent color scheme for relating codon positions to colors.\n",
    "#\n",
    "# Uses a hopefully colorblind-friendly grayscale scheme, where CP 1 = black, CP 2 = gray, and CP 3 = white\n",
    "# (This should be accompanied by black borders around each color so that the gray/white are easily visible on\n",
    "# a light background).\n",
    "#\n",
    "# We also say CP 4 = green (which is not quite as colorblind-friendly... in the future would be good to add a\n",
    "# pattern or something). I know what you're thinking -- \"CP 4? The computer scientists have gone too\n",
    "# far this time.\" But don't worry, I know basic biology (hopefully) -- CP 4 here is shorthand that means\n",
    "# \"positions in non-coding regions of the genome\", since it can be interesting to summarize these positions\n",
    "# alongside the CPs.\n",
    "cp2color = {1: \"#000000\", 2: \"#888888\", 3: \"#ffffff\", 4: \"#29D321\"}\n",
    "BORDERCOLOR = \"#000000\"\n",
    "\n",
    "# This is the blue color used in the ggplot-style matplotlib color settings: from\n",
    "# https://gist.github.com/huyng/816622\n",
    "SCATTERPLOT_PT_COLOR = \"#348ABD\"\n",
    "\n",
    "# Maps seq names to easier-to-remember names used in the report.\n",
    "seq2name = {\"edge_6104\": \"CAMP\", \"edge_1671\": \"BACTERIA\", \"edge_2358\": \"BACTEROIDALES\"}\n",
    "SEQS = [\"edge_6104\", \"edge_1671\", \"edge_2358\"]\n",
    "\n",
    "# used in a few places, notably gene stats and gene utils\n",
    "seq2len = {\"edge_1671\": 2153394, \"edge_2358\": 2806161, \"edge_6104\": 1289244}\n",
    "\n",
    "# Whether or not this edge is in an isolated component of the graph. Used in the phasing analysis.\n",
    "seq2iscircular = {\"edge_1671\": True, \"edge_2358\": True, \"edge_6104\": False}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def use_thousands_sep(mpl_axis):\n",
    "    # Use thousands separators for positions.\n",
    "    # For an arbitrary pyplot figure, you can call this function with pyplot.gca().xaxis or pyplot.gca().yaxis. \n",
    "    # Modified from https://stackoverflow.com/a/25973637 -- this is modified to work better with integers\n",
    "    # (matplotlib seems to store all values as floats internally, even essentially integral things -- so we can\n",
    "    # use the float.is_integer() method to see if a value is \"close enough\" to an integer, and if so remove the\n",
    "    # trailing \".0\" that happens when you try to format a float of an integer -- see\n",
    "    # https://stackoverflow.com/a/21583817.)\n",
    "    mpl_axis.set_major_formatter(matplotlib.ticker.FuncFormatter(\n",
    "        lambda x, pos: \"{:,}\".format(int(x)) if x.is_integer() else \"{:,}\".format(x)\n",
    "    ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_parent_gene_info_of_position(pos, genes_df):\n",
    "    \"\"\"Returns information about the parent gene(s) of a position.\n",
    "    \n",
    "    Parameters\n",
    "    ==========\n",
    "    \n",
    "    pos: int\n",
    "        1-indexed position in a genome.\n",
    "        \n",
    "    genes_df: pd.DataFrame\n",
    "        Result from calling parse_sco().\n",
    "        \n",
    "    Returns\n",
    "    =======\n",
    "    \n",
    "    parent_info: list\n",
    "        List of gene numbers that contain this position. Will usually contain just a single element\n",
    "        (for positions located in a single gene), but may contain multiple elements if genes overlap\n",
    "        at this position. May also be an empty list if this position is located in an intergenic region.\n",
    "    \"\"\"\n",
    "    parent_genes = []\n",
    "    for gene in genes_df.itertuples():\n",
    "        if pos >= int(gene.LeftEnd) and pos <= int(gene.RightEnd):\n",
    "            parent_genes.append(int(gene.Index))\n",
    "    return parent_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def color_positions_by_gene_info(posleft, posright, genes_df,\n",
    "                                 color_multi_genes=\"#888888\", color_no_genes=\"#ffffff\"):\n",
    "    \"\"\"Returns information about the parent gene(s) of all positions.\n",
    "    \n",
    "    Parameters\n",
    "    ==========\n",
    "    \n",
    "    posleft: int\n",
    "        1-indexed position in a genome. Inclusive.\n",
    "        \n",
    "    posright: int\n",
    "        1-indexed position in a genome. Inclusive.\n",
    "        \n",
    "    genes_df: pd.DataFrame\n",
    "        Result from calling parse_sco().\n",
    "        \n",
    "    Returns\n",
    "    =======\n",
    "    \n",
    "    parent_info: list\n",
    "        List (one element per position in the range), where the i-th element corresponds\n",
    "        to the (i-1)-th position in the range's color.\n",
    "    \"\"\"\n",
    "  \n",
    "    num_genes = len(genes_df.index)\n",
    "    \n",
    "    # From https://matplotlib.org/3.1.1/tutorials/colors/colormap-manipulation.html -- adapted from\n",
    "    # phasing graph (call&plot) notebook\n",
    "    cmap = matplotlib.cm.get_cmap(\"viridis\", num_genes)\n",
    "    hexcolors = [matplotlib.colors.to_hex(cmap(c)) for c in range(num_genes)]  \n",
    "    \n",
    "    posm1_to_genes = []\n",
    "    for pos in range(posleft, posright + 1):\n",
    "        posm1_to_genes.append([])\n",
    "        \n",
    "    for gene in genes_df.itertuples():\n",
    "        gene_left = int(gene.LeftEnd)\n",
    "        gene_right = int(gene.RightEnd)\n",
    "        gene_num = int(gene.Index)\n",
    "        \n",
    "        for pos in range(gene_left, gene_right + 1):\n",
    "            posm1_to_genes[pos - 1].append(gene_num)\n",
    "            \n",
    "    posm1_to_color = []\n",
    "    for posm1, genes in enumerate(posm1_to_genes):\n",
    "        if len(genes) == 0:\n",
    "            posm1_to_color.append(color_no_genes)\n",
    "        elif len(genes) > 1:\n",
    "            posm1_to_color.append(color_multi_genes)\n",
    "        else:\n",
    "            posm1_to_color.append(hexcolors[genes[0] - 1])\n",
    "            \n",
    "    return posm1_to_color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
