{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Some general utilities used throughout these notebooks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "from matplotlib import pyplot\n",
    "from statistics import median, mean\n",
    "from collections import defaultdict\n",
    "\n",
    "# Make the plots used here look kinda like ggplots\n",
    "pyplot.style.use(\"ggplot\")\n",
    "\n",
    "# Use a consistent color scheme for relating codon positions to colors.\n",
    "#\n",
    "# Uses a hopefully colorblind-friendly grayscale scheme, where CP 1 = black, CP 2 = gray, and CP 3 = white\n",
    "# (This should be accompanied by black borders around each color so that the gray/white are easily visible on\n",
    "# a light background).\n",
    "#\n",
    "# We also say CP 4 = green (which is not quite as colorblind-friendly... in the future would be good to add a\n",
    "# pattern or something). I know what you're thinking -- \"CP 4? The computer scientists have gone too\n",
    "# far this time.\" But don't worry, I know basic biology (hopefully) -- CP 4 here is shorthand that means\n",
    "# \"positions in non-coding regions of the genome\", since it can be interesting to summarize these positions\n",
    "# alongside the CPs.\n",
    "cp2color = {1: \"#000000\", 2: \"#888888\", 3: \"#ffffff\", 4: \"#29D321\"}\n",
    "BORDERCOLOR = \"#000000\"\n",
    "\n",
    "# This is the blue color used in the ggplot-style matplotlib color settings: from\n",
    "# https://gist.github.com/huyng/816622\n",
    "SCATTERPLOT_PT_COLOR = \"#348ABD\"\n",
    "\n",
    "# Maps seq names to easier-to-remember names used in the report.\n",
    "seq2name = {\"edge_6104\": \"CAMP\", \"edge_1671\": \"BACT1\", \"edge_2358\": \"BACT2\"}\n",
    "SEQS = [\"edge_6104\", \"edge_1671\", \"edge_2358\"]\n",
    "\n",
    "# used in a few places, notably gene stats and gene utils\n",
    "seq2len = {\"edge_1671\": 2153394, \"edge_2358\": 2806161, \"edge_6104\": 1289244}\n",
    "\n",
    "# Whether or not this edge is in an isolated component of the graph. Used in the phasing analysis.\n",
    "seq2iscircular = {\"edge_1671\": True, \"edge_2358\": True, \"edge_6104\": False}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Utility functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def use_thousands_sep(mpl_axis):\n",
    "    # Use thousands separators for positions.\n",
    "    # For an arbitrary pyplot figure, you can call this function with pyplot.gca().xaxis or pyplot.gca().yaxis. \n",
    "    # Modified from https://stackoverflow.com/a/25973637 -- this is modified to work better with integers\n",
    "    # (matplotlib seems to store all values as floats internally, even essentially integral things -- so we can\n",
    "    # use the float.is_integer() method to see if a value is \"close enough\" to an integer, and if so remove the\n",
    "    # trailing \".0\" that happens when you try to format a float of an integer -- see\n",
    "    # https://stackoverflow.com/a/21583817.)\n",
    "    mpl_axis.set_major_formatter(matplotlib.ticker.FuncFormatter(\n",
    "        lambda x, pos: \"{:,}\".format(int(x)) if x.is_integer() else \"{:,}\".format(x)\n",
    "    ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def next_cp(cp):\n",
    "    # \"Trust, but verify.\" --Gregor Mendel or someone idk\n",
    "    if cp == 1:\n",
    "        return 2\n",
    "    elif cp == 2:\n",
    "        return 3\n",
    "    elif cp == 3:\n",
    "        return 1\n",
    "    else:\n",
    "        raise ValueError(\"Invalid CP\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_parent_gene_info_of_position(pos, genes_df):\n",
    "    \"\"\"Returns information about the parent gene(s) of a position.\n",
    "    \n",
    "    Note that this is pretty slow when you call it thousands of times in a row. If operating on\n",
    "    lots of positions at once, you should probably just call get_parent_gene_info_of_many_positions(),\n",
    "    defined also in this file, instead.\n",
    "    \n",
    "    Parameters\n",
    "    ==========\n",
    "    \n",
    "    pos: int\n",
    "        1-indexed position in a genome.\n",
    "        \n",
    "    genes_df: pd.DataFrame\n",
    "        Result from calling parse_sco().\n",
    "        \n",
    "    Returns\n",
    "    =======\n",
    "    \n",
    "    parent_info: list\n",
    "        List of gene numbers that contain this position. Will usually contain just a single element\n",
    "        (for positions located in a single gene), but may contain multiple elements if genes overlap\n",
    "        at this position. May also be an empty list if this position is located in an intergenic region.\n",
    "    \"\"\"\n",
    "    parent_genes = []\n",
    "    for gene in genes_df.itertuples():\n",
    "        if pos >= int(gene.LeftEnd) and pos <= int(gene.RightEnd):\n",
    "            parent_genes.append(int(gene.Index))\n",
    "    return parent_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_parent_gene_info_of_many_positions(genes_df):\n",
    "    \"\"\"Returns information about the parent gene(s) of all positions in a genome.\n",
    "    \n",
    "    This is much faster than calling get_parent_gene_info_of_position() lots of times.\n",
    "    \n",
    "    Parameters\n",
    "    ==========\n",
    "        \n",
    "    genes_df: pd.DataFrame\n",
    "        Result from calling parse_sco() for this genome's genes.\n",
    "        \n",
    "    Returns\n",
    "    =======\n",
    "    \n",
    "    pos_to_genes: defaultdict\n",
    "        Maps position numbers (also 1-indexed) to a list of the gene number(s) of the genes\n",
    "        overlapping this position. This list can have 0, 1, or more elements depending on\n",
    "        how many genes overlap a position.\n",
    "        \n",
    "        Note that positions not in genes just straight-up won't get assigned an entry in this;\n",
    "        this doesn't make a difference, since this is a defaultdict, and these positions' entries\n",
    "        would be [] either way (which is the default value of this defaultdict).\n",
    "    \"\"\"  \n",
    "    pos_to_genes = defaultdict(list)\n",
    "    \n",
    "    for gene in genes_df.itertuples():\n",
    "        gene_left = int(gene.LeftEnd)\n",
    "        gene_right = int(gene.RightEnd)\n",
    "        gene_num = int(gene.Index)\n",
    "        \n",
    "        for pos in range(gene_left, gene_right + 1):\n",
    "            pos_to_genes[pos].append(gene_num)\n",
    "            \n",
    "    return pos_to_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_to_runs(positions):\n",
    "    \"\"\"Converts a (sorted) list of integers to a list of 2-tuple \"runs\" of consecutive integers.\n",
    "    \n",
    "    Parameters\n",
    "    ==========\n",
    "    \n",
    "    positions: list of ints\n",
    "        Assumed to be sorted.\n",
    "        \n",
    "    Returns\n",
    "    =======\n",
    "    \n",
    "    runs: list of 2-tuples of ints\n",
    "        Each entry in this list is a 2-tuple of the format (p1, p2), where p1 and p2 are both integers\n",
    "        such that [p1, p1 + 1, p1 + 2, ..., p2] are all present in positions. If positions was empty,\n",
    "        this will be an empty list.\n",
    "    \"\"\"\n",
    "    \n",
    "    runs = []\n",
    "    if len(positions) > 1:\n",
    "        prev_pos = positions[0]\n",
    "        run_start_pos = positions[0]\n",
    "        for sp in positions[1:]:\n",
    "            if prev_pos == sp - 1:\n",
    "                prev_pos = sp\n",
    "            else:\n",
    "                runs.append((run_start_pos, prev_pos))\n",
    "                run_start_pos = sp\n",
    "                prev_pos = sp\n",
    "\n",
    "        runs.append((run_start_pos, sp))\n",
    "    elif len(positions) == 1:\n",
    "        runs.append((positions[0], positions[0]))\n",
    "    return runs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for when we set up formal tests\n",
    "assert convert_to_runs([1,2,3,5,6,7,10,20,31,32,33,34, 35, 40]) == \\\n",
    "       [(1, 3), (5, 7), (10, 10), (20, 20), (31, 35), (40, 40)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_p2mincov(percentages, minReadNumber):\n",
    "    # percentages should be values of p in the range (0, 50].\n",
    "    # minReadNumber indicates the minimum number of reads supporting a p-mutation we'd need to see.\n",
    "    #\n",
    "    # so, if minReadNumber is 5 and p = 50%, then the minimum coverage is (100 * 5) / 50 = 10x: this\n",
    "    # makes sense, since if this mutation is present at 50% frequency, then on average we'd need to see a coverage\n",
    "    # of 10x in order to see 5 copies of the mutated allele. Lowering p or increasing minReadNumber\n",
    "    # results in the min coverage increasing.\n",
    "    numerator = 100 * minReadNumber\n",
    "    return {p: (numerator / p) for p in percentages}\n",
    "\n",
    "assert get_p2mincov([1, 2, 3, 50, 20, 50], 5) == {\n",
    "    1: 500, 2: 250, 3: (500/3), 50: 10, 20: 25, 50: 10\n",
    "}\n",
    "\n",
    "def get_p2pct(percentages):\n",
    "    return {p: p / 100 for p in percentages}\n",
    "\n",
    "assert get_p2pct([1, 2, 50]) == {1: 0.01, 2: 0.02, 50: 0.5}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_meancovs():\n",
    "    # I guess we could output the median coverages also if we wanted to\n",
    "    import pileup\n",
    "    seq2pos2pileup = pileup.load()\n",
    "    \n",
    "    seq2meancov = {}\n",
    "    for seq in SEQS:\n",
    "        covs = []\n",
    "        for pos in range(1, seq2len[seq] + 1):\n",
    "            covs.append(pileup.get_cov(seq2pos2pileup[seq][pos]))\n",
    "        mean_cov = mean(covs)\n",
    "        median_cov = median(covs)\n",
    "        print(f\"Sequence {seq} has average coverage {mean_cov:,.2f} and median coverage {median_cov:,.2f}.\")\n",
    "        seq2meancov[seq] = mean_cov\n",
    "    return seq2meancov"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
