{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "competent-field",
   "metadata": {},
   "source": [
    "# Highly Mutated Gene Tables\n",
    "\n",
    "- For every sequence:\n",
    "  - Initialize a list that we'll use to store \"gene\" objects.\n",
    "  - For every gene in this sequence:\n",
    "    - Store \"number of mutated positions\", \"number of nonsynonymous mutated positions\"\n",
    "    - For every position in this gene:\n",
    "      - See if this position is mutated (> 0.5% mutation rate; same as elsewhere in the code)\n",
    "      - If this position is mutated:\n",
    "        - Increment number of mutated positions\n",
    "        - Get the codon of this position (will depend on gene being on +/- strand). Call it \"C1\".\n",
    "        - Look at `seq2pos2mismatches` JSON for this sequence and position. Get the maximum-frequency alternate nucleotide (__we could change this behavior if desired__).\n",
    "        - Change the codon of this position, substituting the max-frequency alternate nucleotide we just found. Call it \"C2\".\n",
    "        - If C1 and C2 don't code for the same amino acid / stop codon, this is a nonsynonymous mutation. Increment the number of nonsynonymous mutated positions.\n",
    "    - Compute nonsynonymous mutation rate and total mutation rate by just dividing the numbers we figured out by the total number of positions in this gene.\n",
    "    - Update the list of \"gene\" objects for this sequence with a dict containing some gene info (gene name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "vocational-channel",
   "metadata": {},
   "source": [
    "## Define `Gene` class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "significant-poetry",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "import skbio\n",
    "\n",
    "class Gene:\n",
    "    def __init__(self, gene_data, seq, fasta, minfreq=0.5):\n",
    "        self.leftend = gene_data.LeftEnd\n",
    "        self.rightend = gene_data.RightEnd\n",
    "        if self.leftend >= self.rightend:\n",
    "            raise ValueError(\"Left end of gene must be < right end of gene\")\n",
    "        \n",
    "        self.positions = get_pos_interval_from_gene(gene_data)\n",
    "        \n",
    "        # Code is from within-gene mutation spectrum notebook\n",
    "        self.length = len(self.positions)\n",
    "        if self.length % 3 != 0:\n",
    "            raise ValueError(\"Gene length not divisible by 3\")\n",
    "\n",
    "        # Name of sequence this gene is on (e.g. \"edge_6104\")\n",
    "        self.seq = seq\n",
    "        \n",
    "        # Actual sequence of the, well, sequence -- loaded by skbio\n",
    "        self.fasta = fasta\n",
    "        \n",
    "        # Orientation of gene\n",
    "        #\n",
    "        # if the strand is +, then this gene should look something like\n",
    "        # ATG...TGA\n",
    "        # (start codon)...(stop codon)\n",
    "        #\n",
    "        # and if the strand is -, then this gene should look something like\n",
    "        # AGT...GTA\n",
    "        # (stop codon in reverse)...(start codon in reverse)\n",
    "        #\n",
    "        # (of course there are multiple possible stop [and, more rarely, start] codons; this is just an example)\n",
    "        self.strand = gene_data.Strand\n",
    "        \n",
    "        # Number of this gene in the Prodigal .sco file\n",
    "        self.num = gene_data.Index\n",
    "        \n",
    "        # Numbers of mutations in this gene - we'll compute these soon\n",
    "        self.num_nonsyn_mutations = None\n",
    "        self.num_total_mutations = None\n",
    "        \n",
    "        # Mutation rates - we'll also compute these after computing the numbers, listed above\n",
    "        self.nonsyn_mutation_rate = None\n",
    "        self.total_mutation_rate = None\n",
    "        \n",
    "        self._count_mutations(minfreq)\n",
    "        \n",
    "    def _count_mutations(self, minfreq):\n",
    "        # \"minfreq\" is as described in the paper -- a percentage, where a given position has to have more\n",
    "        # than this percentage of mismatches in the alignment in order for this position to be considered\n",
    "        # \"mutated\" or not.\n",
    "        \n",
    "        self.num_total_mutations = 0\n",
    "        self.num_nonsyn_mutations = 0\n",
    "        minfreq_pct = minfreq / 100\n",
    "        for pos in self.positions:\n",
    "            is_mut = get_val(seq, pos, lambda cov, mismatches: 1 if (mismatches / cov) > minfreq_pct else 0)\n",
    "            if is_mut == 1:\n",
    "                self.num_total_mutations += 1\n",
    "                codon = self.get_codon(pos)\n",
    "                alts = seq2pos2mismatches[seq][str(pos)]\n",
    "                # Retrieve max-freq alternate nucleotide.\n",
    "                # Based on https://stackoverflow.com/a/280156.\n",
    "                # (Note that if there's a tie, the result is arbitrary. Shouldn't be a big deal. Making note\n",
    "                # of in the paper.)\n",
    "                max_freq_alt_nt = max(alts, key=alts.get)\n",
    "                alt_codon = self.get_codon(pos, mutate_to=max_freq_alt_nt)\n",
    "                \n",
    "                if str(codon.translate()) != str(alt_codon.translate()):\n",
    "                    self.num_nonsyn_mutations += 1\n",
    "                    \n",
    "        self.nonsyn_mutation_rate = self.num_nonsyn_mutations / self.length\n",
    "        self.total_mutation_rate = self.num_total_mutations / self.length\n",
    "                \n",
    "    def get_codon(self, pos, mutate_to=None):\n",
    "        # pos is 1-indexed\n",
    "        \n",
    "        def replace_pos_in_codon(codon_dna, codon_pos):                      \n",
    "            # skbio.DNA objects are immutable, so -- to avoid converting from DNA to str and\n",
    "            # then back again -- we use the skbio.Sequence.replace() method. This is the simplest\n",
    "            # way of doing this that I know of in \"pure\" skbio (...though I could totes be missing\n",
    "            # something obvious).\n",
    "            bl = [False, False, False]\n",
    "            bl[codon_pos] = True\n",
    "            return codon_dna.replace(bl, mutate_to)\n",
    "        \n",
    "        if self.strand == \"+\":\n",
    "            # Follows pattern 0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 3, 3, ...\n",
    "            codon_num = math.floor((pos - 1) / 3)\n",
    "            \n",
    "            codon_left = codon_num * 3\n",
    "            codon_right = codon_left + 3\n",
    "            \n",
    "            # Produces a skbio.DNA object\n",
    "            codon = self.fasta[codon_left: codon_right]\n",
    "            \n",
    "            if mutate_to is not None:\n",
    "                # Follows pattern 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, ...\n",
    "                codon_pos = (pos - 1) % 3\n",
    "                return replace_pos_in_codon(codon, codon_pos)\n",
    "            else:\n",
    "                return codon\n",
    "                \n",
    "        elif self.strand == \"-\":\n",
    "            seqlen = len(self.fasta)\n",
    "            # Follows pattern 5, 5, 5, 4, 4, 4, 3, 3, 3, 2, 2, 2, 1, 1, 1, 0, 0, 0\n",
    "            codon_num = math.floor((seqlen - pos) / 3)\n",
    "            \n",
    "            codon_right = seqlen - (codon_num * 3)\n",
    "            codon_left = codon_right - 3\n",
    "            \n",
    "            # Reverse the codon\n",
    "            codon = self.fasta[codon_left: codon_right][::-1]\n",
    "            \n",
    "            if mutate_to is not None:\n",
    "                # Follows pattern 2, 1, 0, 2, 1, 0, ...\n",
    "                codon_pos = (seqlen - pos) % 3\n",
    "                return replace_pos_in_codon(codon, codon_pos)\n",
    "            else:\n",
    "                return codon\n",
    "        \n",
    "        else:\n",
    "            raise ValueError(\"Invalid gene strand: {}\".format(self.strand))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "included-communist",
   "metadata": {},
   "source": [
    "## Actually go through and output highly mutated genes (in LaTeX code!)\n",
    "\n",
    "Settled on having this code actually output the LaTeX code for the rows in a table. Kinda fancy!\n",
    "\n",
    "Idea for this based on https://stackoverflow.com/questions/49223962#comment85460602_49223962, sort of -- I was thinking of just outputting raw data and having LaTeX mess around and load it somehow, but just straight-up outputting the LaTeX from Python is probably much less of a pain than whatever I would've wound up doing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "disabled-timeline",
   "metadata": {},
   "outputs": [],
   "source": [
    "for seq in SEQS:\n",
    "    fasta = skbio.DNA.read(\"../seqs/{}.fasta\".format(seq))\n",
    "    genes = []\n",
    "    df = parse_sco(\"../seqs/genes/{}.sco\".format(seq))\n",
    "    for gene_data in df.itertuples():\n",
    "        genes.append(Gene(gene_data, seq, fasta))\n",
    "    genes_by_mutrate = sorted(genes, key=lambda g: g.total_mutation_rate)\n",
    "    \n",
    "    with open(\"gene-tables/{}-highest-mutated-genes.tex\".format(seq), \"w\") as tblfile:\n",
    "        # Select top 10 genes for each sequence, going by total mutation rate.\n",
    "        # We reverse the order so that the highest-mutation-rate gene is written first, etc.\n",
    "        # (like a \"leaderboard\")\n",
    "        for g in genes_by_mutrate[-10:][::-1]:\n",
    "            tblfile.write(\"{} & {} & {} & {} \\\\\\\\ \\\\hline\\n\".format(\n",
    "                g.num, g.length, g.nonsyn_mutation_rate, g.total_mutation_rate\n",
    "            ))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (qiime2-2021.2)",
   "language": "python",
   "name": "myenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
