{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate the data for \"mutation matrices\"\n",
    "\n",
    "Both for codons (e.g. ATG, TGC) and amino acids (e.g. M, C).\n",
    "\n",
    "**NOTE 1:** Because this boils down to calling [`AlignedSegment.get_aligned_pairs()`](https://pysam.readthedocs.io/en/latest/api.html#pysam.AlignedSegment.get_aligned_pairs) once for every linear alignment to the selected MAGs, **this notebook is currently pretty slow**! I've optimized things to the point where this notebook takes around 20 hours on our cluster.\n",
    "\n",
    "If necessary / desired it should be possible to speed this up even more, using stuff like parallelization / writing this in a faster language like C / etc. There may also be methods in pysam I've overlooked that will help do this faster (from what I can tell the only way to do this in pysam involves calling `get_aligned_pairs()`, but maybe I've missed something...)\n",
    "\n",
    "**NOTE 2:** This doesn't actually generate the figures for the matrices -- it just outputs JSON files to a folder named `matrix-jsons/`, and another notebook will generate the figures based on these JSON files. This is intended to make it easier to regenerate the figures using different styles / etc. without having to wait hours for this stuff to finish.\n",
    "\n",
    "**NOTE 3:** We're redoing things so that the overall matrices are only based on individual MAGs -- so it should be possible to either parallelize this on the level of each MAGs (could probably parallelize things even further), or just only run this for individual MAGs. HOWEVER: for now, I'm going to keep generating this data for all of the three selected MAGs, because (i) we need this data for the Syn/NonSyn and Non-nonsense / Nonsense barplots, and (ii) I don't really want to rewrite this entire notebook since this is currently a one-off analysis..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run \"Header.ipynb\"\n",
    "%run \"GeneUtils.ipynb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import copy\n",
    "import time\n",
    "import json\n",
    "import pysam\n",
    "import skbio\n",
    "from collections import defaultdict, Counter\n",
    "from statistics import mean\n",
    "from parse_sco import parse_sco"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Go through all alignments to each genome and figure out which genes they intersect and which codons in these genes they fully cover\n",
    "\n",
    "__This is the main bottleneck of this notebook__, at least as of writing.\n",
    "\n",
    "Define a dict which we'll use to keep track of aligned codon frequencies for each codon, for each gene, for each genome.\n",
    "\n",
    "- For each alignment, see which predicted genes (if any) this alignment intersects within the genome. Note that \"intersects\" doesn't mean \"fully covers\".\n",
    "\n",
    "- For each of these genes, see which codons (if any) this alignment fully covers within the gene.\n",
    "\n",
    "- Increment aligned codon frequencies for all codons accordingly.\n",
    "\n",
    "The reason we do things this way, as opposed to iterating over just the alignments overlapping each codon in each gene, is that doing things that way is really slow! I'm pretty sure it's because \"find out which alignments overlap this region\" is a pretty slow operation when working with large datasets -- and also since these are long reads, doing this on the level of each codon means we're effectively doing a lot of redundant work (you can imagine that, for a given codon, the odds are pretty good that most alignments overlapping it will also overlap adjacent codon(s)).\n",
    "\n",
    "**Note that we purposefully say \"alignment\" instead of \"read\"** since a single read can correspond to multiple distinct linear alignments (e.g. if this read is aligned to both the start and end of a genome, due to a chimeric / supplementary alignment). We assume that no two linear alignments from a read cover the _same_ region of the genome (the upstream filtering stuff should already guarantee this), but this is about as far as we go. (We could also probably filter out supplementary alignments entirely and I suspect that wouldn't change the results here much.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# There's probably a fancier way of generating this list, but this is fine.\n",
    "codons = []\n",
    "# Also, we figure out the reverse complements of each of the 64 3-mers in advance -- this avoids\n",
    "# us having to call str(skbio.DNA(c).reverse_complement()) every time we see a codon, and saves a tiny\n",
    "# amount of time per alignment (the skbio approach took ~9e-5 seconds every time; the new approach takes ~9e-7\n",
    "# seconds every time). Considering we're going through well over a million alignments, the time savings comes out\n",
    "# to ... 130.977 seconds, aka 2 minutes 10 seconds, if I'm computing this correctly. So, not much, but it's\n",
    "# something!\n",
    "codon2revcomp = {}\n",
    "nts = \"ACGT\"\n",
    "for i in nts:\n",
    "    for j in nts:\n",
    "        for k in nts:\n",
    "            c = \"{}{}{}\".format(i, j, k)\n",
    "            codons.append(c)\n",
    "            codon2revcomp[c] = str(skbio.DNA(c).reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Maps sequence IDs to genes (keyed by their Index in the .sco file) to codons (keyed by (0-indexed!)\n",
    "# left end, i.e. the lower of the two positional boundaries of the codon, regardless of if its gene\n",
    "# is on the + or - strand) to observed aligned codon frequencies (keyed by just the triplet, e.g. \"AAA\").\n",
    "#\n",
    "# Example:\n",
    "# {\"edge_6104\":                                Sequence\n",
    "#     {1:                                      Gene index in the .sco file\n",
    "#         {265:                                Left codon position\n",
    "#             {\"TTA\": 1000, \"TTT\": 1, ... }    Aligned codon frequencies for this particular codon\n",
    "#         }\n",
    "#     }\n",
    "# }\n",
    "seq2gene2codon2alignedcodons = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished initialization for seq = edge_6104\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-10-9acece282b1f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     95\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     96\u001b[0m                 \u001b[0mhavent_checked_next_pairs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 97\u001b[0;31m                 \u001b[0;32mfor\u001b[0m \u001b[0mgene_data\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgenes_overlapping_aln\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     98\u001b[0m                     \u001b[0mgl\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgene_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mLeftEnd\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     99\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "bf = pysam.AlignmentFile(\"../main-workflow/output/fully-filtered-and-sorted-aln.bam\", \"rb\")\n",
    "\n",
    "tT1 = time.time()\n",
    "for seq in SEQS:\n",
    "    df = parse_sco(\"../seqs/genes/{}.sco\".format(seq))\n",
    "    \n",
    "    # We don't actually store any results in this, but we do use it for a slight optimization\n",
    "    gene2isrev = {}\n",
    "    \n",
    "    # Initialize some of the data structures\n",
    "    # NOTE: this is kind of slow. However, it still finishes within a few seconds, so not the most\n",
    "    # important thing to optimize\n",
    "    seq2gene2codon2alignedcodons[seq] = {}\n",
    "    for gene_data in df.itertuples():\n",
    "        \n",
    "        validate_gene_coords(gene_data)        \n",
    "        seq2gene2codon2alignedcodons[seq][gene_data.Index] = {}\n",
    "        gene2isrev[gene_data.Index] = (gene_data.Strand == \"-\")\n",
    "        \n",
    "        codon_positions = get_gene_left_codon_positions(gene_data)\n",
    "\n",
    "        # For each codon in this gene, keep track of all the codons spanning it from the various\n",
    "        # alignments to this genome.\n",
    "        for cpleft in codon_positions:\n",
    "            seq2gene2codon2alignedcodons[seq][gene_data.Index][cpleft] = defaultdict(int)\n",
    "            \n",
    "    print(\"Finished initialization for seq = {}\".format(seq))\n",
    "    alntimes = []\n",
    "    \n",
    "    for ri, aln in enumerate(bf.fetch(seq), 1):\n",
    "        \n",
    "        t1 = time.time()\n",
    "        \n",
    "        # Find all genes that this aln intersects in this genome\n",
    "        \n",
    "        # These are 0-indexed coordinates (and segright is offset to the right by one; see\n",
    "        # https://pysam.readthedocs.io/en/latest/api.html#pysam.AlignedSegment.reference_end)\n",
    "        segleft = aln.reference_start\n",
    "        segright = aln.reference_end\n",
    "        \n",
    "        if segleft is None or segright is None:\n",
    "            raise ValueError(\"Read {} is unmapped? This shouldn't happen!\".format(seg.query_name))\n",
    "        \n",
    "        if segleft >= segright:\n",
    "            raise ValueError(\"Read {}'s coordinates in pysam seem messed up: left = {}, right = {}\".format(\n",
    "                seg.query_name, segleft, segright\n",
    "            ))\n",
    "        \n",
    "        # Convert aligned segment boundaries to 1-indexed coordinates to make comparing with gene\n",
    "        # coordinates from the .sco file easier.\n",
    "        # Since segright was already offset to the right by 1, we don't need to do anything for it\n",
    "        # (the gene coordinates are exact: a gene from [266, 712] starts at base 266 and ends at base 712,\n",
    "        # using 1-indexing. So in order to make the aln boundaries match, we'd add 1 for segright and then\n",
    "        # subract 1 since segright was already 1 base off -- and n + 1 - 1 = n. (...math is hard)\n",
    "        segleft += 1\n",
    "        \n",
    "        # Use vectorization to find genes overlapping this aln: see https://stackoverflow.com/a/17071908\n",
    "        # for details on why parentheses, etc., and\n",
    "        # https://engineering.upside.com/a-beginners-guide-to-optimizing-pandas-code-for-speed-c09ef2c6a4d6\n",
    "        # for justification on why this is useful (tldr: makes code go fast)\n",
    "        genes_overlapping_aln = list(\n",
    "            df.loc[(df[\"RightEnd\"] >= segleft) & (df[\"LeftEnd\"] <= segright)].itertuples()\n",
    "        )\n",
    "        # Note about the above thing that just happened: you may be shaking your fist and saying \"wait\n",
    "        # itertuples is slow!\" And yeah, kinda. But for whatever reason I've tried multiple times to keep\n",
    "        # genes_overlapping_aln as a DataFrame (and then later vectorize stuff like checking that a given\n",
    "        # aligned pair covers a codon within the genes, etc) and the overhead costs seem to slow things down.\n",
    "        # I am sure it's possible to speed things up more, but right now things seem good enough.\n",
    "\n",
    "        # (Debugging code)\n",
    "        # print(\"{} genes overlap aln {}\".format(len(genes_overlapping_aln), ri))\n",
    "        # print(\"Read {}, which ranges from {} to {}, overlaps these genes:\".format(ri, segleft, segright))\n",
    "        # print(genes_overlapping_aln)\n",
    "                \n",
    "        # If no genes overlap this aln, we are free to move on to the next aln.\n",
    "        if len(genes_overlapping_aln) > 0:\n",
    "            \n",
    "            # Computing this is relatively slow, which is why we jump through so many hoops before we do this.\n",
    "            # Each entry in get_aligned_pairs() is a tuple with 2 elements:\n",
    "            # the first is the query/read pos and the second is the reference pos.\n",
    "            # TODO: would it be possible to only do this for certain positions we care about? get_aligned_pairs()\n",
    "            # returns a lot of stuff we don't need, e.g. regions of the aln that don't intersect with any genes.\n",
    "            ap = aln.get_aligned_pairs(matches_only=True)\n",
    "            \n",
    "            # Doesn't look like getting this in advance saves much time, but I don't think it hurts.\n",
    "            read_seq = aln.query_sequence\n",
    "            \n",
    "            # We only consider the leftmost position of each codon, so we don't need to bother checking the last\n",
    "            # two pairs of positions (since neither could be the leftmost position of a codon that this aln\n",
    "            # fully covers).\n",
    "            for api, pair1 in enumerate(ap[:-2]):\n",
    "\n",
    "                # Convert to 1-indexed position for ease of comparison with gene coordinates\n",
    "                pair1_refpos = pair1[1] + 1\n",
    "                  \n",
    "                havent_checked_next_pairs = True\n",
    "                for gene_data in genes_overlapping_aln:\n",
    "                    gl = gene_data.LeftEnd\n",
    "                    \n",
    "                    # Check that this pair is located within this gene and is the leftmost position of a\n",
    "                    # codon in the gene. (Note that check works for both + or - strand genes.\n",
    "                    # Whether the leftmost position is the \"start\" [i.e. CP 1] or \"end\" [i.e. CP 3] of\n",
    "                    # the gene changes with the strand of the gene, but we'll account for that later on\n",
    "                    # when we reverse-complement the codon if needed.)                  \n",
    "                    if pair1_refpos >= gl and pair1_refpos <= gene_data.RightEnd - 2 and ((pair1_refpos - gl) % 3 == 0):\n",
    "                        \n",
    "                        # Nice! Looks like this aln fully covers this codon.\n",
    "                        \n",
    "                        # If we haven't yet, check that this aln doesn't skip over parts of the codon,\n",
    "                        # or stuff like that. The reason this check is located *here* (and not\n",
    "                        # before we loop over the genes) is that it seems like this is a faster strategy:\n",
    "                        # only run these checks once we KNOW that this pair looks like it fully covers a\n",
    "                        # codon, since many pairs might not meet that criteria.\n",
    "                        #\n",
    "                        # (And by recording that we've run this check once, in havent_checked_next_pairs,\n",
    "                        # we can save the time cost of running the check multiple times.)\n",
    "                        #\n",
    "                        # I feel like an insane person trying to optimize this so much lmao.\n",
    "                        if havent_checked_next_pairs:\n",
    "                            # Check that the pairs are all consecutive (i.e. no \"jumps\" in the read,\n",
    "                            # and no \"jumps\" in the reference)\n",
    "                            # Since we don't consider the last two pairs in ap, pair2 and pair3 should\n",
    "                            # always be available.\n",
    "                            pair2 = ap[api + 1]\n",
    "                            pair3 = ap[api + 2]\n",
    "\n",
    "                            # For an aligned pair, [0] is the read pos and [1] is the reference pos.\n",
    "                            # Result of caching this is probably negligible but... may as well\n",
    "                            p10 = pair1[0]\n",
    "                            p20 = pair2[0]\n",
    "                            readpos_consecutive = p20 == (p10 + 1) and pair3[0] == (p20 + 1)\n",
    "                            if not readpos_consecutive:\n",
    "                                break\n",
    "\n",
    "                            # (pair1_refpos is already off by 1 so no need to redo the addition operation)\n",
    "                            p21 = pair2[1]\n",
    "                            refpos_consecutive = p21 == pair1_refpos and pair3[1] == (p21 + 1)\n",
    "                            if not refpos_consecutive:\n",
    "                                break\n",
    "                            havent_checked_next_pairs = False\n",
    "                        \n",
    "                        # Figure out what the read actually *says* in the alignment here.\n",
    "                        # (It'll probably be a complete match most of the time, but there will\n",
    "                        # be some occasional mismatches -- and seeing those is ... the whole point\n",
    "                        # of this notebook.)\n",
    "\n",
    "                        # We make sure to index the read by read coords, not reference coords!\n",
    "                        aligned_codon = read_seq[p10: p10 + 3]\n",
    "\n",
    "                        # Finally, update information about codon frequencies.\n",
    "                        gi = gene_data.Index\n",
    "                        if gene2isrev[gi]:\n",
    "                            seq2gene2codon2alignedcodons[seq][gi][pair1_refpos][codon2revcomp[aligned_codon]] += 1\n",
    "                        else:\n",
    "                            seq2gene2codon2alignedcodons[seq][gi][pair1_refpos][aligned_codon] += 1\n",
    "\n",
    "        t2 = time.time()\n",
    "        alntimes.append(t2 - t1)\n",
    "        if ri % 100 == 0:\n",
    "            print(f\"Seen {ri:,} alignments so far in {seq}.\")\n",
    "            print(f\"Total time taken thus far:               {t2-tT1:,.4f} sec.\")\n",
    "            print(f\"Average time per alignment for this seq: {mean(alntimes):,.4f} sec.\")\n",
    "\n",
    "    # At this point, we've seen all the alignments to all the codons in this genome.\n",
    "    # We can now \"call\" mutations based on the frequencies we've counted.\n",
    "tT2 = time.time()\n",
    "\n",
    "print(\"Figuring all that out took a total of {} seconds.\".format(tT2 - tT1))\n",
    "\n",
    "bf.close()\n",
    "# TODO: It'd probably be safer to output this after each sequence, rather than at the very end.\n",
    "with open(\"matrix-jsons/seq2gene2codon2alignedcodons.json\", \"w\") as dumpster:\n",
    "    dumpster.write(json.dumps(seq2gene2codon2alignedcodons))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using the information we just computed for each genome, \"call\" mutations and store this information in the frequency data structures we set up earlier\n",
    "\n",
    "This section is kind of slow, but on the order of \"takes a few minutes on the cluster\" and not on the order of \"takes literally hours to run\". We could make it more efficient if desired; most of my energy on optimization here thus far has been spent on the earlier step in this notebook.\n",
    "\n",
    "Note that the main output of the above section (`seq2gene2codon2alignedcodons`) has already been written out to a JSON file -- in practice, it's useful to be able to start execution again from this bottom section after stopping the above section (e.g. if something goes wrong _here_, then we don't have to rerun the top part of this notebook again). Even if the entire notebook is run in a single shot, we still save and then load the `seq2gene2codon2alignedcodons` object anyway."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initialize data structures that we'll store frequency data in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 64x63 dict: each key is a triplet of {A, C, G, T}, and each value is another dict with all the other codons\n",
    "codon2codon2freq = {}\n",
    "\n",
    "# 21x20 dict: each key is a proteinogenic amino acid (A, C, D, E, F, ...), limited to just\n",
    "# stuff in the standard genetic code (i.e. ignoring selenocystine and pyrrolsine) but including\n",
    "# \"*\", representing a stop codon.\n",
    "aa2aa2freq = {}\n",
    "\n",
    "# 64-key dict: maps each triplet to an integer indicating how frequently this triplet occurs in all genes\n",
    "# in the genomes (i.e. not counting mutations into this triplet).\n",
    "codon2freq = {}\n",
    "\n",
    "# 21-key dict: maps amino acid/stop codon to integer indicating frequency across all genes.\n",
    "aa2freq = {}\n",
    "\n",
    "aas = set([])\n",
    "for c in codons:\n",
    "    aas.add(str(skbio.DNA(c).translate()))\n",
    "    \n",
    "# Initialize dicts to 0s\n",
    "for c1 in codons:\n",
    "    codon2codon2freq[c1] = {c2: 0 for c2 in set(codons) - set([c1])}\n",
    "    codon2freq[c1] = 0\n",
    "    \n",
    "for aa1 in aas:\n",
    "    aa2aa2freq[aa1] = {aa2: 0 for aa2 in set(aas) - set([aa1])}\n",
    "    aa2freq[aa1] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"matrix-jsons/seq2gene2codon2alignedcodons.json\", \"r\") as loadster:\n",
    "    seq2gene2codon2alignedcodons = json.load(loadster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_mutmatrix_data(seq):\n",
    "    fasta = skbio.DNA.read(\"../seqs/{}.fasta\".format(seq))\n",
    "    df = parse_sco(\"../seqs/genes/{}.sco\".format(seq))\n",
    "    for gene_data in df.itertuples():\n",
    "        print(\"On gene {} in seq {}.\".format(gene_data.Index, seq))\n",
    "        for cpleft in range(gene_data.LeftEnd, gene_data.RightEnd + 1, 3):\n",
    "            \n",
    "            # Make note of the codon sequence and amino acid encoded by this codon in the \"reference\" genome.\n",
    "            # (Keep in mind that the gene data in the .sco file uses 1-indexed coords, so we need to convert\n",
    "            # accordingly.)\n",
    "            codon_dna = fasta[cpleft - 1: cpleft + 2]\n",
    "            if gene_data.Strand == \"-\":\n",
    "                codon_dna = codon_dna.reverse_complement()\n",
    "\n",
    "            codon_seq = str(codon_dna)\n",
    "            aa = str(codon_dna.translate())\n",
    "            \n",
    "            # Update frequencies accordingly.\n",
    "            codon2freq[codon_seq] += 1\n",
    "            aa2freq[aa] += 1\n",
    "            \n",
    "            # We can finally compute stats re: number of mismatching and matching codons.\n",
    "            aligned_codons = seq2gene2codon2alignedcodons[seq][str(gene_data.Index)][str(cpleft)]\n",
    "            \n",
    "            # Ignore weird, low-coverage cases. Copying from the SynAndNonsense barplots notebook, which\n",
    "            # uses similar logic:\n",
    "            #  If the reference codon sequence isn't even included in the aligned codons (but if\n",
    "            #  there are still other codons included in the alignment???), we're likely\n",
    "            #  at a weird low-coverage portion of the alignment. Ignore these cases (there are 116 such\n",
    "            #  problematic codons as of writing this, mostly in edge_1671 -- something might be wrong with\n",
    "            #  the alignment, which I had to recover from a backup recently -- will rerun and check).\n",
    "            #\n",
    "            #  This implicitly accounts for the case where 0 codons are present in aligned_codons, as well.\n",
    "            if codon_seq in aligned_codons:\n",
    "                \n",
    "                num_aligned_codons = sum(aligned_codons.values())\n",
    "                \n",
    "                # Only call a mutation using the max-freq alt codon, not the sum of all alternate codon freqs.\n",
    "                # The max(d, key=d.get) trick is from https://stackoverflow.com/a/280156 (I've used it a lot\n",
    "                # here :) Notably, this breaks ties arbitrarily.\n",
    "                alt_codons = {c: aligned_codons[c] for c in aligned_codons if c != codon_seq}\n",
    "                if len(alt_codons) > 0:\n",
    "                    max_freq_alt_codon = max(alt_codons, key=aligned_codons.get)\n",
    "                    max_freq_alt_codon_frac = alt_codons[max_freq_alt_codon] / num_aligned_codons\n",
    "                else:\n",
    "                    max_freq_alt_codon_frac = 0\n",
    "                    \n",
    "                # print(\"sum of vals of ac is {}\".format(sum(aligned_codons.values())))\n",
    "                # print(\"Codon {} from {} to {} in gene {} in seq {} has mutations: {}\".format(\n",
    "                #     codon_seq, cpleft, cpleft + 2, gene_data.Index, seq, aligned_codons\n",
    "                # ))\n",
    "\n",
    "                # Just for reference: this will print about cases where aggregating alternate codons results in\n",
    "                # \"false positives,\" at least compared to the method of only considering the maximum frequency\n",
    "                # alternate codon.\n",
    "                #\n",
    "                #if max_freq_alt_codon_frac <= 0.005 and sum(alt_codons.values()) / num_aligned_codons > 0.005:\n",
    "                #    print(\"Found contrary case.\")\n",
    "                #    print(f\"{codon_seq}, {aligned_codons}, only mutation in aggregate... max freq alt codon is {max_freq_alt_codon}\")\n",
    "                    \n",
    "                # Using p = 0.5%\n",
    "                if max_freq_alt_codon_frac > 0.005:\n",
    "\n",
    "                    codon2codon2freq[codon_seq][max_freq_alt_codon] += 1\n",
    "\n",
    "                    print(f\"{codon_seq}, {aligned_codons}, is mutation! max freq alt codon is {max_freq_alt_codon}\")\n",
    "\n",
    "                    # NOTE: I guess you could argue that we should do this another way, where we actually compute\n",
    "                    # the translations of all the alt codons and then pick the most common AA/stop codon from there?\n",
    "                    #\n",
    "                    # You could argue this either way: doing it based on just the mutated codon keeps the matrices\n",
    "                    # consistent and lessens the impact of small errors, while taking into account all alt codon\n",
    "                    # translations could help show weird things where multiple mutations have similar consequences.\n",
    "                    # Hmm.\n",
    "                    alt_codon_aa = str(skbio.DNA(max_freq_alt_codon).translate())\n",
    "                    if alt_codon_aa != aa:\n",
    "                        aa2aa2freq[aa][alt_codon_aa] += 1\n",
    "                        # print(\"Is nonsyn mutation! Alt {} codes for {}; orig coded for {}\".format(\n",
    "                        #     max_freq_alt_codon, alt_codon_aa, aa\n",
    "                        # ))\n",
    "\n",
    "    # Write out stuff for further analysis / in case of crisis\n",
    "    with open(f\"matrix-jsons/{seq}-codon2codon2freq.json\", \"w\") as dumpster:\n",
    "        dumpster.write(json.dumps(codon2codon2freq))\n",
    "\n",
    "    with open(f\"matrix-jsons/{seq}-codon2freq.json\", \"w\") as dumpster:\n",
    "        dumpster.write(json.dumps(codon2freq))\n",
    "\n",
    "    with open(f\"matrix-jsons/{seq}-aa2aa2freq.json\", \"w\") as dumpster:\n",
    "        dumpster.write(json.dumps(aa2aa2freq))\n",
    "\n",
    "    with open(f\"matrix-jsons/{seq}-aa2freq.json\", \"w\") as dumpster:\n",
    "        dumpster.write(json.dumps(aa2freq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "On gene 1 in seq edge_1671.\n",
      "TCA, {'TCA': 1394, 'TCG': 14}, is mutation! max freq alt codon is TCG\n",
      "ACT, {'ACT': 1407, 'ACG': 14, 'AAT': 1}, is mutation! max freq alt codon is ACG\n",
      "CAT, {'CAT': 1409, 'CAC': 14}, is mutation! max freq alt codon is CAC\n",
      "Found contrary case.\n",
      "AGC, {'AGC': 1384, 'AGT': 6, 'GCC': 1}, only mutation in aggregate... max freq alt codon is AGT\n",
      "Found contrary case.\n",
      "CCC, {'CCC': 1416, 'CCA': 1, 'CCT': 7}, only mutation in aggregate... max freq alt codon is CCT\n",
      "Found contrary case.\n",
      "ATG, {'ATG': 1419, 'ATC': 7, 'CTG': 1}, only mutation in aggregate... max freq alt codon is ATC\n",
      "Found contrary case.\n",
      "CTT, {'CTT': 1428, 'TTT': 7, 'CTG': 1, 'CTA': 1}, only mutation in aggregate... max freq alt codon is TTT\n",
      "Found contrary case.\n",
      "GCT, {'GCT': 1426, 'GTT': 7, 'CCT': 1}, only mutation in aggregate... max freq alt codon is GTT\n",
      "CAC, {'CAC': 1423, 'CAT': 15}, is mutation! max freq alt codon is CAT\n",
      "GGG, {'GGG': 1327, 'GGA': 15}, is mutation! max freq alt codon is GGA\n",
      "GGT, {'GGT': 1424, 'TGC': 14, 'GCC': 1}, is mutation! max freq alt codon is TGC\n",
      "AGA, {'AGA': 1427, 'AAA': 15}, is mutation! max freq alt codon is AAA\n",
      "GGG, {'GGG': 1410, 'GGC': 28}, is mutation! max freq alt codon is GGC\n",
      "GGA, {'GGA': 1426, 'GGT': 15}, is mutation! max freq alt codon is GGT\n",
      "AAC, {'AAC': 1430, 'AAT': 15}, is mutation! max freq alt codon is AAT\n",
      "GCC, {'GCC': 1422, 'ACC': 15}, is mutation! max freq alt codon is ACC\n",
      "Found contrary case.\n",
      "GCG, {'GCG': 1380, 'TGC': 4, 'CGG': 2, 'GGG': 1, 'TCG': 1, 'GGC': 1}, only mutation in aggregate... max freq alt codon is TGC\n",
      "GGC, {'GGC': 1433, 'GGT': 9, 'GCC': 1}, is mutation! max freq alt codon is GGT\n",
      "GAA, {'GAA': 1426, 'GAG': 9}, is mutation! max freq alt codon is GAG\n",
      "CTG, {'CTG': 1420, 'CTT': 10, 'CAG': 1}, is mutation! max freq alt codon is CTT\n",
      "AGT, {'AGT': 1425, 'AGC': 10}, is mutation! max freq alt codon is AGC\n",
      "GTG, {'GTG': 1422, 'GTT': 10}, is mutation! max freq alt codon is GTT\n",
      "GAC, {'GAC': 1421, 'GAT': 11, 'GTC': 1}, is mutation! max freq alt codon is GAT\n",
      "TTC, {'TTC': 1415, 'TTT': 13, 'TGC': 1, 'TTA': 1}, is mutation! max freq alt codon is TTT\n",
      "TGC, {'TGC': 1419, 'TGT': 8, 'TGG': 1}, is mutation! max freq alt codon is TGT\n",
      "CCT, {'CCT': 1398, 'CCA': 22, 'ACT': 1, 'TCT': 1}, is mutation! max freq alt codon is CCA\n",
      "TAT, {'TAT': 1418, 'TAC': 8, 'TAG': 1}, is mutation! max freq alt codon is TAC\n",
      "GAG, {'GAG': 1409, 'GGG': 10, 'GTG': 1}, is mutation! max freq alt codon is GGG\n",
      "TAC, {'TAC': 1424, 'TAT': 8}, is mutation! max freq alt codon is TAT\n",
      "TAC, {'TAC': 1420, 'TCC': 2, 'TAT': 8}, is mutation! max freq alt codon is TAT\n",
      "AGC, {'AGC': 1412, 'AGT': 8}, is mutation! max freq alt codon is AGT\n",
      "CCT, {'CCT': 1414, 'CCC': 8}, is mutation! max freq alt codon is CCC\n",
      "TAC, {'TAC': 1421, 'TAT': 8}, is mutation! max freq alt codon is TAT\n",
      "AAA, {'CAA': 1, 'AAA': 1402, 'GAA': 8}, is mutation! max freq alt codon is GAA\n",
      "GAT, {'GAT': 1416, 'GCT': 1, 'GAC': 8}, is mutation! max freq alt codon is GAC\n",
      "AAT, {'AAT': 1415, 'AAC': 8, 'ATT': 1}, is mutation! max freq alt codon is AAC\n",
      "GGT, {'GGT': 1319, 'GGA': 8, 'GTT': 1}, is mutation! max freq alt codon is GGA\n",
      "GAG, {'GAG': 1397, 'GAA': 20}, is mutation! max freq alt codon is GAA\n",
      "AAC, {'AAC': 1390, 'AAT': 23, 'ACC': 1, 'ATC': 1}, is mutation! max freq alt codon is AAT\n",
      "GGC, {'GGC': 1404, 'GGT': 13, 'TGC': 1, 'GCC': 1, 'GGA': 1}, is mutation! max freq alt codon is GGT\n",
      "AAC, {'AAC': 1406, 'AAT': 13, 'CAC': 1, 'ACC': 1}, is mutation! max freq alt codon is AAT\n",
      "ATT, {'ATT': 1411, 'ATC': 15}, is mutation! max freq alt codon is ATC\n",
      "GAG, {'GAG': 1417, 'GTG': 9}, is mutation! max freq alt codon is GTG\n",
      "CGG, {'CGG': 1388, 'CGT': 32, 'CTG': 1, 'GGG': 1}, is mutation! max freq alt codon is CGT\n",
      "CGC, {'CGC': 1407, 'CGT': 17, 'GCC': 1}, is mutation! max freq alt codon is CGT\n",
      "CCA, {'CCA': 1411, 'CCT': 18}, is mutation! max freq alt codon is CCT\n",
      "AAG, {'AAG': 1425, 'AAT': 20}, is mutation! max freq alt codon is AAT\n",
      "Found contrary case.\n",
      "TAA, {'TAA': 1386, 'AAA': 3, 'TTA': 4, 'ATA': 1}, only mutation in aggregate... max freq alt codon is TTA\n",
      "On gene 2 in seq edge_1671.\n",
      "GAA, {'GAA': 1423, 'GAG': 23}, is mutation! max freq alt codon is GAG\n",
      "CAT, {'CAT': 1417, 'CAC': 52}, is mutation! max freq alt codon is CAC\n",
      "TGT, {'TGT': 1376, 'TAT': 50, 'TGA': 1, 'TTT': 1}, is mutation! max freq alt codon is TAT\n",
      "ACT, {'ACT': 1450, 'ACG': 9}, is mutation! max freq alt codon is ACG\n",
      "GTC, {'GTC': 1436, 'GCC': 19}, is mutation! max freq alt codon is GCC\n",
      "GTT, {'GTT': 1448, 'ATT': 8, 'GTA': 1}, is mutation! max freq alt codon is ATT\n",
      "ACT, {'ACT': 1410, 'ACC': 55}, is mutation! max freq alt codon is ACC\n",
      "On gene 3 in seq edge_1671.\n",
      "AAG, {'AAG': 1443, 'AAA': 11, 'ACG': 1}, is mutation! max freq alt codon is AAA\n",
      "CGT, {'CGT': 1434, 'CGA': 25, 'CGG': 1}, is mutation! max freq alt codon is CGA\n",
      "CCA, {'CCA': 1436, 'CAA': 12}, is mutation! max freq alt codon is CAA\n",
      "AAG, {'AAG': 1402, 'AAA': 45, 'GGG': 1}, is mutation! max freq alt codon is AAA\n",
      "GAG, {'GAG': 1400, 'GAA': 45}, is mutation! max freq alt codon is GAA\n",
      "GTC, {'GTC': 1423, 'ATA': 26}, is mutation! max freq alt codon is ATA\n",
      "CTT, {'CTT': 1425, 'CTA': 26}, is mutation! max freq alt codon is CTA\n",
      "ACA, {'ACA': 1366, 'ACC': 24, 'AAC': 1, 'CCA': 1}, is mutation! max freq alt codon is ACC\n",
      "TGT, {'TGT': 1430, 'TGC': 19, 'TGG': 1}, is mutation! max freq alt codon is TGC\n",
      "TCT, {'TCT': 1432, 'TCG': 20}, is mutation! max freq alt codon is TCG\n",
      "CCA, {'CCA': 1429, 'CCG': 18}, is mutation! max freq alt codon is CCG\n",
      "CCT, {'CCT': 1426, 'CCA': 19, 'CAT': 1}, is mutation! max freq alt codon is CCA\n",
      "GCA, {'GCA': 1434, 'GCG': 19, 'TCA': 1}, is mutation! max freq alt codon is GCG\n",
      "GGC, {'GGC': 1432, 'AGC': 20}, is mutation! max freq alt codon is AGC\n",
      "GCG, {'GCG': 1429, 'GCT': 2, 'GCA': 19}, is mutation! max freq alt codon is GCA\n",
      "GTT, {'GTT': 1388, 'ATT': 64, 'GGT': 1, 'TGT': 1, 'TTT': 1}, is mutation! max freq alt codon is ATT\n",
      "CGC, {'CGC': 1432, 'CGT': 31, 'GGC': 2, 'GCC': 1, 'CGG': 1}, is mutation! max freq alt codon is CGT\n",
      "CAG, {'CAG': 1441, 'CAA': 35}, is mutation! max freq alt codon is CAA\n",
      "CGT, {'CGT': 1426, 'CGC': 49, 'CGA': 1}, is mutation! max freq alt codon is CGC\n",
      "TAT, {'TAT': 1452, 'CAT': 23, 'TAA': 1}, is mutation! max freq alt codon is CAT\n",
      "GGA, {'GGA': 1440, 'GGG': 25}, is mutation! max freq alt codon is GGG\n",
      "GAC, {'GAC': 1447, 'GAG': 25, 'TAC': 1}, is mutation! max freq alt codon is GAG\n",
      "On gene 4 in seq edge_1671.\n",
      "TCT, {'TCT': 1471, 'TCC': 13, 'TCA': 1}, is mutation! max freq alt codon is TCC\n",
      "AGC, {'AGC': 1427, 'ACC': 61}, is mutation! max freq alt codon is ACC\n",
      "TTG, {'TTG': 1444, 'GTG': 1, 'TTA': 24}, is mutation! max freq alt codon is TTA\n",
      "GAC, {'GAC': 1435, 'TAC': 1, 'GAG': 24}, is mutation! max freq alt codon is GAG\n",
      "On gene 5 in seq edge_1671.\n",
      "ATG, {'ATG': 1425, 'ATC': 47, 'AAT': 1, 'ATA': 1, 'TGT': 2, 'TTG': 1, 'TCT': 1}, is mutation! max freq alt codon is ATC\n",
      "GGG, {'GGG': 1438, 'GGA': 25}, is mutation! max freq alt codon is GGA\n",
      "GCT, {'GCT': 1421, 'TCT': 49, 'TAT': 1}, is mutation! max freq alt codon is TCT\n",
      "AGT, {'AGT': 1374, 'AGC': 75, 'ATT': 1, 'CGT': 1}, is mutation! max freq alt codon is AGC\n",
      "TAT, {'TAT': 1373, 'GAT': 91, 'AAT': 2, 'CAT': 1, 'ATA': 1}, is mutation! max freq alt codon is GAT\n",
      "AGC, {'AGC': 1439, 'AAC': 27, 'ATC': 1}, is mutation! max freq alt codon is AAC\n",
      "CTG, {'CTG': 1443, 'CTT': 15, 'ATG': 1}, is mutation! max freq alt codon is CTT\n",
      "CGG, {'CGG': 1436, 'CGC': 15, 'CTG': 2, 'GGG': 1}, is mutation! max freq alt codon is CGC\n",
      "On gene 6 in seq edge_1671.\n",
      "GAT, {'GAT': 1355, 'GCG': 16}, is mutation! max freq alt codon is GCG\n",
      "TTT, {'TTT': 1386, 'TTC': 17, 'TGT': 1}, is mutation! max freq alt codon is TTC\n",
      "ATT, {'ATT': 1383, 'GTT': 17, 'AAT': 1}, is mutation! max freq alt codon is GTT\n",
      "Found contrary case.\n",
      "ACA, {'ACA': 1382, 'ACC': 1, 'AAC': 5, 'CAA': 2}, only mutation in aggregate... max freq alt codon is AAC\n",
      "CTA, {'CTA': 1399, 'TTA': 19, 'CTT': 1}, is mutation! max freq alt codon is TTA\n",
      "On gene 7 in seq edge_1671.\n",
      "On gene 8 in seq edge_1671.\n",
      "CCT, {'CCT': 1502, 'CCC': 29}, is mutation! max freq alt codon is CCC\n",
      "GGT, {'GGT': 1508, 'GGC': 25, 'GGG': 1, 'GTT': 1}, is mutation! max freq alt codon is GGC\n",
      "On gene 9 in seq edge_1671.\n",
      "On gene 10 in seq edge_1671.\n",
      "CGT, {'CGT': 1526, 'CGC': 30, 'GGT': 1}, is mutation! max freq alt codon is CGC\n",
      "GTC, {'GTC': 1426, 'GTT': 95}, is mutation! max freq alt codon is GTT\n",
      "GTA, {'GTA': 1464, 'GTT': 95}, is mutation! max freq alt codon is GTT\n",
      "GAG, {'GAG': 1457, 'GAA': 93, 'GAT': 1}, is mutation! max freq alt codon is GAA\n",
      "ATT, {'ATT': 1454, 'ATA': 97}, is mutation! max freq alt codon is ATA\n",
      "ATA, {'ATA': 1455, 'ACA': 99, 'CTA': 1}, is mutation! max freq alt codon is ACA\n",
      "CGC, {'CGC': 1448, 'CGT': 96}, is mutation! max freq alt codon is CGT\n",
      "CCC, {'CCC': 1479, 'CCA': 73}, is mutation! max freq alt codon is CCA\n",
      "GGA, {'GGA': 1333, 'GGT': 102, 'GAA': 2}, is mutation! max freq alt codon is GGT\n",
      "GGT, {'GGT': 1498, 'GTT': 1, 'GGG': 32}, is mutation! max freq alt codon is GGG\n",
      "ATC, {'ATC': 1427, 'ATT': 128}, is mutation! max freq alt codon is ATT\n",
      "GAA, {'GAA': 1405, 'GAG': 128, 'AAA': 1, 'TAA': 1, 'GCA': 1}, is mutation! max freq alt codon is GAG\n",
      "ATA, {'ATA': 1417, 'ATC': 127, 'TTA': 1}, is mutation! max freq alt codon is ATC\n",
      "ATC, {'ATC': 1521, 'ATT': 34}, is mutation! max freq alt codon is ATT\n",
      "GAA, {'GAA': 1521, 'GAG': 26}, is mutation! max freq alt codon is GAG\n",
      "AAG, {'AAG': 1522, 'AAA': 26}, is mutation! max freq alt codon is AAA\n",
      "GTT, {'GTT': 1509, 'GTA': 34, 'TTT': 1}, is mutation! max freq alt codon is GTA\n",
      "CAA, {'CAA': 1487, 'CAG': 59}, is mutation! max freq alt codon is CAG\n",
      "AAA, {'AAA': 1417, 'AAG': 138}, is mutation! max freq alt codon is AAG\n",
      "GAC, {'GAC': 1529, 'GAT': 33, 'GAA': 1}, is mutation! max freq alt codon is GAT\n",
      "CGA, {'CGA': 1481, 'AGA': 79}, is mutation! max freq alt codon is AGA\n",
      "On gene 11 in seq edge_1671.\n",
      "On gene 12 in seq edge_1671.\n",
      "CAC, {'CAC': 1539, 'CAT': 33}, is mutation! max freq alt codon is CAT\n",
      "On gene 13 in seq edge_1671.\n",
      "GGA, {'GGA': 1534, 'GGC': 35, 'TGA': 1}, is mutation! max freq alt codon is GGC\n",
      "CCC, {'CCC': 1537, 'CCT': 35, 'CCA': 1}, is mutation! max freq alt codon is CCT\n",
      "On gene 14 in seq edge_1671.\n",
      "On gene 15 in seq edge_1671.\n",
      "AAC, {'AAC': 1529, 'AAT': 33}, is mutation! max freq alt codon is AAT\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GGG, {'GGG': 1430, 'GGA': 116}, is mutation! max freq alt codon is GGA\n",
      "On gene 16 in seq edge_1671.\n",
      "TAC, {'TAC': 1522, 'TAT': 28, 'GAC': 1}, is mutation! max freq alt codon is TAT\n",
      "On gene 17 in seq edge_1671.\n",
      "On gene 18 in seq edge_1671.\n",
      "On gene 19 in seq edge_1671.\n",
      "CCA, {'CCA': 1441, 'CCT': 121}, is mutation! max freq alt codon is CCT\n",
      "GGA, {'GGA': 1524, 'GGG': 28, 'GAA': 2, 'GGT': 1, 'AGA': 1}, is mutation! max freq alt codon is GGG\n",
      "On gene 20 in seq edge_1671.\n",
      "ATC, {'ATC': 1536, 'ATT': 27, 'AAT': 1}, is mutation! max freq alt codon is ATT\n",
      "On gene 21 in seq edge_1671.\n",
      "On gene 22 in seq edge_1671.\n",
      "On gene 23 in seq edge_1671.\n",
      "On gene 24 in seq edge_1671.\n",
      "ATC, {'ATC': 1505, 'ATT': 24}, is mutation! max freq alt codon is ATT\n",
      "On gene 25 in seq edge_1671.\n",
      "On gene 26 in seq edge_1671.\n",
      "On gene 27 in seq edge_1671.\n",
      "GAG, {'GAG': 1421, 'GAA': 86, 'GGG': 1}, is mutation! max freq alt codon is GAA\n",
      "On gene 28 in seq edge_1671.\n",
      "GAC, {'GAC': 1432, 'GAT': 87}, is mutation! max freq alt codon is GAT\n",
      "On gene 29 in seq edge_1671.\n",
      "CCT, {'CCT': 1355, 'CCC': 135}, is mutation! max freq alt codon is CCC\n",
      "GTA, {'GTA': 1462, 'GTG': 32}, is mutation! max freq alt codon is GTG\n",
      "GAT, {'GAT': 1344, 'GAC': 136}, is mutation! max freq alt codon is GAC\n",
      "On gene 30 in seq edge_1671.\n",
      "On gene 31 in seq edge_1671.\n",
      "GTA, {'GTA': 1497, 'GTC': 19, 'TTA': 1}, is mutation! max freq alt codon is GTC\n",
      "GGG, {'GGG': 1409, 'GGA': 103, 'CGG': 1}, is mutation! max freq alt codon is GGA\n",
      "ATA, {'ATA': 1382, 'ATC': 136, 'TTA': 2}, is mutation! max freq alt codon is ATC\n",
      "On gene 32 in seq edge_1671.\n",
      "CCG, {'CCG': 1492, 'CCC': 29}, is mutation! max freq alt codon is CCC\n",
      "CTG, {'CTG': 1412, 'CTA': 106}, is mutation! max freq alt codon is CTA\n",
      "CTC, {'CTC': 1493, 'CTT': 29}, is mutation! max freq alt codon is CTT\n",
      "ATT, {'ATT': 1440, 'ATC': 75}, is mutation! max freq alt codon is ATC\n",
      "CCC, {'CCC': 1423, 'CCA': 75, 'CAC': 1}, is mutation! max freq alt codon is CCA\n",
      "ATA, {'ATA': 1441, 'ATT': 73}, is mutation! max freq alt codon is ATT\n",
      "GCA, {'GCA': 1438, 'GGC': 1, 'GCT': 70, 'GCC': 1}, is mutation! max freq alt codon is GCT\n",
      "On gene 33 in seq edge_1671.\n",
      "TAC, {'TAC': 1445, 'TAT': 67}, is mutation! max freq alt codon is TAT\n",
      "On gene 34 in seq edge_1671.\n",
      "On gene 35 in seq edge_1671.\n",
      "On gene 36 in seq edge_1671.\n",
      "On gene 37 in seq edge_1671.\n",
      "On gene 38 in seq edge_1671.\n",
      "GTG, {'GTG': 1436, 'GTA': 88, 'ATG': 1}, is mutation! max freq alt codon is GTA\n",
      "On gene 39 in seq edge_1671.\n",
      "On gene 40 in seq edge_1671.\n",
      "CGT, {'CGT': 1499, 'CGC': 28}, is mutation! max freq alt codon is CGC\n",
      "On gene 41 in seq edge_1671.\n",
      "TTA, {'TTA': 1482, 'ATA': 33}, is mutation! max freq alt codon is ATA\n",
      "CGC, {'CGC': 1496, 'CTC': 31, 'AGC': 1}, is mutation! max freq alt codon is CTC\n",
      "On gene 42 in seq edge_1671.\n",
      "CCG, {'CCG': 1469, 'TCG': 35, 'CGT': 1}, is mutation! max freq alt codon is TCG\n",
      "AGA, {'AGA': 1467, 'ATA': 35}, is mutation! max freq alt codon is ATA\n",
      "GGT, {'GGT': 1398, 'GGA': 117, 'GAT': 1, 'TGT': 1}, is mutation! max freq alt codon is GGA\n",
      "GGA, {'GGA': 1476, 'GTA': 24, 'GGT': 1}, is mutation! max freq alt codon is GTA\n",
      "ACT, {'ACT': 1404, 'TCT': 84, 'ATT': 2, 'ACA': 1}, is mutation! max freq alt codon is TCT\n",
      "CAT, {'CAT': 1407, 'ACT': 84, 'CCT': 1}, is mutation! max freq alt codon is ACT\n",
      "ACA, {'ACA': 1406, 'TCA': 83}, is mutation! max freq alt codon is TCA\n",
      "On gene 43 in seq edge_1671.\n",
      "AAT, {'AAT': 1403, 'AAA': 84}, is mutation! max freq alt codon is AAA\n",
      "AAT, {'AAT': 1402, 'ACT': 85}, is mutation! max freq alt codon is ACT\n",
      "AGA, {'AGA': 1352, 'ATA': 85, 'AGC': 1, 'AAA': 1, 'GAA': 1, 'CTA': 1}, is mutation! max freq alt codon is ATA\n",
      "AGC, {'AGC': 1410, 'AGG': 86}, is mutation! max freq alt codon is AGG\n",
      "AAT, {'AAT': 1407, 'AGT': 85, 'CAT': 2, 'ATG': 1}, is mutation! max freq alt codon is AGT\n",
      "CAG, {'CAG': 1409, 'CCT': 85, 'CTG': 1}, is mutation! max freq alt codon is CCT\n",
      "CAA, {'CAA': 1403, 'GAA': 85, 'CCA': 1}, is mutation! max freq alt codon is GAA\n",
      "CTA, {'CTA': 1399, 'CAA': 86}, is mutation! max freq alt codon is CAA\n",
      "On gene 44 in seq edge_1671.\n",
      "On gene 45 in seq edge_1671.\n",
      "On gene 46 in seq edge_1671.\n",
      "CTG, {'TTG': 31, 'CTG': 1495}, is mutation! max freq alt codon is TTG\n",
      "AGA, {'AAA': 30, 'AGA': 1509}, is mutation! max freq alt codon is AAA\n",
      "GAT, {'GAT': 1431, 'GAC': 97}, is mutation! max freq alt codon is GAC\n",
      "AAT, {'AAT': 1468, 'AAG': 65, 'AAA': 1, 'ATT': 2}, is mutation! max freq alt codon is AAG\n",
      "TAC, {'TAC': 1474, 'TAT': 65}, is mutation! max freq alt codon is TAT\n",
      "On gene 47 in seq edge_1671.\n",
      "GAC, {'GAC': 1502, 'GAT': 36}, is mutation! max freq alt codon is GAT\n",
      "TCG, {'TCA': 132, 'TCG': 1406}, is mutation! max freq alt codon is TCA\n",
      "TTA, {'TTA': 1393, 'CTG': 126}, is mutation! max freq alt codon is CTG\n",
      "TCT, {'TCT': 1459, 'CCT': 65, 'TCA': 2, 'TAT': 1}, is mutation! max freq alt codon is CCT\n",
      "GCA, {'GCG': 29, 'GCA': 1493, 'GAA': 1, 'GCC': 1}, is mutation! max freq alt codon is GCG\n",
      "GAG, {'GAG': 1464, 'GAA': 65, 'GAC': 1}, is mutation! max freq alt codon is GAA\n",
      "AGC, {'AGT': 32, 'AGC': 1496}, is mutation! max freq alt codon is AGT\n",
      "ACA, {'ACC': 99, 'ACA': 1426}, is mutation! max freq alt codon is ACC\n",
      "ACG, {'ACG': 1448, 'ACA': 68, 'ACT': 1, 'ACC': 1}, is mutation! max freq alt codon is ACA\n",
      "On gene 48 in seq edge_1671.\n",
      "GAT, {'GAG': 105, 'GAT': 1434}, is mutation! max freq alt codon is GAG\n",
      "ACT, {'ACC': 98, 'ACT': 1413}, is mutation! max freq alt codon is ACC\n",
      "GTA, {'GTT': 101, 'GTA': 1410}, is mutation! max freq alt codon is GTT\n",
      "CCT, {'CCG': 102, 'CCT': 1406}, is mutation! max freq alt codon is CCG\n",
      "AAA, {'AAC': 103, 'AAA': 1413}, is mutation! max freq alt codon is AAC\n",
      "TTG, {'CTG': 104, 'TTG': 1395}, is mutation! max freq alt codon is CTG\n",
      "AGT, {'AGC': 103, 'AGT': 1392, 'AGG': 1}, is mutation! max freq alt codon is AGC\n",
      "On gene 49 in seq edge_1671.\n",
      "GCT, {'GCG': 104, 'GCT': 1391}, is mutation! max freq alt codon is GCG\n",
      "ATA, {'ATC': 105, 'ATA': 1400, 'TTA': 1, 'CAT': 1, 'TAA': 1}, is mutation! max freq alt codon is ATC\n",
      "On gene 50 in seq edge_1671.\n",
      "ACT, {'ACC': 102, 'ACT': 1379, 'AAC': 1, 'AAT': 1}, is mutation! max freq alt codon is ACC\n",
      "GGG, {'GGA': 103, 'GGG': 1385}, is mutation! max freq alt codon is GGA\n",
      "AGC, {'AGT': 100, 'AGC': 1347}, is mutation! max freq alt codon is AGT\n",
      "AAA, {'AGA': 97, 'AAA': 1392, 'CAA': 2}, is mutation! max freq alt codon is AGA\n",
      "On gene 51 in seq edge_1671.\n",
      "AGT, {'GGT': 94, 'AGT': 1388, 'AGG': 1}, is mutation! max freq alt codon is GGT\n",
      "GTT, {'GTC': 100, 'GTT': 1394, 'TTT': 2, 'TTA': 1, 'GGT': 1, 'TAT': 1}, is mutation! max freq alt codon is GTC\n",
      "GAG, {'GAA': 101, 'GAG': 1404}, is mutation! max freq alt codon is GAA\n",
      "ACT, {'ACC': 101, 'ACT': 1412, 'CCT': 1}, is mutation! max freq alt codon is ACC\n",
      "AAG, {'AAA': 99, 'AAG': 1382, 'TAA': 1}, is mutation! max freq alt codon is AAA\n",
      "TCG, {'TCA': 36, 'TCG': 1490}, is mutation! max freq alt codon is TCA\n",
      "AGC, {'AGC': 1460, 'AGT': 61}, is mutation! max freq alt codon is AGT\n",
      "CCG, {'CCT': 35, 'CCG': 1477, 'TCC': 1, 'ACC': 1, 'TCG': 1}, is mutation! max freq alt codon is CCT\n",
      "ATT, {'ATC': 36, 'ATT': 1481}, is mutation! max freq alt codon is ATC\n",
      "On gene 52 in seq edge_1671.\n",
      "TTC, {'TTC': 1462, 'TTT': 61}, is mutation! max freq alt codon is TTT\n",
      "GCA, {'GCA': 1459, 'GCT': 61, 'GCG': 1}, is mutation! max freq alt codon is GCT\n",
      "CAC, {'CAT': 100, 'CAC': 1437, 'CAA': 1}, is mutation! max freq alt codon is CAT\n",
      "GGT, {'GGT': 1437, 'GGC': 102, 'TGT': 1, 'GTT': 1}, is mutation! max freq alt codon is GGC\n",
      "AAC, {'AAC': 1442, 'AAT': 100, 'ACC': 2}, is mutation! max freq alt codon is AAT\n",
      "GCT, {'GCT': 1446, 'GCC': 101}, is mutation! max freq alt codon is GCC\n",
      "ATT, {'ATT': 1479, 'GTT': 65}, is mutation! max freq alt codon is GTT\n",
      "GCA, {'GCA': 1476, 'GCT': 67, 'TCA': 1}, is mutation! max freq alt codon is GCT\n",
      "GAC, {'GAC': 1439, 'AAC': 105, 'ACC': 1}, is mutation! max freq alt codon is AAC\n",
      "CCT, {'CCT': 1439, 'CCC': 105, 'CCG': 1}, is mutation! max freq alt codon is CCC\n",
      "GAG, {'GAG': 1493, 'GAA': 70}, is mutation! max freq alt codon is GAA\n",
      "GAA, {'GAA': 1494, 'AAA': 70}, is mutation! max freq alt codon is AAA\n",
      "TCT, {'TCT': 1459, 'TCC': 101, 'TCG': 1}, is mutation! max freq alt codon is TCC\n",
      "CCT, {'CCT': 1453, 'CCC': 101, 'CTA': 1}, is mutation! max freq alt codon is CCC\n",
      "CAC, {'CAC': 1514, 'CAT': 36}, is mutation! max freq alt codon is CAT\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GGC, {'GGC': 1517, 'GGT': 37}, is mutation! max freq alt codon is GGT\n",
      "On gene 53 in seq edge_1671.\n",
      "ATT, {'ATT': 1535, 'ATC': 36}, is mutation! max freq alt codon is ATC\n",
      "GTT, {'GTT': 1457, 'ATT': 98, 'CTT': 1, 'TTT': 1, 'AAT': 1}, is mutation! max freq alt codon is ATT\n",
      "ATA, {'ATA': 1458, 'GTA': 98, 'TTA': 1}, is mutation! max freq alt codon is GTA\n",
      "TTC, {'TTT': 64, 'TTC': 1491, 'CTC': 1}, is mutation! max freq alt codon is TTT\n",
      "TTT, {'TTT': 1519, 'CTT': 33}, is mutation! max freq alt codon is CTT\n",
      "TCA, {'TCA': 1460, 'TCC': 97, 'TCG': 1, 'TCT': 1}, is mutation! max freq alt codon is TCC\n",
      "TTT, {'TTT': 1490, 'TTC': 62, 'TTA': 2, 'TCA': 1, 'GTT': 1}, is mutation! max freq alt codon is TTC\n",
      "ATT, {'ATT': 1474, 'TCA': 63, 'TTT': 1}, is mutation! max freq alt codon is TCA\n",
      "GGT, {'GGT': 1461, 'GAT': 101}, is mutation! max freq alt codon is GAT\n",
      "GGT, {'GGT': 1438, 'AGT': 101, 'TGT': 1}, is mutation! max freq alt codon is AGT\n",
      "GGT, {'GGT': 1424, 'TGT': 101, 'GTT': 2}, is mutation! max freq alt codon is TGT\n",
      "On gene 54 in seq edge_1671.\n",
      "AGT, {'AGT': 1429, 'AGC': 101}, is mutation! max freq alt codon is AGC\n",
      "GTA, {'GTA': 1423, 'GTT': 102}, is mutation! max freq alt codon is GTT\n",
      "CAG, {'CAG': 1424, 'CAA': 102}, is mutation! max freq alt codon is CAA\n",
      "TTC, {'TTC': 1422, 'TTT': 102}, is mutation! max freq alt codon is TTT\n",
      "CTA, {'CTA': 1431, 'TTA': 101, 'CAA': 1, 'ATA': 1}, is mutation! max freq alt codon is TTA\n",
      "CGC, {'CGC': 1416, 'CGT': 101, 'CGG': 1}, is mutation! max freq alt codon is CGT\n",
      "CAT, {'CAT': 1457, 'GAT': 63, 'CAA': 1}, is mutation! max freq alt codon is GAT\n",
      "AAA, {'AAA': 1411, 'AAG': 100}, is mutation! max freq alt codon is AAG\n",
      "On gene 55 in seq edge_1671.\n",
      "TCG, {'TCG': 1482, 'TCA': 38}, is mutation! max freq alt codon is TCA\n",
      "GGA, {'GGA': 1382, 'GGC': 100, 'GGG': 2, 'GAG': 1, 'GTA': 1}, is mutation! max freq alt codon is GGC\n",
      "GGA, {'GGA': 1409, 'GGG': 97, 'GTA': 1, 'TGA': 1, 'GAA': 2, 'GGC': 1}, is mutation! max freq alt codon is GGG\n",
      "TCC, {'CCC': 101, 'TCC': 1430}, is mutation! max freq alt codon is CCC\n",
      "AAT, {'AAC': 101, 'AAT': 1428, 'TAT': 1, 'CAT': 1}, is mutation! max freq alt codon is AAC\n",
      "On gene 56 in seq edge_1671.\n",
      "GTA, {'GTT': 39, 'GTA': 1484}, is mutation! max freq alt codon is GTT\n",
      "TCA, {'TCG': 106, 'TCA': 1402}, is mutation! max freq alt codon is TCG\n",
      "GAT, {'GAC': 107, 'GAT': 1409, 'GAG': 1}, is mutation! max freq alt codon is GAC\n",
      "On gene 57 in seq edge_1671.\n",
      "Found contrary case.\n",
      "TTA, {'TTA': 1499, 'TTT': 5, 'GTA': 1, 'TAA': 1, 'GTT': 1}, only mutation in aggregate... max freq alt codon is TTT\n",
      "AAG, {'AAG': 1480, 'AAC': 36, 'CAG': 1, 'AAA': 1}, is mutation! max freq alt codon is AAC\n",
      "GAC, {'GAC': 1483, 'GGC': 37, 'GCC': 1}, is mutation! max freq alt codon is GGC\n",
      "On gene 58 in seq edge_1671.\n",
      "CCG, {'CCG': 1460, 'CCA': 38, 'CGC': 1, 'CCC': 1}, is mutation! max freq alt codon is CCA\n",
      "Found contrary case.\n",
      "TTC, {'TTC': 1467, 'TTT': 5, 'GTC': 1, 'TCA': 2, 'TTA': 1, 'TAC': 1}, only mutation in aggregate... max freq alt codon is TTT\n",
      "ATA, {'ATA': 1488, 'ATT': 37}, is mutation! max freq alt codon is ATT\n",
      "AGA, {'AGA': 1494, 'AGG': 35}, is mutation! max freq alt codon is AGG\n",
      "Found contrary case.\n",
      "AAC, {'AAC': 1471, 'AAA': 6, 'ATC': 1, 'TAC': 1}, only mutation in aggregate... max freq alt codon is AAA\n",
      "On gene 59 in seq edge_1671.\n",
      "ATC, {'ATC': 1423, 'AAT': 2, 'ATT': 99, 'CTC': 1}, is mutation! max freq alt codon is ATT\n",
      "CCG, {'CCG': 1380, 'CCA': 145, 'CAG': 2}, is mutation! max freq alt codon is CCA\n",
      "ATA, {'ATA': 1491, 'ATC': 36, 'TAA': 1}, is mutation! max freq alt codon is ATC\n",
      "GGT, {'GGT': 1407, 'GGC': 106, 'TGT': 1}, is mutation! max freq alt codon is GGC\n",
      "GCA, {'GCA': 1417, 'GCG': 107, 'TCA': 1}, is mutation! max freq alt codon is GCG\n",
      "TTG, {'TTG': 1414, 'TTA': 105, 'GTA': 1}, is mutation! max freq alt codon is TTA\n",
      "CCA, {'CCA': 1405, 'CCG': 107, 'ACG': 1}, is mutation! max freq alt codon is CCG\n",
      "GTG, {'GTG': 1415, 'GTT': 107}, is mutation! max freq alt codon is GTT\n",
      "GAG, {'GAG': 1401, 'GAA': 107, 'GAT': 1}, is mutation! max freq alt codon is GAA\n",
      "TTC, {'TTC': 1482, 'TTT': 35}, is mutation! max freq alt codon is TTT\n",
      "CTG, {'CTG': 1380, 'TTG': 143, 'CCG': 1}, is mutation! max freq alt codon is TTG\n",
      "GAG, {'GAG': 1371, 'GAA': 139}, is mutation! max freq alt codon is GAA\n",
      "TAA, {'TAA': 1400, 'TTA': 107, 'AAA': 1, 'TAT': 2, 'TCA': 1}, is mutation! max freq alt codon is TTA\n",
      "On gene 60 in seq edge_1671.\n",
      "CTA, {'CTA': 1409, 'CTG': 103}, is mutation! max freq alt codon is CTG\n",
      "ATC, {'ATC': 1386, 'ATA': 140}, is mutation! max freq alt codon is ATA\n",
      "CTC, {'CTC': 1488, 'CTT': 34, 'CCC': 1}, is mutation! max freq alt codon is CTT\n",
      "On gene 61 in seq edge_1671.\n",
      "CGA, {'CGA': 1408, 'AGA': 99, 'GGA': 1}, is mutation! max freq alt codon is AGA\n",
      "ATC, {'ATC': 1459, 'ATT': 32, 'TTC': 2, 'AAT': 1}, is mutation! max freq alt codon is ATT\n",
      "CAG, {'CAG': 1358, 'CAA': 134, 'CTG': 1, 'CCG': 1}, is mutation! max freq alt codon is CAA\n",
      "GGC, {'GGC': 1467, 'GGT': 33}, is mutation! max freq alt codon is GGT\n",
      "On gene 62 in seq edge_1671.\n",
      "AAT, {'AAT': 1360, 'AAC': 136}, is mutation! max freq alt codon is AAC\n",
      "GGT, {'GGT': 1383, 'GGA': 106}, is mutation! max freq alt codon is GGA\n",
      "AAT, {'AAT': 1354, 'GAT': 138, 'ATT': 2}, is mutation! max freq alt codon is GAT\n",
      "ATC, {'ATC': 1408, 'ATT': 108}, is mutation! max freq alt codon is ATT\n",
      "GAT, {'GAT': 1370, 'GCT': 139, 'GAG': 1}, is mutation! max freq alt codon is GCT\n",
      "GCG, {'GCG': 1371, 'GCT': 140}, is mutation! max freq alt codon is GCT\n",
      "On gene 63 in seq edge_1671.\n",
      "GTA, {'GTA': 1380, 'ATA': 110, 'GGT': 1}, is mutation! max freq alt codon is ATA\n",
      "CCC, {'CCC': 1345, 'CCT': 138}, is mutation! max freq alt codon is CCT\n",
      "GGC, {'GGC': 1382, 'GGT': 137, 'GGA': 1, 'TGG': 1, 'CGG': 1, 'TGT': 1}, is mutation! max freq alt codon is GGT\n",
      "GCG, {'GCG': 1492, 'GCA': 27, 'CCG': 1}, is mutation! max freq alt codon is GCA\n",
      "On gene 64 in seq edge_1671.\n",
      "GGT, {'GGT': 1377, 'GAT': 143, 'TGT': 1, 'CGT': 1}, is mutation! max freq alt codon is GAT\n",
      "AGC, {'AGC': 1495, 'AGT': 30, 'AGG': 1}, is mutation! max freq alt codon is AGT\n",
      "GGC, {'GGC': 1498, 'TGC': 1, 'AGC': 28}, is mutation! max freq alt codon is AGC\n",
      "GGG, {'GGG': 1484, 'GGA': 28}, is mutation! max freq alt codon is GGA\n",
      "ATT, {'ATT': 1375, 'TTT': 140, 'AAT': 2}, is mutation! max freq alt codon is TTT\n",
      "On gene 65 in seq edge_1671.\n",
      "GAG, {'GAG': 1514, 'AAG': 9}, is mutation! max freq alt codon is AAG\n",
      "GGG, {'GGG': 1488, 'GGT': 30, 'TGG': 1}, is mutation! max freq alt codon is GGT\n",
      "GGA, {'GGA': 1484, 'GGG': 26, 'AGA': 1}, is mutation! max freq alt codon is GGG\n",
      "AGA, {'AGA': 1398, 'AAA': 107, 'ACA': 1}, is mutation! max freq alt codon is AAA\n",
      "On gene 66 in seq edge_1671.\n",
      "ACC, {'ACC': 1394, 'ACT': 108, 'TAC': 1, 'GAC': 1}, is mutation! max freq alt codon is ACT\n",
      "TTG, {'TTG': 1399, 'CTG': 111, 'TGT': 1, 'TGG': 1}, is mutation! max freq alt codon is CTG\n",
      "AAG, {'AAG': 1359, 'AAC': 142, 'TAG': 1}, is mutation! max freq alt codon is AAC\n",
      "AGC, {'AGC': 1390, 'AGT': 116}, is mutation! max freq alt codon is AGT\n",
      "GAA, {'GAA': 1385, 'GAG': 116, 'GCA': 1}, is mutation! max freq alt codon is GAG\n",
      "ATC, {'ATC': 1390, 'GTC': 115, 'TTC': 1}, is mutation! max freq alt codon is GTC\n",
      "ACG, {'ACG': 1484, 'AGG': 24}, is mutation! max freq alt codon is AGG\n",
      "AAT, {'AAT': 1478, 'GAT': 26, 'AAA': 1, 'TAT': 1, 'ACT': 1}, is mutation! max freq alt codon is GAT\n",
      "GGC, {'GGT': 143, 'GGC': 1361, 'GGG': 1}, is mutation! max freq alt codon is GGT\n",
      "CGT, {'CGT': 1487, 'CGC': 26}, is mutation! max freq alt codon is CGC\n",
      "Found contrary case.\n",
      "AAA, {'AAA': 1466, 'CAA': 6, 'TAA': 1, 'ACA': 1}, only mutation in aggregate... max freq alt codon is CAA\n",
      "GCC, {'ACC': 116, 'GCC': 1387}, is mutation! max freq alt codon is ACC\n",
      "GAA, {'GGA': 144, 'GAA': 1347, 'ATA': 1, 'AAA': 1, 'AGA': 2, 'GTA': 1, 'GAC': 1, 'GGG': 1}, is mutation! max freq alt codon is GGA\n",
      "TCA, {'TCA': 1480, 'TCG': 24}, is mutation! max freq alt codon is TCG\n",
      "AGT, {'AAT': 118, 'AGT': 1392, 'ACT': 1}, is mutation! max freq alt codon is AAT\n",
      "On gene 67 in seq edge_1671.\n",
      "GGA, {'GGG': 140, 'GGA': 1373, 'CGA': 1}, is mutation! max freq alt codon is GGG\n",
      "CCA, {'CCG': 120, 'CCA': 1398}, is mutation! max freq alt codon is CCG\n",
      "ACA, {'ACG': 146, 'ACA': 1355}, is mutation! max freq alt codon is ACG\n",
      "GTC, {'GTC': 1476, 'GTT': 23}, is mutation! max freq alt codon is GTT\n",
      "TTT, {'TTC': 147, 'TTT': 1351, 'CTT': 1}, is mutation! max freq alt codon is TTC\n",
      "AAG, {'AAG': 1476, 'AAA': 22, 'ATG': 1}, is mutation! max freq alt codon is AAA\n",
      "ACC, {'ACA': 145, 'ACC': 1356, 'AAC': 2}, is mutation! max freq alt codon is ACA\n",
      "CCC, {'TCC': 143, 'CCC': 1358, 'CCG': 1}, is mutation! max freq alt codon is TCC\n",
      "CCT, {'CCC': 143, 'CCT': 1345}, is mutation! max freq alt codon is CCC\n",
      "On gene 68 in seq edge_1671.\n",
      "GTT, {'ATT': 123, 'GTT': 1386, 'TTT': 1, 'AGT': 1}, is mutation! max freq alt codon is ATT\n",
      "GCA, {'GCG': 129, 'GCA': 1384}, is mutation! max freq alt codon is GCG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCA, {'GCA': 1473, 'TCA': 23}, is mutation! max freq alt codon is TCA\n",
      "GCT, {'GCT': 1492, 'GAT': 1, 'GCC': 22}, is mutation! max freq alt codon is GCC\n",
      "GGA, {'GGT': 151, 'GGA': 1368, 'AGA': 1, 'GTG': 1}, is mutation! max freq alt codon is GGT\n",
      "ACC, {'ACC': 1486, 'ACT': 19, 'AAC': 1}, is mutation! max freq alt codon is ACT\n",
      "GGA, {'GGA': 1372, 'GGT': 148, 'GGG': 1}, is mutation! max freq alt codon is GGT\n",
      "AAA, {'AAA': 1373, 'AAG': 147, 'AGA': 1, 'TAA': 1}, is mutation! max freq alt codon is AAG\n",
      "GCC, {'GCC': 1499, 'GCT': 19}, is mutation! max freq alt codon is GCT\n",
      "GGA, {'GGA': 1357, 'GGC': 149, 'GCA': 1}, is mutation! max freq alt codon is GGC\n",
      "ACT, {'ACT': 1359, 'ACA': 151, 'CCT': 1}, is mutation! max freq alt codon is ACA\n",
      "GGA, {'GGA': 1503, 'GGC': 21, 'GAA': 1, 'AGA': 1}, is mutation! max freq alt codon is GGC\n",
      "GCC, {'GCC': 1366, 'GCA': 155, 'GCG': 1}, is mutation! max freq alt codon is GCA\n",
      "On gene 69 in seq edge_1671.\n",
      "ATA, {'ATA': 1371, 'GTA': 126, 'CTA': 1, 'AAT': 1, 'TTA': 1, 'ATT': 1, 'AGA': 1}, is mutation! max freq alt codon is GTA\n",
      "TCA, {'TCA': 1499, 'TCG': 23}, is mutation! max freq alt codon is TCG\n",
      "On gene 70 in seq edge_1671.\n",
      "AAT, {'AAT': 1374, 'AAC': 148}, is mutation! max freq alt codon is AAC\n",
      "On gene 71 in seq edge_1671.\n",
      "CAA, {'CAA': 1494, 'CAT': 17}, is mutation! max freq alt codon is CAT\n",
      "GTA, {'GTA': 1500, 'ATA': 17}, is mutation! max freq alt codon is ATA\n",
      "On gene 72 in seq edge_1671.\n",
      "GAG, {'GAG': 1355, 'GAA': 143}, is mutation! max freq alt codon is GAA\n",
      "GAA, {'GAA': 1355, 'GAG': 122}, is mutation! max freq alt codon is GAG\n",
      "CCT, {'CCT': 1474, 'CCA': 18}, is mutation! max freq alt codon is CCA\n",
      "On gene 73 in seq edge_1671.\n",
      "ACT, {'ACT': 1495, 'ACC': 20}, is mutation! max freq alt codon is ACC\n",
      "CTA, {'CTA': 1483, 'TTA': 16, 'ACT': 1}, is mutation! max freq alt codon is TTA\n",
      "On gene 74 in seq edge_1671.\n",
      "AAT, {'AAT': 1508, 'ATG': 1, 'AAC': 21, 'CAT': 1}, is mutation! max freq alt codon is AAC\n",
      "GGA, {'GGA': 1506, 'GGG': 21, 'AGA': 1}, is mutation! max freq alt codon is GGG\n",
      "TTA, {'TTA': 1502, 'ATA': 20}, is mutation! max freq alt codon is ATA\n",
      "TTC, {'TTC': 1497, 'TTT': 21, 'TCC': 1}, is mutation! max freq alt codon is TTT\n",
      "CCC, {'CCC': 1472, 'CCT': 19, 'TCC': 1, 'ACC': 1, 'CTC': 1}, is mutation! max freq alt codon is CCT\n",
      "AGT, {'AGT': 1381, 'AGC': 145}, is mutation! max freq alt codon is AGC\n",
      "GGG, {'GGG': 1368, 'GGT': 148, 'GGA': 2}, is mutation! max freq alt codon is GGT\n",
      "GTT, {'GTT': 1505, 'GTG': 18}, is mutation! max freq alt codon is GTG\n",
      "On gene 75 in seq edge_1671.\n",
      "ATT, {'ATT': 1404, 'GTT': 130}, is mutation! max freq alt codon is GTT\n",
      "GAA, {'GAA': 1386, 'GGA': 2, 'GAG': 149}, is mutation! max freq alt codon is GAG\n",
      "AAC, {'AAC': 1384, 'AAT': 130, 'AAA': 1, 'CAC': 1}, is mutation! max freq alt codon is AAT\n",
      "GGA, {'GGA': 1516, 'GAA': 22}, is mutation! max freq alt codon is GAA\n",
      "On gene 76 in seq edge_1671.\n",
      "GAG, {'GAG': 1509, 'GAA': 20, 'TAG': 1, 'GCG': 1, 'AGG': 1}, is mutation! max freq alt codon is GAA\n",
      "On gene 77 in seq edge_1671.\n",
      "CTT, {'CTT': 1504, 'ATT': 2, 'CAT': 2, 'GTT': 19}, is mutation! max freq alt codon is GTT\n",
      "AAT, {'AAT': 1374, 'AAC': 152, 'ATG': 1, 'AGC': 1, 'CAT': 1}, is mutation! max freq alt codon is AAC\n",
      "GAT, {'GAT': 1371, 'GAC': 153}, is mutation! max freq alt codon is GAC\n",
      "Found contrary case.\n",
      "ACA, {'ACA': 1449, 'ACT': 1, 'TAC': 3, 'CAA': 1, 'AAA': 1, 'AAC': 1, 'ACC': 1}, only mutation in aggregate... max freq alt codon is TAC\n",
      "On gene 78 in seq edge_1671.\n",
      "GTT, {'GTT': 1357, 'GCT': 149}, is mutation! max freq alt codon is GCT\n",
      "GCA, {'GCA': 1485, 'GCT': 20}, is mutation! max freq alt codon is GCT\n",
      "CAA, {'CAA': 1350, 'CAG': 147, 'CAC': 1}, is mutation! max freq alt codon is CAG\n",
      "ATG, {'ATG': 1348, 'GTG': 147, 'ATT': 1, 'GTA': 1}, is mutation! max freq alt codon is GTG\n",
      "AAG, {'AAG': 1353, 'AAA': 145}, is mutation! max freq alt codon is AAA\n",
      "Found contrary case.\n",
      "GAA, {'GAA': 1430, 'AAA': 5, 'GGA': 2, 'AGA': 1}, only mutation in aggregate... max freq alt codon is AAA\n",
      "Found contrary case.\n",
      "ATG, {'ATG': 1470, 'AAG': 3, 'AAT': 2, 'GTG': 2, 'TGG': 1}, only mutation in aggregate... max freq alt codon is AAG\n",
      "GAT, {'GAT': 1357, 'GAA': 146}, is mutation! max freq alt codon is GAA\n",
      "On gene 79 in seq edge_1671.\n",
      "GGT, {'GGT': 1479, 'GGA': 22}, is mutation! max freq alt codon is GGA\n",
      "GTA, {'GTA': 1478, 'GTC': 23}, is mutation! max freq alt codon is GTC\n",
      "ATA, {'ATA': 1479, 'ATT': 21, 'CTA': 1}, is mutation! max freq alt codon is ATT\n",
      "ATA, {'ATA': 1474, 'ATT': 22, 'AGG': 1}, is mutation! max freq alt codon is ATT\n",
      "GGG, {'GGG': 1461, 'GGT': 22, 'CGG': 1, 'GGC': 2, 'GTG': 1}, is mutation! max freq alt codon is GGT\n",
      "GTA, {'GTA': 1472, 'GTG': 22, 'GAA': 2}, is mutation! max freq alt codon is GTG\n",
      "CCT, {'CCT': 1469, 'CCA': 22}, is mutation! max freq alt codon is CCA\n",
      "TTG, {'TTG': 1471, 'CTG': 22, 'TGC': 1, 'ATG': 1, 'TGG': 1}, is mutation! max freq alt codon is CTG\n",
      "ACT, {'ACT': 1476, 'ACA': 21}, is mutation! max freq alt codon is ACA\n",
      "CTC, {'CTC': 1476, 'CTA': 21}, is mutation! max freq alt codon is CTA\n",
      "GCT, {'GCT': 1475, 'GCA': 21}, is mutation! max freq alt codon is GCA\n",
      "CGC, {'CGC': 1477, 'CGA': 21}, is mutation! max freq alt codon is CGA\n",
      "CCG, {'CCG': 1436, 'CCA': 21, 'CGG': 2, 'ACC': 1}, is mutation! max freq alt codon is CCA\n",
      "CCC, {'CCC': 1464, 'CCT': 20}, is mutation! max freq alt codon is CCT\n",
      "GTT, {'GTT': 1465, 'CGT': 1, 'GTA': 21, 'TTT': 1, 'GTG': 1}, is mutation! max freq alt codon is GTA\n",
      "CAA, {'CAA': 1351, 'CAG': 21, 'CCA': 2, 'CAC': 1}, is mutation! max freq alt codon is CAG\n",
      "On gene 80 in seq edge_1671.\n",
      "TAC, {'TAC': 1464, 'TAT': 21}, is mutation! max freq alt codon is TAT\n",
      "TTT, {'TTT': 1328, 'CTT': 147, 'CTG': 1}, is mutation! max freq alt codon is CTT\n",
      "GCA, {'GCA': 1357, 'GCT': 125}, is mutation! max freq alt codon is GCT\n",
      "TCG, {'TCG': 1320, 'TCA': 142}, is mutation! max freq alt codon is TCA\n",
      "CAT, {'CAT': 1348, 'CAC': 124}, is mutation! max freq alt codon is CAC\n",
      "GGA, {'GGA': 1427, 'GGG': 22, 'GAC': 1, 'GGC': 1}, is mutation! max freq alt codon is GGG\n",
      "AGC, {'AGC': 1449, 'AGT': 22}, is mutation! max freq alt codon is AGT\n",
      "GCC, {'GCC': 1445, 'GCA': 22}, is mutation! max freq alt codon is GCA\n",
      "GTA, {'GTA': 1345, 'GTT': 119, 'TTA': 1}, is mutation! max freq alt codon is GTT\n",
      "TTG, {'TTG': 1335, 'TTA': 121, 'TAG': 1}, is mutation! max freq alt codon is TTA\n",
      "GGC, {'GGC': 1339, 'GGT': 122, 'TGC': 1}, is mutation! max freq alt codon is GGT\n",
      "GTC, {'GTC': 1351, 'GTT': 121}, is mutation! max freq alt codon is GTT\n",
      "GTA, {'GTA': 1352, 'GTG': 121}, is mutation! max freq alt codon is GTG\n",
      "On gene 81 in seq edge_1671.\n",
      "CGC, {'CGC': 1432, 'CGT': 22}, is mutation! max freq alt codon is CGT\n",
      "GTT, {'GTT': 1339, 'GTA': 112, 'TGT': 1}, is mutation! max freq alt codon is GTA\n",
      "GGA, {'GGA': 1334, 'GGG': 109}, is mutation! max freq alt codon is GGG\n",
      "ATC, {'ATC': 1332, 'ATA': 114}, is mutation! max freq alt codon is ATA\n",
      "GAA, {'GAA': 1332, 'AAA': 113}, is mutation! max freq alt codon is AAA\n",
      "AAG, {'AAG': 1418, 'AAA': 23}, is mutation! max freq alt codon is AAA\n",
      "On gene 82 in seq edge_1671.\n",
      "ATT, {'ATT': 1306, 'ATA': 140}, is mutation! max freq alt codon is ATA\n",
      "GGT, {'GGT': 1362, 'GGC': 60, 'GGG': 1}, is mutation! max freq alt codon is GGC\n",
      "GAC, {'GAC': 1390, 'GAT': 49, 'AAC': 1}, is mutation! max freq alt codon is GAT\n",
      "ATC, {'ATC': 1407, 'ATT': 50, 'CTC': 1}, is mutation! max freq alt codon is ATT\n",
      "GGG, {'GGG': 1339, 'GGA': 106, 'AGG': 7}, is mutation! max freq alt codon is GGA\n",
      "On gene 83 in seq edge_1671.\n",
      "ATA, {'ATA': 1376, 'TTA': 58}, is mutation! max freq alt codon is TTA\n",
      "CAC, {'CAC': 1388, 'TAC': 49}, is mutation! max freq alt codon is TAC\n",
      "ACG, {'ACG': 1310, 'ACA': 130}, is mutation! max freq alt codon is ACA\n",
      "GGT, {'GGT': 1406, 'AGT': 23, 'GGA': 1, 'TGT': 1}, is mutation! max freq alt codon is AGT\n",
      "GCA, {'GCA': 1389, 'ACA': 41}, is mutation! max freq alt codon is ACA\n",
      "GCG, {'GCG': 1376, 'ACG': 46, 'GGC': 1}, is mutation! max freq alt codon is ACG\n",
      "ACA, {'ACA': 1296, 'GCA': 128}, is mutation! max freq alt codon is GCA\n",
      "GAG, {'GAG': 1311, 'GCG': 126}, is mutation! max freq alt codon is GCG\n",
      "On gene 84 in seq edge_1671.\n",
      "TTG, {'TTG': 1367, 'TTA': 45}, is mutation! max freq alt codon is TTA\n",
      "TTG, {'TTG': 1349, 'CTG': 60, 'TTT': 1}, is mutation! max freq alt codon is CTG\n",
      "GAA, {'GAA': 1381, 'GGA': 1, 'AAA': 24}, is mutation! max freq alt codon is AAA\n",
      "TGT, {'TGT': 1272, 'TGC': 126}, is mutation! max freq alt codon is TGC\n",
      "On gene 85 in seq edge_1671.\n",
      "AGT, {'AGT': 1342, 'AGC': 69, 'AGG': 2}, is mutation! max freq alt codon is AGC\n",
      "GCA, {'GCA': 1272, 'GTA': 129}, is mutation! max freq alt codon is GTA\n",
      "CCA, {'CCA': 1351, 'CCT': 59, 'ACA': 1}, is mutation! max freq alt codon is CCT\n",
      "AAG, {'AAG': 1353, 'AAA': 57}, is mutation! max freq alt codon is AAA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACA, {'ACA': 1368, 'ACC': 41, 'CCA': 1, 'AAA': 1}, is mutation! max freq alt codon is ACC\n",
      "GTG, {'GTG': 1280, 'TTG': 40, 'CTG': 80, 'GGG': 1, 'GAG': 1}, is mutation! max freq alt codon is CTG\n",
      "TTG, {'TTG': 1364, 'TTA': 41, 'GTG': 1, 'TAG': 1}, is mutation! max freq alt codon is TTA\n",
      "AAG, {'AAG': 1363, 'AAA': 41, 'CAA': 1}, is mutation! max freq alt codon is AAA\n",
      "AGA, {'AGA': 1369, 'AGT': 42}, is mutation! max freq alt codon is AGT\n",
      "ATA, {'ATA': 1366, 'ATT': 41}, is mutation! max freq alt codon is ATT\n",
      "CGC, {'CGC': 1362, 'TGC': 41, 'GCT': 1}, is mutation! max freq alt codon is TGC\n",
      "AGT, {'AGT': 1362, 'ACA': 41}, is mutation! max freq alt codon is ACA\n",
      "CGT, {'CGT': 1353, 'TGT': 53}, is mutation! max freq alt codon is TGT\n",
      "CAC, {'CAC': 1285, 'CAT': 122}, is mutation! max freq alt codon is CAT\n",
      "AAG, {'AAG': 1272, 'GAG': 121, 'AGT': 1}, is mutation! max freq alt codon is GAG\n",
      "AAG, {'AAG': 1297, 'AAA': 54}, is mutation! max freq alt codon is AAA\n",
      "AAA, {'AAA': 1345, 'AAG': 55}, is mutation! max freq alt codon is AAG\n",
      "GAC, {'GAC': 1358, 'CAC': 1, 'AAC': 26}, is mutation! max freq alt codon is AAC\n",
      "GGC, {'GGC': 1356, 'GGA': 25}, is mutation! max freq alt codon is GGA\n",
      "GTT, {'GTT': 1356, 'GTG': 24, 'GTA': 1}, is mutation! max freq alt codon is GTG\n",
      "GCG, {'GCG': 1332, 'GCA': 51, 'GGC': 1}, is mutation! max freq alt codon is GCA\n",
      "ATT, {'ATT': 1327, 'ATC': 47, 'AGT': 1}, is mutation! max freq alt codon is ATC\n",
      "On gene 86 in seq edge_1671.\n",
      "TAA, {'TAA': 1334, 'TCA': 1, 'TTA': 32}, is mutation! max freq alt codon is TTA\n",
      "GAG, {'GAG': 1278, 'GAT': 38}, is mutation! max freq alt codon is GAT\n",
      "ATT, {'ATT': 1253, 'GTT': 37, 'AAT': 1, 'CAT': 1, 'CTT': 1}, is mutation! max freq alt codon is GTT\n",
      "GTG, {'GTG': 1280, 'GTA': 38}, is mutation! max freq alt codon is GTA\n",
      "GGT, {'GGT': 1279, 'GGA': 38}, is mutation! max freq alt codon is GGA\n",
      "TCC, {'TCC': 1276, 'GCC': 38}, is mutation! max freq alt codon is GCC\n",
      "AAC, {'AAC': 1269, 'AGC': 37}, is mutation! max freq alt codon is AGC\n",
      "TCA, {'TCA': 1273, 'ACA': 37}, is mutation! max freq alt codon is ACA\n",
      "TCT, {'TCT': 1256, 'CCT': 36}, is mutation! max freq alt codon is CCT\n",
      "On gene 87 in seq edge_1671.\n",
      "GAG, {'GAG': 1280, 'GGA': 36}, is mutation! max freq alt codon is GGA\n",
      "ATT, {'ATT': 1279, 'GTT': 37}, is mutation! max freq alt codon is GTT\n",
      "ACC, {'ACC': 1276, 'ACG': 37, 'ACA': 1, 'AAC': 2}, is mutation! max freq alt codon is ACG\n",
      "On gene 88 in seq edge_1671.\n",
      "AAG, {'AAG': 1291, 'AAA': 39}, is mutation! max freq alt codon is AAA\n",
      "AAC, {'AAC': 1297, 'AAT': 40, 'AAA': 1}, is mutation! max freq alt codon is AAT\n",
      "On gene 89 in seq edge_1671.\n",
      "On gene 90 in seq edge_1671.\n",
      "Found contrary case.\n",
      "CTT, {'CTT': 1386, 'TTT': 1, 'CCT': 1, 'ACT': 4, 'CTG': 1}, only mutation in aggregate... max freq alt codon is ACT\n",
      "ACG, {'ACG': 1331, 'ACA': 43, 'CAG': 1, 'ATG': 41}, is mutation! max freq alt codon is ACA\n",
      "GTC, {'GTC': 1381, 'GTT': 41, 'GGC': 1}, is mutation! max freq alt codon is GTT\n",
      "AAA, {'AAA': 1306, 'CAA': 106, 'TAA': 1}, is mutation! max freq alt codon is CAA\n",
      "GCC, {'GCC': 1377, 'GCT': 41, 'GCA': 1, 'GCG': 1}, is mutation! max freq alt codon is GCT\n",
      "GCT, {'GCT': 1382, 'GCC': 42}, is mutation! max freq alt codon is GCC\n",
      "ACA, {'ACA': 1316, 'GCA': 105}, is mutation! max freq alt codon is GCA\n",
      "GCA, {'GCA': 1369, 'GCG': 40}, is mutation! max freq alt codon is GCG\n",
      "GCC, {'GCC': 1386, 'GGC': 1, 'GCT': 25}, is mutation! max freq alt codon is GCT\n",
      "TTC, {'TTC': 1368, 'TTT': 38}, is mutation! max freq alt codon is TTT\n",
      "GCG, {'GCG': 1367, 'GTG': 40, 'GGC': 1}, is mutation! max freq alt codon is GTG\n",
      "GCC, {'GCC': 1340, 'TCT': 62, 'GCA': 1}, is mutation! max freq alt codon is TCT\n",
      "ACT, {'ACT': 1343, 'ACA': 62}, is mutation! max freq alt codon is ACA\n",
      "ACC, {'ACC': 1340, 'ACT': 62}, is mutation! max freq alt codon is ACT\n",
      "AAC, {'AAC': 1313, 'AAT': 62}, is mutation! max freq alt codon is AAT\n",
      "CCC, {'CCC': 1344, 'CCT': 62, 'ACC': 1}, is mutation! max freq alt codon is CCT\n",
      "GAC, {'GAC': 1335, 'GAT': 62, 'CGC': 1}, is mutation! max freq alt codon is GAT\n",
      "AAC, {'AAC': 1340, 'AAT': 59, 'ACT': 1}, is mutation! max freq alt codon is AAT\n",
      "ACT, {'ACT': 1369, 'ACC': 40, 'ACG': 1, 'CCT': 1}, is mutation! max freq alt codon is ACC\n",
      "GGT, {'GGT': 1365, 'GGC': 40}, is mutation! max freq alt codon is GGC\n",
      "ACT, {'ACT': 1365, 'ACC': 38, 'CCT': 1}, is mutation! max freq alt codon is ACC\n",
      "GTT, {'GTT': 1278, 'GTC': 109, 'TTT': 1}, is mutation! max freq alt codon is GTC\n",
      "GAT, {'GAT': 1316, 'GAC': 112}, is mutation! max freq alt codon is GAC\n",
      "CCA, {'CCA': 1378, 'CCC': 40, 'CAA': 2}, is mutation! max freq alt codon is CCC\n",
      "GAC, {'GAC': 1393, 'GAT': 41}, is mutation! max freq alt codon is GAT\n",
      "GTC, {'GTC': 1353, 'GTA': 91}, is mutation! max freq alt codon is GTA\n",
      "TAT, {'TAT': 1322, 'TAC': 113}, is mutation! max freq alt codon is TAC\n",
      "TCA, {'TCA': 1392, 'GCA': 49}, is mutation! max freq alt codon is GCA\n",
      "GCT, {'GCA': 41, 'GCT': 1399}, is mutation! max freq alt codon is GCA\n",
      "AAA, {'AAA': 1337, 'GAA': 111, 'CAA': 1}, is mutation! max freq alt codon is GAA\n",
      "AGC, {'AGC': 1337, 'AGT': 108}, is mutation! max freq alt codon is AGT\n",
      "On gene 91 in seq edge_1671.\n",
      "CCG, {'CCG': 1418, 'GCG': 1, 'TCG': 49}, is mutation! max freq alt codon is TCG\n",
      "CCG, {'CCG': 1439, 'CCT': 43}, is mutation! max freq alt codon is CCT\n",
      "CGT, {'CGT': 1456, 'TGT': 19}, is mutation! max freq alt codon is TGT\n",
      "On gene 92 in seq edge_1671.\n",
      "GAT, {'GAT': 1372, 'GAC': 92}, is mutation! max freq alt codon is GAC\n",
      "ACA, {'ACA': 1380, 'ACT': 90}, is mutation! max freq alt codon is ACT\n",
      "ATA, {'ATA': 1364, 'ATC': 110}, is mutation! max freq alt codon is ATC\n",
      "GCT, {'GCT': 1387, 'GCA': 90}, is mutation! max freq alt codon is GCA\n",
      "AAA, {'AAA': 1357, 'AGA': 111, 'AAT': 1, 'AAG': 1}, is mutation! max freq alt codon is AGA\n",
      "GAC, {'GAC': 1461, 'GAG': 21}, is mutation! max freq alt codon is GAG\n",
      "On gene 93 in seq edge_1671.\n",
      "ACT, {'ACT': 1385, 'TCT': 117}, is mutation! max freq alt codon is TCT\n",
      "GAT, {'GAT': 1404, 'AAT': 95}, is mutation! max freq alt codon is AAT\n",
      "AAC, {'AAC': 1471, 'ACC': 1, 'AGC': 17}, is mutation! max freq alt codon is AGC\n",
      "On gene 94 in seq edge_1671.\n",
      "On gene 95 in seq edge_1671.\n",
      "CTT, {'CTT': 1381, 'TTA': 97, 'TTT': 1, 'ATT': 1}, is mutation! max freq alt codon is TTA\n",
      "CCA, {'CCA': 1471, 'CCG': 17}, is mutation! max freq alt codon is CCG\n",
      "ATT, {'ATT': 1387, 'GTT': 117, 'ATG': 1, 'TTT': 1}, is mutation! max freq alt codon is GTT\n",
      "GCG, {'GCG': 1404, 'ACG': 99}, is mutation! max freq alt codon is ACG\n",
      "CCG, {'CCG': 1420, 'CCT': 19, 'CCC': 1}, is mutation! max freq alt codon is CCT\n",
      "GGA, {'GGA': 1460, 'TGA': 2, 'GTA': 1, 'GGG': 19, 'GAA': 1}, is mutation! max freq alt codon is GGG\n",
      "CGG, {'CGG': 1464, 'CGA': 18}, is mutation! max freq alt codon is CGA\n",
      "ACC, {'ACC': 1375, 'GCC': 113, 'CAC': 1, 'CCC': 1}, is mutation! max freq alt codon is GCC\n",
      "ATA, {'ATA': 1389, 'ATT': 114}, is mutation! max freq alt codon is ATT\n",
      "GAC, {'GAC': 1402, 'GAT': 96, 'ACC': 1}, is mutation! max freq alt codon is GAT\n",
      "ATA, {'ATA': 1406, 'GTA': 96}, is mutation! max freq alt codon is GTA\n",
      "GAA, {'GAA': 1396, 'GGG': 96, 'AAA': 1}, is mutation! max freq alt codon is GGG\n",
      "GTA, {'GTA': 1404, 'GCA': 96, 'GAA': 1}, is mutation! max freq alt codon is GCA\n",
      "TTT, {'TTT': 1411, 'TTC': 94, 'GTC': 1}, is mutation! max freq alt codon is TTC\n",
      "GTT, {'GTT': 1402, 'GTC': 94}, is mutation! max freq alt codon is GTC\n",
      "GTC, {'GTC': 1481, 'ATC': 17}, is mutation! max freq alt codon is ATC\n",
      "TCA, {'TCA': 1394, 'TCT': 94, 'TCC': 1}, is mutation! max freq alt codon is TCT\n",
      "GAT, {'GAT': 1386, 'GAC': 112, 'AAT': 1}, is mutation! max freq alt codon is GAC\n",
      "ATC, {'ATC': 1389, 'ATT': 111}, is mutation! max freq alt codon is ATT\n",
      "GGT, {'GGT': 1476, 'GGC': 18}, is mutation! max freq alt codon is GGC\n",
      "GTG, {'GTG': 1392, 'GCG': 1, 'GTA': 93}, is mutation! max freq alt codon is GTA\n",
      "TTT, {'TTT': 1358, 'TCT': 107, 'ATT': 1}, is mutation! max freq alt codon is TCT\n",
      "GGA, {'GGA': 1398, 'GAA': 1, 'GGG': 92}, is mutation! max freq alt codon is GGG\n",
      "ACT, {'ACT': 1467, 'AAC': 2, 'ACC': 17, 'ACA': 1}, is mutation! max freq alt codon is ACC\n",
      "TTG, {'TTG': 1348, 'GTG': 1, 'CTG': 108, 'ATT': 1}, is mutation! max freq alt codon is CTG\n",
      "CGT, {'CGT': 1484, 'CGA': 17}, is mutation! max freq alt codon is CGA\n",
      "CGT, {'CGT': 1410, 'CGC': 94}, is mutation! max freq alt codon is CGC\n",
      "ACA, {'ACA': 1447, 'ACT': 20, 'ACC': 2}, is mutation! max freq alt codon is ACT\n",
      "AAA, {'AAA': 1468, 'AAG': 19}, is mutation! max freq alt codon is AAG\n",
      "GGG, {'GGG': 1362, 'GGA': 111, 'TGG': 1}, is mutation! max freq alt codon is GGA\n",
      "TTG, {'TTG': 1361, 'TTA': 111, 'ATA': 1}, is mutation! max freq alt codon is TTA\n",
      "GAC, {'GAC': 1364, 'GAT': 108, 'GCC': 1}, is mutation! max freq alt codon is GAT\n",
      "CGT, {'CGT': 1371, 'CGC': 112}, is mutation! max freq alt codon is CGC\n",
      "GGC, {'GGC': 1364, 'GGA': 111, 'TGC': 1, 'GAG': 1}, is mutation! max freq alt codon is GGA\n",
      "GTA, {'GTA': 1463, 'GTC': 19, 'TAA': 1}, is mutation! max freq alt codon is GTC\n",
      "GTT, {'GTT': 1361, 'GTG': 113, 'GTA': 1, 'GGT': 1}, is mutation! max freq alt codon is GTG\n",
      "GGC, {'GGC': 1363, 'GGT': 112, 'TGC': 2, 'GCA': 1}, is mutation! max freq alt codon is GGT\n",
      "GGT, {'GGT': 1386, 'GGA': 92, 'TGA': 1}, is mutation! max freq alt codon is GGA\n",
      "GGC, {'GGC': 1365, 'GGA': 111, 'GGG': 1}, is mutation! max freq alt codon is GGA\n",
      "ATT, {'ATT': 1375, 'ATC': 110}, is mutation! max freq alt codon is ATC\n",
      "ACA, {'ACA': 1371, 'ACT': 106, 'ACC': 1, 'GCA': 1}, is mutation! max freq alt codon is ACT\n",
      "TCG, {'TCG': 1372, 'TCA': 105}, is mutation! max freq alt codon is TCA\n",
      "CAA, {'CAA': 1372, 'CAG': 106}, is mutation! max freq alt codon is CAG\n",
      "GTT, {'GTT': 1368, 'GCA': 106}, is mutation! max freq alt codon is GCA\n",
      "GAG, {'GAG': 1365, 'GAC': 106, 'TAG': 1}, is mutation! max freq alt codon is GAC\n",
      "CCA, {'CCA': 1363, 'CCT': 106, 'CCC': 1}, is mutation! max freq alt codon is CCT\n",
      "ACA, {'ACA': 1373, 'ACC': 1, 'AAT': 105}, is mutation! max freq alt codon is AAT\n",
      "TTG, {'TTG': 1372, 'CTG': 106}, is mutation! max freq alt codon is CTG\n",
      "TTA, {'TTA': 1372, 'CTT': 104, 'CTA': 1}, is mutation! max freq alt codon is CTT\n",
      "AGT, {'AGT': 1371, 'AAC': 104}, is mutation! max freq alt codon is AAC\n",
      "GAT, {'GAT': 1367, 'GAC': 106, 'GAA': 1}, is mutation! max freq alt codon is GAC\n",
      "TAT, {'TAT': 1372, 'TAC': 106}, is mutation! max freq alt codon is TAC\n",
      "AAT, {'AAT': 1375, 'AAC': 106}, is mutation! max freq alt codon is AAC\n",
      "ACT, {'ACT': 1374, 'ACA': 106}, is mutation! max freq alt codon is ACA\n",
      "CAG, {'CAG': 1372, 'CAA': 105}, is mutation! max freq alt codon is CAA\n",
      "CTT, {'CTT': 1374, 'CTC': 106}, is mutation! max freq alt codon is CTC\n",
      "GTT, {'GTT': 1371, 'ATC': 102, 'ATT': 3}, is mutation! max freq alt codon is ATC\n",
      "AGG, {'AGG': 1374, 'CGT': 106}, is mutation! max freq alt codon is CGT\n",
      "TAT, {'TAT': 1374, 'TAC': 105}, is mutation! max freq alt codon is TAC\n",
      "CCA, {'CCA': 1378, 'CCC': 104}, is mutation! max freq alt codon is CCC\n",
      "AGA, {'AGA': 1375, 'AGG': 105}, is mutation! max freq alt codon is AGG\n",
      "ATA, {'ATA': 1341, 'ATT': 106}, is mutation! max freq alt codon is ATT\n",
      "ATA, {'ATA': 1368, 'GTT': 102, 'AGC': 1, 'CTA': 1, 'TTT': 1, 'ATC': 1}, is mutation! max freq alt codon is GTT\n",
      "TAT, {'TAT': 1368, 'TTT': 106}, is mutation! max freq alt codon is TTT\n",
      "ATA, {'ATA': 1335, 'GTT': 106, 'TAG': 1}, is mutation! max freq alt codon is GTT\n",
      "GTA, {'GTA': 1376, 'GAG': 105}, is mutation! max freq alt codon is GAG\n",
      "AGA, {'AGA': 1377, 'AGC': 106}, is mutation! max freq alt codon is AGC\n",
      "CAA, {'CAA': 1375, 'CAC': 2, 'AAA': 105}, is mutation! max freq alt codon is AAA\n",
      "AAA, {'AAA': 1469, 'AAG': 18}, is mutation! max freq alt codon is AAG\n",
      "TAT, {'TAT': 1467, 'TAC': 18}, is mutation! max freq alt codon is TAC\n",
      "GAC, {'GAC': 1465, 'GAT': 18}, is mutation! max freq alt codon is GAT\n",
      "CTA, {'CTA': 1469, 'TTA': 18}, is mutation! max freq alt codon is TTA\n",
      "GCA, {'GCA': 1458, 'AAT': 18}, is mutation! max freq alt codon is AAT\n",
      "GGT, {'GGT': 1459, 'GGA': 18}, is mutation! max freq alt codon is GGA\n",
      "CCC, {'CCC': 1451, 'CCT': 18}, is mutation! max freq alt codon is CCT\n",
      "GTT, {'GTT': 1452, 'ATA': 18, 'GAT': 1}, is mutation! max freq alt codon is ATA\n",
      "ATT, {'ATT': 1460, 'ATA': 18, 'TTT': 1}, is mutation! max freq alt codon is ATA\n",
      "GTA, {'GTA': 1461, 'GTT': 18}, is mutation! max freq alt codon is GTT\n",
      "TAT, {'TAT': 1458, 'TTA': 2, 'TTT': 18}, is mutation! max freq alt codon is TTT\n",
      "AAA, {'AAA': 1460, 'AAG': 18, 'AAT': 1}, is mutation! max freq alt codon is AAG\n",
      "TTA, {'TTA': 1459, 'TTC': 18}, is mutation! max freq alt codon is TTC\n",
      "On gene 96 in seq edge_1671.\n",
      "On gene 97 in seq edge_1671.\n",
      "GGT, {'GGT': 1363, 'GTT': 1, 'GGA': 71}, is mutation! max freq alt codon is GGA\n",
      "GCT, {'GCT': 1377, 'TCT': 70, 'ACT': 1}, is mutation! max freq alt codon is TCT\n",
      "ATT, {'ATT': 1374, 'ATC': 71}, is mutation! max freq alt codon is ATC\n",
      "GAA, {'GAA': 1371, 'AAA': 1, 'AGA': 1, 'GAG': 69}, is mutation! max freq alt codon is GAG\n",
      "GAA, {'GAA': 1376, 'GAT': 71}, is mutation! max freq alt codon is GAT\n",
      "AAG, {'AAG': 1364, 'CAG': 1, 'AAA': 71}, is mutation! max freq alt codon is AAA\n",
      "ACT, {'ACT': 1374, 'ACA': 71}, is mutation! max freq alt codon is ACA\n",
      "GGT, {'GGT': 1366, 'GTT': 1, 'GTA': 1, 'GGA': 71}, is mutation! max freq alt codon is GGA\n",
      "ACA, {'ACA': 1351, 'ACC': 72, 'AAA': 1}, is mutation! max freq alt codon is ACC\n",
      "GAC, {'GAC': 1373, 'GCC': 1, 'GAT': 72}, is mutation! max freq alt codon is GAT\n",
      "ATA, {'ATA': 1388, 'ATT': 50}, is mutation! max freq alt codon is ATT\n",
      "GGG, {'GGG': 1387, 'TGG': 1, 'GGA': 49}, is mutation! max freq alt codon is GGA\n",
      "AAA, {'AAA': 1384, 'AAG': 49, 'TAA': 1}, is mutation! max freq alt codon is AAG\n",
      "AGC, {'AGC': 1388, 'AGT': 49}, is mutation! max freq alt codon is AGT\n",
      "GTA, {'GTA': 1390, 'GTT': 49}, is mutation! max freq alt codon is GTT\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATA, {'ATA': 1369, 'ATT': 70}, is mutation! max freq alt codon is ATT\n",
      "TTT, {'TTT': 1364, 'TTC': 71}, is mutation! max freq alt codon is TTC\n",
      "ATT, {'ATT': 1366, 'GTT': 71}, is mutation! max freq alt codon is GTT\n",
      "TAC, {'TAC': 1368, 'TAT': 71}, is mutation! max freq alt codon is TAT\n",
      "GTT, {'GTT': 1361, 'GAT': 1, 'GTA': 75}, is mutation! max freq alt codon is GTA\n",
      "GGA, {'GGA': 1362, 'GGG': 73, 'GGC': 1}, is mutation! max freq alt codon is GGG\n",
      "ACT, {'ACT': 1370, 'ACA': 74}, is mutation! max freq alt codon is ACA\n",
      "ACT, {'ACT': 1370, 'ACA': 73}, is mutation! max freq alt codon is ACA\n",
      "ATT, {'ATT': 1366, 'AAT': 1, 'ATA': 73}, is mutation! max freq alt codon is ATA\n",
      "CCT, {'CCT': 1354, 'CCC': 75}, is mutation! max freq alt codon is CCC\n",
      "ACT, {'ACT': 1366, 'ACA': 74}, is mutation! max freq alt codon is ACA\n",
      "GCT, {'GCT': 1364, 'CTT': 1, 'GCC': 73}, is mutation! max freq alt codon is GCC\n",
      "ACT, {'ACT': 1369, 'ACC': 74}, is mutation! max freq alt codon is ACC\n",
      "GGT, {'GGT': 1363, 'GGG': 1, 'GGA': 74}, is mutation! max freq alt codon is GGA\n",
      "CCC, {'CCC': 1353, 'CCG': 1, 'CCT': 74}, is mutation! max freq alt codon is CCT\n",
      "GCT, {'GCT': 1363, 'CCT': 2, 'GTA': 51, 'GCA': 23}, is mutation! max freq alt codon is GTA\n",
      "TTA, {'TTA': 1362, 'ATA': 1, 'TGA': 1, 'TTG': 74}, is mutation! max freq alt codon is TTG\n",
      "GCT, {'GCT': 1363, 'GCA': 75}, is mutation! max freq alt codon is GCA\n",
      "GAT, {'GAT': 1361, 'GAC': 76}, is mutation! max freq alt codon is GAC\n",
      "GGT, {'GGT': 1357, 'GGA': 75}, is mutation! max freq alt codon is GGA\n",
      "GCT, {'GCT': 1360, 'GCA': 75}, is mutation! max freq alt codon is GCA\n",
      "ACT, {'ACT': 1357, 'ACA': 76}, is mutation! max freq alt codon is ACA\n",
      "CTA, {'CTA': 1355, 'TTG': 76}, is mutation! max freq alt codon is TTG\n",
      "GGA, {'GGA': 1354, 'GGT': 75}, is mutation! max freq alt codon is GGT\n",
      "TTG, {'TTG': 1358, 'CTT': 76}, is mutation! max freq alt codon is CTT\n",
      "CTT, {'CTT': 1361, 'CTC': 76}, is mutation! max freq alt codon is CTC\n",
      "ACT, {'ACT': 1355, 'CCT': 1, 'ACA': 76}, is mutation! max freq alt codon is ACA\n",
      "GCT, {'GCT': 1355, 'GCA': 76}, is mutation! max freq alt codon is GCA\n",
      "TAT, {'TAT': 1358, 'TAC': 75}, is mutation! max freq alt codon is TAC\n",
      "GCT, {'GCT': 1361, 'GCA': 74}, is mutation! max freq alt codon is GCA\n",
      "GGT, {'GGT': 1346, 'CGT': 1, 'GGA': 76, 'GTT': 1}, is mutation! max freq alt codon is GGA\n",
      "GCT, {'GCT': 1354, 'GCA': 76}, is mutation! max freq alt codon is GCA\n",
      "GCT, {'GCT': 1356, 'GCA': 76}, is mutation! max freq alt codon is GCA\n",
      "GCT, {'GCT': 1378, 'TCT': 53}, is mutation! max freq alt codon is TCT\n",
      "TTG, {'TTG': 1336, 'ATG': 1, 'TTA': 76}, is mutation! max freq alt codon is TTA\n",
      "GGC, {'GGC': 1355, 'GGA': 76}, is mutation! max freq alt codon is GGA\n",
      "ATT, {'ATT': 1351, 'ATA': 76}, is mutation! max freq alt codon is ATA\n",
      "TGC, {'TGC': 1352, 'TGT': 75}, is mutation! max freq alt codon is TGT\n",
      "ATC, {'ATC': 1384, 'ATA': 52}, is mutation! max freq alt codon is ATA\n",
      "On gene 98 in seq edge_1671.\n",
      "On gene 99 in seq edge_1671.\n",
      "ACG, {'ACG': 1343, 'GCG': 40, 'CCG': 2}, is mutation! max freq alt codon is GCG\n",
      "On gene 100 in seq edge_1671.\n",
      "CGC, {'CGC': 1245, 'CGT': 140}, is mutation! max freq alt codon is CGT\n",
      "GGA, {'GGA': 1316, 'GAA': 72, 'GGG': 1}, is mutation! max freq alt codon is GAA\n",
      "GGT, {'GGT': 1314, 'GGG': 72, 'TGT': 1}, is mutation! max freq alt codon is GGG\n",
      "TTA, {'TTA': 1289, 'TAA': 1, 'ATA': 70, 'TTT': 1}, is mutation! max freq alt codon is ATA\n",
      "GAC, {'GAC': 1318, 'GAT': 70}, is mutation! max freq alt codon is GAT\n",
      "AGC, {'AGC': 1319, 'AGT': 68, 'TGC': 1}, is mutation! max freq alt codon is AGT\n",
      "GCT, {'GCT': 1313, 'ACT': 67, 'GCG': 1}, is mutation! max freq alt codon is ACT\n",
      "GGG, {'GGG': 1304, 'GCG': 1, 'GGT': 71}, is mutation! max freq alt codon is GGT\n",
      "GAA, {'GAA': 1283, 'GGA': 72, 'AAA': 2}, is mutation! max freq alt codon is GGA\n",
      "On gene 101 in seq edge_1671.\n",
      "GCT, {'GCT': 1320, 'GCA': 72}, is mutation! max freq alt codon is GCA\n",
      "TTA, {'TTA': 1306, 'CTA': 73, 'TAA': 1, 'CTT': 1}, is mutation! max freq alt codon is CTA\n",
      "ACC, {'ACC': 1316, 'ACA': 72, 'ACT': 1, 'AAC': 1}, is mutation! max freq alt codon is ACA\n",
      "GGA, {'GGA': 1314, 'GGC': 73, 'GAA': 1}, is mutation! max freq alt codon is GGC\n",
      "TTA, {'TTA': 1315, 'ATA': 73}, is mutation! max freq alt codon is ATA\n",
      "ATA, {'ATA': 1312, 'ATG': 73}, is mutation! max freq alt codon is ATG\n",
      "AAT, {'AAT': 1309, 'AAC': 71, 'ATC': 1}, is mutation! max freq alt codon is AAC\n",
      "AGC, {'AGC': 1319, 'AGT': 73}, is mutation! max freq alt codon is AGT\n",
      "GCA, {'GCA': 1305, 'CTA': 1, 'GCC': 72}, is mutation! max freq alt codon is GCC\n",
      "CCC, {'CCC': 1302, 'CCT': 73}, is mutation! max freq alt codon is CCT\n",
      "GAA, {'GAA': 1311, 'GAG': 74}, is mutation! max freq alt codon is GAG\n",
      "GGT, {'GGT': 1300, 'GGA': 74, 'GGG': 1}, is mutation! max freq alt codon is GGA\n",
      "ATT, {'ATT': 1310, 'ATA': 75}, is mutation! max freq alt codon is ATA\n",
      "GAG, {'GAG': 1313, 'GAA': 74}, is mutation! max freq alt codon is GAA\n",
      "CGT, {'CGT': 1310, 'CGC': 74, 'GAC': 1}, is mutation! max freq alt codon is CGC\n",
      "GAC, {'GAC': 1303, 'GAT': 75}, is mutation! max freq alt codon is GAT\n",
      "CAG, {'CAG': 1312, 'CAA': 75}, is mutation! max freq alt codon is CAA\n",
      "AAT, {'AAT': 1309, 'GAC': 75}, is mutation! max freq alt codon is GAC\n",
      "GGC, {'GGC': 1307, 'GGA': 75}, is mutation! max freq alt codon is GGA\n",
      "GAC, {'GAC': 1306, 'GAT': 77}, is mutation! max freq alt codon is GAT\n",
      "GCT, {'GCT': 1312, 'GCC': 74}, is mutation! max freq alt codon is GCC\n",
      "TTC, {'TTC': 1311, 'TTT': 76}, is mutation! max freq alt codon is TTT\n",
      "CAC, {'CAC': 1309, 'CAT': 76}, is mutation! max freq alt codon is CAT\n",
      "CCT, {'CCT': 1306, 'CCC': 74, 'TCT': 1}, is mutation! max freq alt codon is CCC\n",
      "TGT, {'TGT': 1307, 'TGC': 74, 'TGG': 1}, is mutation! max freq alt codon is TGC\n",
      "TAC, {'TAC': 1304, 'TAA': 1, 'TAT': 76}, is mutation! max freq alt codon is TAT\n",
      "CGC, {'CGC': 1302, 'CGT': 76}, is mutation! max freq alt codon is CGT\n",
      "GTA, {'GTA': 1306, 'GTC': 76}, is mutation! max freq alt codon is GTC\n",
      "GGA, {'GGA': 1305, 'GGG': 76}, is mutation! max freq alt codon is GGG\n",
      "GGT, {'GGT': 1351, 'GGC': 36, 'GTT': 1}, is mutation! max freq alt codon is GGC\n",
      "GCA, {'GCA': 1318, 'GCG': 86}, is mutation! max freq alt codon is GCG\n",
      "GGT, {'GGT': 1313, 'GTT': 1, 'GGC': 86}, is mutation! max freq alt codon is GGC\n",
      "GGA, {'GGA': 1315, 'GGC': 86}, is mutation! max freq alt codon is GGC\n",
      "CAA, {'CAA': 1319, 'TAA': 87}, is mutation! max freq alt codon is TAA\n",
      "On gene 102 in seq edge_1671.\n",
      "GGT, {'GGT': 1281, 'GTT': 113}, is mutation! max freq alt codon is GTT\n",
      "TCG, {'TCG': 1290, 'TTG': 115}, is mutation! max freq alt codon is TTG\n",
      "AGC, {'AGC': 1285, 'AGT': 113, 'AAG': 1, 'ATT': 1, 'GGC': 1}, is mutation! max freq alt codon is AGT\n",
      "Found contrary case.\n",
      "AGA, {'AGA': 1181, 'GAA': 5, 'AAA': 1, 'GTA': 1, 'CAG': 1}, only mutation in aggregate... max freq alt codon is GAA\n",
      "GTA, {'GTA': 1299, 'GCA': 113}, is mutation! max freq alt codon is GCA\n",
      "On gene 103 in seq edge_1671.\n",
      "On gene 104 in seq edge_1671.\n",
      "GAA, {'GAA': 1318, 'GAG': 111, 'AAA': 1}, is mutation! max freq alt codon is GAG\n",
      "GAA, {'GAA': 1315, 'GAG': 111}, is mutation! max freq alt codon is GAG\n",
      "GGA, {'GGA': 1285, 'GGT': 146, 'CGA': 1, 'GGG': 1}, is mutation! max freq alt codon is GGT\n",
      "On gene 105 in seq edge_1671.\n",
      "TCT, {'TCT': 1328, 'TCA': 101, 'TTC': 1}, is mutation! max freq alt codon is TCA\n",
      "TCA, {'TCA': 1323, 'ACA': 104}, is mutation! max freq alt codon is ACA\n",
      "On gene 106 in seq edge_1671.\n",
      "GCG, {'GCG': 1353, 'GCA': 80}, is mutation! max freq alt codon is GCA\n",
      "CCT, {'CCT': 1321, 'CCC': 100}, is mutation! max freq alt codon is CCC\n",
      "On gene 107 in seq edge_1671.\n",
      "CTG, {'CTG': 1275, 'TTG': 137}, is mutation! max freq alt codon is TTG\n",
      "CTA, {'CTA': 1387, 'CTT': 21, 'TTA': 1}, is mutation! max freq alt codon is CTT\n",
      "AAC, {'AAC': 1272, 'AAT': 132, 'CAC': 1, 'ACA': 1}, is mutation! max freq alt codon is AAT\n",
      "CTA, {'CTA': 1285, 'CTG': 137}, is mutation! max freq alt codon is CTG\n",
      "ACG, {'ACG': 1386, 'AAG': 35}, is mutation! max freq alt codon is AAG\n",
      "CGA, {'CGA': 1390, 'CGG': 35}, is mutation! max freq alt codon is CGG\n",
      "ATC, {'ATC': 1256, 'ACC': 130, 'TAT': 1}, is mutation! max freq alt codon is ACC\n",
      "TTA, {'TTA': 1266, 'CTA': 139}, is mutation! max freq alt codon is CTA\n",
      "ATA, {'ATA': 1269, 'ATT': 138, 'AAT': 1}, is mutation! max freq alt codon is ATT\n",
      "GAG, {'GAG': 1271, 'GAT': 138}, is mutation! max freq alt codon is GAT\n",
      "ATC, {'ATC': 1272, 'ATT': 137}, is mutation! max freq alt codon is ATT\n",
      "GTT, {'GTT': 1268, 'GTA': 138}, is mutation! max freq alt codon is GTA\n",
      "ATA, {'ATA': 1377, 'ATG': 35, 'TAG': 1}, is mutation! max freq alt codon is ATG\n",
      "GAA, {'GAA': 1376, 'GAG': 35}, is mutation! max freq alt codon is GAG\n",
      "GAT, {'GAT': 1277, 'GAC': 102, 'GAG': 34, 'AAT': 1}, is mutation! max freq alt codon is GAC\n",
      "Found contrary case.\n",
      "TTA, {'TTA': 1356, 'TTT': 5, 'TAA': 5}, only mutation in aggregate... max freq alt codon is TTT\n",
      "TTA, {'TTA': 1383, 'TTG': 34, 'ATA': 2, 'TTC': 1, 'TGA': 1}, is mutation! max freq alt codon is TTG\n",
      "On gene 108 in seq edge_1671.\n",
      "TGT, {'TGT': 1311, 'TGC': 101, 'TTT': 1, 'GCA': 1}, is mutation! max freq alt codon is TGC\n",
      "GGA, {'GGA': 1295, 'GGT': 100, 'GAA': 1}, is mutation! max freq alt codon is GGT\n",
      "GAG, {'GAG': 1290, 'GAA': 103, 'TGA': 1, 'GAT': 1}, is mutation! max freq alt codon is GAA\n",
      "GGA, {'GGA': 1315, 'GGG': 102, 'TGA': 1}, is mutation! max freq alt codon is GGG\n",
      "TGT, {'TGT': 1311, 'TGC': 104, 'GGT': 1}, is mutation! max freq alt codon is TGC\n",
      "AAT, {'AAT': 1308, 'GAT': 103, 'ATT': 1, 'ACT': 1}, is mutation! max freq alt codon is GAT\n",
      "TTA, {'TTA': 1310, 'TTG': 104, 'GTA': 1}, is mutation! max freq alt codon is TTG\n",
      "GAT, {'GAT': 1309, 'GAC': 104}, is mutation! max freq alt codon is GAC\n",
      "GAT, {'GAT': 1310, 'GAC': 104}, is mutation! max freq alt codon is GAC\n",
      "CCT, {'CCT': 1305, 'CCG': 102, 'ACT': 1}, is mutation! max freq alt codon is CCG\n",
      "On gene 109 in seq edge_1671.\n",
      "GCT, {'GCT': 1302, 'GCA': 101, 'GCG': 1}, is mutation! max freq alt codon is GCA\n",
      "On gene 110 in seq edge_1671.\n",
      "GCG, {'GCG': 1293, 'GCA': 103}, is mutation! max freq alt codon is GCA\n",
      "TAC, {'TAC': 1378, 'TAT': 19}, is mutation! max freq alt codon is TAT\n",
      "GGA, {'GGA': 1312, 'GGT': 86}, is mutation! max freq alt codon is GGT\n",
      "GCA, {'GCA': 1292, 'GCT': 106}, is mutation! max freq alt codon is GCT\n",
      "On gene 111 in seq edge_1671.\n",
      "TTA, {'TTA': 1250, 'ATA': 7, 'TAT': 1, 'TTT': 5}, is mutation! max freq alt codon is ATA\n",
      "CAC, {'CAC': 1271, 'CAT': 135}, is mutation! max freq alt codon is CAT\n",
      "GTT, {'GTT': 1363, 'GTA': 35, 'AGT': 1}, is mutation! max freq alt codon is GTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCC, {'CCC': 1351, 'CCA': 37, 'CGC': 1}, is mutation! max freq alt codon is CCA\n",
      "ATT, {'ATT': 1373, 'ATA': 36}, is mutation! max freq alt codon is ATA\n",
      "CCT, {'CCT': 1295, 'CCC': 98, 'ACT': 1}, is mutation! max freq alt codon is CCC\n",
      "TTC, {'TTC': 1309, 'TTT': 98, 'TCT': 1}, is mutation! max freq alt codon is TTT\n",
      "GAT, {'GAT': 1275, 'GAC': 135}, is mutation! max freq alt codon is GAC\n",
      "CCT, {'CCT': 1376, 'CCA': 36}, is mutation! max freq alt codon is CCA\n",
      "GCG, {'GCG': 1375, 'GCT': 1, 'GCA': 35}, is mutation! max freq alt codon is GCA\n",
      "CGT, {'CGT': 1358, 'CGA': 35, 'CGG': 2}, is mutation! max freq alt codon is CGA\n",
      "AAA, {'AAA': 1375, 'AAG': 35}, is mutation! max freq alt codon is AAG\n",
      "TTG, {'TTG': 1308, 'TTA': 100}, is mutation! max freq alt codon is TTA\n",
      "GCG, {'GCG': 1311, 'GCC': 100}, is mutation! max freq alt codon is GCC\n",
      "GAG, {'GAG': 1296, 'GAA': 98, 'AAG': 1}, is mutation! max freq alt codon is GAA\n",
      "On gene 112 in seq edge_1671.\n",
      "GAG, {'GAG': 1256, 'GAA': 134, 'GAT': 1}, is mutation! max freq alt codon is GAA\n",
      "GCT, {'GCT': 1268, 'GCA': 134}, is mutation! max freq alt codon is GCA\n",
      "GAC, {'GAC': 1256, 'GAT': 132, 'GCC': 1}, is mutation! max freq alt codon is GAT\n",
      "GAT, {'GAT': 1265, 'GAC': 133}, is mutation! max freq alt codon is GAC\n",
      "TTT, {'TTT': 1287, 'TTC': 99}, is mutation! max freq alt codon is TTC\n",
      "AGC, {'AGC': 1298, 'AGT': 101}, is mutation! max freq alt codon is AGT\n",
      "ATT, {'ATT': 1291, 'ATA': 101}, is mutation! max freq alt codon is ATA\n",
      "TCA, {'TCA': 1299, 'TCG': 101, 'TCT': 1}, is mutation! max freq alt codon is TCG\n",
      "AAC, {'AAC': 1303, 'AAT': 101}, is mutation! max freq alt codon is AAT\n",
      "GTC, {'GTC': 1289, 'GTA': 100}, is mutation! max freq alt codon is GTA\n",
      "GAC, {'GAC': 1270, 'AAT': 138, 'AAA': 1}, is mutation! max freq alt codon is AAT\n",
      "GGG, {'GGG': 1302, 'GGA': 100}, is mutation! max freq alt codon is GGA\n",
      "ACA, {'ACA': 1307, 'ACG': 100}, is mutation! max freq alt codon is ACG\n",
      "ATA, {'ATA': 1364, 'ATC': 37, 'AGA': 1}, is mutation! max freq alt codon is ATC\n",
      "CGT, {'CGT': 1289, 'CGC': 100}, is mutation! max freq alt codon is CGC\n",
      "Found contrary case.\n",
      "AAC, {'AAC': 1376, 'AAA': 6, 'ACA': 1, 'ATC': 1}, only mutation in aggregate... max freq alt codon is AAA\n",
      "AAA, {'AAA': 1351, 'CAA': 12, 'TAA': 1}, is mutation! max freq alt codon is CAA\n",
      "GGA, {'GGA': 1260, 'GGT': 138, 'GAC': 1}, is mutation! max freq alt codon is GGT\n",
      "CCT, {'CCT': 1255, 'CCA': 137}, is mutation! max freq alt codon is CCA\n",
      "CCT, {'CCT': 1248, 'CCA': 135}, is mutation! max freq alt codon is CCA\n",
      "TCG, {'TCG': 1255, 'TCA': 136, 'TTG': 1}, is mutation! max freq alt codon is TCA\n",
      "ACT, {'ACT': 1231, 'ACC': 137, 'CTT': 1}, is mutation! max freq alt codon is ACC\n",
      "CCA, {'CCA': 1351, 'CCG': 37, 'CAC': 1, 'ACA': 1}, is mutation! max freq alt codon is CCG\n",
      "CGT, {'CGT': 1250, 'CGA': 136}, is mutation! max freq alt codon is CGA\n",
      "CTC, {'CTC': 1259, 'CTT': 135}, is mutation! max freq alt codon is CTT\n",
      "ATC, {'ATC': 1350, 'ATT': 36, 'AAT': 1}, is mutation! max freq alt codon is ATT\n",
      "CGT, {'CGT': 1259, 'CGC': 137}, is mutation! max freq alt codon is CGC\n",
      "CAG, {'CAG': 1359, 'CAA': 37, 'CGG': 1}, is mutation! max freq alt codon is CAA\n",
      "Found contrary case.\n",
      "TAA, {'TAA': 1331, 'ATA': 3, 'TTA': 1, 'AAA': 4}, only mutation in aggregate... max freq alt codon is AAA\n",
      "On gene 113 in seq edge_1671.\n",
      "GGT, {'GGT': 1311, 'GGG': 95}, is mutation! max freq alt codon is GGG\n",
      "ACC, {'ACC': 1359, 'ACA': 1, 'ACT': 37}, is mutation! max freq alt codon is ACT\n",
      "ACC, {'ACC': 1244, 'ACT': 134, 'AAC': 2, 'CCC': 1}, is mutation! max freq alt codon is ACT\n",
      "AGC, {'AGC': 1310, 'TGC': 99}, is mutation! max freq alt codon is TGC\n",
      "CGT, {'CGT': 1397, 'AGT': 17, 'TGA': 1, 'GCT': 1}, is mutation! max freq alt codon is AGT\n",
      "GTA, {'GTA': 1314, 'GTG': 95}, is mutation! max freq alt codon is GTG\n",
      "TCT, {'TCT': 1270, 'TCA': 135, 'ACA': 1}, is mutation! max freq alt codon is TCA\n",
      "GTA, {'GTA': 1271, 'GCA': 135, 'ATA': 1, 'GAA': 2}, is mutation! max freq alt codon is GCA\n",
      "On gene 114 in seq edge_1671.\n",
      "CGT, {'CGC': 137, 'CGT': 1283, 'CGA': 1}, is mutation! max freq alt codon is CGC\n",
      "TCG, {'TCA': 140, 'TCG': 1277, 'GTC': 1, 'ACG': 1}, is mutation! max freq alt codon is TCA\n",
      "TTG, {'TTA': 137, 'TTG': 1261, 'TGG': 2, 'CTT': 1}, is mutation! max freq alt codon is TTA\n",
      "GCT, {'GCT': 1406, 'GCC': 16}, is mutation! max freq alt codon is GCC\n",
      "GAT, {'GAC': 142, 'GAT': 1289}, is mutation! max freq alt codon is GAC\n",
      "TTA, {'TTG': 142, 'TTA': 1290, 'GTA': 1}, is mutation! max freq alt codon is TTG\n",
      "Found contrary case.\n",
      "ACA, {'ACA': 1366, 'CAA': 2, 'AAC': 2, 'CCA': 2, 'TCA': 1}, only mutation in aggregate... max freq alt codon is CAA\n",
      "GGC, {'GGC': 1399, 'GGT': 43, 'GCC': 1}, is mutation! max freq alt codon is GGT\n",
      "CCT, {'CCT': 1401, 'CCC': 43}, is mutation! max freq alt codon is CCC\n",
      "GAC, {'GAC': 1405, 'GAT': 43}, is mutation! max freq alt codon is GAT\n",
      "GGA, {'GGA': 1405, 'GGG': 42, 'GGC': 1}, is mutation! max freq alt codon is GGG\n",
      "GTG, {'GTG': 1429, 'GTA': 18, 'GGG': 1}, is mutation! max freq alt codon is GTA\n",
      "GGA, {'GGA': 1418, 'GGG': 18, 'TGA': 1}, is mutation! max freq alt codon is GGG\n",
      "AAT, {'AAC': 146, 'AAT': 1307}, is mutation! max freq alt codon is AAC\n",
      "ACC, {'ACC': 1404, 'ACT': 42, 'ACA': 1}, is mutation! max freq alt codon is ACT\n",
      "CCG, {'CCG': 1348, 'CCT': 82, 'ACG': 1, 'CGG': 1}, is mutation! max freq alt codon is CCT\n",
      "GTT, {'GTT': 1295, 'GTC': 144}, is mutation! max freq alt codon is GTC\n",
      "GAC, {'GAC': 1359, 'GAT': 80, 'GAA': 1}, is mutation! max freq alt codon is GAT\n",
      "GCA, {'GCA': 1337, 'GCG': 80}, is mutation! max freq alt codon is GCG\n",
      "On gene 115 in seq edge_1671.\n",
      "TCA, {'TCA': 1285, 'TCG': 136, 'GCA': 1}, is mutation! max freq alt codon is TCG\n",
      "GGT, {'GGT': 1291, 'GGC': 139}, is mutation! max freq alt codon is GGC\n",
      "Found contrary case.\n",
      "GGG, {'GGG': 1385, 'GAG': 2, 'AGG': 5, 'GTG': 1, 'AGT': 1}, only mutation in aggregate... max freq alt codon is AGG\n",
      "ACG, {'ACG': 1384, 'ATG': 42}, is mutation! max freq alt codon is ATG\n",
      "TTG, {'TTG': 1313, 'CTG': 118}, is mutation! max freq alt codon is CTG\n",
      "CCA, {'CCA': 1378, 'CCT': 42, 'CAA': 2, 'CTA': 1}, is mutation! max freq alt codon is CCT\n",
      "CGC, {'CGC': 1280, 'CGT': 140}, is mutation! max freq alt codon is CGT\n",
      "GCA, {'GCA': 1288, 'ACA': 138}, is mutation! max freq alt codon is ACA\n",
      "On gene 116 in seq edge_1671.\n",
      "TCG, {'TCG': 1358, 'TCT': 78}, is mutation! max freq alt codon is TCT\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-29-7fce27460942>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mgenerate_mutmatrix_data\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"edge_1671\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mgenerate_mutmatrix_data\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"edge_6104\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mgenerate_mutmatrix_data\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"edge_2358\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-28-9e6c74037507>\u001b[0m in \u001b[0;36mgenerate_mutmatrix_data\u001b[0;34m(seq)\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m             \u001b[0mcodon_seq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcodon_dna\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m             \u001b[0maa\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcodon_dna\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtranslate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m             \u001b[0;31m# Update frequencies accordingly.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/mut-analyses/lib/python3.7/site-packages/skbio/sequence/_sequence.py\u001b[0m in \u001b[0;36m__str__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1073\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1074\u001b[0m         \"\"\"\n\u001b[0;32m-> 1075\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_string\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdecode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"ascii\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1076\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1077\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mstable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mas_of\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"0.4.0\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "generate_mutmatrix_data(\"edge_1671\")\n",
    "generate_mutmatrix_data(\"edge_6104\")\n",
    "generate_mutmatrix_data(\"edge_2358\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
