{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Perform read smoothing then assemble with LJA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-02-07T11:06:12.386442Z",
     "iopub.status.busy": "2022-02-07T11:06:12.385488Z",
     "iopub.status.idle": "2022-02-07T11:06:12.717083Z",
     "shell.execute_reply": "2022-02-07T11:06:12.716149Z"
    }
   },
   "outputs": [],
   "source": [
    "%run \"Header.ipynb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-02-07T11:06:12.721285Z",
     "iopub.status.busy": "2022-02-07T11:06:12.720458Z",
     "iopub.status.idle": "2022-02-07T11:06:13.238408Z",
     "shell.execute_reply": "2022-02-07T11:06:13.239092Z"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import time\n",
    "import pickle\n",
    "import pysam\n",
    "import skbio\n",
    "from collections import defaultdict, Counter\n",
    "from linked_mutations_utils import find_mutated_positions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0. Quick sanity check: ensure that all $k$-mers ($k$ = 5,001) are unique in each MAG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "On start pos 0 in CAMP.\n",
      "On start pos 1,000,000 in CAMP.\n",
      "The most common k = 5,001-mer in CAMP occurred 1 time(s).\n",
      "On start pos 0 in BACT1.\n",
      "On start pos 1,000,000 in BACT1.\n",
      "On start pos 2,000,000 in BACT1.\n",
      "The most common k = 5,001-mer in BACT1 occurred 1 time(s).\n",
      "On start pos 0 in BACT2.\n",
      "On start pos 1,000,000 in BACT2.\n",
      "On start pos 2,000,000 in BACT2.\n",
      "The most common k = 5,001-mer in BACT2 occurred 1 time(s).\n"
     ]
    }
   ],
   "source": [
    "# I know there are actual k-mer counting tools you can use but no reason to overcomplicate things for now\n",
    "\n",
    "k = 5001\n",
    "\n",
    "for seq in SEQS:\n",
    "    fasta = skbio.DNA.read(f\"../seqs/{seq}.fasta\")\n",
    "    bargain_bin_kmer_counter = Counter()\n",
    "    \n",
    "    # The skbio.DNA object is 0-indexed, so 0 is the leftmost k-mer start position and\n",
    "    # ((seq length) - k) is the rightmost k-mer start position. The + 1 is because python ranges don't include\n",
    "    # the right endpoint.\n",
    "    for start_pos in range(0, seq2len[seq] - k + 1):\n",
    "        \n",
    "        # NOTE: this is a terrible no good very bad way to do this; it's more efficient to use a \"sliding window\"\n",
    "        # approach where you store the entire k-mer and then, with each step, just remove the first character and\n",
    "        # add on a new last character. \"But, uh, this code will only be run on these three MAGs, so I'm gonna\n",
    "        # prioritize clarity over optimization,\" says me, the insane person who just spent like a minute writing\n",
    "        # this comment when I could've been optimizing this code instead look WHATEVER this counts k-mers and it's\n",
    "        # 4am let's not overcomplicate it, look if you're on GitHub right now and you see this inane comment\n",
    "        # we can both just pretend that you were looking at some really optimized code and we'll both walk away\n",
    "        # satisfied, capisce\n",
    "        kmer = fasta[start_pos : start_pos + k]\n",
    "        \n",
    "        bargain_bin_kmer_counter[str(kmer)] += 1\n",
    "        if start_pos % 1000000 == 0: print(f\"On start pos {start_pos:,} in {seq2name[seq]}.\")\n",
    "    \n",
    "    mckc = bargain_bin_kmer_counter.most_common(1)[0][1]\n",
    "    print(f\"The most common k = {k:,}-mer in {seq2name[seq]} occurred {mckc:,} time(s).\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Smooth reads\n",
    "\n",
    "Lots of this code is duplicated from the `Phasing-01-MakeGraph.ipynb` notebook in this folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set this to True to actually generate ordinary smoothed reads that include called mutations;\n",
    "# set this to False to generate \"sanity check\" perfect smoothed reads, where no mutations are included\n",
    "# and the read entirely matches the reference\n",
    "actually_include_mutations_in_the_smoothed_reads = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-02-07T11:06:13.258951Z",
     "iopub.status.busy": "2022-02-07T11:06:13.257971Z",
     "iopub.status.idle": "2022-02-07T11:08:10.160475Z",
     "shell.execute_reply": "2022-02-07T11:08:10.159569Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Identifying mutated positions (p = 1%) in genome CAMP...\n",
      "Found 83 mutated positions (p = 1%) in CAMP.\n",
      "Note that this tally is higher than you'd see in e.g. the CP1/2/3 plots, because now we're including both 'rare' and non-rare mutations. Just so you don't waste five minutes sanity-checking this like I did.\n",
      "Going through these positions...\n",
      "\tOn aln 1,000 in seq CAMP. Time spent on CAMP so far: 0.06 sec.\n",
      "\tOn aln 2,000 in seq CAMP. Time spent on CAMP so far: 0.09 sec.\n",
      "\tOn aln 3,000 in seq CAMP. Time spent on CAMP so far: 0.14 sec.\n",
      "\tOn aln 4,000 in seq CAMP. Time spent on CAMP so far: 0.17 sec.\n",
      "\tOn aln 5,000 in seq CAMP. Time spent on CAMP so far: 0.21 sec.\n",
      "\tOn aln 6,000 in seq CAMP. Time spent on CAMP so far: 0.26 sec.\n",
      "\tOn aln 7,000 in seq CAMP. Time spent on CAMP so far: 0.31 sec.\n",
      "\tOn aln 8,000 in seq CAMP. Time spent on CAMP so far: 0.36 sec.\n",
      "\tOn aln 9,000 in seq CAMP. Time spent on CAMP so far: 0.41 sec.\n",
      "\tOn aln 10,000 in seq CAMP. Time spent on CAMP so far: 0.46 sec.\n",
      "\tOn aln 11,000 in seq CAMP. Time spent on CAMP so far: 0.51 sec.\n",
      "\tOn aln 12,000 in seq CAMP. Time spent on CAMP so far: 0.56 sec.\n",
      "\tOn aln 13,000 in seq CAMP. Time spent on CAMP so far: 0.61 sec.\n",
      "\tOn aln 14,000 in seq CAMP. Time spent on CAMP so far: 0.66 sec.\n",
      "\tOn aln 15,000 in seq CAMP. Time spent on CAMP so far: 0.71 sec.\n",
      "\tOn aln 16,000 in seq CAMP. Time spent on CAMP so far: 0.76 sec.\n",
      "\tOn aln 17,000 in seq CAMP. Time spent on CAMP so far: 0.81 sec.\n",
      "\tOn aln 18,000 in seq CAMP. Time spent on CAMP so far: 0.86 sec.\n",
      "\tOn aln 19,000 in seq CAMP. Time spent on CAMP so far: 0.91 sec.\n",
      "\tOn aln 20,000 in seq CAMP. Time spent on CAMP so far: 0.96 sec.\n",
      "\tOn aln 21,000 in seq CAMP. Time spent on CAMP so far: 1.01 sec.\n",
      "\tOn aln 22,000 in seq CAMP. Time spent on CAMP so far: 1.07 sec.\n",
      "\tOn aln 23,000 in seq CAMP. Time spent on CAMP so far: 1.12 sec.\n",
      "\tOn aln 24,000 in seq CAMP. Time spent on CAMP so far: 1.17 sec.\n",
      "\tOn aln 25,000 in seq CAMP. Time spent on CAMP so far: 1.22 sec.\n",
      "\tOn aln 26,000 in seq CAMP. Time spent on CAMP so far: 1.27 sec.\n",
      "\tOn aln 27,000 in seq CAMP. Time spent on CAMP so far: 1.32 sec.\n",
      "\tOn aln 28,000 in seq CAMP. Time spent on CAMP so far: 1.37 sec.\n",
      "\tOn aln 29,000 in seq CAMP. Time spent on CAMP so far: 1.42 sec.\n",
      "\tOn aln 30,000 in seq CAMP. Time spent on CAMP so far: 1.47 sec.\n",
      "\tOn aln 31,000 in seq CAMP. Time spent on CAMP so far: 1.52 sec.\n",
      "\tOn aln 32,000 in seq CAMP. Time spent on CAMP so far: 1.56 sec.\n",
      "\tOn aln 33,000 in seq CAMP. Time spent on CAMP so far: 1.61 sec.\n",
      "\tOn aln 34,000 in seq CAMP. Time spent on CAMP so far: 1.66 sec.\n",
      "\tOn aln 35,000 in seq CAMP. Time spent on CAMP so far: 1.71 sec.\n",
      "\tOn aln 36,000 in seq CAMP. Time spent on CAMP so far: 1.76 sec.\n",
      "\tOn aln 37,000 in seq CAMP. Time spent on CAMP so far: 1.81 sec.\n",
      "\tOn aln 38,000 in seq CAMP. Time spent on CAMP so far: 1.86 sec.\n",
      "\tOn aln 39,000 in seq CAMP. Time spent on CAMP so far: 1.91 sec.\n",
      "\tOn aln 40,000 in seq CAMP. Time spent on CAMP so far: 1.96 sec.\n",
      "\tOn aln 41,000 in seq CAMP. Time spent on CAMP so far: 2.01 sec.\n",
      "\tOn aln 42,000 in seq CAMP. Time spent on CAMP so far: 2.06 sec.\n",
      "\tOn aln 43,000 in seq CAMP. Time spent on CAMP so far: 2.11 sec.\n",
      "\tOn aln 44,000 in seq CAMP. Time spent on CAMP so far: 2.16 sec.\n",
      "\tOn aln 45,000 in seq CAMP. Time spent on CAMP so far: 2.21 sec.\n",
      "\tOn aln 46,000 in seq CAMP. Time spent on CAMP so far: 2.26 sec.\n",
      "\tOn aln 47,000 in seq CAMP. Time spent on CAMP so far: 2.30 sec.\n",
      "\tOn aln 48,000 in seq CAMP. Time spent on CAMP so far: 2.35 sec.\n",
      "\tOn aln 49,000 in seq CAMP. Time spent on CAMP so far: 2.40 sec.\n",
      "\tOn aln 50,000 in seq CAMP. Time spent on CAMP so far: 2.45 sec.\n",
      "\tOn aln 51,000 in seq CAMP. Time spent on CAMP so far: 2.50 sec.\n",
      "\tOn aln 52,000 in seq CAMP. Time spent on CAMP so far: 2.55 sec.\n",
      "\tOn aln 53,000 in seq CAMP. Time spent on CAMP so far: 2.60 sec.\n",
      "\tOn aln 54,000 in seq CAMP. Time spent on CAMP so far: 2.65 sec.\n",
      "\tOn aln 55,000 in seq CAMP. Time spent on CAMP so far: 2.70 sec.\n",
      "\tOn aln 56,000 in seq CAMP. Time spent on CAMP so far: 2.75 sec.\n",
      "\tOn aln 57,000 in seq CAMP. Time spent on CAMP so far: 2.80 sec.\n",
      "\tOn aln 58,000 in seq CAMP. Time spent on CAMP so far: 2.85 sec.\n",
      "\tOn aln 59,000 in seq CAMP. Time spent on CAMP so far: 2.90 sec.\n",
      "\tOn aln 60,000 in seq CAMP. Time spent on CAMP so far: 2.95 sec.\n",
      "\tOn aln 61,000 in seq CAMP. Time spent on CAMP so far: 3.00 sec.\n",
      "\tOn aln 62,000 in seq CAMP. Time spent on CAMP so far: 3.05 sec.\n",
      "\tOn aln 63,000 in seq CAMP. Time spent on CAMP so far: 3.09 sec.\n",
      "\tOn aln 64,000 in seq CAMP. Time spent on CAMP so far: 3.14 sec.\n",
      "\tOn aln 65,000 in seq CAMP. Time spent on CAMP so far: 3.19 sec.\n",
      "\tOn aln 66,000 in seq CAMP. Time spent on CAMP so far: 3.25 sec.\n",
      "\tOn aln 67,000 in seq CAMP. Time spent on CAMP so far: 3.30 sec.\n",
      "\tOn aln 68,000 in seq CAMP. Time spent on CAMP so far: 3.35 sec.\n",
      "\tOn aln 69,000 in seq CAMP. Time spent on CAMP so far: 3.40 sec.\n",
      "\tOn aln 70,000 in seq CAMP. Time spent on CAMP so far: 3.45 sec.\n",
      "\tOn aln 71,000 in seq CAMP. Time spent on CAMP so far: 3.50 sec.\n",
      "\tOn aln 72,000 in seq CAMP. Time spent on CAMP so far: 3.55 sec.\n",
      "\tOn aln 73,000 in seq CAMP. Time spent on CAMP so far: 3.60 sec.\n",
      "\tOn aln 74,000 in seq CAMP. Time spent on CAMP so far: 3.65 sec.\n",
      "\tOn aln 75,000 in seq CAMP. Time spent on CAMP so far: 3.70 sec.\n",
      "\tOn aln 76,000 in seq CAMP. Time spent on CAMP so far: 3.75 sec.\n",
      "\tOn aln 77,000 in seq CAMP. Time spent on CAMP so far: 3.80 sec.\n",
      "\tOn aln 78,000 in seq CAMP. Time spent on CAMP so far: 3.86 sec.\n",
      "\tOn aln 79,000 in seq CAMP. Time spent on CAMP so far: 3.91 sec.\n",
      "\tOn aln 80,000 in seq CAMP. Time spent on CAMP so far: 3.95 sec.\n",
      "\tOn aln 81,000 in seq CAMP. Time spent on CAMP so far: 4.00 sec.\n",
      "\tOn aln 82,000 in seq CAMP. Time spent on CAMP so far: 4.05 sec.\n",
      "\tOn aln 83,000 in seq CAMP. Time spent on CAMP so far: 4.10 sec.\n",
      "\tOn aln 84,000 in seq CAMP. Time spent on CAMP so far: 4.15 sec.\n",
      "\tOn aln 85,000 in seq CAMP. Time spent on CAMP so far: 4.21 sec.\n",
      "\tOn aln 86,000 in seq CAMP. Time spent on CAMP so far: 4.26 sec.\n",
      "\tOn aln 87,000 in seq CAMP. Time spent on CAMP so far: 4.31 sec.\n",
      "\tOn aln 88,000 in seq CAMP. Time spent on CAMP so far: 4.36 sec.\n",
      "\tOn aln 89,000 in seq CAMP. Time spent on CAMP so far: 4.41 sec.\n",
      "\tOn aln 90,000 in seq CAMP. Time spent on CAMP so far: 4.46 sec.\n",
      "\tOn aln 91,000 in seq CAMP. Time spent on CAMP so far: 4.51 sec.\n",
      "\tOn aln 92,000 in seq CAMP. Time spent on CAMP so far: 4.56 sec.\n",
      "\tOn aln 93,000 in seq CAMP. Time spent on CAMP so far: 4.60 sec.\n",
      "\tOn aln 94,000 in seq CAMP. Time spent on CAMP so far: 4.65 sec.\n",
      "\tOn aln 95,000 in seq CAMP. Time spent on CAMP so far: 4.70 sec.\n",
      "\tOn aln 96,000 in seq CAMP. Time spent on CAMP so far: 4.75 sec.\n",
      "\tOn aln 97,000 in seq CAMP. Time spent on CAMP so far: 4.80 sec.\n",
      "\tOn aln 98,000 in seq CAMP. Time spent on CAMP so far: 4.85 sec.\n",
      "\tOn aln 99,000 in seq CAMP. Time spent on CAMP so far: 4.90 sec.\n",
      "\tOn aln 100,000 in seq CAMP. Time spent on CAMP so far: 4.95 sec.\n",
      "\tOn aln 101,000 in seq CAMP. Time spent on CAMP so far: 5.00 sec.\n",
      "\tOn aln 102,000 in seq CAMP. Time spent on CAMP so far: 5.05 sec.\n",
      "\tOn aln 103,000 in seq CAMP. Time spent on CAMP so far: 5.10 sec.\n",
      "\tOn aln 104,000 in seq CAMP. Time spent on CAMP so far: 5.15 sec.\n",
      "\tOn aln 105,000 in seq CAMP. Time spent on CAMP so far: 5.20 sec.\n",
      "\tOn aln 106,000 in seq CAMP. Time spent on CAMP so far: 5.25 sec.\n",
      "\tOn aln 107,000 in seq CAMP. Time spent on CAMP so far: 5.30 sec.\n",
      "\tOn aln 108,000 in seq CAMP. Time spent on CAMP so far: 5.35 sec.\n",
      "\tOn aln 109,000 in seq CAMP. Time spent on CAMP so far: 5.40 sec.\n",
      "\tOn aln 110,000 in seq CAMP. Time spent on CAMP so far: 5.45 sec.\n",
      "\tOn aln 111,000 in seq CAMP. Time spent on CAMP so far: 5.50 sec.\n",
      "\tOn aln 112,000 in seq CAMP. Time spent on CAMP so far: 5.55 sec.\n",
      "\tOn aln 113,000 in seq CAMP. Time spent on CAMP so far: 5.60 sec.\n",
      "\tOn aln 114,000 in seq CAMP. Time spent on CAMP so far: 5.65 sec.\n",
      "\tOn aln 115,000 in seq CAMP. Time spent on CAMP so far: 5.70 sec.\n",
      "\tOn aln 116,000 in seq CAMP. Time spent on CAMP so far: 5.75 sec.\n",
      "\tOn aln 117,000 in seq CAMP. Time spent on CAMP so far: 5.80 sec.\n",
      "\tOn aln 118,000 in seq CAMP. Time spent on CAMP so far: 5.85 sec.\n",
      "\tOn aln 119,000 in seq CAMP. Time spent on CAMP so far: 5.90 sec.\n",
      "\tOn aln 120,000 in seq CAMP. Time spent on CAMP so far: 5.94 sec.\n",
      "\tOn aln 121,000 in seq CAMP. Time spent on CAMP so far: 5.99 sec.\n",
      "\tOn aln 122,000 in seq CAMP. Time spent on CAMP so far: 6.04 sec.\n",
      "\tOn aln 123,000 in seq CAMP. Time spent on CAMP so far: 6.09 sec.\n",
      "\tOn aln 124,000 in seq CAMP. Time spent on CAMP so far: 6.14 sec.\n",
      "\tOn aln 125,000 in seq CAMP. Time spent on CAMP so far: 6.19 sec.\n",
      "\tOn aln 126,000 in seq CAMP. Time spent on CAMP so far: 6.24 sec.\n",
      "\tOn aln 127,000 in seq CAMP. Time spent on CAMP so far: 6.29 sec.\n",
      "\tOn aln 128,000 in seq CAMP. Time spent on CAMP so far: 6.34 sec.\n",
      "\tOn aln 129,000 in seq CAMP. Time spent on CAMP so far: 6.39 sec.\n",
      "\tOn aln 130,000 in seq CAMP. Time spent on CAMP so far: 6.44 sec.\n",
      "\tOn aln 131,000 in seq CAMP. Time spent on CAMP so far: 6.49 sec.\n",
      "\tOn aln 132,000 in seq CAMP. Time spent on CAMP so far: 6.54 sec.\n",
      "\tOn aln 133,000 in seq CAMP. Time spent on CAMP so far: 6.59 sec.\n",
      "\tOn aln 134,000 in seq CAMP. Time spent on CAMP so far: 6.64 sec.\n",
      "\tOn aln 135,000 in seq CAMP. Time spent on CAMP so far: 6.69 sec.\n",
      "\tOn aln 136,000 in seq CAMP. Time spent on CAMP so far: 6.74 sec.\n",
      "\tOn aln 137,000 in seq CAMP. Time spent on CAMP so far: 6.79 sec.\n",
      "\tOn aln 138,000 in seq CAMP. Time spent on CAMP so far: 6.84 sec.\n",
      "\tOn aln 139,000 in seq CAMP. Time spent on CAMP so far: 6.89 sec.\n",
      "\tOn aln 140,000 in seq CAMP. Time spent on CAMP so far: 6.94 sec.\n",
      "\tOn aln 141,000 in seq CAMP. Time spent on CAMP so far: 6.99 sec.\n",
      "\tOn aln 142,000 in seq CAMP. Time spent on CAMP so far: 7.04 sec.\n",
      "\tOn aln 143,000 in seq CAMP. Time spent on CAMP so far: 7.09 sec.\n",
      "\tOn aln 144,000 in seq CAMP. Time spent on CAMP so far: 7.14 sec.\n",
      "\tOn aln 145,000 in seq CAMP. Time spent on CAMP so far: 7.19 sec.\n",
      "\tOn aln 146,000 in seq CAMP. Time spent on CAMP so far: 7.24 sec.\n",
      "\tOn aln 147,000 in seq CAMP. Time spent on CAMP so far: 7.29 sec.\n",
      "\tOn aln 148,000 in seq CAMP. Time spent on CAMP so far: 7.34 sec.\n",
      "\tOn aln 149,000 in seq CAMP. Time spent on CAMP so far: 7.38 sec.\n",
      "\tOn aln 150,000 in seq CAMP. Time spent on CAMP so far: 7.43 sec.\n",
      "\tOn aln 151,000 in seq CAMP. Time spent on CAMP so far: 7.48 sec.\n",
      "\tOn aln 152,000 in seq CAMP. Time spent on CAMP so far: 7.53 sec.\n",
      "\tOn aln 153,000 in seq CAMP. Time spent on CAMP so far: 7.58 sec.\n",
      "\tOn aln 154,000 in seq CAMP. Time spent on CAMP so far: 7.63 sec.\n",
      "\tOn aln 155,000 in seq CAMP. Time spent on CAMP so far: 7.68 sec.\n",
      "\tOn aln 156,000 in seq CAMP. Time spent on CAMP so far: 7.73 sec.\n",
      "\tOn aln 157,000 in seq CAMP. Time spent on CAMP so far: 7.78 sec.\n",
      "\tOn aln 158,000 in seq CAMP. Time spent on CAMP so far: 7.83 sec.\n",
      "\tOn aln 159,000 in seq CAMP. Time spent on CAMP so far: 7.88 sec.\n",
      "\tOn aln 160,000 in seq CAMP. Time spent on CAMP so far: 7.93 sec.\n",
      "\tOn aln 161,000 in seq CAMP. Time spent on CAMP so far: 7.98 sec.\n",
      "\tOn aln 162,000 in seq CAMP. Time spent on CAMP so far: 8.03 sec.\n",
      "\tOn aln 163,000 in seq CAMP. Time spent on CAMP so far: 8.08 sec.\n",
      "\tOn aln 164,000 in seq CAMP. Time spent on CAMP so far: 8.13 sec.\n",
      "\tOn aln 165,000 in seq CAMP. Time spent on CAMP so far: 8.18 sec.\n",
      "\tOn aln 166,000 in seq CAMP. Time spent on CAMP so far: 8.23 sec.\n",
      "\tOn aln 167,000 in seq CAMP. Time spent on CAMP so far: 8.28 sec.\n",
      "\tOn aln 168,000 in seq CAMP. Time spent on CAMP so far: 8.33 sec.\n",
      "\tOn aln 169,000 in seq CAMP. Time spent on CAMP so far: 8.38 sec.\n",
      "\tOn aln 170,000 in seq CAMP. Time spent on CAMP so far: 8.43 sec.\n",
      "\tOn aln 171,000 in seq CAMP. Time spent on CAMP so far: 8.48 sec.\n",
      "\tOn aln 172,000 in seq CAMP. Time spent on CAMP so far: 8.53 sec.\n",
      "\tOn aln 173,000 in seq CAMP. Time spent on CAMP so far: 8.58 sec.\n",
      "\tOn aln 174,000 in seq CAMP. Time spent on CAMP so far: 8.63 sec.\n",
      "\tOn aln 175,000 in seq CAMP. Time spent on CAMP so far: 8.68 sec.\n",
      "\tOn aln 176,000 in seq CAMP. Time spent on CAMP so far: 8.73 sec.\n",
      "\tOn aln 177,000 in seq CAMP. Time spent on CAMP so far: 8.78 sec.\n",
      "\tOn aln 178,000 in seq CAMP. Time spent on CAMP so far: 8.83 sec.\n",
      "\tOn aln 179,000 in seq CAMP. Time spent on CAMP so far: 8.88 sec.\n",
      "\tOn aln 180,000 in seq CAMP. Time spent on CAMP so far: 8.93 sec.\n",
      "\tOn aln 181,000 in seq CAMP. Time spent on CAMP so far: 8.98 sec.\n",
      "\tOn aln 182,000 in seq CAMP. Time spent on CAMP so far: 9.03 sec.\n",
      "\tOn aln 183,000 in seq CAMP. Time spent on CAMP so far: 9.08 sec.\n",
      "\tOn aln 184,000 in seq CAMP. Time spent on CAMP so far: 9.13 sec.\n",
      "\tOn aln 185,000 in seq CAMP. Time spent on CAMP so far: 9.18 sec.\n",
      "\tOn aln 186,000 in seq CAMP. Time spent on CAMP so far: 9.23 sec.\n",
      "\tOn aln 187,000 in seq CAMP. Time spent on CAMP so far: 9.28 sec.\n",
      "\tOn aln 188,000 in seq CAMP. Time spent on CAMP so far: 9.33 sec.\n",
      "\tOn aln 189,000 in seq CAMP. Time spent on CAMP so far: 9.37 sec.\n",
      "\tOn aln 190,000 in seq CAMP. Time spent on CAMP so far: 9.42 sec.\n",
      "\tOn aln 191,000 in seq CAMP. Time spent on CAMP so far: 9.47 sec.\n",
      "\tOn aln 192,000 in seq CAMP. Time spent on CAMP so far: 9.52 sec.\n",
      "\tOn aln 193,000 in seq CAMP. Time spent on CAMP so far: 9.57 sec.\n",
      "\tOn aln 194,000 in seq CAMP. Time spent on CAMP so far: 9.62 sec.\n",
      "\tOn aln 195,000 in seq CAMP. Time spent on CAMP so far: 9.67 sec.\n",
      "\tOn aln 196,000 in seq CAMP. Time spent on CAMP so far: 9.72 sec.\n",
      "\tOn aln 197,000 in seq CAMP. Time spent on CAMP so far: 9.77 sec.\n",
      "\tOn aln 198,000 in seq CAMP. Time spent on CAMP so far: 9.82 sec.\n",
      "\tOn aln 199,000 in seq CAMP. Time spent on CAMP so far: 9.87 sec.\n",
      "\tOn aln 200,000 in seq CAMP. Time spent on CAMP so far: 9.92 sec.\n",
      "\tOn aln 201,000 in seq CAMP. Time spent on CAMP so far: 9.97 sec.\n",
      "\tOn aln 202,000 in seq CAMP. Time spent on CAMP so far: 10.02 sec.\n",
      "\tOn aln 203,000 in seq CAMP. Time spent on CAMP so far: 10.07 sec.\n",
      "\tOn aln 204,000 in seq CAMP. Time spent on CAMP so far: 10.12 sec.\n",
      "\tOn aln 205,000 in seq CAMP. Time spent on CAMP so far: 10.17 sec.\n",
      "\tOn aln 206,000 in seq CAMP. Time spent on CAMP so far: 10.21 sec.\n",
      "\tOn aln 207,000 in seq CAMP. Time spent on CAMP so far: 10.27 sec.\n",
      "\tOn aln 208,000 in seq CAMP. Time spent on CAMP so far: 10.32 sec.\n",
      "\tOn aln 209,000 in seq CAMP. Time spent on CAMP so far: 10.37 sec.\n",
      "\tOn aln 210,000 in seq CAMP. Time spent on CAMP so far: 10.42 sec.\n",
      "\tOn aln 211,000 in seq CAMP. Time spent on CAMP so far: 10.47 sec.\n",
      "\tOn aln 212,000 in seq CAMP. Time spent on CAMP so far: 10.52 sec.\n",
      "\tOn aln 213,000 in seq CAMP. Time spent on CAMP so far: 10.57 sec.\n",
      "\tOn aln 214,000 in seq CAMP. Time spent on CAMP so far: 10.62 sec.\n",
      "\tOn aln 215,000 in seq CAMP. Time spent on CAMP so far: 10.67 sec.\n",
      "\tOn aln 216,000 in seq CAMP. Time spent on CAMP so far: 10.72 sec.\n",
      "\tOn aln 217,000 in seq CAMP. Time spent on CAMP so far: 10.77 sec.\n",
      "\tOn aln 218,000 in seq CAMP. Time spent on CAMP so far: 10.82 sec.\n",
      "\tOn aln 219,000 in seq CAMP. Time spent on CAMP so far: 10.87 sec.\n",
      "\tOn aln 220,000 in seq CAMP. Time spent on CAMP so far: 10.93 sec.\n",
      "\tOn aln 221,000 in seq CAMP. Time spent on CAMP so far: 10.98 sec.\n",
      "\tOn aln 222,000 in seq CAMP. Time spent on CAMP so far: 11.03 sec.\n",
      "\tOn aln 223,000 in seq CAMP. Time spent on CAMP so far: 11.08 sec.\n",
      "\tOn aln 224,000 in seq CAMP. Time spent on CAMP so far: 11.13 sec.\n",
      "\tOn aln 225,000 in seq CAMP. Time spent on CAMP so far: 11.18 sec.\n",
      "\tOn aln 226,000 in seq CAMP. Time spent on CAMP so far: 11.23 sec.\n",
      "\tOn aln 227,000 in seq CAMP. Time spent on CAMP so far: 11.28 sec.\n",
      "\tOn aln 228,000 in seq CAMP. Time spent on CAMP so far: 11.33 sec.\n",
      "\tOn aln 229,000 in seq CAMP. Time spent on CAMP so far: 11.38 sec.\n",
      "\tOn aln 230,000 in seq CAMP. Time spent on CAMP so far: 11.43 sec.\n",
      "\tOn aln 231,000 in seq CAMP. Time spent on CAMP so far: 11.48 sec.\n",
      "\tOn aln 232,000 in seq CAMP. Time spent on CAMP so far: 11.53 sec.\n",
      "\tOn aln 233,000 in seq CAMP. Time spent on CAMP so far: 11.58 sec.\n",
      "\tOn aln 234,000 in seq CAMP. Time spent on CAMP so far: 11.63 sec.\n",
      "\tOn aln 235,000 in seq CAMP. Time spent on CAMP so far: 11.69 sec.\n",
      "\tOn aln 236,000 in seq CAMP. Time spent on CAMP so far: 11.74 sec.\n",
      "\tOn aln 237,000 in seq CAMP. Time spent on CAMP so far: 11.79 sec.\n",
      "\tOn aln 238,000 in seq CAMP. Time spent on CAMP so far: 11.84 sec.\n",
      "\tOn aln 239,000 in seq CAMP. Time spent on CAMP so far: 11.89 sec.\n",
      "\tOn aln 240,000 in seq CAMP. Time spent on CAMP so far: 11.94 sec.\n",
      "\tOn aln 241,000 in seq CAMP. Time spent on CAMP so far: 11.99 sec.\n",
      "\tOn aln 242,000 in seq CAMP. Time spent on CAMP so far: 12.04 sec.\n",
      "\tOn aln 243,000 in seq CAMP. Time spent on CAMP so far: 12.09 sec.\n",
      "\tOn aln 244,000 in seq CAMP. Time spent on CAMP so far: 12.14 sec.\n",
      "\tOn aln 245,000 in seq CAMP. Time spent on CAMP so far: 12.19 sec.\n",
      "\tOn aln 246,000 in seq CAMP. Time spent on CAMP so far: 12.24 sec.\n",
      "\tOn aln 247,000 in seq CAMP. Time spent on CAMP so far: 12.29 sec.\n",
      "\tOn aln 248,000 in seq CAMP. Time spent on CAMP so far: 12.34 sec.\n",
      "\tOn aln 249,000 in seq CAMP. Time spent on CAMP so far: 12.39 sec.\n",
      "\tOn aln 250,000 in seq CAMP. Time spent on CAMP so far: 12.44 sec.\n",
      "\tOn aln 251,000 in seq CAMP. Time spent on CAMP so far: 12.49 sec.\n",
      "\tOn aln 252,000 in seq CAMP. Time spent on CAMP so far: 12.54 sec.\n",
      "\tOn aln 253,000 in seq CAMP. Time spent on CAMP so far: 12.59 sec.\n",
      "\tOn aln 254,000 in seq CAMP. Time spent on CAMP so far: 12.64 sec.\n",
      "\tOn aln 255,000 in seq CAMP. Time spent on CAMP so far: 12.69 sec.\n",
      "\tOn aln 256,000 in seq CAMP. Time spent on CAMP so far: 12.74 sec.\n",
      "\tOn aln 257,000 in seq CAMP. Time spent on CAMP so far: 12.79 sec.\n",
      "\tOn aln 258,000 in seq CAMP. Time spent on CAMP so far: 12.84 sec.\n",
      "\tOn aln 259,000 in seq CAMP. Time spent on CAMP so far: 12.89 sec.\n",
      "\tOn aln 260,000 in seq CAMP. Time spent on CAMP so far: 12.94 sec.\n",
      "\tOn aln 261,000 in seq CAMP. Time spent on CAMP so far: 12.99 sec.\n",
      "\tOn aln 262,000 in seq CAMP. Time spent on CAMP so far: 13.04 sec.\n",
      "\tOn aln 263,000 in seq CAMP. Time spent on CAMP so far: 13.09 sec.\n",
      "\tOn aln 264,000 in seq CAMP. Time spent on CAMP so far: 13.14 sec.\n",
      "\tOn aln 265,000 in seq CAMP. Time spent on CAMP so far: 13.19 sec.\n",
      "\tOn aln 266,000 in seq CAMP. Time spent on CAMP so far: 13.24 sec.\n",
      "\tOn aln 267,000 in seq CAMP. Time spent on CAMP so far: 13.28 sec.\n",
      "\tOn aln 268,000 in seq CAMP. Time spent on CAMP so far: 13.33 sec.\n",
      "\tOn aln 269,000 in seq CAMP. Time spent on CAMP so far: 13.38 sec.\n",
      "\tOn aln 270,000 in seq CAMP. Time spent on CAMP so far: 13.43 sec.\n",
      "\tOn aln 271,000 in seq CAMP. Time spent on CAMP so far: 13.48 sec.\n",
      "\tOn aln 272,000 in seq CAMP. Time spent on CAMP so far: 13.53 sec.\n",
      "\tOn aln 273,000 in seq CAMP. Time spent on CAMP so far: 13.58 sec.\n",
      "\tOn aln 274,000 in seq CAMP. Time spent on CAMP so far: 13.63 sec.\n",
      "\tOn aln 275,000 in seq CAMP. Time spent on CAMP so far: 13.68 sec.\n",
      "\tOn aln 276,000 in seq CAMP. Time spent on CAMP so far: 13.73 sec.\n",
      "\tOn aln 277,000 in seq CAMP. Time spent on CAMP so far: 13.78 sec.\n",
      "\tOn aln 278,000 in seq CAMP. Time spent on CAMP so far: 13.83 sec.\n",
      "\tOn aln 279,000 in seq CAMP. Time spent on CAMP so far: 13.88 sec.\n",
      "\tOn aln 280,000 in seq CAMP. Time spent on CAMP so far: 13.93 sec.\n",
      "\tOn aln 281,000 in seq CAMP. Time spent on CAMP so far: 13.98 sec.\n",
      "\tOn aln 282,000 in seq CAMP. Time spent on CAMP so far: 14.03 sec.\n",
      "\tOn aln 283,000 in seq CAMP. Time spent on CAMP so far: 14.08 sec.\n",
      "\tOn aln 284,000 in seq CAMP. Time spent on CAMP so far: 14.13 sec.\n",
      "\tOn aln 285,000 in seq CAMP. Time spent on CAMP so far: 14.18 sec.\n",
      "\tOn aln 286,000 in seq CAMP. Time spent on CAMP so far: 14.23 sec.\n",
      "\tOn aln 287,000 in seq CAMP. Time spent on CAMP so far: 14.27 sec.\n",
      "\tOn aln 288,000 in seq CAMP. Time spent on CAMP so far: 14.32 sec.\n",
      "\tOn aln 289,000 in seq CAMP. Time spent on CAMP so far: 14.37 sec.\n",
      "\tOn aln 290,000 in seq CAMP. Time spent on CAMP so far: 14.42 sec.\n",
      "\tOn aln 291,000 in seq CAMP. Time spent on CAMP so far: 14.47 sec.\n",
      "\tOn aln 292,000 in seq CAMP. Time spent on CAMP so far: 14.52 sec.\n",
      "\tOn aln 293,000 in seq CAMP. Time spent on CAMP so far: 14.57 sec.\n",
      "\tOn aln 294,000 in seq CAMP. Time spent on CAMP so far: 14.62 sec.\n",
      "\tOn aln 295,000 in seq CAMP. Time spent on CAMP so far: 14.66 sec.\n",
      "\tOn aln 296,000 in seq CAMP. Time spent on CAMP so far: 14.71 sec.\n",
      "\tOn aln 297,000 in seq CAMP. Time spent on CAMP so far: 14.76 sec.\n",
      "\tOn aln 298,000 in seq CAMP. Time spent on CAMP so far: 14.81 sec.\n",
      "\tOn aln 299,000 in seq CAMP. Time spent on CAMP so far: 14.86 sec.\n",
      "\tOn aln 300,000 in seq CAMP. Time spent on CAMP so far: 14.91 sec.\n",
      "\tOn aln 301,000 in seq CAMP. Time spent on CAMP so far: 14.96 sec.\n",
      "\tOn aln 302,000 in seq CAMP. Time spent on CAMP so far: 15.01 sec.\n",
      "\tOn aln 303,000 in seq CAMP. Time spent on CAMP so far: 15.06 sec.\n",
      "\tOn aln 304,000 in seq CAMP. Time spent on CAMP so far: 15.11 sec.\n",
      "\tOn aln 305,000 in seq CAMP. Time spent on CAMP so far: 15.16 sec.\n",
      "\tOn aln 306,000 in seq CAMP. Time spent on CAMP so far: 15.21 sec.\n",
      "\tOn aln 307,000 in seq CAMP. Time spent on CAMP so far: 15.26 sec.\n",
      "\tOn aln 308,000 in seq CAMP. Time spent on CAMP so far: 15.31 sec.\n",
      "\tOn aln 309,000 in seq CAMP. Time spent on CAMP so far: 15.36 sec.\n",
      "\tOn aln 310,000 in seq CAMP. Time spent on CAMP so far: 15.41 sec.\n",
      "\tOn aln 311,000 in seq CAMP. Time spent on CAMP so far: 15.46 sec.\n",
      "\tOn aln 312,000 in seq CAMP. Time spent on CAMP so far: 15.51 sec.\n",
      "\tOn aln 313,000 in seq CAMP. Time spent on CAMP so far: 15.56 sec.\n",
      "\tOn aln 314,000 in seq CAMP. Time spent on CAMP so far: 15.61 sec.\n",
      "\tOn aln 315,000 in seq CAMP. Time spent on CAMP so far: 15.66 sec.\n",
      "\tOn aln 316,000 in seq CAMP. Time spent on CAMP so far: 15.71 sec.\n",
      "\tOn aln 317,000 in seq CAMP. Time spent on CAMP so far: 15.76 sec.\n",
      "\tOn aln 318,000 in seq CAMP. Time spent on CAMP so far: 15.81 sec.\n",
      "\tOn aln 319,000 in seq CAMP. Time spent on CAMP so far: 15.86 sec.\n",
      "\tOn aln 320,000 in seq CAMP. Time spent on CAMP so far: 15.90 sec.\n",
      "\tOn aln 321,000 in seq CAMP. Time spent on CAMP so far: 15.95 sec.\n",
      "\tOn aln 322,000 in seq CAMP. Time spent on CAMP so far: 16.00 sec.\n",
      "\tOn aln 323,000 in seq CAMP. Time spent on CAMP so far: 16.05 sec.\n",
      "\tOn aln 324,000 in seq CAMP. Time spent on CAMP so far: 16.10 sec.\n",
      "\tOn aln 325,000 in seq CAMP. Time spent on CAMP so far: 16.15 sec.\n",
      "\tOn aln 326,000 in seq CAMP. Time spent on CAMP so far: 16.20 sec.\n",
      "\tOn aln 327,000 in seq CAMP. Time spent on CAMP so far: 16.25 sec.\n",
      "\tOn aln 328,000 in seq CAMP. Time spent on CAMP so far: 16.30 sec.\n",
      "\tOn aln 329,000 in seq CAMP. Time spent on CAMP so far: 16.35 sec.\n",
      "\tOn aln 330,000 in seq CAMP. Time spent on CAMP so far: 16.40 sec.\n",
      "\tOn aln 331,000 in seq CAMP. Time spent on CAMP so far: 16.45 sec.\n",
      "\tOn aln 332,000 in seq CAMP. Time spent on CAMP so far: 16.49 sec.\n",
      "\tOn aln 333,000 in seq CAMP. Time spent on CAMP so far: 16.54 sec.\n",
      "\tOn aln 334,000 in seq CAMP. Time spent on CAMP so far: 16.59 sec.\n",
      "\tOn aln 335,000 in seq CAMP. Time spent on CAMP so far: 16.65 sec.\n",
      "\tOn aln 336,000 in seq CAMP. Time spent on CAMP so far: 16.70 sec.\n",
      "\tOn aln 337,000 in seq CAMP. Time spent on CAMP so far: 16.75 sec.\n",
      "\tOn aln 338,000 in seq CAMP. Time spent on CAMP so far: 16.80 sec.\n",
      "\tOn aln 339,000 in seq CAMP. Time spent on CAMP so far: 16.85 sec.\n",
      "\tOn aln 340,000 in seq CAMP. Time spent on CAMP so far: 16.90 sec.\n",
      "\tOn aln 341,000 in seq CAMP. Time spent on CAMP so far: 16.95 sec.\n",
      "\tOn aln 342,000 in seq CAMP. Time spent on CAMP so far: 17.00 sec.\n",
      "\tOn aln 343,000 in seq CAMP. Time spent on CAMP so far: 17.05 sec.\n",
      "\tOn aln 344,000 in seq CAMP. Time spent on CAMP so far: 17.10 sec.\n",
      "\tOn aln 345,000 in seq CAMP. Time spent on CAMP so far: 17.14 sec.\n",
      "\tOn aln 346,000 in seq CAMP. Time spent on CAMP so far: 17.19 sec.\n",
      "\tOn aln 347,000 in seq CAMP. Time spent on CAMP so far: 17.24 sec.\n",
      "\tOn aln 348,000 in seq CAMP. Time spent on CAMP so far: 17.29 sec.\n",
      "\tOn aln 349,000 in seq CAMP. Time spent on CAMP so far: 17.34 sec.\n",
      "\tOn aln 350,000 in seq CAMP. Time spent on CAMP so far: 17.40 sec.\n",
      "\tOn aln 351,000 in seq CAMP. Time spent on CAMP so far: 17.45 sec.\n",
      "\tOn aln 352,000 in seq CAMP. Time spent on CAMP so far: 17.50 sec.\n",
      "\tOn aln 353,000 in seq CAMP. Time spent on CAMP so far: 17.55 sec.\n",
      "\tOn aln 354,000 in seq CAMP. Time spent on CAMP so far: 17.60 sec.\n",
      "\tOn aln 355,000 in seq CAMP. Time spent on CAMP so far: 17.65 sec.\n",
      "\tOn aln 356,000 in seq CAMP. Time spent on CAMP so far: 17.70 sec.\n",
      "\tOn aln 357,000 in seq CAMP. Time spent on CAMP so far: 17.75 sec.\n",
      "\tOn aln 358,000 in seq CAMP. Time spent on CAMP so far: 17.80 sec.\n",
      "\tOn aln 359,000 in seq CAMP. Time spent on CAMP so far: 17.85 sec.\n",
      "\tOn aln 360,000 in seq CAMP. Time spent on CAMP so far: 17.90 sec.\n",
      "\tOn aln 361,000 in seq CAMP. Time spent on CAMP so far: 17.95 sec.\n",
      "\tOn aln 362,000 in seq CAMP. Time spent on CAMP so far: 18.00 sec.\n",
      "\tOn aln 363,000 in seq CAMP. Time spent on CAMP so far: 18.05 sec.\n",
      "\tOn aln 364,000 in seq CAMP. Time spent on CAMP so far: 18.10 sec.\n",
      "\tOn aln 365,000 in seq CAMP. Time spent on CAMP so far: 18.15 sec.\n",
      "\tOn aln 366,000 in seq CAMP. Time spent on CAMP so far: 18.21 sec.\n",
      "\tOn aln 367,000 in seq CAMP. Time spent on CAMP so far: 18.26 sec.\n",
      "\tOn aln 368,000 in seq CAMP. Time spent on CAMP so far: 18.31 sec.\n",
      "\tOn aln 369,000 in seq CAMP. Time spent on CAMP so far: 18.36 sec.\n",
      "\tOn aln 370,000 in seq CAMP. Time spent on CAMP so far: 18.41 sec.\n",
      "\tOn aln 371,000 in seq CAMP. Time spent on CAMP so far: 18.46 sec.\n",
      "\tOn aln 372,000 in seq CAMP. Time spent on CAMP so far: 18.51 sec.\n",
      "\tOn aln 373,000 in seq CAMP. Time spent on CAMP so far: 18.56 sec.\n",
      "\tOn aln 374,000 in seq CAMP. Time spent on CAMP so far: 18.61 sec.\n",
      "\tOn aln 375,000 in seq CAMP. Time spent on CAMP so far: 18.66 sec.\n",
      "\tOn aln 376,000 in seq CAMP. Time spent on CAMP so far: 18.71 sec.\n",
      "\tOn aln 377,000 in seq CAMP. Time spent on CAMP so far: 18.76 sec.\n",
      "\tOn aln 378,000 in seq CAMP. Time spent on CAMP so far: 18.81 sec.\n",
      "\tOn aln 379,000 in seq CAMP. Time spent on CAMP so far: 18.86 sec.\n",
      "\tOn aln 380,000 in seq CAMP. Time spent on CAMP so far: 18.90 sec.\n",
      "\tOn aln 381,000 in seq CAMP. Time spent on CAMP so far: 18.95 sec.\n",
      "\tOn aln 382,000 in seq CAMP. Time spent on CAMP so far: 19.00 sec.\n",
      "\tOn aln 383,000 in seq CAMP. Time spent on CAMP so far: 19.04 sec.\n",
      "\tOn aln 384,000 in seq CAMP. Time spent on CAMP so far: 19.09 sec.\n",
      "\tOn aln 385,000 in seq CAMP. Time spent on CAMP so far: 19.14 sec.\n",
      "\tOn aln 386,000 in seq CAMP. Time spent on CAMP so far: 19.19 sec.\n",
      "\tOn aln 387,000 in seq CAMP. Time spent on CAMP so far: 19.24 sec.\n",
      "\tOn aln 388,000 in seq CAMP. Time spent on CAMP so far: 19.29 sec.\n",
      "\tOn aln 389,000 in seq CAMP. Time spent on CAMP so far: 19.34 sec.\n",
      "\tOn aln 390,000 in seq CAMP. Time spent on CAMP so far: 19.39 sec.\n",
      "\tOn aln 391,000 in seq CAMP. Time spent on CAMP so far: 19.44 sec.\n",
      "\tOn aln 392,000 in seq CAMP. Time spent on CAMP so far: 19.49 sec.\n",
      "\tOn aln 393,000 in seq CAMP. Time spent on CAMP so far: 19.54 sec.\n",
      "\tOn aln 394,000 in seq CAMP. Time spent on CAMP so far: 19.59 sec.\n",
      "\tOn aln 395,000 in seq CAMP. Time spent on CAMP so far: 19.64 sec.\n",
      "\tOn aln 396,000 in seq CAMP. Time spent on CAMP so far: 19.69 sec.\n",
      "\tOn aln 397,000 in seq CAMP. Time spent on CAMP so far: 19.74 sec.\n",
      "\tOn aln 398,000 in seq CAMP. Time spent on CAMP so far: 19.79 sec.\n",
      "\tOn aln 399,000 in seq CAMP. Time spent on CAMP so far: 19.84 sec.\n",
      "\tOn aln 400,000 in seq CAMP. Time spent on CAMP so far: 19.89 sec.\n",
      "\tOn aln 401,000 in seq CAMP. Time spent on CAMP so far: 19.94 sec.\n",
      "\tOn aln 402,000 in seq CAMP. Time spent on CAMP so far: 19.99 sec.\n",
      "\tOn aln 403,000 in seq CAMP. Time spent on CAMP so far: 20.03 sec.\n",
      "\tOn aln 404,000 in seq CAMP. Time spent on CAMP so far: 20.08 sec.\n",
      "\tOn aln 405,000 in seq CAMP. Time spent on CAMP so far: 20.13 sec.\n",
      "\tOn aln 406,000 in seq CAMP. Time spent on CAMP so far: 20.18 sec.\n",
      "\tOn aln 407,000 in seq CAMP. Time spent on CAMP so far: 20.23 sec.\n",
      "\tOn aln 408,000 in seq CAMP. Time spent on CAMP so far: 20.28 sec.\n",
      "\tOn aln 409,000 in seq CAMP. Time spent on CAMP so far: 20.33 sec.\n",
      "\tOn aln 410,000 in seq CAMP. Time spent on CAMP so far: 20.38 sec.\n",
      "\tOn aln 411,000 in seq CAMP. Time spent on CAMP so far: 20.43 sec.\n",
      "\tOn aln 412,000 in seq CAMP. Time spent on CAMP so far: 20.47 sec.\n",
      "\tOn aln 413,000 in seq CAMP. Time spent on CAMP so far: 20.52 sec.\n",
      "\tOn aln 414,000 in seq CAMP. Time spent on CAMP so far: 20.57 sec.\n",
      "\tOn aln 415,000 in seq CAMP. Time spent on CAMP so far: 20.62 sec.\n",
      "\tOn aln 416,000 in seq CAMP. Time spent on CAMP so far: 20.67 sec.\n",
      "\tOn aln 417,000 in seq CAMP. Time spent on CAMP so far: 20.72 sec.\n",
      "\tOn aln 418,000 in seq CAMP. Time spent on CAMP so far: 20.77 sec.\n",
      "\tOn aln 419,000 in seq CAMP. Time spent on CAMP so far: 20.82 sec.\n",
      "\tOn aln 420,000 in seq CAMP. Time spent on CAMP so far: 20.87 sec.\n",
      "\tOn aln 421,000 in seq CAMP. Time spent on CAMP so far: 20.92 sec.\n",
      "\tOn aln 422,000 in seq CAMP. Time spent on CAMP so far: 20.97 sec.\n",
      "\tOn aln 423,000 in seq CAMP. Time spent on CAMP so far: 21.02 sec.\n",
      "\tOn aln 424,000 in seq CAMP. Time spent on CAMP so far: 21.07 sec.\n",
      "\tOn aln 425,000 in seq CAMP. Time spent on CAMP so far: 21.12 sec.\n",
      "\tOn aln 426,000 in seq CAMP. Time spent on CAMP so far: 21.16 sec.\n",
      "\tOn aln 427,000 in seq CAMP. Time spent on CAMP so far: 21.21 sec.\n",
      "\tOn aln 428,000 in seq CAMP. Time spent on CAMP so far: 21.26 sec.\n",
      "\tOn aln 429,000 in seq CAMP. Time spent on CAMP so far: 21.31 sec.\n",
      "\tOn aln 430,000 in seq CAMP. Time spent on CAMP so far: 21.36 sec.\n",
      "\tOn aln 431,000 in seq CAMP. Time spent on CAMP so far: 21.41 sec.\n",
      "\tOn aln 432,000 in seq CAMP. Time spent on CAMP so far: 21.46 sec.\n",
      "\tOn aln 433,000 in seq CAMP. Time spent on CAMP so far: 21.51 sec.\n",
      "\tOn aln 434,000 in seq CAMP. Time spent on CAMP so far: 21.56 sec.\n",
      "\tOn aln 435,000 in seq CAMP. Time spent on CAMP so far: 21.61 sec.\n",
      "\tOn aln 436,000 in seq CAMP. Time spent on CAMP so far: 21.66 sec.\n",
      "\tOn aln 437,000 in seq CAMP. Time spent on CAMP so far: 21.71 sec.\n",
      "\tOn aln 438,000 in seq CAMP. Time spent on CAMP so far: 21.76 sec.\n",
      "\tOn aln 439,000 in seq CAMP. Time spent on CAMP so far: 21.81 sec.\n",
      "\tOn aln 440,000 in seq CAMP. Time spent on CAMP so far: 21.86 sec.\n",
      "\tOn aln 441,000 in seq CAMP. Time spent on CAMP so far: 21.91 sec.\n",
      "\tOn aln 442,000 in seq CAMP. Time spent on CAMP so far: 21.96 sec.\n",
      "\tOn aln 443,000 in seq CAMP. Time spent on CAMP so far: 22.01 sec.\n",
      "\tOn aln 444,000 in seq CAMP. Time spent on CAMP so far: 22.06 sec.\n",
      "\tOn aln 445,000 in seq CAMP. Time spent on CAMP so far: 22.11 sec.\n",
      "\tOn aln 446,000 in seq CAMP. Time spent on CAMP so far: 22.16 sec.\n",
      "\tOn aln 447,000 in seq CAMP. Time spent on CAMP so far: 22.20 sec.\n",
      "\tOn aln 448,000 in seq CAMP. Time spent on CAMP so far: 22.25 sec.\n",
      "\tOn aln 449,000 in seq CAMP. Time spent on CAMP so far: 22.30 sec.\n",
      "\tOn aln 450,000 in seq CAMP. Time spent on CAMP so far: 22.35 sec.\n",
      "\tOn aln 451,000 in seq CAMP. Time spent on CAMP so far: 22.40 sec.\n",
      "\tOn aln 452,000 in seq CAMP. Time spent on CAMP so far: 22.45 sec.\n",
      "\tOn aln 453,000 in seq CAMP. Time spent on CAMP so far: 22.50 sec.\n",
      "\tOn aln 454,000 in seq CAMP. Time spent on CAMP so far: 22.55 sec.\n",
      "\tOn aln 455,000 in seq CAMP. Time spent on CAMP so far: 22.60 sec.\n",
      "\tOn aln 456,000 in seq CAMP. Time spent on CAMP so far: 22.65 sec.\n",
      "\tOn aln 457,000 in seq CAMP. Time spent on CAMP so far: 22.70 sec.\n",
      "\tOn aln 458,000 in seq CAMP. Time spent on CAMP so far: 22.75 sec.\n",
      "\tOn aln 459,000 in seq CAMP. Time spent on CAMP so far: 22.80 sec.\n",
      "\tOn aln 460,000 in seq CAMP. Time spent on CAMP so far: 22.85 sec.\n",
      "\tOn aln 461,000 in seq CAMP. Time spent on CAMP so far: 22.90 sec.\n",
      "\tOn aln 462,000 in seq CAMP. Time spent on CAMP so far: 22.95 sec.\n",
      "\tOn aln 463,000 in seq CAMP. Time spent on CAMP so far: 23.00 sec.\n",
      "\tOn aln 464,000 in seq CAMP. Time spent on CAMP so far: 23.05 sec.\n",
      "\tOn aln 465,000 in seq CAMP. Time spent on CAMP so far: 23.10 sec.\n",
      "\tOn aln 466,000 in seq CAMP. Time spent on CAMP so far: 23.15 sec.\n",
      "\tOn aln 467,000 in seq CAMP. Time spent on CAMP so far: 23.20 sec.\n",
      "\tOn aln 468,000 in seq CAMP. Time spent on CAMP so far: 23.25 sec.\n",
      "\tOn aln 469,000 in seq CAMP. Time spent on CAMP so far: 23.30 sec.\n",
      "\tOn aln 470,000 in seq CAMP. Time spent on CAMP so far: 23.35 sec.\n",
      "\tOn aln 471,000 in seq CAMP. Time spent on CAMP so far: 23.40 sec.\n",
      "\tOn aln 472,000 in seq CAMP. Time spent on CAMP so far: 23.45 sec.\n",
      "\tOn aln 473,000 in seq CAMP. Time spent on CAMP so far: 23.50 sec.\n",
      "\tOn aln 474,000 in seq CAMP. Time spent on CAMP so far: 23.55 sec.\n",
      "\tOn aln 475,000 in seq CAMP. Time spent on CAMP so far: 23.60 sec.\n",
      "\tOn aln 476,000 in seq CAMP. Time spent on CAMP so far: 23.64 sec.\n",
      "Done with edge_6104! Took 23.66 sec.\n",
      "Identifying mutated positions (p = 1%) in genome BACT1...\n",
      "Found 22,144 mutated positions (p = 1%) in BACT1.\n",
      "Note that this tally is higher than you'd see in e.g. the CP1/2/3 plots, because now we're including both 'rare' and non-rare mutations. Just so you don't waste five minutes sanity-checking this like I did.\n",
      "Going through these positions...\n",
      "\tOn aln 1,000 in seq BACT1. Time spent on BACT1 so far: 0.04 sec.\n",
      "\tOn aln 2,000 in seq BACT1. Time spent on BACT1 so far: 0.08 sec.\n",
      "\tOn aln 3,000 in seq BACT1. Time spent on BACT1 so far: 0.13 sec.\n",
      "\tOn aln 4,000 in seq BACT1. Time spent on BACT1 so far: 0.18 sec.\n",
      "\tOn aln 5,000 in seq BACT1. Time spent on BACT1 so far: 0.23 sec.\n",
      "\tOn aln 6,000 in seq BACT1. Time spent on BACT1 so far: 0.28 sec.\n",
      "\tOn aln 7,000 in seq BACT1. Time spent on BACT1 so far: 0.33 sec.\n",
      "\tOn aln 8,000 in seq BACT1. Time spent on BACT1 so far: 0.39 sec.\n",
      "\tOn aln 9,000 in seq BACT1. Time spent on BACT1 so far: 0.44 sec.\n",
      "\tOn aln 10,000 in seq BACT1. Time spent on BACT1 so far: 0.49 sec.\n",
      "\tOn aln 11,000 in seq BACT1. Time spent on BACT1 so far: 0.54 sec.\n",
      "\tOn aln 12,000 in seq BACT1. Time spent on BACT1 so far: 0.59 sec.\n",
      "\tOn aln 13,000 in seq BACT1. Time spent on BACT1 so far: 0.64 sec.\n",
      "\tOn aln 14,000 in seq BACT1. Time spent on BACT1 so far: 0.69 sec.\n",
      "\tOn aln 15,000 in seq BACT1. Time spent on BACT1 so far: 0.74 sec.\n",
      "\tOn aln 16,000 in seq BACT1. Time spent on BACT1 so far: 0.79 sec.\n",
      "\tOn aln 17,000 in seq BACT1. Time spent on BACT1 so far: 0.84 sec.\n",
      "\tOn aln 18,000 in seq BACT1. Time spent on BACT1 so far: 0.89 sec.\n",
      "\tOn aln 19,000 in seq BACT1. Time spent on BACT1 so far: 0.95 sec.\n",
      "\tOn aln 20,000 in seq BACT1. Time spent on BACT1 so far: 1.00 sec.\n",
      "\tOn aln 21,000 in seq BACT1. Time spent on BACT1 so far: 1.05 sec.\n",
      "\tOn aln 22,000 in seq BACT1. Time spent on BACT1 so far: 1.11 sec.\n",
      "\tOn aln 23,000 in seq BACT1. Time spent on BACT1 so far: 1.16 sec.\n",
      "\tOn aln 24,000 in seq BACT1. Time spent on BACT1 so far: 1.21 sec.\n",
      "\tOn aln 25,000 in seq BACT1. Time spent on BACT1 so far: 1.26 sec.\n",
      "\tOn aln 26,000 in seq BACT1. Time spent on BACT1 so far: 1.31 sec.\n",
      "\tOn aln 27,000 in seq BACT1. Time spent on BACT1 so far: 1.36 sec.\n",
      "\tOn aln 28,000 in seq BACT1. Time spent on BACT1 so far: 1.42 sec.\n",
      "\tOn aln 29,000 in seq BACT1. Time spent on BACT1 so far: 1.47 sec.\n",
      "\tOn aln 30,000 in seq BACT1. Time spent on BACT1 so far: 1.52 sec.\n",
      "\tOn aln 31,000 in seq BACT1. Time spent on BACT1 so far: 1.57 sec.\n",
      "\tOn aln 32,000 in seq BACT1. Time spent on BACT1 so far: 1.62 sec.\n",
      "\tOn aln 33,000 in seq BACT1. Time spent on BACT1 so far: 1.67 sec.\n",
      "\tOn aln 34,000 in seq BACT1. Time spent on BACT1 so far: 1.73 sec.\n",
      "\tOn aln 35,000 in seq BACT1. Time spent on BACT1 so far: 1.78 sec.\n",
      "\tOn aln 36,000 in seq BACT1. Time spent on BACT1 so far: 1.83 sec.\n",
      "\tOn aln 37,000 in seq BACT1. Time spent on BACT1 so far: 1.88 sec.\n",
      "\tOn aln 38,000 in seq BACT1. Time spent on BACT1 so far: 1.93 sec.\n",
      "\tOn aln 39,000 in seq BACT1. Time spent on BACT1 so far: 1.99 sec.\n",
      "\tOn aln 40,000 in seq BACT1. Time spent on BACT1 so far: 2.04 sec.\n",
      "\tOn aln 41,000 in seq BACT1. Time spent on BACT1 so far: 2.09 sec.\n",
      "\tOn aln 42,000 in seq BACT1. Time spent on BACT1 so far: 2.14 sec.\n",
      "\tOn aln 43,000 in seq BACT1. Time spent on BACT1 so far: 2.19 sec.\n",
      "\tOn aln 44,000 in seq BACT1. Time spent on BACT1 so far: 2.24 sec.\n",
      "\tOn aln 45,000 in seq BACT1. Time spent on BACT1 so far: 2.30 sec.\n",
      "\tOn aln 46,000 in seq BACT1. Time spent on BACT1 so far: 2.35 sec.\n",
      "\tOn aln 47,000 in seq BACT1. Time spent on BACT1 so far: 2.40 sec.\n",
      "\tOn aln 48,000 in seq BACT1. Time spent on BACT1 so far: 2.45 sec.\n",
      "\tOn aln 49,000 in seq BACT1. Time spent on BACT1 so far: 2.50 sec.\n",
      "\tOn aln 50,000 in seq BACT1. Time spent on BACT1 so far: 2.55 sec.\n",
      "\tOn aln 51,000 in seq BACT1. Time spent on BACT1 so far: 2.61 sec.\n",
      "\tOn aln 52,000 in seq BACT1. Time spent on BACT1 so far: 2.66 sec.\n",
      "\tOn aln 53,000 in seq BACT1. Time spent on BACT1 so far: 2.71 sec.\n",
      "\tOn aln 54,000 in seq BACT1. Time spent on BACT1 so far: 2.76 sec.\n",
      "\tOn aln 55,000 in seq BACT1. Time spent on BACT1 so far: 2.82 sec.\n",
      "\tOn aln 56,000 in seq BACT1. Time spent on BACT1 so far: 2.87 sec.\n",
      "\tOn aln 57,000 in seq BACT1. Time spent on BACT1 so far: 2.92 sec.\n",
      "\tOn aln 58,000 in seq BACT1. Time spent on BACT1 so far: 2.97 sec.\n",
      "\tOn aln 59,000 in seq BACT1. Time spent on BACT1 so far: 3.02 sec.\n",
      "\tOn aln 60,000 in seq BACT1. Time spent on BACT1 so far: 3.07 sec.\n",
      "\tOn aln 61,000 in seq BACT1. Time spent on BACT1 so far: 3.13 sec.\n",
      "\tOn aln 62,000 in seq BACT1. Time spent on BACT1 so far: 3.18 sec.\n",
      "\tOn aln 63,000 in seq BACT1. Time spent on BACT1 so far: 3.23 sec.\n",
      "\tOn aln 64,000 in seq BACT1. Time spent on BACT1 so far: 3.28 sec.\n",
      "\tOn aln 65,000 in seq BACT1. Time spent on BACT1 so far: 3.33 sec.\n",
      "\tOn aln 66,000 in seq BACT1. Time spent on BACT1 so far: 3.38 sec.\n",
      "\tOn aln 67,000 in seq BACT1. Time spent on BACT1 so far: 3.43 sec.\n",
      "\tOn aln 68,000 in seq BACT1. Time spent on BACT1 so far: 3.48 sec.\n",
      "\tOn aln 69,000 in seq BACT1. Time spent on BACT1 so far: 3.53 sec.\n",
      "\tOn aln 70,000 in seq BACT1. Time spent on BACT1 so far: 3.58 sec.\n",
      "\tOn aln 71,000 in seq BACT1. Time spent on BACT1 so far: 3.63 sec.\n",
      "\tOn aln 72,000 in seq BACT1. Time spent on BACT1 so far: 3.69 sec.\n",
      "\tOn aln 73,000 in seq BACT1. Time spent on BACT1 so far: 3.74 sec.\n",
      "\tOn aln 74,000 in seq BACT1. Time spent on BACT1 so far: 3.79 sec.\n",
      "\tOn aln 75,000 in seq BACT1. Time spent on BACT1 so far: 3.84 sec.\n",
      "\tOn aln 76,000 in seq BACT1. Time spent on BACT1 so far: 3.90 sec.\n",
      "\tOn aln 77,000 in seq BACT1. Time spent on BACT1 so far: 3.95 sec.\n",
      "\tOn aln 78,000 in seq BACT1. Time spent on BACT1 so far: 4.00 sec.\n",
      "\tOn aln 79,000 in seq BACT1. Time spent on BACT1 so far: 4.05 sec.\n",
      "\tOn aln 80,000 in seq BACT1. Time spent on BACT1 so far: 4.10 sec.\n",
      "\tOn aln 81,000 in seq BACT1. Time spent on BACT1 so far: 4.16 sec.\n",
      "\tOn aln 82,000 in seq BACT1. Time spent on BACT1 so far: 4.21 sec.\n",
      "\tOn aln 83,000 in seq BACT1. Time spent on BACT1 so far: 4.26 sec.\n",
      "\tOn aln 84,000 in seq BACT1. Time spent on BACT1 so far: 4.31 sec.\n",
      "\tOn aln 85,000 in seq BACT1. Time spent on BACT1 so far: 4.37 sec.\n",
      "\tOn aln 86,000 in seq BACT1. Time spent on BACT1 so far: 4.42 sec.\n",
      "\tOn aln 87,000 in seq BACT1. Time spent on BACT1 so far: 4.47 sec.\n",
      "\tOn aln 88,000 in seq BACT1. Time spent on BACT1 so far: 4.52 sec.\n",
      "\tOn aln 89,000 in seq BACT1. Time spent on BACT1 so far: 4.58 sec.\n",
      "\tOn aln 90,000 in seq BACT1. Time spent on BACT1 so far: 4.63 sec.\n",
      "\tOn aln 91,000 in seq BACT1. Time spent on BACT1 so far: 4.68 sec.\n",
      "\tOn aln 92,000 in seq BACT1. Time spent on BACT1 so far: 4.73 sec.\n",
      "\tOn aln 93,000 in seq BACT1. Time spent on BACT1 so far: 4.78 sec.\n",
      "\tOn aln 94,000 in seq BACT1. Time spent on BACT1 so far: 4.83 sec.\n",
      "\tOn aln 95,000 in seq BACT1. Time spent on BACT1 so far: 4.88 sec.\n",
      "\tOn aln 96,000 in seq BACT1. Time spent on BACT1 so far: 4.94 sec.\n",
      "\tOn aln 97,000 in seq BACT1. Time spent on BACT1 so far: 4.99 sec.\n",
      "\tOn aln 98,000 in seq BACT1. Time spent on BACT1 so far: 5.04 sec.\n",
      "\tOn aln 99,000 in seq BACT1. Time spent on BACT1 so far: 5.10 sec.\n",
      "\tOn aln 100,000 in seq BACT1. Time spent on BACT1 so far: 5.15 sec.\n",
      "\tOn aln 101,000 in seq BACT1. Time spent on BACT1 so far: 5.20 sec.\n",
      "\tOn aln 102,000 in seq BACT1. Time spent on BACT1 so far: 5.25 sec.\n",
      "\tOn aln 103,000 in seq BACT1. Time spent on BACT1 so far: 5.30 sec.\n",
      "\tOn aln 104,000 in seq BACT1. Time spent on BACT1 so far: 5.35 sec.\n",
      "\tOn aln 105,000 in seq BACT1. Time spent on BACT1 so far: 5.41 sec.\n",
      "\tOn aln 106,000 in seq BACT1. Time spent on BACT1 so far: 5.46 sec.\n",
      "\tOn aln 107,000 in seq BACT1. Time spent on BACT1 so far: 5.51 sec.\n",
      "\tOn aln 108,000 in seq BACT1. Time spent on BACT1 so far: 5.56 sec.\n",
      "\tOn aln 109,000 in seq BACT1. Time spent on BACT1 so far: 5.61 sec.\n",
      "\tOn aln 110,000 in seq BACT1. Time spent on BACT1 so far: 5.67 sec.\n",
      "\tOn aln 111,000 in seq BACT1. Time spent on BACT1 so far: 5.72 sec.\n",
      "\tOn aln 112,000 in seq BACT1. Time spent on BACT1 so far: 5.77 sec.\n",
      "\tOn aln 113,000 in seq BACT1. Time spent on BACT1 so far: 5.82 sec.\n",
      "\tOn aln 114,000 in seq BACT1. Time spent on BACT1 so far: 5.88 sec.\n",
      "\tOn aln 115,000 in seq BACT1. Time spent on BACT1 so far: 5.93 sec.\n",
      "\tOn aln 116,000 in seq BACT1. Time spent on BACT1 so far: 5.98 sec.\n",
      "\tOn aln 117,000 in seq BACT1. Time spent on BACT1 so far: 6.03 sec.\n",
      "\tOn aln 118,000 in seq BACT1. Time spent on BACT1 so far: 6.08 sec.\n",
      "\tOn aln 119,000 in seq BACT1. Time spent on BACT1 so far: 6.14 sec.\n",
      "\tOn aln 120,000 in seq BACT1. Time spent on BACT1 so far: 6.19 sec.\n",
      "\tOn aln 121,000 in seq BACT1. Time spent on BACT1 so far: 6.24 sec.\n",
      "\tOn aln 122,000 in seq BACT1. Time spent on BACT1 so far: 6.29 sec.\n",
      "\tOn aln 123,000 in seq BACT1. Time spent on BACT1 so far: 6.34 sec.\n",
      "\tOn aln 124,000 in seq BACT1. Time spent on BACT1 so far: 6.40 sec.\n",
      "\tOn aln 125,000 in seq BACT1. Time spent on BACT1 so far: 6.45 sec.\n",
      "\tOn aln 126,000 in seq BACT1. Time spent on BACT1 so far: 6.50 sec.\n",
      "\tOn aln 127,000 in seq BACT1. Time spent on BACT1 so far: 6.55 sec.\n",
      "\tOn aln 128,000 in seq BACT1. Time spent on BACT1 so far: 6.60 sec.\n",
      "\tOn aln 129,000 in seq BACT1. Time spent on BACT1 so far: 6.65 sec.\n",
      "\tOn aln 130,000 in seq BACT1. Time spent on BACT1 so far: 6.70 sec.\n",
      "\tOn aln 131,000 in seq BACT1. Time spent on BACT1 so far: 6.75 sec.\n",
      "\tOn aln 132,000 in seq BACT1. Time spent on BACT1 so far: 6.81 sec.\n",
      "\tOn aln 133,000 in seq BACT1. Time spent on BACT1 so far: 6.86 sec.\n",
      "\tOn aln 134,000 in seq BACT1. Time spent on BACT1 so far: 6.91 sec.\n",
      "\tOn aln 135,000 in seq BACT1. Time spent on BACT1 so far: 6.96 sec.\n",
      "\tOn aln 136,000 in seq BACT1. Time spent on BACT1 so far: 7.01 sec.\n",
      "\tOn aln 137,000 in seq BACT1. Time spent on BACT1 so far: 7.06 sec.\n",
      "\tOn aln 138,000 in seq BACT1. Time spent on BACT1 so far: 7.11 sec.\n",
      "\tOn aln 139,000 in seq BACT1. Time spent on BACT1 so far: 7.17 sec.\n",
      "\tOn aln 140,000 in seq BACT1. Time spent on BACT1 so far: 7.22 sec.\n",
      "\tOn aln 141,000 in seq BACT1. Time spent on BACT1 so far: 7.27 sec.\n",
      "\tOn aln 142,000 in seq BACT1. Time spent on BACT1 so far: 7.32 sec.\n",
      "\tOn aln 143,000 in seq BACT1. Time spent on BACT1 so far: 7.37 sec.\n",
      "\tOn aln 144,000 in seq BACT1. Time spent on BACT1 so far: 7.42 sec.\n",
      "\tOn aln 145,000 in seq BACT1. Time spent on BACT1 so far: 7.47 sec.\n",
      "\tOn aln 146,000 in seq BACT1. Time spent on BACT1 so far: 7.52 sec.\n",
      "\tOn aln 147,000 in seq BACT1. Time spent on BACT1 so far: 7.57 sec.\n",
      "\tOn aln 148,000 in seq BACT1. Time spent on BACT1 so far: 7.62 sec.\n",
      "\tOn aln 149,000 in seq BACT1. Time spent on BACT1 so far: 7.68 sec.\n",
      "\tOn aln 150,000 in seq BACT1. Time spent on BACT1 so far: 7.73 sec.\n",
      "\tOn aln 151,000 in seq BACT1. Time spent on BACT1 so far: 7.78 sec.\n",
      "\tOn aln 152,000 in seq BACT1. Time spent on BACT1 so far: 7.83 sec.\n",
      "\tOn aln 153,000 in seq BACT1. Time spent on BACT1 so far: 7.88 sec.\n",
      "\tOn aln 154,000 in seq BACT1. Time spent on BACT1 so far: 7.94 sec.\n",
      "\tOn aln 155,000 in seq BACT1. Time spent on BACT1 so far: 7.99 sec.\n",
      "\tOn aln 156,000 in seq BACT1. Time spent on BACT1 so far: 8.04 sec.\n",
      "\tOn aln 157,000 in seq BACT1. Time spent on BACT1 so far: 8.09 sec.\n",
      "\tOn aln 158,000 in seq BACT1. Time spent on BACT1 so far: 8.14 sec.\n",
      "\tOn aln 159,000 in seq BACT1. Time spent on BACT1 so far: 8.19 sec.\n",
      "\tOn aln 160,000 in seq BACT1. Time spent on BACT1 so far: 8.25 sec.\n",
      "\tOn aln 161,000 in seq BACT1. Time spent on BACT1 so far: 8.30 sec.\n",
      "\tOn aln 162,000 in seq BACT1. Time spent on BACT1 so far: 8.35 sec.\n",
      "\tOn aln 163,000 in seq BACT1. Time spent on BACT1 so far: 8.40 sec.\n",
      "\tOn aln 164,000 in seq BACT1. Time spent on BACT1 so far: 8.45 sec.\n",
      "\tOn aln 165,000 in seq BACT1. Time spent on BACT1 so far: 8.51 sec.\n",
      "\tOn aln 166,000 in seq BACT1. Time spent on BACT1 so far: 8.56 sec.\n",
      "\tOn aln 167,000 in seq BACT1. Time spent on BACT1 so far: 8.61 sec.\n",
      "\tOn aln 168,000 in seq BACT1. Time spent on BACT1 so far: 8.67 sec.\n",
      "\tOn aln 169,000 in seq BACT1. Time spent on BACT1 so far: 8.72 sec.\n",
      "\tOn aln 170,000 in seq BACT1. Time spent on BACT1 so far: 8.77 sec.\n",
      "\tOn aln 171,000 in seq BACT1. Time spent on BACT1 so far: 8.82 sec.\n",
      "\tOn aln 172,000 in seq BACT1. Time spent on BACT1 so far: 8.87 sec.\n",
      "\tOn aln 173,000 in seq BACT1. Time spent on BACT1 so far: 8.93 sec.\n",
      "\tOn aln 174,000 in seq BACT1. Time spent on BACT1 so far: 8.98 sec.\n",
      "\tOn aln 175,000 in seq BACT1. Time spent on BACT1 so far: 9.03 sec.\n",
      "\tOn aln 176,000 in seq BACT1. Time spent on BACT1 so far: 9.08 sec.\n",
      "\tOn aln 177,000 in seq BACT1. Time spent on BACT1 so far: 9.14 sec.\n",
      "\tOn aln 178,000 in seq BACT1. Time spent on BACT1 so far: 9.19 sec.\n",
      "\tOn aln 179,000 in seq BACT1. Time spent on BACT1 so far: 9.24 sec.\n",
      "\tOn aln 180,000 in seq BACT1. Time spent on BACT1 so far: 9.29 sec.\n",
      "\tOn aln 181,000 in seq BACT1. Time spent on BACT1 so far: 9.35 sec.\n",
      "\tOn aln 182,000 in seq BACT1. Time spent on BACT1 so far: 9.40 sec.\n",
      "\tOn aln 183,000 in seq BACT1. Time spent on BACT1 so far: 9.45 sec.\n",
      "\tOn aln 184,000 in seq BACT1. Time spent on BACT1 so far: 9.50 sec.\n",
      "\tOn aln 185,000 in seq BACT1. Time spent on BACT1 so far: 9.55 sec.\n",
      "\tOn aln 186,000 in seq BACT1. Time spent on BACT1 so far: 9.61 sec.\n",
      "\tOn aln 187,000 in seq BACT1. Time spent on BACT1 so far: 9.66 sec.\n",
      "\tOn aln 188,000 in seq BACT1. Time spent on BACT1 so far: 9.71 sec.\n",
      "\tOn aln 189,000 in seq BACT1. Time spent on BACT1 so far: 9.76 sec.\n",
      "\tOn aln 190,000 in seq BACT1. Time spent on BACT1 so far: 9.81 sec.\n",
      "\tOn aln 191,000 in seq BACT1. Time spent on BACT1 so far: 9.86 sec.\n",
      "\tOn aln 192,000 in seq BACT1. Time spent on BACT1 so far: 9.91 sec.\n",
      "\tOn aln 193,000 in seq BACT1. Time spent on BACT1 so far: 9.96 sec.\n",
      "\tOn aln 194,000 in seq BACT1. Time spent on BACT1 so far: 10.01 sec.\n",
      "\tOn aln 195,000 in seq BACT1. Time spent on BACT1 so far: 10.07 sec.\n",
      "\tOn aln 196,000 in seq BACT1. Time spent on BACT1 so far: 10.11 sec.\n",
      "\tOn aln 197,000 in seq BACT1. Time spent on BACT1 so far: 10.16 sec.\n",
      "\tOn aln 198,000 in seq BACT1. Time spent on BACT1 so far: 10.22 sec.\n",
      "\tOn aln 199,000 in seq BACT1. Time spent on BACT1 so far: 10.27 sec.\n",
      "\tOn aln 200,000 in seq BACT1. Time spent on BACT1 so far: 10.32 sec.\n",
      "\tOn aln 201,000 in seq BACT1. Time spent on BACT1 so far: 10.37 sec.\n",
      "\tOn aln 202,000 in seq BACT1. Time spent on BACT1 so far: 10.42 sec.\n",
      "\tOn aln 203,000 in seq BACT1. Time spent on BACT1 so far: 10.48 sec.\n",
      "\tOn aln 204,000 in seq BACT1. Time spent on BACT1 so far: 10.53 sec.\n",
      "\tOn aln 205,000 in seq BACT1. Time spent on BACT1 so far: 10.58 sec.\n",
      "\tOn aln 206,000 in seq BACT1. Time spent on BACT1 so far: 10.63 sec.\n",
      "\tOn aln 207,000 in seq BACT1. Time spent on BACT1 so far: 10.68 sec.\n",
      "\tOn aln 208,000 in seq BACT1. Time spent on BACT1 so far: 10.73 sec.\n",
      "\tOn aln 209,000 in seq BACT1. Time spent on BACT1 so far: 10.78 sec.\n",
      "\tOn aln 210,000 in seq BACT1. Time spent on BACT1 so far: 10.83 sec.\n",
      "\tOn aln 211,000 in seq BACT1. Time spent on BACT1 so far: 10.89 sec.\n",
      "\tOn aln 212,000 in seq BACT1. Time spent on BACT1 so far: 10.94 sec.\n",
      "\tOn aln 213,000 in seq BACT1. Time spent on BACT1 so far: 10.99 sec.\n",
      "\tOn aln 214,000 in seq BACT1. Time spent on BACT1 so far: 11.04 sec.\n",
      "\tOn aln 215,000 in seq BACT1. Time spent on BACT1 so far: 11.09 sec.\n",
      "\tOn aln 216,000 in seq BACT1. Time spent on BACT1 so far: 11.14 sec.\n",
      "\tOn aln 217,000 in seq BACT1. Time spent on BACT1 so far: 11.20 sec.\n",
      "\tOn aln 218,000 in seq BACT1. Time spent on BACT1 so far: 11.25 sec.\n",
      "\tOn aln 219,000 in seq BACT1. Time spent on BACT1 so far: 11.30 sec.\n",
      "\tOn aln 220,000 in seq BACT1. Time spent on BACT1 so far: 11.35 sec.\n",
      "\tOn aln 221,000 in seq BACT1. Time spent on BACT1 so far: 11.41 sec.\n",
      "\tOn aln 222,000 in seq BACT1. Time spent on BACT1 so far: 11.46 sec.\n",
      "\tOn aln 223,000 in seq BACT1. Time spent on BACT1 so far: 11.51 sec.\n",
      "\tOn aln 224,000 in seq BACT1. Time spent on BACT1 so far: 11.56 sec.\n",
      "\tOn aln 225,000 in seq BACT1. Time spent on BACT1 so far: 11.61 sec.\n",
      "\tOn aln 226,000 in seq BACT1. Time spent on BACT1 so far: 11.67 sec.\n",
      "\tOn aln 227,000 in seq BACT1. Time spent on BACT1 so far: 11.72 sec.\n",
      "\tOn aln 228,000 in seq BACT1. Time spent on BACT1 so far: 11.77 sec.\n",
      "\tOn aln 229,000 in seq BACT1. Time spent on BACT1 so far: 11.82 sec.\n",
      "\tOn aln 230,000 in seq BACT1. Time spent on BACT1 so far: 11.87 sec.\n",
      "\tOn aln 231,000 in seq BACT1. Time spent on BACT1 so far: 11.93 sec.\n",
      "\tOn aln 232,000 in seq BACT1. Time spent on BACT1 so far: 11.98 sec.\n",
      "\tOn aln 233,000 in seq BACT1. Time spent on BACT1 so far: 12.03 sec.\n",
      "\tOn aln 234,000 in seq BACT1. Time spent on BACT1 so far: 12.08 sec.\n",
      "\tOn aln 235,000 in seq BACT1. Time spent on BACT1 so far: 12.13 sec.\n",
      "\tOn aln 236,000 in seq BACT1. Time spent on BACT1 so far: 12.18 sec.\n",
      "\tOn aln 237,000 in seq BACT1. Time spent on BACT1 so far: 12.24 sec.\n",
      "\tOn aln 238,000 in seq BACT1. Time spent on BACT1 so far: 12.29 sec.\n",
      "\tOn aln 239,000 in seq BACT1. Time spent on BACT1 so far: 12.34 sec.\n",
      "\tOn aln 240,000 in seq BACT1. Time spent on BACT1 so far: 12.39 sec.\n",
      "\tOn aln 241,000 in seq BACT1. Time spent on BACT1 so far: 12.44 sec.\n",
      "\tOn aln 242,000 in seq BACT1. Time spent on BACT1 so far: 12.49 sec.\n",
      "\tOn aln 243,000 in seq BACT1. Time spent on BACT1 so far: 12.55 sec.\n",
      "\tOn aln 244,000 in seq BACT1. Time spent on BACT1 so far: 12.60 sec.\n",
      "\tOn aln 245,000 in seq BACT1. Time spent on BACT1 so far: 12.65 sec.\n",
      "\tOn aln 246,000 in seq BACT1. Time spent on BACT1 so far: 12.70 sec.\n",
      "\tOn aln 247,000 in seq BACT1. Time spent on BACT1 so far: 12.75 sec.\n",
      "\tOn aln 248,000 in seq BACT1. Time spent on BACT1 so far: 12.81 sec.\n",
      "\tOn aln 249,000 in seq BACT1. Time spent on BACT1 so far: 12.86 sec.\n",
      "\tOn aln 250,000 in seq BACT1. Time spent on BACT1 so far: 12.91 sec.\n",
      "\tOn aln 251,000 in seq BACT1. Time spent on BACT1 so far: 12.96 sec.\n",
      "\tOn aln 252,000 in seq BACT1. Time spent on BACT1 so far: 13.01 sec.\n",
      "\tOn aln 253,000 in seq BACT1. Time spent on BACT1 so far: 13.07 sec.\n",
      "\tOn aln 254,000 in seq BACT1. Time spent on BACT1 so far: 13.12 sec.\n",
      "\tOn aln 255,000 in seq BACT1. Time spent on BACT1 so far: 13.17 sec.\n",
      "\tOn aln 256,000 in seq BACT1. Time spent on BACT1 so far: 13.22 sec.\n",
      "\tOn aln 257,000 in seq BACT1. Time spent on BACT1 so far: 13.27 sec.\n",
      "\tOn aln 258,000 in seq BACT1. Time spent on BACT1 so far: 13.33 sec.\n",
      "\tOn aln 259,000 in seq BACT1. Time spent on BACT1 so far: 13.38 sec.\n",
      "\tOn aln 260,000 in seq BACT1. Time spent on BACT1 so far: 13.43 sec.\n",
      "\tOn aln 261,000 in seq BACT1. Time spent on BACT1 so far: 13.48 sec.\n",
      "\tOn aln 262,000 in seq BACT1. Time spent on BACT1 so far: 13.54 sec.\n",
      "\tOn aln 263,000 in seq BACT1. Time spent on BACT1 so far: 13.57 sec.\n",
      "Done with edge_1671! Took 13.58 sec.\n",
      "Identifying mutated positions (p = 1%) in genome BACT2...\n",
      "Found 372 mutated positions (p = 1%) in BACT2.\n",
      "Note that this tally is higher than you'd see in e.g. the CP1/2/3 plots, because now we're including both 'rare' and non-rare mutations. Just so you don't waste five minutes sanity-checking this like I did.\n",
      "Going through these positions...\n",
      "\tOn aln 1,000 in seq BACT2. Time spent on BACT2 so far: 0.04 sec.\n",
      "\tOn aln 2,000 in seq BACT2. Time spent on BACT2 so far: 0.08 sec.\n",
      "\tOn aln 3,000 in seq BACT2. Time spent on BACT2 so far: 0.11 sec.\n",
      "\tOn aln 4,000 in seq BACT2. Time spent on BACT2 so far: 0.16 sec.\n",
      "\tOn aln 5,000 in seq BACT2. Time spent on BACT2 so far: 0.21 sec.\n",
      "\tOn aln 6,000 in seq BACT2. Time spent on BACT2 so far: 0.27 sec.\n",
      "\tOn aln 7,000 in seq BACT2. Time spent on BACT2 so far: 0.32 sec.\n",
      "\tOn aln 8,000 in seq BACT2. Time spent on BACT2 so far: 0.36 sec.\n",
      "\tOn aln 9,000 in seq BACT2. Time spent on BACT2 so far: 0.41 sec.\n",
      "\tOn aln 10,000 in seq BACT2. Time spent on BACT2 so far: 0.47 sec.\n",
      "\tOn aln 11,000 in seq BACT2. Time spent on BACT2 so far: 0.52 sec.\n",
      "\tOn aln 12,000 in seq BACT2. Time spent on BACT2 so far: 0.56 sec.\n",
      "\tOn aln 13,000 in seq BACT2. Time spent on BACT2 so far: 0.61 sec.\n",
      "\tOn aln 14,000 in seq BACT2. Time spent on BACT2 so far: 0.66 sec.\n",
      "\tOn aln 15,000 in seq BACT2. Time spent on BACT2 so far: 0.71 sec.\n",
      "\tOn aln 16,000 in seq BACT2. Time spent on BACT2 so far: 0.76 sec.\n",
      "\tOn aln 17,000 in seq BACT2. Time spent on BACT2 so far: 0.81 sec.\n",
      "\tOn aln 18,000 in seq BACT2. Time spent on BACT2 so far: 0.86 sec.\n",
      "\tOn aln 19,000 in seq BACT2. Time spent on BACT2 so far: 0.91 sec.\n",
      "\tOn aln 20,000 in seq BACT2. Time spent on BACT2 so far: 0.96 sec.\n",
      "\tOn aln 21,000 in seq BACT2. Time spent on BACT2 so far: 1.01 sec.\n",
      "\tOn aln 22,000 in seq BACT2. Time spent on BACT2 so far: 1.06 sec.\n",
      "\tOn aln 23,000 in seq BACT2. Time spent on BACT2 so far: 1.12 sec.\n",
      "\tOn aln 24,000 in seq BACT2. Time spent on BACT2 so far: 1.17 sec.\n",
      "\tOn aln 25,000 in seq BACT2. Time spent on BACT2 so far: 1.22 sec.\n",
      "\tOn aln 26,000 in seq BACT2. Time spent on BACT2 so far: 1.26 sec.\n",
      "\tOn aln 27,000 in seq BACT2. Time spent on BACT2 so far: 1.31 sec.\n",
      "\tOn aln 28,000 in seq BACT2. Time spent on BACT2 so far: 1.37 sec.\n",
      "\tOn aln 29,000 in seq BACT2. Time spent on BACT2 so far: 1.41 sec.\n",
      "\tOn aln 30,000 in seq BACT2. Time spent on BACT2 so far: 1.46 sec.\n",
      "\tOn aln 31,000 in seq BACT2. Time spent on BACT2 so far: 1.51 sec.\n",
      "\tOn aln 32,000 in seq BACT2. Time spent on BACT2 so far: 1.57 sec.\n",
      "\tOn aln 33,000 in seq BACT2. Time spent on BACT2 so far: 1.62 sec.\n",
      "\tOn aln 34,000 in seq BACT2. Time spent on BACT2 so far: 1.66 sec.\n",
      "\tOn aln 35,000 in seq BACT2. Time spent on BACT2 so far: 1.72 sec.\n",
      "\tOn aln 36,000 in seq BACT2. Time spent on BACT2 so far: 1.77 sec.\n",
      "\tOn aln 37,000 in seq BACT2. Time spent on BACT2 so far: 1.82 sec.\n",
      "\tOn aln 38,000 in seq BACT2. Time spent on BACT2 so far: 1.87 sec.\n",
      "\tOn aln 39,000 in seq BACT2. Time spent on BACT2 so far: 1.91 sec.\n",
      "\tOn aln 40,000 in seq BACT2. Time spent on BACT2 so far: 1.96 sec.\n",
      "\tOn aln 41,000 in seq BACT2. Time spent on BACT2 so far: 2.02 sec.\n",
      "\tOn aln 42,000 in seq BACT2. Time spent on BACT2 so far: 2.07 sec.\n",
      "\tOn aln 43,000 in seq BACT2. Time spent on BACT2 so far: 2.12 sec.\n",
      "\tOn aln 44,000 in seq BACT2. Time spent on BACT2 so far: 2.17 sec.\n",
      "\tOn aln 45,000 in seq BACT2. Time spent on BACT2 so far: 2.22 sec.\n",
      "\tOn aln 46,000 in seq BACT2. Time spent on BACT2 so far: 2.27 sec.\n",
      "\tOn aln 47,000 in seq BACT2. Time spent on BACT2 so far: 2.32 sec.\n",
      "\tOn aln 48,000 in seq BACT2. Time spent on BACT2 so far: 2.37 sec.\n",
      "\tOn aln 49,000 in seq BACT2. Time spent on BACT2 so far: 2.42 sec.\n",
      "\tOn aln 50,000 in seq BACT2. Time spent on BACT2 so far: 2.47 sec.\n",
      "\tOn aln 51,000 in seq BACT2. Time spent on BACT2 so far: 2.52 sec.\n",
      "\tOn aln 52,000 in seq BACT2. Time spent on BACT2 so far: 2.57 sec.\n",
      "\tOn aln 53,000 in seq BACT2. Time spent on BACT2 so far: 2.62 sec.\n",
      "\tOn aln 54,000 in seq BACT2. Time spent on BACT2 so far: 2.67 sec.\n",
      "\tOn aln 55,000 in seq BACT2. Time spent on BACT2 so far: 2.72 sec.\n",
      "\tOn aln 56,000 in seq BACT2. Time spent on BACT2 so far: 2.77 sec.\n",
      "\tOn aln 57,000 in seq BACT2. Time spent on BACT2 so far: 2.82 sec.\n",
      "\tOn aln 58,000 in seq BACT2. Time spent on BACT2 so far: 2.87 sec.\n",
      "\tOn aln 59,000 in seq BACT2. Time spent on BACT2 so far: 2.92 sec.\n",
      "\tOn aln 60,000 in seq BACT2. Time spent on BACT2 so far: 2.97 sec.\n",
      "\tOn aln 61,000 in seq BACT2. Time spent on BACT2 so far: 3.02 sec.\n",
      "\tOn aln 62,000 in seq BACT2. Time spent on BACT2 so far: 3.07 sec.\n",
      "\tOn aln 63,000 in seq BACT2. Time spent on BACT2 so far: 3.12 sec.\n",
      "\tOn aln 64,000 in seq BACT2. Time spent on BACT2 so far: 3.17 sec.\n",
      "\tOn aln 65,000 in seq BACT2. Time spent on BACT2 so far: 3.22 sec.\n",
      "\tOn aln 66,000 in seq BACT2. Time spent on BACT2 so far: 3.27 sec.\n",
      "\tOn aln 67,000 in seq BACT2. Time spent on BACT2 so far: 3.32 sec.\n",
      "\tOn aln 68,000 in seq BACT2. Time spent on BACT2 so far: 3.37 sec.\n",
      "\tOn aln 69,000 in seq BACT2. Time spent on BACT2 so far: 3.42 sec.\n",
      "\tOn aln 70,000 in seq BACT2. Time spent on BACT2 so far: 3.47 sec.\n",
      "\tOn aln 71,000 in seq BACT2. Time spent on BACT2 so far: 3.52 sec.\n",
      "\tOn aln 72,000 in seq BACT2. Time spent on BACT2 so far: 3.57 sec.\n",
      "\tOn aln 73,000 in seq BACT2. Time spent on BACT2 so far: 3.62 sec.\n",
      "\tOn aln 74,000 in seq BACT2. Time spent on BACT2 so far: 3.68 sec.\n",
      "\tOn aln 75,000 in seq BACT2. Time spent on BACT2 so far: 3.72 sec.\n",
      "\tOn aln 76,000 in seq BACT2. Time spent on BACT2 so far: 3.78 sec.\n",
      "\tOn aln 77,000 in seq BACT2. Time spent on BACT2 so far: 3.83 sec.\n",
      "\tOn aln 78,000 in seq BACT2. Time spent on BACT2 so far: 3.88 sec.\n",
      "\tOn aln 79,000 in seq BACT2. Time spent on BACT2 so far: 3.93 sec.\n",
      "\tOn aln 80,000 in seq BACT2. Time spent on BACT2 so far: 3.98 sec.\n",
      "\tOn aln 81,000 in seq BACT2. Time spent on BACT2 so far: 4.03 sec.\n",
      "\tOn aln 82,000 in seq BACT2. Time spent on BACT2 so far: 4.08 sec.\n",
      "\tOn aln 83,000 in seq BACT2. Time spent on BACT2 so far: 4.13 sec.\n",
      "\tOn aln 84,000 in seq BACT2. Time spent on BACT2 so far: 4.18 sec.\n",
      "\tOn aln 85,000 in seq BACT2. Time spent on BACT2 so far: 4.23 sec.\n",
      "\tOn aln 86,000 in seq BACT2. Time spent on BACT2 so far: 4.28 sec.\n",
      "\tOn aln 87,000 in seq BACT2. Time spent on BACT2 so far: 4.33 sec.\n",
      "\tOn aln 88,000 in seq BACT2. Time spent on BACT2 so far: 4.38 sec.\n",
      "\tOn aln 89,000 in seq BACT2. Time spent on BACT2 so far: 4.43 sec.\n",
      "\tOn aln 90,000 in seq BACT2. Time spent on BACT2 so far: 4.48 sec.\n",
      "\tOn aln 91,000 in seq BACT2. Time spent on BACT2 so far: 4.53 sec.\n",
      "\tOn aln 92,000 in seq BACT2. Time spent on BACT2 so far: 4.58 sec.\n",
      "\tOn aln 93,000 in seq BACT2. Time spent on BACT2 so far: 4.63 sec.\n",
      "\tOn aln 94,000 in seq BACT2. Time spent on BACT2 so far: 4.68 sec.\n",
      "\tOn aln 95,000 in seq BACT2. Time spent on BACT2 so far: 4.73 sec.\n",
      "\tOn aln 96,000 in seq BACT2. Time spent on BACT2 so far: 4.78 sec.\n",
      "\tOn aln 97,000 in seq BACT2. Time spent on BACT2 so far: 4.83 sec.\n",
      "\tOn aln 98,000 in seq BACT2. Time spent on BACT2 so far: 4.88 sec.\n",
      "\tOn aln 99,000 in seq BACT2. Time spent on BACT2 so far: 4.93 sec.\n",
      "\tOn aln 100,000 in seq BACT2. Time spent on BACT2 so far: 4.98 sec.\n",
      "\tOn aln 101,000 in seq BACT2. Time spent on BACT2 so far: 5.03 sec.\n",
      "\tOn aln 102,000 in seq BACT2. Time spent on BACT2 so far: 5.08 sec.\n",
      "\tOn aln 103,000 in seq BACT2. Time spent on BACT2 so far: 5.13 sec.\n",
      "\tOn aln 104,000 in seq BACT2. Time spent on BACT2 so far: 5.18 sec.\n",
      "\tOn aln 105,000 in seq BACT2. Time spent on BACT2 so far: 5.23 sec.\n",
      "\tOn aln 106,000 in seq BACT2. Time spent on BACT2 so far: 5.28 sec.\n",
      "\tOn aln 107,000 in seq BACT2. Time spent on BACT2 so far: 5.33 sec.\n",
      "\tOn aln 108,000 in seq BACT2. Time spent on BACT2 so far: 5.38 sec.\n",
      "\tOn aln 109,000 in seq BACT2. Time spent on BACT2 so far: 5.43 sec.\n",
      "\tOn aln 110,000 in seq BACT2. Time spent on BACT2 so far: 5.48 sec.\n",
      "\tOn aln 111,000 in seq BACT2. Time spent on BACT2 so far: 5.53 sec.\n",
      "\tOn aln 112,000 in seq BACT2. Time spent on BACT2 so far: 5.58 sec.\n",
      "\tOn aln 113,000 in seq BACT2. Time spent on BACT2 so far: 5.63 sec.\n",
      "\tOn aln 114,000 in seq BACT2. Time spent on BACT2 so far: 5.68 sec.\n",
      "\tOn aln 115,000 in seq BACT2. Time spent on BACT2 so far: 5.73 sec.\n",
      "\tOn aln 116,000 in seq BACT2. Time spent on BACT2 so far: 5.78 sec.\n",
      "\tOn aln 117,000 in seq BACT2. Time spent on BACT2 so far: 5.83 sec.\n",
      "\tOn aln 118,000 in seq BACT2. Time spent on BACT2 so far: 5.88 sec.\n",
      "\tOn aln 119,000 in seq BACT2. Time spent on BACT2 so far: 5.94 sec.\n",
      "\tOn aln 120,000 in seq BACT2. Time spent on BACT2 so far: 5.99 sec.\n",
      "\tOn aln 121,000 in seq BACT2. Time spent on BACT2 so far: 6.04 sec.\n",
      "\tOn aln 122,000 in seq BACT2. Time spent on BACT2 so far: 6.09 sec.\n",
      "\tOn aln 123,000 in seq BACT2. Time spent on BACT2 so far: 6.14 sec.\n",
      "\tOn aln 124,000 in seq BACT2. Time spent on BACT2 so far: 6.20 sec.\n",
      "\tOn aln 125,000 in seq BACT2. Time spent on BACT2 so far: 6.25 sec.\n",
      "\tOn aln 126,000 in seq BACT2. Time spent on BACT2 so far: 6.30 sec.\n",
      "\tOn aln 127,000 in seq BACT2. Time spent on BACT2 so far: 6.35 sec.\n",
      "\tOn aln 128,000 in seq BACT2. Time spent on BACT2 so far: 6.40 sec.\n",
      "\tOn aln 129,000 in seq BACT2. Time spent on BACT2 so far: 6.45 sec.\n",
      "\tOn aln 130,000 in seq BACT2. Time spent on BACT2 so far: 6.50 sec.\n",
      "\tOn aln 131,000 in seq BACT2. Time spent on BACT2 so far: 6.55 sec.\n",
      "\tOn aln 132,000 in seq BACT2. Time spent on BACT2 so far: 6.60 sec.\n",
      "\tOn aln 133,000 in seq BACT2. Time spent on BACT2 so far: 6.65 sec.\n",
      "\tOn aln 134,000 in seq BACT2. Time spent on BACT2 so far: 6.70 sec.\n",
      "\tOn aln 135,000 in seq BACT2. Time spent on BACT2 so far: 6.75 sec.\n",
      "\tOn aln 136,000 in seq BACT2. Time spent on BACT2 so far: 6.80 sec.\n",
      "\tOn aln 137,000 in seq BACT2. Time spent on BACT2 so far: 6.86 sec.\n",
      "\tOn aln 138,000 in seq BACT2. Time spent on BACT2 so far: 6.91 sec.\n",
      "\tOn aln 139,000 in seq BACT2. Time spent on BACT2 so far: 6.96 sec.\n",
      "\tOn aln 140,000 in seq BACT2. Time spent on BACT2 so far: 7.02 sec.\n",
      "\tOn aln 141,000 in seq BACT2. Time spent on BACT2 so far: 7.07 sec.\n",
      "\tOn aln 142,000 in seq BACT2. Time spent on BACT2 so far: 7.13 sec.\n",
      "\tOn aln 143,000 in seq BACT2. Time spent on BACT2 so far: 7.18 sec.\n",
      "\tOn aln 144,000 in seq BACT2. Time spent on BACT2 so far: 7.23 sec.\n",
      "\tOn aln 145,000 in seq BACT2. Time spent on BACT2 so far: 7.28 sec.\n",
      "\tOn aln 146,000 in seq BACT2. Time spent on BACT2 so far: 7.33 sec.\n",
      "\tOn aln 147,000 in seq BACT2. Time spent on BACT2 so far: 7.38 sec.\n",
      "\tOn aln 148,000 in seq BACT2. Time spent on BACT2 so far: 7.43 sec.\n",
      "\tOn aln 149,000 in seq BACT2. Time spent on BACT2 so far: 7.48 sec.\n",
      "\tOn aln 150,000 in seq BACT2. Time spent on BACT2 so far: 7.54 sec.\n",
      "\tOn aln 151,000 in seq BACT2. Time spent on BACT2 so far: 7.59 sec.\n",
      "\tOn aln 152,000 in seq BACT2. Time spent on BACT2 so far: 7.65 sec.\n",
      "\tOn aln 153,000 in seq BACT2. Time spent on BACT2 so far: 7.70 sec.\n",
      "\tOn aln 154,000 in seq BACT2. Time spent on BACT2 so far: 7.75 sec.\n",
      "\tOn aln 155,000 in seq BACT2. Time spent on BACT2 so far: 7.80 sec.\n",
      "\tOn aln 156,000 in seq BACT2. Time spent on BACT2 so far: 7.85 sec.\n",
      "\tOn aln 157,000 in seq BACT2. Time spent on BACT2 so far: 7.90 sec.\n",
      "\tOn aln 158,000 in seq BACT2. Time spent on BACT2 so far: 7.95 sec.\n",
      "\tOn aln 159,000 in seq BACT2. Time spent on BACT2 so far: 8.00 sec.\n",
      "\tOn aln 160,000 in seq BACT2. Time spent on BACT2 so far: 8.05 sec.\n",
      "\tOn aln 161,000 in seq BACT2. Time spent on BACT2 so far: 8.10 sec.\n",
      "\tOn aln 162,000 in seq BACT2. Time spent on BACT2 so far: 8.16 sec.\n",
      "\tOn aln 163,000 in seq BACT2. Time spent on BACT2 so far: 8.21 sec.\n",
      "\tOn aln 164,000 in seq BACT2. Time spent on BACT2 so far: 8.26 sec.\n",
      "\tOn aln 165,000 in seq BACT2. Time spent on BACT2 so far: 8.31 sec.\n",
      "\tOn aln 166,000 in seq BACT2. Time spent on BACT2 so far: 8.36 sec.\n",
      "\tOn aln 167,000 in seq BACT2. Time spent on BACT2 so far: 8.41 sec.\n",
      "\tOn aln 168,000 in seq BACT2. Time spent on BACT2 so far: 8.46 sec.\n",
      "\tOn aln 169,000 in seq BACT2. Time spent on BACT2 so far: 8.51 sec.\n",
      "\tOn aln 170,000 in seq BACT2. Time spent on BACT2 so far: 8.56 sec.\n",
      "\tOn aln 171,000 in seq BACT2. Time spent on BACT2 so far: 8.61 sec.\n",
      "\tOn aln 172,000 in seq BACT2. Time spent on BACT2 so far: 8.66 sec.\n",
      "\tOn aln 173,000 in seq BACT2. Time spent on BACT2 so far: 8.71 sec.\n",
      "\tOn aln 174,000 in seq BACT2. Time spent on BACT2 so far: 8.76 sec.\n",
      "\tOn aln 175,000 in seq BACT2. Time spent on BACT2 so far: 8.82 sec.\n",
      "\tOn aln 176,000 in seq BACT2. Time spent on BACT2 so far: 8.87 sec.\n",
      "\tOn aln 177,000 in seq BACT2. Time spent on BACT2 so far: 8.92 sec.\n",
      "\tOn aln 178,000 in seq BACT2. Time spent on BACT2 so far: 8.97 sec.\n",
      "\tOn aln 179,000 in seq BACT2. Time spent on BACT2 so far: 9.02 sec.\n",
      "\tOn aln 180,000 in seq BACT2. Time spent on BACT2 so far: 9.07 sec.\n",
      "\tOn aln 181,000 in seq BACT2. Time spent on BACT2 so far: 9.12 sec.\n",
      "\tOn aln 182,000 in seq BACT2. Time spent on BACT2 so far: 9.17 sec.\n",
      "\tOn aln 183,000 in seq BACT2. Time spent on BACT2 so far: 9.23 sec.\n",
      "\tOn aln 184,000 in seq BACT2. Time spent on BACT2 so far: 9.28 sec.\n",
      "\tOn aln 185,000 in seq BACT2. Time spent on BACT2 so far: 9.33 sec.\n",
      "\tOn aln 186,000 in seq BACT2. Time spent on BACT2 so far: 9.37 sec.\n",
      "\tOn aln 187,000 in seq BACT2. Time spent on BACT2 so far: 9.42 sec.\n",
      "\tOn aln 188,000 in seq BACT2. Time spent on BACT2 so far: 9.47 sec.\n",
      "\tOn aln 189,000 in seq BACT2. Time spent on BACT2 so far: 9.52 sec.\n",
      "\tOn aln 190,000 in seq BACT2. Time spent on BACT2 so far: 9.58 sec.\n",
      "\tOn aln 191,000 in seq BACT2. Time spent on BACT2 so far: 9.63 sec.\n",
      "\tOn aln 192,000 in seq BACT2. Time spent on BACT2 so far: 9.68 sec.\n",
      "\tOn aln 193,000 in seq BACT2. Time spent on BACT2 so far: 9.73 sec.\n",
      "\tOn aln 194,000 in seq BACT2. Time spent on BACT2 so far: 9.78 sec.\n",
      "\tOn aln 195,000 in seq BACT2. Time spent on BACT2 so far: 9.83 sec.\n",
      "\tOn aln 196,000 in seq BACT2. Time spent on BACT2 so far: 9.88 sec.\n",
      "\tOn aln 197,000 in seq BACT2. Time spent on BACT2 so far: 9.93 sec.\n",
      "\tOn aln 198,000 in seq BACT2. Time spent on BACT2 so far: 9.99 sec.\n",
      "\tOn aln 199,000 in seq BACT2. Time spent on BACT2 so far: 10.04 sec.\n",
      "\tOn aln 200,000 in seq BACT2. Time spent on BACT2 so far: 10.08 sec.\n",
      "\tOn aln 201,000 in seq BACT2. Time spent on BACT2 so far: 10.13 sec.\n",
      "\tOn aln 202,000 in seq BACT2. Time spent on BACT2 so far: 10.18 sec.\n",
      "\tOn aln 203,000 in seq BACT2. Time spent on BACT2 so far: 10.24 sec.\n",
      "\tOn aln 204,000 in seq BACT2. Time spent on BACT2 so far: 10.28 sec.\n",
      "\tOn aln 205,000 in seq BACT2. Time spent on BACT2 so far: 10.34 sec.\n",
      "\tOn aln 206,000 in seq BACT2. Time spent on BACT2 so far: 10.39 sec.\n",
      "\tOn aln 207,000 in seq BACT2. Time spent on BACT2 so far: 10.43 sec.\n",
      "\tOn aln 208,000 in seq BACT2. Time spent on BACT2 so far: 10.48 sec.\n",
      "\tOn aln 209,000 in seq BACT2. Time spent on BACT2 so far: 10.53 sec.\n",
      "\tOn aln 210,000 in seq BACT2. Time spent on BACT2 so far: 10.58 sec.\n",
      "\tOn aln 211,000 in seq BACT2. Time spent on BACT2 so far: 10.63 sec.\n",
      "\tOn aln 212,000 in seq BACT2. Time spent on BACT2 so far: 10.68 sec.\n",
      "\tOn aln 213,000 in seq BACT2. Time spent on BACT2 so far: 10.73 sec.\n",
      "\tOn aln 214,000 in seq BACT2. Time spent on BACT2 so far: 10.78 sec.\n",
      "\tOn aln 215,000 in seq BACT2. Time spent on BACT2 so far: 10.83 sec.\n",
      "\tOn aln 216,000 in seq BACT2. Time spent on BACT2 so far: 10.88 sec.\n",
      "\tOn aln 217,000 in seq BACT2. Time spent on BACT2 so far: 10.94 sec.\n",
      "\tOn aln 218,000 in seq BACT2. Time spent on BACT2 so far: 10.99 sec.\n",
      "\tOn aln 219,000 in seq BACT2. Time spent on BACT2 so far: 11.04 sec.\n",
      "\tOn aln 220,000 in seq BACT2. Time spent on BACT2 so far: 11.09 sec.\n",
      "\tOn aln 221,000 in seq BACT2. Time spent on BACT2 so far: 11.14 sec.\n",
      "\tOn aln 222,000 in seq BACT2. Time spent on BACT2 so far: 11.19 sec.\n",
      "\tOn aln 223,000 in seq BACT2. Time spent on BACT2 so far: 11.24 sec.\n",
      "\tOn aln 224,000 in seq BACT2. Time spent on BACT2 so far: 11.29 sec.\n",
      "\tOn aln 225,000 in seq BACT2. Time spent on BACT2 so far: 11.34 sec.\n",
      "\tOn aln 226,000 in seq BACT2. Time spent on BACT2 so far: 11.39 sec.\n",
      "\tOn aln 227,000 in seq BACT2. Time spent on BACT2 so far: 11.44 sec.\n",
      "\tOn aln 228,000 in seq BACT2. Time spent on BACT2 so far: 11.49 sec.\n",
      "\tOn aln 229,000 in seq BACT2. Time spent on BACT2 so far: 11.54 sec.\n",
      "\tOn aln 230,000 in seq BACT2. Time spent on BACT2 so far: 11.60 sec.\n",
      "\tOn aln 231,000 in seq BACT2. Time spent on BACT2 so far: 11.65 sec.\n",
      "\tOn aln 232,000 in seq BACT2. Time spent on BACT2 so far: 11.70 sec.\n",
      "\tOn aln 233,000 in seq BACT2. Time spent on BACT2 so far: 11.75 sec.\n",
      "\tOn aln 234,000 in seq BACT2. Time spent on BACT2 so far: 11.80 sec.\n",
      "\tOn aln 235,000 in seq BACT2. Time spent on BACT2 so far: 11.85 sec.\n",
      "\tOn aln 236,000 in seq BACT2. Time spent on BACT2 so far: 11.90 sec.\n",
      "\tOn aln 237,000 in seq BACT2. Time spent on BACT2 so far: 11.95 sec.\n",
      "\tOn aln 238,000 in seq BACT2. Time spent on BACT2 so far: 12.00 sec.\n",
      "\tOn aln 239,000 in seq BACT2. Time spent on BACT2 so far: 12.05 sec.\n",
      "\tOn aln 240,000 in seq BACT2. Time spent on BACT2 so far: 12.10 sec.\n",
      "\tOn aln 241,000 in seq BACT2. Time spent on BACT2 so far: 12.15 sec.\n",
      "\tOn aln 242,000 in seq BACT2. Time spent on BACT2 so far: 12.20 sec.\n",
      "\tOn aln 243,000 in seq BACT2. Time spent on BACT2 so far: 12.25 sec.\n",
      "\tOn aln 244,000 in seq BACT2. Time spent on BACT2 so far: 12.30 sec.\n",
      "\tOn aln 245,000 in seq BACT2. Time spent on BACT2 so far: 12.35 sec.\n",
      "\tOn aln 246,000 in seq BACT2. Time spent on BACT2 so far: 12.40 sec.\n",
      "\tOn aln 247,000 in seq BACT2. Time spent on BACT2 so far: 12.45 sec.\n",
      "\tOn aln 248,000 in seq BACT2. Time spent on BACT2 so far: 12.50 sec.\n",
      "\tOn aln 249,000 in seq BACT2. Time spent on BACT2 so far: 12.55 sec.\n",
      "\tOn aln 250,000 in seq BACT2. Time spent on BACT2 so far: 12.60 sec.\n",
      "\tOn aln 251,000 in seq BACT2. Time spent on BACT2 so far: 12.65 sec.\n",
      "\tOn aln 252,000 in seq BACT2. Time spent on BACT2 so far: 12.70 sec.\n",
      "\tOn aln 253,000 in seq BACT2. Time spent on BACT2 so far: 12.75 sec.\n",
      "\tOn aln 254,000 in seq BACT2. Time spent on BACT2 so far: 12.80 sec.\n",
      "\tOn aln 255,000 in seq BACT2. Time spent on BACT2 so far: 12.85 sec.\n",
      "\tOn aln 256,000 in seq BACT2. Time spent on BACT2 so far: 12.90 sec.\n",
      "\tOn aln 257,000 in seq BACT2. Time spent on BACT2 so far: 12.95 sec.\n",
      "\tOn aln 258,000 in seq BACT2. Time spent on BACT2 so far: 13.00 sec.\n",
      "\tOn aln 259,000 in seq BACT2. Time spent on BACT2 so far: 13.05 sec.\n",
      "\tOn aln 260,000 in seq BACT2. Time spent on BACT2 so far: 13.11 sec.\n",
      "\tOn aln 261,000 in seq BACT2. Time spent on BACT2 so far: 13.16 sec.\n",
      "\tOn aln 262,000 in seq BACT2. Time spent on BACT2 so far: 13.21 sec.\n",
      "\tOn aln 263,000 in seq BACT2. Time spent on BACT2 so far: 13.26 sec.\n",
      "\tOn aln 264,000 in seq BACT2. Time spent on BACT2 so far: 13.31 sec.\n",
      "\tOn aln 265,000 in seq BACT2. Time spent on BACT2 so far: 13.36 sec.\n",
      "\tOn aln 266,000 in seq BACT2. Time spent on BACT2 so far: 13.41 sec.\n",
      "\tOn aln 267,000 in seq BACT2. Time spent on BACT2 so far: 13.46 sec.\n",
      "\tOn aln 268,000 in seq BACT2. Time spent on BACT2 so far: 13.51 sec.\n",
      "\tOn aln 269,000 in seq BACT2. Time spent on BACT2 so far: 13.56 sec.\n",
      "\tOn aln 270,000 in seq BACT2. Time spent on BACT2 so far: 13.61 sec.\n",
      "\tOn aln 271,000 in seq BACT2. Time spent on BACT2 so far: 13.66 sec.\n",
      "\tOn aln 272,000 in seq BACT2. Time spent on BACT2 so far: 13.71 sec.\n",
      "\tOn aln 273,000 in seq BACT2. Time spent on BACT2 so far: 13.76 sec.\n",
      "\tOn aln 274,000 in seq BACT2. Time spent on BACT2 so far: 13.81 sec.\n",
      "\tOn aln 275,000 in seq BACT2. Time spent on BACT2 so far: 13.86 sec.\n",
      "\tOn aln 276,000 in seq BACT2. Time spent on BACT2 so far: 13.91 sec.\n",
      "\tOn aln 277,000 in seq BACT2. Time spent on BACT2 so far: 13.96 sec.\n",
      "\tOn aln 278,000 in seq BACT2. Time spent on BACT2 so far: 14.01 sec.\n",
      "\tOn aln 279,000 in seq BACT2. Time spent on BACT2 so far: 14.06 sec.\n",
      "\tOn aln 280,000 in seq BACT2. Time spent on BACT2 so far: 14.11 sec.\n",
      "\tOn aln 281,000 in seq BACT2. Time spent on BACT2 so far: 14.16 sec.\n",
      "\tOn aln 282,000 in seq BACT2. Time spent on BACT2 so far: 14.21 sec.\n",
      "\tOn aln 283,000 in seq BACT2. Time spent on BACT2 so far: 14.26 sec.\n",
      "\tOn aln 284,000 in seq BACT2. Time spent on BACT2 so far: 14.31 sec.\n",
      "\tOn aln 285,000 in seq BACT2. Time spent on BACT2 so far: 14.36 sec.\n",
      "\tOn aln 286,000 in seq BACT2. Time spent on BACT2 so far: 14.41 sec.\n",
      "\tOn aln 287,000 in seq BACT2. Time spent on BACT2 so far: 14.46 sec.\n",
      "\tOn aln 288,000 in seq BACT2. Time spent on BACT2 so far: 14.51 sec.\n",
      "\tOn aln 289,000 in seq BACT2. Time spent on BACT2 so far: 14.56 sec.\n",
      "\tOn aln 290,000 in seq BACT2. Time spent on BACT2 so far: 14.61 sec.\n",
      "\tOn aln 291,000 in seq BACT2. Time spent on BACT2 so far: 14.66 sec.\n",
      "\tOn aln 292,000 in seq BACT2. Time spent on BACT2 so far: 14.71 sec.\n",
      "\tOn aln 293,000 in seq BACT2. Time spent on BACT2 so far: 14.76 sec.\n",
      "\tOn aln 294,000 in seq BACT2. Time spent on BACT2 so far: 14.81 sec.\n",
      "\tOn aln 295,000 in seq BACT2. Time spent on BACT2 so far: 14.86 sec.\n",
      "\tOn aln 296,000 in seq BACT2. Time spent on BACT2 so far: 14.91 sec.\n",
      "\tOn aln 297,000 in seq BACT2. Time spent on BACT2 so far: 14.96 sec.\n",
      "\tOn aln 298,000 in seq BACT2. Time spent on BACT2 so far: 15.01 sec.\n",
      "\tOn aln 299,000 in seq BACT2. Time spent on BACT2 so far: 15.06 sec.\n",
      "\tOn aln 300,000 in seq BACT2. Time spent on BACT2 so far: 15.11 sec.\n",
      "\tOn aln 301,000 in seq BACT2. Time spent on BACT2 so far: 15.16 sec.\n",
      "\tOn aln 302,000 in seq BACT2. Time spent on BACT2 so far: 15.21 sec.\n",
      "\tOn aln 303,000 in seq BACT2. Time spent on BACT2 so far: 15.26 sec.\n",
      "\tOn aln 304,000 in seq BACT2. Time spent on BACT2 so far: 15.31 sec.\n",
      "\tOn aln 305,000 in seq BACT2. Time spent on BACT2 so far: 15.36 sec.\n",
      "\tOn aln 306,000 in seq BACT2. Time spent on BACT2 so far: 15.41 sec.\n",
      "\tOn aln 307,000 in seq BACT2. Time spent on BACT2 so far: 15.46 sec.\n",
      "\tOn aln 308,000 in seq BACT2. Time spent on BACT2 so far: 15.51 sec.\n",
      "\tOn aln 309,000 in seq BACT2. Time spent on BACT2 so far: 15.56 sec.\n",
      "\tOn aln 310,000 in seq BACT2. Time spent on BACT2 so far: 15.61 sec.\n",
      "\tOn aln 311,000 in seq BACT2. Time spent on BACT2 so far: 15.66 sec.\n",
      "\tOn aln 312,000 in seq BACT2. Time spent on BACT2 so far: 15.71 sec.\n",
      "\tOn aln 313,000 in seq BACT2. Time spent on BACT2 so far: 15.76 sec.\n",
      "\tOn aln 314,000 in seq BACT2. Time spent on BACT2 so far: 15.81 sec.\n",
      "\tOn aln 315,000 in seq BACT2. Time spent on BACT2 so far: 15.86 sec.\n",
      "\tOn aln 316,000 in seq BACT2. Time spent on BACT2 so far: 15.91 sec.\n",
      "\tOn aln 317,000 in seq BACT2. Time spent on BACT2 so far: 15.96 sec.\n",
      "\tOn aln 318,000 in seq BACT2. Time spent on BACT2 so far: 16.01 sec.\n",
      "\tOn aln 319,000 in seq BACT2. Time spent on BACT2 so far: 16.06 sec.\n",
      "\tOn aln 320,000 in seq BACT2. Time spent on BACT2 so far: 16.11 sec.\n",
      "\tOn aln 321,000 in seq BACT2. Time spent on BACT2 so far: 16.16 sec.\n",
      "\tOn aln 322,000 in seq BACT2. Time spent on BACT2 so far: 16.21 sec.\n",
      "\tOn aln 323,000 in seq BACT2. Time spent on BACT2 so far: 16.26 sec.\n",
      "\tOn aln 324,000 in seq BACT2. Time spent on BACT2 so far: 16.31 sec.\n",
      "\tOn aln 325,000 in seq BACT2. Time spent on BACT2 so far: 16.36 sec.\n",
      "\tOn aln 326,000 in seq BACT2. Time spent on BACT2 so far: 16.41 sec.\n",
      "\tOn aln 327,000 in seq BACT2. Time spent on BACT2 so far: 16.46 sec.\n",
      "\tOn aln 328,000 in seq BACT2. Time spent on BACT2 so far: 16.51 sec.\n",
      "\tOn aln 329,000 in seq BACT2. Time spent on BACT2 so far: 16.56 sec.\n",
      "\tOn aln 330,000 in seq BACT2. Time spent on BACT2 so far: 16.61 sec.\n",
      "\tOn aln 331,000 in seq BACT2. Time spent on BACT2 so far: 16.66 sec.\n",
      "\tOn aln 332,000 in seq BACT2. Time spent on BACT2 so far: 16.71 sec.\n",
      "\tOn aln 333,000 in seq BACT2. Time spent on BACT2 so far: 16.76 sec.\n",
      "\tOn aln 334,000 in seq BACT2. Time spent on BACT2 so far: 16.81 sec.\n",
      "\tOn aln 335,000 in seq BACT2. Time spent on BACT2 so far: 16.86 sec.\n",
      "\tOn aln 336,000 in seq BACT2. Time spent on BACT2 so far: 16.91 sec.\n",
      "\tOn aln 337,000 in seq BACT2. Time spent on BACT2 so far: 16.96 sec.\n",
      "\tOn aln 338,000 in seq BACT2. Time spent on BACT2 so far: 17.01 sec.\n",
      "\tOn aln 339,000 in seq BACT2. Time spent on BACT2 so far: 17.06 sec.\n",
      "\tOn aln 340,000 in seq BACT2. Time spent on BACT2 so far: 17.11 sec.\n",
      "\tOn aln 341,000 in seq BACT2. Time spent on BACT2 so far: 17.16 sec.\n",
      "\tOn aln 342,000 in seq BACT2. Time spent on BACT2 so far: 17.21 sec.\n",
      "\tOn aln 343,000 in seq BACT2. Time spent on BACT2 so far: 17.26 sec.\n",
      "\tOn aln 344,000 in seq BACT2. Time spent on BACT2 so far: 17.31 sec.\n",
      "\tOn aln 345,000 in seq BACT2. Time spent on BACT2 so far: 17.36 sec.\n",
      "\tOn aln 346,000 in seq BACT2. Time spent on BACT2 so far: 17.41 sec.\n",
      "\tOn aln 347,000 in seq BACT2. Time spent on BACT2 so far: 17.46 sec.\n",
      "\tOn aln 348,000 in seq BACT2. Time spent on BACT2 so far: 17.51 sec.\n",
      "\tOn aln 349,000 in seq BACT2. Time spent on BACT2 so far: 17.56 sec.\n",
      "\tOn aln 350,000 in seq BACT2. Time spent on BACT2 so far: 17.63 sec.\n",
      "\tOn aln 351,000 in seq BACT2. Time spent on BACT2 so far: 17.68 sec.\n",
      "\tOn aln 352,000 in seq BACT2. Time spent on BACT2 so far: 17.73 sec.\n",
      "\tOn aln 353,000 in seq BACT2. Time spent on BACT2 so far: 17.78 sec.\n",
      "\tOn aln 354,000 in seq BACT2. Time spent on BACT2 so far: 17.83 sec.\n",
      "\tOn aln 355,000 in seq BACT2. Time spent on BACT2 so far: 17.88 sec.\n",
      "\tOn aln 356,000 in seq BACT2. Time spent on BACT2 so far: 17.93 sec.\n",
      "\tOn aln 357,000 in seq BACT2. Time spent on BACT2 so far: 17.98 sec.\n",
      "\tOn aln 358,000 in seq BACT2. Time spent on BACT2 so far: 18.03 sec.\n",
      "\tOn aln 359,000 in seq BACT2. Time spent on BACT2 so far: 18.08 sec.\n",
      "\tOn aln 360,000 in seq BACT2. Time spent on BACT2 so far: 18.13 sec.\n",
      "\tOn aln 361,000 in seq BACT2. Time spent on BACT2 so far: 18.18 sec.\n",
      "\tOn aln 362,000 in seq BACT2. Time spent on BACT2 so far: 18.23 sec.\n",
      "\tOn aln 363,000 in seq BACT2. Time spent on BACT2 so far: 18.28 sec.\n",
      "\tOn aln 364,000 in seq BACT2. Time spent on BACT2 so far: 18.33 sec.\n",
      "\tOn aln 365,000 in seq BACT2. Time spent on BACT2 so far: 18.39 sec.\n",
      "\tOn aln 366,000 in seq BACT2. Time spent on BACT2 so far: 18.44 sec.\n",
      "\tOn aln 367,000 in seq BACT2. Time spent on BACT2 so far: 18.49 sec.\n",
      "\tOn aln 368,000 in seq BACT2. Time spent on BACT2 so far: 18.54 sec.\n",
      "\tOn aln 369,000 in seq BACT2. Time spent on BACT2 so far: 18.59 sec.\n",
      "\tOn aln 370,000 in seq BACT2. Time spent on BACT2 so far: 18.64 sec.\n",
      "\tOn aln 371,000 in seq BACT2. Time spent on BACT2 so far: 18.69 sec.\n",
      "\tOn aln 372,000 in seq BACT2. Time spent on BACT2 so far: 18.74 sec.\n",
      "\tOn aln 373,000 in seq BACT2. Time spent on BACT2 so far: 18.79 sec.\n",
      "\tOn aln 374,000 in seq BACT2. Time spent on BACT2 so far: 18.84 sec.\n",
      "\tOn aln 375,000 in seq BACT2. Time spent on BACT2 so far: 18.89 sec.\n",
      "\tOn aln 376,000 in seq BACT2. Time spent on BACT2 so far: 18.94 sec.\n",
      "\tOn aln 377,000 in seq BACT2. Time spent on BACT2 so far: 18.99 sec.\n",
      "\tOn aln 378,000 in seq BACT2. Time spent on BACT2 so far: 19.04 sec.\n",
      "\tOn aln 379,000 in seq BACT2. Time spent on BACT2 so far: 19.09 sec.\n",
      "\tOn aln 380,000 in seq BACT2. Time spent on BACT2 so far: 19.14 sec.\n",
      "\tOn aln 381,000 in seq BACT2. Time spent on BACT2 so far: 19.19 sec.\n",
      "\tOn aln 382,000 in seq BACT2. Time spent on BACT2 so far: 19.24 sec.\n",
      "\tOn aln 383,000 in seq BACT2. Time spent on BACT2 so far: 19.29 sec.\n",
      "\tOn aln 384,000 in seq BACT2. Time spent on BACT2 so far: 19.34 sec.\n",
      "\tOn aln 385,000 in seq BACT2. Time spent on BACT2 so far: 19.39 sec.\n",
      "\tOn aln 386,000 in seq BACT2. Time spent on BACT2 so far: 19.44 sec.\n",
      "\tOn aln 387,000 in seq BACT2. Time spent on BACT2 so far: 19.49 sec.\n",
      "\tOn aln 388,000 in seq BACT2. Time spent on BACT2 so far: 19.55 sec.\n",
      "\tOn aln 389,000 in seq BACT2. Time spent on BACT2 so far: 19.60 sec.\n",
      "\tOn aln 390,000 in seq BACT2. Time spent on BACT2 so far: 19.65 sec.\n",
      "\tOn aln 391,000 in seq BACT2. Time spent on BACT2 so far: 19.70 sec.\n",
      "\tOn aln 392,000 in seq BACT2. Time spent on BACT2 so far: 19.75 sec.\n",
      "\tOn aln 393,000 in seq BACT2. Time spent on BACT2 so far: 19.80 sec.\n",
      "\tOn aln 394,000 in seq BACT2. Time spent on BACT2 so far: 19.85 sec.\n",
      "\tOn aln 395,000 in seq BACT2. Time spent on BACT2 so far: 19.90 sec.\n",
      "\tOn aln 396,000 in seq BACT2. Time spent on BACT2 so far: 19.95 sec.\n",
      "\tOn aln 397,000 in seq BACT2. Time spent on BACT2 so far: 20.00 sec.\n",
      "\tOn aln 398,000 in seq BACT2. Time spent on BACT2 so far: 20.05 sec.\n",
      "\tOn aln 399,000 in seq BACT2. Time spent on BACT2 so far: 20.10 sec.\n",
      "\tOn aln 400,000 in seq BACT2. Time spent on BACT2 so far: 20.15 sec.\n",
      "\tOn aln 401,000 in seq BACT2. Time spent on BACT2 so far: 20.20 sec.\n",
      "\tOn aln 402,000 in seq BACT2. Time spent on BACT2 so far: 20.25 sec.\n",
      "\tOn aln 403,000 in seq BACT2. Time spent on BACT2 so far: 20.30 sec.\n",
      "\tOn aln 404,000 in seq BACT2. Time spent on BACT2 so far: 20.35 sec.\n",
      "\tOn aln 405,000 in seq BACT2. Time spent on BACT2 so far: 20.40 sec.\n",
      "\tOn aln 406,000 in seq BACT2. Time spent on BACT2 so far: 20.45 sec.\n",
      "\tOn aln 407,000 in seq BACT2. Time spent on BACT2 so far: 20.50 sec.\n",
      "\tOn aln 408,000 in seq BACT2. Time spent on BACT2 so far: 20.56 sec.\n",
      "\tOn aln 409,000 in seq BACT2. Time spent on BACT2 so far: 20.61 sec.\n",
      "\tOn aln 410,000 in seq BACT2. Time spent on BACT2 so far: 20.66 sec.\n",
      "\tOn aln 411,000 in seq BACT2. Time spent on BACT2 so far: 20.71 sec.\n",
      "\tOn aln 412,000 in seq BACT2. Time spent on BACT2 so far: 20.76 sec.\n",
      "\tOn aln 413,000 in seq BACT2. Time spent on BACT2 so far: 20.81 sec.\n",
      "\tOn aln 414,000 in seq BACT2. Time spent on BACT2 so far: 20.86 sec.\n",
      "\tOn aln 415,000 in seq BACT2. Time spent on BACT2 so far: 20.91 sec.\n",
      "\tOn aln 416,000 in seq BACT2. Time spent on BACT2 so far: 20.96 sec.\n",
      "\tOn aln 417,000 in seq BACT2. Time spent on BACT2 so far: 21.01 sec.\n",
      "\tOn aln 418,000 in seq BACT2. Time spent on BACT2 so far: 21.06 sec.\n",
      "\tOn aln 419,000 in seq BACT2. Time spent on BACT2 so far: 21.11 sec.\n",
      "\tOn aln 420,000 in seq BACT2. Time spent on BACT2 so far: 21.16 sec.\n",
      "\tOn aln 421,000 in seq BACT2. Time spent on BACT2 so far: 21.21 sec.\n",
      "\tOn aln 422,000 in seq BACT2. Time spent on BACT2 so far: 21.26 sec.\n",
      "\tOn aln 423,000 in seq BACT2. Time spent on BACT2 so far: 21.31 sec.\n",
      "\tOn aln 424,000 in seq BACT2. Time spent on BACT2 so far: 21.36 sec.\n",
      "\tOn aln 425,000 in seq BACT2. Time spent on BACT2 so far: 21.41 sec.\n",
      "\tOn aln 426,000 in seq BACT2. Time spent on BACT2 so far: 21.46 sec.\n",
      "\tOn aln 427,000 in seq BACT2. Time spent on BACT2 so far: 21.51 sec.\n",
      "\tOn aln 428,000 in seq BACT2. Time spent on BACT2 so far: 21.57 sec.\n",
      "\tOn aln 429,000 in seq BACT2. Time spent on BACT2 so far: 21.62 sec.\n",
      "\tOn aln 430,000 in seq BACT2. Time spent on BACT2 so far: 21.67 sec.\n",
      "\tOn aln 431,000 in seq BACT2. Time spent on BACT2 so far: 21.72 sec.\n",
      "\tOn aln 432,000 in seq BACT2. Time spent on BACT2 so far: 21.77 sec.\n",
      "\tOn aln 433,000 in seq BACT2. Time spent on BACT2 so far: 21.82 sec.\n",
      "\tOn aln 434,000 in seq BACT2. Time spent on BACT2 so far: 21.87 sec.\n",
      "\tOn aln 435,000 in seq BACT2. Time spent on BACT2 so far: 21.92 sec.\n",
      "\tOn aln 436,000 in seq BACT2. Time spent on BACT2 so far: 21.97 sec.\n",
      "\tOn aln 437,000 in seq BACT2. Time spent on BACT2 so far: 22.02 sec.\n",
      "\tOn aln 438,000 in seq BACT2. Time spent on BACT2 so far: 22.07 sec.\n",
      "\tOn aln 439,000 in seq BACT2. Time spent on BACT2 so far: 22.12 sec.\n",
      "\tOn aln 440,000 in seq BACT2. Time spent on BACT2 so far: 22.17 sec.\n",
      "\tOn aln 441,000 in seq BACT2. Time spent on BACT2 so far: 22.22 sec.\n",
      "\tOn aln 442,000 in seq BACT2. Time spent on BACT2 so far: 22.27 sec.\n",
      "\tOn aln 443,000 in seq BACT2. Time spent on BACT2 so far: 22.32 sec.\n",
      "\tOn aln 444,000 in seq BACT2. Time spent on BACT2 so far: 22.37 sec.\n",
      "\tOn aln 445,000 in seq BACT2. Time spent on BACT2 so far: 22.42 sec.\n",
      "\tOn aln 446,000 in seq BACT2. Time spent on BACT2 so far: 22.47 sec.\n",
      "\tOn aln 447,000 in seq BACT2. Time spent on BACT2 so far: 22.52 sec.\n",
      "\tOn aln 448,000 in seq BACT2. Time spent on BACT2 so far: 22.57 sec.\n",
      "\tOn aln 449,000 in seq BACT2. Time spent on BACT2 so far: 22.62 sec.\n",
      "\tOn aln 450,000 in seq BACT2. Time spent on BACT2 so far: 22.67 sec.\n",
      "\tOn aln 451,000 in seq BACT2. Time spent on BACT2 so far: 22.72 sec.\n",
      "\tOn aln 452,000 in seq BACT2. Time spent on BACT2 so far: 22.77 sec.\n",
      "\tOn aln 453,000 in seq BACT2. Time spent on BACT2 so far: 22.83 sec.\n",
      "\tOn aln 454,000 in seq BACT2. Time spent on BACT2 so far: 22.88 sec.\n",
      "\tOn aln 455,000 in seq BACT2. Time spent on BACT2 so far: 22.93 sec.\n",
      "\tOn aln 456,000 in seq BACT2. Time spent on BACT2 so far: 22.98 sec.\n",
      "\tOn aln 457,000 in seq BACT2. Time spent on BACT2 so far: 23.03 sec.\n",
      "\tOn aln 458,000 in seq BACT2. Time spent on BACT2 so far: 23.08 sec.\n",
      "\tOn aln 459,000 in seq BACT2. Time spent on BACT2 so far: 23.13 sec.\n",
      "\tOn aln 460,000 in seq BACT2. Time spent on BACT2 so far: 23.18 sec.\n",
      "\tOn aln 461,000 in seq BACT2. Time spent on BACT2 so far: 23.23 sec.\n",
      "\tOn aln 462,000 in seq BACT2. Time spent on BACT2 so far: 23.28 sec.\n",
      "\tOn aln 463,000 in seq BACT2. Time spent on BACT2 so far: 23.33 sec.\n",
      "\tOn aln 464,000 in seq BACT2. Time spent on BACT2 so far: 23.38 sec.\n",
      "\tOn aln 465,000 in seq BACT2. Time spent on BACT2 so far: 23.43 sec.\n",
      "\tOn aln 466,000 in seq BACT2. Time spent on BACT2 so far: 23.48 sec.\n",
      "\tOn aln 467,000 in seq BACT2. Time spent on BACT2 so far: 23.53 sec.\n",
      "\tOn aln 468,000 in seq BACT2. Time spent on BACT2 so far: 23.59 sec.\n",
      "\tOn aln 469,000 in seq BACT2. Time spent on BACT2 so far: 23.64 sec.\n",
      "\tOn aln 470,000 in seq BACT2. Time spent on BACT2 so far: 23.69 sec.\n",
      "\tOn aln 471,000 in seq BACT2. Time spent on BACT2 so far: 23.74 sec.\n",
      "\tOn aln 472,000 in seq BACT2. Time spent on BACT2 so far: 23.79 sec.\n",
      "\tOn aln 473,000 in seq BACT2. Time spent on BACT2 so far: 23.84 sec.\n",
      "\tOn aln 474,000 in seq BACT2. Time spent on BACT2 so far: 23.89 sec.\n",
      "\tOn aln 475,000 in seq BACT2. Time spent on BACT2 so far: 23.94 sec.\n",
      "\tOn aln 476,000 in seq BACT2. Time spent on BACT2 so far: 23.99 sec.\n",
      "\tOn aln 477,000 in seq BACT2. Time spent on BACT2 so far: 24.04 sec.\n",
      "\tOn aln 478,000 in seq BACT2. Time spent on BACT2 so far: 24.09 sec.\n",
      "\tOn aln 479,000 in seq BACT2. Time spent on BACT2 so far: 24.14 sec.\n",
      "\tOn aln 480,000 in seq BACT2. Time spent on BACT2 so far: 24.19 sec.\n",
      "\tOn aln 481,000 in seq BACT2. Time spent on BACT2 so far: 24.24 sec.\n",
      "\tOn aln 482,000 in seq BACT2. Time spent on BACT2 so far: 24.29 sec.\n",
      "\tOn aln 483,000 in seq BACT2. Time spent on BACT2 so far: 24.34 sec.\n",
      "\tOn aln 484,000 in seq BACT2. Time spent on BACT2 so far: 24.40 sec.\n",
      "\tOn aln 485,000 in seq BACT2. Time spent on BACT2 so far: 24.45 sec.\n",
      "\tOn aln 486,000 in seq BACT2. Time spent on BACT2 so far: 24.50 sec.\n",
      "\tOn aln 487,000 in seq BACT2. Time spent on BACT2 so far: 24.55 sec.\n",
      "\tOn aln 488,000 in seq BACT2. Time spent on BACT2 so far: 24.60 sec.\n",
      "\tOn aln 489,000 in seq BACT2. Time spent on BACT2 so far: 24.64 sec.\n",
      "\tOn aln 490,000 in seq BACT2. Time spent on BACT2 so far: 24.70 sec.\n",
      "\tOn aln 491,000 in seq BACT2. Time spent on BACT2 so far: 24.74 sec.\n",
      "\tOn aln 492,000 in seq BACT2. Time spent on BACT2 so far: 24.79 sec.\n",
      "\tOn aln 493,000 in seq BACT2. Time spent on BACT2 so far: 24.84 sec.\n",
      "\tOn aln 494,000 in seq BACT2. Time spent on BACT2 so far: 24.89 sec.\n",
      "\tOn aln 495,000 in seq BACT2. Time spent on BACT2 so far: 24.94 sec.\n",
      "\tOn aln 496,000 in seq BACT2. Time spent on BACT2 so far: 24.99 sec.\n",
      "\tOn aln 497,000 in seq BACT2. Time spent on BACT2 so far: 25.04 sec.\n",
      "\tOn aln 498,000 in seq BACT2. Time spent on BACT2 so far: 25.09 sec.\n",
      "\tOn aln 499,000 in seq BACT2. Time spent on BACT2 so far: 25.15 sec.\n",
      "\tOn aln 500,000 in seq BACT2. Time spent on BACT2 so far: 25.20 sec.\n",
      "\tOn aln 501,000 in seq BACT2. Time spent on BACT2 so far: 25.25 sec.\n",
      "\tOn aln 502,000 in seq BACT2. Time spent on BACT2 so far: 25.30 sec.\n",
      "\tOn aln 503,000 in seq BACT2. Time spent on BACT2 so far: 25.35 sec.\n",
      "\tOn aln 504,000 in seq BACT2. Time spent on BACT2 so far: 25.40 sec.\n",
      "\tOn aln 505,000 in seq BACT2. Time spent on BACT2 so far: 25.45 sec.\n",
      "\tOn aln 506,000 in seq BACT2. Time spent on BACT2 so far: 25.50 sec.\n",
      "\tOn aln 507,000 in seq BACT2. Time spent on BACT2 so far: 25.55 sec.\n",
      "\tOn aln 508,000 in seq BACT2. Time spent on BACT2 so far: 25.60 sec.\n",
      "\tOn aln 509,000 in seq BACT2. Time spent on BACT2 so far: 25.66 sec.\n",
      "\tOn aln 510,000 in seq BACT2. Time spent on BACT2 so far: 25.71 sec.\n",
      "\tOn aln 511,000 in seq BACT2. Time spent on BACT2 so far: 25.76 sec.\n",
      "\tOn aln 512,000 in seq BACT2. Time spent on BACT2 so far: 25.81 sec.\n",
      "\tOn aln 513,000 in seq BACT2. Time spent on BACT2 so far: 25.86 sec.\n",
      "\tOn aln 514,000 in seq BACT2. Time spent on BACT2 so far: 25.91 sec.\n",
      "\tOn aln 515,000 in seq BACT2. Time spent on BACT2 so far: 25.96 sec.\n",
      "\tOn aln 516,000 in seq BACT2. Time spent on BACT2 so far: 26.01 sec.\n",
      "\tOn aln 517,000 in seq BACT2. Time spent on BACT2 so far: 26.06 sec.\n",
      "\tOn aln 518,000 in seq BACT2. Time spent on BACT2 so far: 26.11 sec.\n",
      "\tOn aln 519,000 in seq BACT2. Time spent on BACT2 so far: 26.16 sec.\n",
      "\tOn aln 520,000 in seq BACT2. Time spent on BACT2 so far: 26.22 sec.\n",
      "\tOn aln 521,000 in seq BACT2. Time spent on BACT2 so far: 26.27 sec.\n",
      "\tOn aln 522,000 in seq BACT2. Time spent on BACT2 so far: 26.32 sec.\n",
      "\tOn aln 523,000 in seq BACT2. Time spent on BACT2 so far: 26.37 sec.\n",
      "\tOn aln 524,000 in seq BACT2. Time spent on BACT2 so far: 26.42 sec.\n",
      "\tOn aln 525,000 in seq BACT2. Time spent on BACT2 so far: 26.47 sec.\n",
      "\tOn aln 526,000 in seq BACT2. Time spent on BACT2 so far: 26.52 sec.\n",
      "\tOn aln 527,000 in seq BACT2. Time spent on BACT2 so far: 26.57 sec.\n",
      "\tOn aln 528,000 in seq BACT2. Time spent on BACT2 so far: 26.62 sec.\n",
      "\tOn aln 529,000 in seq BACT2. Time spent on BACT2 so far: 26.67 sec.\n",
      "\tOn aln 530,000 in seq BACT2. Time spent on BACT2 so far: 26.72 sec.\n",
      "\tOn aln 531,000 in seq BACT2. Time spent on BACT2 so far: 26.77 sec.\n",
      "\tOn aln 532,000 in seq BACT2. Time spent on BACT2 so far: 26.82 sec.\n",
      "\tOn aln 533,000 in seq BACT2. Time spent on BACT2 so far: 26.87 sec.\n",
      "\tOn aln 534,000 in seq BACT2. Time spent on BACT2 so far: 26.92 sec.\n",
      "\tOn aln 535,000 in seq BACT2. Time spent on BACT2 so far: 26.97 sec.\n",
      "\tOn aln 536,000 in seq BACT2. Time spent on BACT2 so far: 27.02 sec.\n",
      "\tOn aln 537,000 in seq BACT2. Time spent on BACT2 so far: 27.07 sec.\n",
      "\tOn aln 538,000 in seq BACT2. Time spent on BACT2 so far: 27.13 sec.\n",
      "\tOn aln 539,000 in seq BACT2. Time spent on BACT2 so far: 27.18 sec.\n",
      "\tOn aln 540,000 in seq BACT2. Time spent on BACT2 so far: 27.23 sec.\n",
      "\tOn aln 541,000 in seq BACT2. Time spent on BACT2 so far: 27.28 sec.\n",
      "\tOn aln 542,000 in seq BACT2. Time spent on BACT2 so far: 27.33 sec.\n",
      "\tOn aln 543,000 in seq BACT2. Time spent on BACT2 so far: 27.37 sec.\n",
      "\tOn aln 544,000 in seq BACT2. Time spent on BACT2 so far: 27.43 sec.\n",
      "\tOn aln 545,000 in seq BACT2. Time spent on BACT2 so far: 27.47 sec.\n",
      "\tOn aln 546,000 in seq BACT2. Time spent on BACT2 so far: 27.53 sec.\n",
      "\tOn aln 547,000 in seq BACT2. Time spent on BACT2 so far: 27.58 sec.\n",
      "\tOn aln 548,000 in seq BACT2. Time spent on BACT2 so far: 27.63 sec.\n",
      "\tOn aln 549,000 in seq BACT2. Time spent on BACT2 so far: 27.68 sec.\n",
      "\tOn aln 550,000 in seq BACT2. Time spent on BACT2 so far: 27.73 sec.\n",
      "\tOn aln 551,000 in seq BACT2. Time spent on BACT2 so far: 27.78 sec.\n",
      "\tOn aln 552,000 in seq BACT2. Time spent on BACT2 so far: 27.84 sec.\n",
      "\tOn aln 553,000 in seq BACT2. Time spent on BACT2 so far: 27.89 sec.\n",
      "\tOn aln 554,000 in seq BACT2. Time spent on BACT2 so far: 27.93 sec.\n",
      "\tOn aln 555,000 in seq BACT2. Time spent on BACT2 so far: 27.99 sec.\n",
      "\tOn aln 556,000 in seq BACT2. Time spent on BACT2 so far: 28.03 sec.\n",
      "\tOn aln 557,000 in seq BACT2. Time spent on BACT2 so far: 28.09 sec.\n",
      "\tOn aln 558,000 in seq BACT2. Time spent on BACT2 so far: 28.14 sec.\n",
      "\tOn aln 559,000 in seq BACT2. Time spent on BACT2 so far: 28.18 sec.\n",
      "\tOn aln 560,000 in seq BACT2. Time spent on BACT2 so far: 28.24 sec.\n",
      "\tOn aln 561,000 in seq BACT2. Time spent on BACT2 so far: 28.28 sec.\n",
      "\tOn aln 562,000 in seq BACT2. Time spent on BACT2 so far: 28.33 sec.\n",
      "\tOn aln 563,000 in seq BACT2. Time spent on BACT2 so far: 28.38 sec.\n",
      "\tOn aln 564,000 in seq BACT2. Time spent on BACT2 so far: 28.44 sec.\n",
      "\tOn aln 565,000 in seq BACT2. Time spent on BACT2 so far: 28.49 sec.\n",
      "\tOn aln 566,000 in seq BACT2. Time spent on BACT2 so far: 28.54 sec.\n",
      "\tOn aln 567,000 in seq BACT2. Time spent on BACT2 so far: 28.59 sec.\n",
      "\tOn aln 568,000 in seq BACT2. Time spent on BACT2 so far: 28.64 sec.\n",
      "\tOn aln 569,000 in seq BACT2. Time spent on BACT2 so far: 28.69 sec.\n",
      "\tOn aln 570,000 in seq BACT2. Time spent on BACT2 so far: 28.74 sec.\n",
      "\tOn aln 571,000 in seq BACT2. Time spent on BACT2 so far: 28.79 sec.\n",
      "\tOn aln 572,000 in seq BACT2. Time spent on BACT2 so far: 28.84 sec.\n",
      "\tOn aln 573,000 in seq BACT2. Time spent on BACT2 so far: 28.89 sec.\n",
      "\tOn aln 574,000 in seq BACT2. Time spent on BACT2 so far: 28.94 sec.\n",
      "\tOn aln 575,000 in seq BACT2. Time spent on BACT2 so far: 28.99 sec.\n",
      "\tOn aln 576,000 in seq BACT2. Time spent on BACT2 so far: 29.04 sec.\n",
      "\tOn aln 577,000 in seq BACT2. Time spent on BACT2 so far: 29.10 sec.\n",
      "\tOn aln 578,000 in seq BACT2. Time spent on BACT2 so far: 29.15 sec.\n",
      "\tOn aln 579,000 in seq BACT2. Time spent on BACT2 so far: 29.20 sec.\n",
      "\tOn aln 580,000 in seq BACT2. Time spent on BACT2 so far: 29.25 sec.\n",
      "\tOn aln 581,000 in seq BACT2. Time spent on BACT2 so far: 29.30 sec.\n",
      "\tOn aln 582,000 in seq BACT2. Time spent on BACT2 so far: 29.35 sec.\n",
      "\tOn aln 583,000 in seq BACT2. Time spent on BACT2 so far: 29.40 sec.\n",
      "\tOn aln 584,000 in seq BACT2. Time spent on BACT2 so far: 29.45 sec.\n",
      "\tOn aln 585,000 in seq BACT2. Time spent on BACT2 so far: 29.50 sec.\n",
      "\tOn aln 586,000 in seq BACT2. Time spent on BACT2 so far: 29.55 sec.\n",
      "\tOn aln 587,000 in seq BACT2. Time spent on BACT2 so far: 29.60 sec.\n",
      "\tOn aln 588,000 in seq BACT2. Time spent on BACT2 so far: 29.65 sec.\n",
      "\tOn aln 589,000 in seq BACT2. Time spent on BACT2 so far: 29.70 sec.\n",
      "\tOn aln 590,000 in seq BACT2. Time spent on BACT2 so far: 29.75 sec.\n",
      "\tOn aln 591,000 in seq BACT2. Time spent on BACT2 so far: 29.81 sec.\n",
      "\tOn aln 592,000 in seq BACT2. Time spent on BACT2 so far: 29.86 sec.\n",
      "\tOn aln 593,000 in seq BACT2. Time spent on BACT2 so far: 29.90 sec.\n",
      "\tOn aln 594,000 in seq BACT2. Time spent on BACT2 so far: 29.95 sec.\n",
      "\tOn aln 595,000 in seq BACT2. Time spent on BACT2 so far: 30.00 sec.\n",
      "\tOn aln 596,000 in seq BACT2. Time spent on BACT2 so far: 30.06 sec.\n",
      "\tOn aln 597,000 in seq BACT2. Time spent on BACT2 so far: 30.11 sec.\n",
      "\tOn aln 598,000 in seq BACT2. Time spent on BACT2 so far: 30.15 sec.\n",
      "\tOn aln 599,000 in seq BACT2. Time spent on BACT2 so far: 30.21 sec.\n",
      "\tOn aln 600,000 in seq BACT2. Time spent on BACT2 so far: 30.26 sec.\n",
      "\tOn aln 601,000 in seq BACT2. Time spent on BACT2 so far: 30.31 sec.\n",
      "\tOn aln 602,000 in seq BACT2. Time spent on BACT2 so far: 30.36 sec.\n",
      "\tOn aln 603,000 in seq BACT2. Time spent on BACT2 so far: 30.41 sec.\n",
      "\tOn aln 604,000 in seq BACT2. Time spent on BACT2 so far: 30.46 sec.\n",
      "\tOn aln 605,000 in seq BACT2. Time spent on BACT2 so far: 30.52 sec.\n",
      "\tOn aln 606,000 in seq BACT2. Time spent on BACT2 so far: 30.57 sec.\n",
      "\tOn aln 607,000 in seq BACT2. Time spent on BACT2 so far: 30.62 sec.\n",
      "\tOn aln 608,000 in seq BACT2. Time spent on BACT2 so far: 30.67 sec.\n",
      "\tOn aln 609,000 in seq BACT2. Time spent on BACT2 so far: 30.72 sec.\n",
      "\tOn aln 610,000 in seq BACT2. Time spent on BACT2 so far: 30.77 sec.\n",
      "\tOn aln 611,000 in seq BACT2. Time spent on BACT2 so far: 30.82 sec.\n",
      "\tOn aln 612,000 in seq BACT2. Time spent on BACT2 so far: 30.87 sec.\n",
      "\tOn aln 613,000 in seq BACT2. Time spent on BACT2 so far: 30.92 sec.\n",
      "\tOn aln 614,000 in seq BACT2. Time spent on BACT2 so far: 30.97 sec.\n",
      "\tOn aln 615,000 in seq BACT2. Time spent on BACT2 so far: 31.03 sec.\n",
      "\tOn aln 616,000 in seq BACT2. Time spent on BACT2 so far: 31.08 sec.\n",
      "\tOn aln 617,000 in seq BACT2. Time spent on BACT2 so far: 31.13 sec.\n",
      "\tOn aln 618,000 in seq BACT2. Time spent on BACT2 so far: 31.18 sec.\n",
      "\tOn aln 619,000 in seq BACT2. Time spent on BACT2 so far: 31.23 sec.\n",
      "\tOn aln 620,000 in seq BACT2. Time spent on BACT2 so far: 31.28 sec.\n",
      "\tOn aln 621,000 in seq BACT2. Time spent on BACT2 so far: 31.33 sec.\n",
      "\tOn aln 622,000 in seq BACT2. Time spent on BACT2 so far: 31.38 sec.\n",
      "\tOn aln 623,000 in seq BACT2. Time spent on BACT2 so far: 31.43 sec.\n",
      "\tOn aln 624,000 in seq BACT2. Time spent on BACT2 so far: 31.48 sec.\n",
      "\tOn aln 625,000 in seq BACT2. Time spent on BACT2 so far: 31.53 sec.\n",
      "\tOn aln 626,000 in seq BACT2. Time spent on BACT2 so far: 31.58 sec.\n",
      "\tOn aln 627,000 in seq BACT2. Time spent on BACT2 so far: 31.63 sec.\n",
      "\tOn aln 628,000 in seq BACT2. Time spent on BACT2 so far: 31.68 sec.\n",
      "\tOn aln 629,000 in seq BACT2. Time spent on BACT2 so far: 31.73 sec.\n",
      "\tOn aln 630,000 in seq BACT2. Time spent on BACT2 so far: 31.78 sec.\n",
      "\tOn aln 631,000 in seq BACT2. Time spent on BACT2 so far: 31.83 sec.\n",
      "\tOn aln 632,000 in seq BACT2. Time spent on BACT2 so far: 31.88 sec.\n",
      "\tOn aln 633,000 in seq BACT2. Time spent on BACT2 so far: 31.93 sec.\n",
      "\tOn aln 634,000 in seq BACT2. Time spent on BACT2 so far: 31.98 sec.\n",
      "\tOn aln 635,000 in seq BACT2. Time spent on BACT2 so far: 32.04 sec.\n",
      "\tOn aln 636,000 in seq BACT2. Time spent on BACT2 so far: 32.09 sec.\n",
      "\tOn aln 637,000 in seq BACT2. Time spent on BACT2 so far: 32.14 sec.\n",
      "\tOn aln 638,000 in seq BACT2. Time spent on BACT2 so far: 32.19 sec.\n",
      "\tOn aln 639,000 in seq BACT2. Time spent on BACT2 so far: 32.24 sec.\n",
      "\tOn aln 640,000 in seq BACT2. Time spent on BACT2 so far: 32.29 sec.\n",
      "\tOn aln 641,000 in seq BACT2. Time spent on BACT2 so far: 32.34 sec.\n",
      "\tOn aln 642,000 in seq BACT2. Time spent on BACT2 so far: 32.39 sec.\n",
      "\tOn aln 643,000 in seq BACT2. Time spent on BACT2 so far: 32.44 sec.\n",
      "\tOn aln 644,000 in seq BACT2. Time spent on BACT2 so far: 32.49 sec.\n",
      "\tOn aln 645,000 in seq BACT2. Time spent on BACT2 so far: 32.54 sec.\n",
      "\tOn aln 646,000 in seq BACT2. Time spent on BACT2 so far: 32.59 sec.\n",
      "\tOn aln 647,000 in seq BACT2. Time spent on BACT2 so far: 32.64 sec.\n",
      "\tOn aln 648,000 in seq BACT2. Time spent on BACT2 so far: 32.69 sec.\n",
      "\tOn aln 649,000 in seq BACT2. Time spent on BACT2 so far: 32.74 sec.\n",
      "\tOn aln 650,000 in seq BACT2. Time spent on BACT2 so far: 32.79 sec.\n",
      "\tOn aln 651,000 in seq BACT2. Time spent on BACT2 so far: 32.84 sec.\n",
      "\tOn aln 652,000 in seq BACT2. Time spent on BACT2 so far: 32.90 sec.\n",
      "\tOn aln 653,000 in seq BACT2. Time spent on BACT2 so far: 32.96 sec.\n",
      "\tOn aln 654,000 in seq BACT2. Time spent on BACT2 so far: 33.01 sec.\n",
      "\tOn aln 655,000 in seq BACT2. Time spent on BACT2 so far: 33.07 sec.\n",
      "\tOn aln 656,000 in seq BACT2. Time spent on BACT2 so far: 33.13 sec.\n",
      "\tOn aln 657,000 in seq BACT2. Time spent on BACT2 so far: 33.18 sec.\n",
      "\tOn aln 658,000 in seq BACT2. Time spent on BACT2 so far: 33.23 sec.\n",
      "\tOn aln 659,000 in seq BACT2. Time spent on BACT2 so far: 33.28 sec.\n",
      "\tOn aln 660,000 in seq BACT2. Time spent on BACT2 so far: 33.33 sec.\n",
      "\tOn aln 661,000 in seq BACT2. Time spent on BACT2 so far: 33.38 sec.\n",
      "\tOn aln 662,000 in seq BACT2. Time spent on BACT2 so far: 33.43 sec.\n",
      "\tOn aln 663,000 in seq BACT2. Time spent on BACT2 so far: 33.48 sec.\n",
      "\tOn aln 664,000 in seq BACT2. Time spent on BACT2 so far: 33.53 sec.\n",
      "\tOn aln 665,000 in seq BACT2. Time spent on BACT2 so far: 33.59 sec.\n",
      "\tOn aln 666,000 in seq BACT2. Time spent on BACT2 so far: 33.64 sec.\n",
      "\tOn aln 667,000 in seq BACT2. Time spent on BACT2 so far: 33.69 sec.\n",
      "\tOn aln 668,000 in seq BACT2. Time spent on BACT2 so far: 33.73 sec.\n",
      "\tOn aln 669,000 in seq BACT2. Time spent on BACT2 so far: 33.78 sec.\n",
      "\tOn aln 670,000 in seq BACT2. Time spent on BACT2 so far: 33.83 sec.\n",
      "\tOn aln 671,000 in seq BACT2. Time spent on BACT2 so far: 33.88 sec.\n",
      "\tOn aln 672,000 in seq BACT2. Time spent on BACT2 so far: 33.93 sec.\n",
      "\tOn aln 673,000 in seq BACT2. Time spent on BACT2 so far: 33.98 sec.\n",
      "\tOn aln 674,000 in seq BACT2. Time spent on BACT2 so far: 34.03 sec.\n",
      "\tOn aln 675,000 in seq BACT2. Time spent on BACT2 so far: 34.08 sec.\n",
      "\tOn aln 676,000 in seq BACT2. Time spent on BACT2 so far: 34.14 sec.\n",
      "\tOn aln 677,000 in seq BACT2. Time spent on BACT2 so far: 34.19 sec.\n",
      "\tOn aln 678,000 in seq BACT2. Time spent on BACT2 so far: 34.24 sec.\n",
      "\tOn aln 679,000 in seq BACT2. Time spent on BACT2 so far: 34.29 sec.\n",
      "\tOn aln 680,000 in seq BACT2. Time spent on BACT2 so far: 34.34 sec.\n",
      "\tOn aln 681,000 in seq BACT2. Time spent on BACT2 so far: 34.39 sec.\n",
      "\tOn aln 682,000 in seq BACT2. Time spent on BACT2 so far: 34.44 sec.\n",
      "\tOn aln 683,000 in seq BACT2. Time spent on BACT2 so far: 34.49 sec.\n",
      "\tOn aln 684,000 in seq BACT2. Time spent on BACT2 so far: 34.54 sec.\n",
      "\tOn aln 685,000 in seq BACT2. Time spent on BACT2 so far: 34.59 sec.\n",
      "\tOn aln 686,000 in seq BACT2. Time spent on BACT2 so far: 34.64 sec.\n",
      "\tOn aln 687,000 in seq BACT2. Time spent on BACT2 so far: 34.70 sec.\n",
      "\tOn aln 688,000 in seq BACT2. Time spent on BACT2 so far: 34.75 sec.\n",
      "\tOn aln 689,000 in seq BACT2. Time spent on BACT2 so far: 34.80 sec.\n",
      "\tOn aln 690,000 in seq BACT2. Time spent on BACT2 so far: 34.85 sec.\n",
      "\tOn aln 691,000 in seq BACT2. Time spent on BACT2 so far: 34.90 sec.\n",
      "\tOn aln 692,000 in seq BACT2. Time spent on BACT2 so far: 34.95 sec.\n",
      "\tOn aln 693,000 in seq BACT2. Time spent on BACT2 so far: 35.00 sec.\n",
      "\tOn aln 694,000 in seq BACT2. Time spent on BACT2 so far: 35.05 sec.\n",
      "\tOn aln 695,000 in seq BACT2. Time spent on BACT2 so far: 35.10 sec.\n",
      "\tOn aln 696,000 in seq BACT2. Time spent on BACT2 so far: 35.15 sec.\n",
      "\tOn aln 697,000 in seq BACT2. Time spent on BACT2 so far: 35.21 sec.\n",
      "\tOn aln 698,000 in seq BACT2. Time spent on BACT2 so far: 35.26 sec.\n",
      "\tOn aln 699,000 in seq BACT2. Time spent on BACT2 so far: 35.31 sec.\n",
      "\tOn aln 700,000 in seq BACT2. Time spent on BACT2 so far: 35.36 sec.\n",
      "\tOn aln 701,000 in seq BACT2. Time spent on BACT2 so far: 35.44 sec.\n",
      "\tOn aln 702,000 in seq BACT2. Time spent on BACT2 so far: 35.50 sec.\n",
      "\tOn aln 703,000 in seq BACT2. Time spent on BACT2 so far: 35.55 sec.\n",
      "\tOn aln 704,000 in seq BACT2. Time spent on BACT2 so far: 35.60 sec.\n",
      "\tOn aln 705,000 in seq BACT2. Time spent on BACT2 so far: 35.65 sec.\n",
      "\tOn aln 706,000 in seq BACT2. Time spent on BACT2 so far: 35.70 sec.\n",
      "\tOn aln 707,000 in seq BACT2. Time spent on BACT2 so far: 35.76 sec.\n",
      "\tOn aln 708,000 in seq BACT2. Time spent on BACT2 so far: 35.81 sec.\n",
      "\tOn aln 709,000 in seq BACT2. Time spent on BACT2 so far: 35.86 sec.\n",
      "\tOn aln 710,000 in seq BACT2. Time spent on BACT2 so far: 35.91 sec.\n",
      "\tOn aln 711,000 in seq BACT2. Time spent on BACT2 so far: 35.96 sec.\n",
      "\tOn aln 712,000 in seq BACT2. Time spent on BACT2 so far: 36.02 sec.\n",
      "\tOn aln 713,000 in seq BACT2. Time spent on BACT2 so far: 36.07 sec.\n",
      "\tOn aln 714,000 in seq BACT2. Time spent on BACT2 so far: 36.12 sec.\n",
      "\tOn aln 715,000 in seq BACT2. Time spent on BACT2 so far: 36.17 sec.\n",
      "\tOn aln 716,000 in seq BACT2. Time spent on BACT2 so far: 36.22 sec.\n",
      "\tOn aln 717,000 in seq BACT2. Time spent on BACT2 so far: 36.27 sec.\n",
      "\tOn aln 718,000 in seq BACT2. Time spent on BACT2 so far: 36.32 sec.\n",
      "\tOn aln 719,000 in seq BACT2. Time spent on BACT2 so far: 36.37 sec.\n",
      "\tOn aln 720,000 in seq BACT2. Time spent on BACT2 so far: 36.43 sec.\n",
      "\tOn aln 721,000 in seq BACT2. Time spent on BACT2 so far: 36.48 sec.\n",
      "\tOn aln 722,000 in seq BACT2. Time spent on BACT2 so far: 36.52 sec.\n",
      "\tOn aln 723,000 in seq BACT2. Time spent on BACT2 so far: 36.57 sec.\n",
      "\tOn aln 724,000 in seq BACT2. Time spent on BACT2 so far: 36.63 sec.\n",
      "\tOn aln 725,000 in seq BACT2. Time spent on BACT2 so far: 36.68 sec.\n",
      "\tOn aln 726,000 in seq BACT2. Time spent on BACT2 so far: 36.73 sec.\n",
      "\tOn aln 727,000 in seq BACT2. Time spent on BACT2 so far: 36.78 sec.\n",
      "\tOn aln 728,000 in seq BACT2. Time spent on BACT2 so far: 36.83 sec.\n",
      "\tOn aln 729,000 in seq BACT2. Time spent on BACT2 so far: 36.88 sec.\n",
      "\tOn aln 730,000 in seq BACT2. Time spent on BACT2 so far: 36.93 sec.\n",
      "\tOn aln 731,000 in seq BACT2. Time spent on BACT2 so far: 36.98 sec.\n",
      "\tOn aln 732,000 in seq BACT2. Time spent on BACT2 so far: 37.03 sec.\n",
      "\tOn aln 733,000 in seq BACT2. Time spent on BACT2 so far: 37.08 sec.\n",
      "\tOn aln 734,000 in seq BACT2. Time spent on BACT2 so far: 37.13 sec.\n",
      "\tOn aln 735,000 in seq BACT2. Time spent on BACT2 so far: 37.19 sec.\n",
      "\tOn aln 736,000 in seq BACT2. Time spent on BACT2 so far: 37.24 sec.\n",
      "\tOn aln 737,000 in seq BACT2. Time spent on BACT2 so far: 37.29 sec.\n",
      "\tOn aln 738,000 in seq BACT2. Time spent on BACT2 so far: 37.35 sec.\n",
      "\tOn aln 739,000 in seq BACT2. Time spent on BACT2 so far: 37.40 sec.\n",
      "\tOn aln 740,000 in seq BACT2. Time spent on BACT2 so far: 37.45 sec.\n",
      "\tOn aln 741,000 in seq BACT2. Time spent on BACT2 so far: 37.50 sec.\n",
      "\tOn aln 742,000 in seq BACT2. Time spent on BACT2 so far: 37.54 sec.\n",
      "Done with edge_2358! Took 37.56 sec.\n",
      "Time taken: 116.8143961429596 sec.\n"
     ]
    }
   ],
   "source": [
    "bf = pysam.AlignmentFile(\"../main-workflow/output/fully-filtered-and-sorted-aln.bam\", \"rb\")\n",
    "output_dir = \"phasing-data/smoothed-reads-nomutsatall/\"\n",
    "\n",
    "# verbose?\n",
    "no_indoor_voice = False\n",
    "\n",
    "def write_out_reads(filepath, readname2seq):\n",
    "    # Notably, this uses the \"a\" (append) method in order to add to the end of a file\n",
    "    with open(filepath, \"a\") as of:\n",
    "        for readname in readname2seq:\n",
    "            # Write out both the header and the sequence for each read\n",
    "            of.write(f\">{readname}\\n{str(readname2seq[readname])}\\n\")\n",
    "            \n",
    "ALN_UPDATE_FREQ = 1000\n",
    "ALN_BUFFER_FREQ = 1000\n",
    "\n",
    "P = 1\n",
    "            \n",
    "t1 = time.time()\n",
    "for seq in SEQS:\n",
    "    fasta = skbio.DNA.read(f\"../seqs/{seq}.fasta\")\n",
    "    \n",
    "    output_smoothed_reads_file = os.path.join(output_dir, f\"{seq}_smoothed_reads.fasta\")\n",
    "    \n",
    "    # Identify all (0-indexed, so compatible with skbio / pysam!)\n",
    "    # mutated positions in this genome up front to save time.\n",
    "    #\n",
    "    # Equivalently, we could also just take in an arbitrary VCF as input\n",
    "    # (e.g. one produced from another variant calling tool), although we'd\n",
    "    # need to be careful to only include SNVs and not indels/etc...\n",
    "    \n",
    "    print(f\"Identifying mutated positions (p = {P}%) in genome {seq2name[seq]}...\")\n",
    "    mutated_positions = find_mutated_positions(seq, p_to_use=P)\n",
    "    print(f\"Found {len(mutated_positions):,} mutated positions (p = {P}%) in {seq2name[seq]}.\")\n",
    "    print(\n",
    "        f\"Note that this tally is higher than you'd see in e.g. the CP1/2/3 plots, because now we're including \"\n",
    "        \"both 'rare' and non-rare mutations. Just so you don't waste five minutes sanity-checking this like I did.\"\n",
    "    )\n",
    "    \n",
    "    print(\"Going through these positions...\")\n",
    "    \n",
    "    # This should already be implicitly sorted, I think, but the code below relies on mutated_positions being\n",
    "    # in the exact same order as expected. So we may as well be paranoid.\n",
    "    mutated_positions = sorted(mutated_positions)\n",
    "    \n",
    "    # Instead of just writing out every smoothed alignment as soon as we generate it, we build up a \"buffer\"\n",
    "    # of these alignments and then write a bunch out at once. This way we limit slowdown due to constantly\n",
    "    # having to open/close files. I don't really have a good source for this as best practice, but I remembered\n",
    "    # to do it while writing this code, so somewhere in College Park the CS faculty at Maryland are smiling\n",
    "    #\n",
    "    # Also fyi this maps read name to smoothed alignment (well, at this point, just read) sequence. The read name\n",
    "    # is useful to preserve in fasta files so we have some idea of provenance (where smoothed reads came from)\n",
    "    smoothed_aln_buffer = {}\n",
    "    \n",
    "    # The first time we see an alignment of a read, it's 1; if we see a supp aln of this read, it's 2; etc.\n",
    "    # Lets us distinguish alignments with different names\n",
    "    readname2freq_so_far = defaultdict(int)\n",
    "    \n",
    "    # Go through all linear alignments of each read to this genome, focusing (for now) on just the primary\n",
    "    # alignments...\n",
    "    ts1 = time.time()\n",
    "    for ai, aln in enumerate(bf.fetch(seq), 1):\n",
    "        \n",
    "        if ai % ALN_UPDATE_FREQ == 0:\n",
    "            print(\n",
    "                f\"\\tOn aln {ai:,} in seq {seq2name[seq]}. \"\n",
    "                f\"Time spent on {seq2name[seq]} so far: {time.time() - ts1:,.2f} sec.\"\n",
    "            )\n",
    "            \n",
    "        if aln.is_secondary:\n",
    "            raise ValueError(\n",
    "                \"Not to get political or anything, but you should've already filtered secondary alns out\"\n",
    "            )\n",
    "            \n",
    "        # Note that supplementary alignments are ok, though! We implicitly handle these here.\n",
    "        #\n",
    "        # Different alignments of the same read will have different new_readnames, because we're gonna\n",
    "        # be treating them as distinct \"reads\". We should have already filtered reference-overlapping\n",
    "        # supp alns so this shouldn't be a problem\n",
    "        \n",
    "        readname = aln.query_name\n",
    "        readname2freq_so_far[readname] += 1\n",
    "        new_readname = f\"{readname}_{readname2freq_so_far[readname]}\"\n",
    "        \n",
    "        # should never happen\n",
    "        if new_readname in smoothed_aln_buffer:\n",
    "            raise ValueError(\"This exact read alignment has already been smoothed? Weird.\")\n",
    "            \n",
    "        # Figure out where on the MAG this alignment \"hits.\" These are 0-indexed positions from Pysam.\n",
    "        # (reference_end points to the position after the actual final position, since these are designed to\n",
    "        # be interoperable with Python's half-open intervals.)\n",
    "        #\n",
    "        # Of course, there likely will be indels within this range: we're purposefully ignoring those here.\n",
    "        ref_start = aln.reference_start\n",
    "        ref_end = aln.reference_end - 1\n",
    "        \n",
    "        # This should never happen (TM)\n",
    "        if ref_start >= ref_end:\n",
    "            # Du sollst jetzt mit Gott sprechen.\n",
    "            raise ValueError(\n",
    "                f\"Ref start {ref_start:,} >= ref end {ref_end:,} for read {new_readname}?\"\n",
    "            )\n",
    "        \n",
    "        # Smoothed sequence; we'll edit this so that if this read has (mis)matches to any called mutated\n",
    "        # positions, these positions are updated with the read's aligned nucleotides at these positions.\n",
    "        smoothed_aln_seq = fasta[ref_start: ref_end + 1]\n",
    "        \n",
    "        if actually_include_mutations_in_the_smoothed_reads:\n",
    "            # just for debugging: track the exact edits made to smoothed_aln_seq\n",
    "            replacements_made = {}\n",
    "\n",
    "            ap = aln.get_aligned_pairs(matches_only=True)\n",
    "\n",
    "            # Iterating through the aligned pairs is expensive. Since read lengths are generally in the thousands\n",
    "            # to tens of thousands of bp (which is much less than the > 1 million bp length of any bacterial genome),\n",
    "            # we set things up so that we only iterate through the aligned pairs once. We maintain an integer, mpi,\n",
    "            # that is a poor man's \"pointer\" to an index in mutated_positions.\n",
    "\n",
    "            mpi = 0\n",
    "\n",
    "            # Go through this aln's aligned pairs. As we see each pair, compare the pair's reference position\n",
    "            # (refpos) to the mpi-th mutated position (herein referred to as \"mutpos\").\n",
    "            #\n",
    "            # If refpos >  mutpos, increment mpi until refpos <= mutpos (stopping as early as possible).\n",
    "            # If refpos == mutpos, we have a match! Update readname2mutpos2ismutated[mutpos] based on\n",
    "            #                      comparing the read to the reference at the aligned positions.\n",
    "            # If refpos <  mutpos, continue to the next pair.\n",
    "\n",
    "            for pair in ap:\n",
    "\n",
    "                refpos = pair[1]\n",
    "                mutpos = mutated_positions[mpi]\n",
    "\n",
    "                no_mutations_to_right_of_here = False\n",
    "\n",
    "                # Increment mpi until we get to the next mutated position at or after the reference pos for this\n",
    "                # aligned pair (or until we run out of mutated positions).\n",
    "                while refpos > mutpos:\n",
    "                    mpi += 1\n",
    "                    if mpi < len(mutated_positions):\n",
    "                        mutpos = mutated_positions[mpi]\n",
    "                    else:\n",
    "                        no_mutations_to_right_of_here = True\n",
    "                        break\n",
    "\n",
    "                # I expect this should happen only for reads aligned near the right end of the genome.\n",
    "                if no_mutations_to_right_of_here:\n",
    "                    break\n",
    "\n",
    "                # If the next mutation occurs after this aligned pair, continue on to a later pair.\n",
    "                if refpos < mutpos:\n",
    "                    continue\n",
    "\n",
    "                # If we've made it here, refpos == mutpos!\n",
    "                # (...unless I messed something up in how I designed this code.)\n",
    "                if refpos != mutpos:\n",
    "                    raise ValueError(\"This should never happen!\")\n",
    "\n",
    "                # Finally, get the nucleotide aligned to this mutated position from this read.\n",
    "                readpos = pair[0]\n",
    "                read_nt = aln.query_sequence[readpos]\n",
    "\n",
    "                ref_nt = str(fasta[mutpos])\n",
    "                # We don't need to do anything if this read already matches the reference MAG at this position\n",
    "                if read_nt == ref_nt:\n",
    "                    if no_indoor_voice:\n",
    "                        print(f\"Read {new_readname} matches ref at mutpos {mutpos + 1:,}: both {read_nt}\")\n",
    "                else:\n",
    "                    # Record this specific \"allele\" for this read.\n",
    "                    relative_pos_on_aln = mutpos - ref_start\n",
    "                    smoothed_aln_seq = smoothed_aln_seq.replace([relative_pos_on_aln], read_nt)\n",
    "                    replacements_made[relative_pos_on_aln] = read_nt\n",
    "                    if no_indoor_voice:\n",
    "                        print(\n",
    "                            f\"Read {new_readname} mismatches ref at mutpos {mutpos + 1:,}: \"\n",
    "                            f\"ref = {ref_nt}, read = {read_nt}\"\n",
    "                        )\n",
    "\n",
    "            if no_indoor_voice:\n",
    "                print(f\"Read {new_readname} required {len(replacements_made):,} replacements!\")\n",
    "        \n",
    "        # Now that we've finished processing all called mutations that this alignment spans, prepare it\n",
    "        # to be written out to a FASTA file. See comments above on smoothed_aln_buffer, and why we don't\n",
    "        # just write everything out as soon as it's ready.\n",
    "        #\n",
    "        # (Also, we've already guaranteed readname isn't already in smoothed_aln_buffer, so no need to worry\n",
    "        # about accidentally overwriting something from earlier.)\n",
    "        smoothed_aln_buffer[new_readname] = smoothed_aln_seq\n",
    "        \n",
    "        # Notably, we don't necessarily write out *exactly* ALN_BUFFER_FREQ reads at once -- skipping alignments\n",
    "        # due to them being supplementary, etc. (actually no need for an \"etc.\", that's literally the only\n",
    "        # possible reason as of writing, but whatever) doesn't stop ai from going up. Shouldn't make a difference\n",
    "        # unless we have a zillion supplementary alignments.\n",
    "        if ai % ALN_BUFFER_FREQ == 0:\n",
    "            write_out_reads(output_smoothed_reads_file, smoothed_aln_buffer)\n",
    "            # Clear the buffer\n",
    "            smoothed_aln_buffer = {}\n",
    "        \n",
    "    # We're probably going to have left over smoothed reads that we still haven't written out, unless things\n",
    "    # worked out so that on the final alignment we saw ai was exactly divisible by ALN_BUFFER_FREQ (and that's\n",
    "    # pretty unlikely unless you set the buffer freq to a low number). So make one last dump of the buffer.\n",
    "    if len(smoothed_aln_buffer) > 0:\n",
    "        write_out_reads(output_smoothed_reads_file, smoothed_aln_buffer)\n",
    "    \n",
    "    print(f\"Done with {seq}! Took {time.time() - ts1:,.2f} sec.\")\n",
    "        \n",
    "print(f\"Time taken: {time.time() - t1:,} sec.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.5. Stats about smoothed read lengths\n",
    "\n",
    "We could have just figured this out while computing the stuff above, but ... I didn't have the foresight to think of this earlier, and I don't want to rerun that stuff for another >1 hour, so we just loop through the FASTA files we just generated quickly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-02-07T11:08:10.170008Z",
     "iopub.status.busy": "2022-02-07T11:08:10.169168Z",
     "iopub.status.idle": "2022-02-07T11:08:22.517122Z",
     "shell.execute_reply": "2022-02-07T11:08:22.516400Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAMP: 476,356 smoothed reads\n",
      "\tmin / mean / median / max length = 200 / 11,264.7 / 11,154.0 / 35,099\n",
      "\tNum of reads with length ≥ 7,001: 438,334 / 476,356 (92.0%)\n",
      "BACT1: 263,145 smoothed reads\n",
      "\tmin / mean / median / max length = 202 / 11,597.6 / 11,515 / 39,723\n",
      "\tNum of reads with length ≥ 7,001: 243,480 / 263,145 (92.5%)\n",
      "BACT2: 742,926 smoothed reads\n",
      "\tmin / mean / median / max length = 200 / 11,318.0 / 11,259.0 / 40,346\n",
      "\tNum of reads with length ≥ 7,001: 673,220 / 742,926 (90.6%)\n"
     ]
    }
   ],
   "source": [
    "for seq in SEQS:\n",
    "    read_lengths = []\n",
    "    \n",
    "    # Parse a FASTA file -- I stole this code from myself in the Diversity Indices notebook\n",
    "    with open(f\"phasing-data/smoothed-reads-nomutsatall/{seq}_smoothed_reads.fasta\", \"r\") as fastafile:\n",
    "\n",
    "        # Assumes that sequences are not split up over multiple lines (so a FASTA file with N sequences\n",
    "        # should have only 2N lines, maybe 2N + 1 if there's an extra empty newline at the bottom of the file)\n",
    "        for linenum, line in enumerate(fastafile):\n",
    "\n",
    "            if line.startswith(\">\"):\n",
    "                if linenum % 2 != 0:\n",
    "                    raise ValueError(\"something weird with > location in all_edges.fasta. Go yell at Marcus.\")\n",
    "            else:\n",
    "                if linenum % 2 != 1:\n",
    "                    raise ValueError(\"something weird with non > location in all_edges.fasta. Go yell at Marcus.\")\n",
    "\n",
    "                read_lengths.append(len(line.strip()))\n",
    "\n",
    "    num_reads = len(read_lengths)\n",
    "    minlen = min(read_lengths)\n",
    "    maxlen = max(read_lengths)\n",
    "    avglen = mean(read_lengths)\n",
    "    medlen = median(read_lengths)\n",
    "    \n",
    "    # Reads with length less than w + k = (threshold) will be ignored by jumboDB when constructing\n",
    "    # the graph, so we output stats about this to verify that we're not dropping a TON of reads\n",
    "    # (ofc ideally we wouldn't drop any tho...)\n",
    "    threshold = 7001\n",
    "    geq_threshold = len([rl for rl in read_lengths if rl >= threshold])\n",
    "    pct = 100 * (geq_threshold / num_reads)\n",
    "    \n",
    "    print(f\"{seq2name[seq]}: {num_reads:,} smoothed reads\")\n",
    "    print(f\"\\tmin / mean / median / max length = {minlen:,} / {avglen:,.1f} / {medlen:,} / {maxlen:,}\")\n",
    "    print(f\"\\tNum of reads with length \\u2265 {threshold:,}: {geq_threshold:,} / {num_reads:,} ({pct:.1f}%)\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Run LJA on these smoothed reads\n",
    "\n",
    "`--noec` is an undocumented flag that avoids the error correction step in LJA, which is what we want.\n",
    "\n",
    "Possible TODO: adjust this to detect the options above (which mutations are we using? did we actually include mutations in the smoothed reads?) and adjust where files get spit out to accordingly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-02-07T11:08:22.524979Z",
     "iopub.status.busy": "2022-02-07T11:08:22.524164Z",
     "iopub.status.idle": "2022-02-07T11:17:31.130849Z",
     "shell.execute_reply": "2022-02-07T11:17:31.131477Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "00:00:00 370Mb  INFO: Hello! You are running La Jolla Assembler (LJA), a tool for genome assembly from PacBio HiFi reads\r\n",
      "00:00:09 370Mb  INFO: 8730b3f45c8822e043b8710a1008f810faf73f63\r\n",
      "00:00:09 370Mb  INFO: LJA pipeline started\r\n",
      "00:00:09 370Mb  INFO: Performing initial correction with k = 5001\r\n",
      "00:00:09 0Mb  INFO: Reading reads\r\n",
      "00:00:09 0Mb  INFO: Extracting minimizers\r\n",
      "00:00:25 5.4Gb  INFO: Finished read processing\r\n",
      "00:00:25 5.4Gb  INFO: 5258239 hashs collected. Starting sorting.\r\n",
      "00:00:26 5.4Gb  INFO: Finished sorting. Total distinct minimizers: 3538\r\n",
      "00:00:26 5.4Gb  INFO: Starting construction of sparse de Bruijn graph\r\n",
      "00:00:26 5.4Gb  INFO: Vertex map constructed.\r\n",
      "00:00:26 5.4Gb  INFO: Filling edge sequences.\r\n",
      "00:00:43 5.4Gb  INFO: Finished sparse de Bruijn graph construction.\r\n",
      "00:00:43 5.4Gb  INFO:  Collecting tips \r\n",
      "00:00:43 5.4Gb  INFO: Added 4 artificial minimizers from tips.\r\n",
      "00:00:43 5.4Gb  INFO: Collected 7072 old edges.\r\n",
      "00:00:43 5.4Gb  INFO: New minimizers added to sparse graph.\r\n",
      "00:00:43 5.4Gb  INFO: Refilling graph with old edges.\r\n",
      "00:00:43 5.4Gb  INFO: Filling graph with new edges.\r\n",
      "00:00:43 5.4Gb  INFO: Finished fixing sparse de Bruijn graph.\r\n",
      "00:00:43 5.4Gb  INFO: Starting to extract disjointigs.\r\n",
      "00:00:43 5.4Gb  INFO: Finished extracting 2 disjointigs of total size 904383\r\n",
      "00:00:43 0Mb  INFO: Loading disjointigs from file \"phasing-data/smoothed-reads-nomutsatall/edge_6104_lja_noec/k5001/disjointigs.fasta\"\r\n",
      "00:00:43 8Mb  INFO: Filling bloom filter with k+1-mers.\r\n",
      "00:00:44 8Mb  INFO: Filled 4141553 bits out of 28620192\r\n",
      "00:00:44 8Mb  INFO: Finished filling bloom filter. Selecting junctions.\r\n",
      "00:00:44 8Mb  INFO: Collected 315 junctions.\r\n",
      "00:00:44 8Mb  INFO: Starting DBG construction.\r\n",
      "00:00:44 8Mb  INFO: Vertices created.\r\n",
      "00:00:44 8Mb  INFO: Filled dbg edges. Adding hanging vertices \r\n",
      "00:00:44 8Mb  INFO: Added 0 hanging vertices\r\n",
      "00:00:44 8Mb  INFO: Merging unbranching paths\r\n",
      "00:00:44 8Mb  INFO: Ended merging edges. Resulting size 4\r\n",
      "00:00:44 9Mb  INFO: Cleaning edge coverages\r\n",
      "00:00:44 9Mb  INFO: Collecting alignments of sequences to the graph\r\n",
      "00:00:58 3.5Gb  INFO: Alignment collection finished. Total length of alignments is 415385\r\n",
      "00:01:00 3.5Gb  INFO: Printing reads to fasta file \"phasing-data/smoothed-reads-nomutsatall/edge_6104_lja_noec/k5001/corrected_reads.fasta\"\r\n",
      "00:01:31 370Mb  INFO: Performing repeat resolution by transforming de Bruijn graph into Multiplex de Bruijn graph\r\n",
      "00:01:31 0Mb  INFO: Loading graph from fasta\r\n",
      "00:01:31 8Mb  INFO: Finished loading graph\r\n",
      "00:01:32 153Mb  INFO: Looking for unique edges\r\n",
      "00:01:32 153Mb  INFO: Marked 4 long edges as unique\r\n",
      "00:01:32 153Mb  INFO: Marking extra edges as unique based on read paths\r\n",
      "00:01:32 153Mb  INFO: Marked 4 edges as unique\r\n",
      "00:01:32 153Mb  INFO: Splitting graph with unique edges\r\n",
      "00:01:32 153Mb  INFO: Processing 4 components\r\n",
      "00:01:32 153Mb  INFO: Finished unique edges search. Found 4 unique edges\r\n",
      "00:01:32 153Mb  INFO: Analysing repeats of multiplicity 2 and looking for additional unique edges\r\n",
      "00:01:32 153Mb  INFO: Finished processing of repeats of multiplicity 2. Found 0 erroneous edges.\r\n",
      "00:01:32 153Mb  INFO: Resolving repeats\r\n",
      "00:01:32 153Mb  INFO: Constructing paths\r\n",
      "00:01:35 299Mb  INFO: Building graph\r\n",
      "00:01:35 299Mb  INFO: Increasing k\r\n",
      "00:01:35 299Mb  INFO: Finished increasing k\r\n",
      "00:01:35 299Mb  INFO: Exporting remaining active transitions\r\n",
      "00:01:35 299Mb  INFO: Export to Dot\r\n",
      "00:01:35 299Mb  INFO: Export to GFA and compressed contigs\r\n",
      "00:01:35 307Mb  INFO: Finished repeat resolution\r\n",
      "00:01:36 370Mb  INFO: Performing polishing and homopolymer uncompression\r\n",
      "00:01:36 5Mb  INFO: Aligning reads back to assembly\r\n",
      "00:01:51 3.8Gb  INFO: Finished alignment.\r\n",
      "00:01:51 3.8Gb  INFO: Printing alignments to \"phasing-data/smoothed-reads-nomutsatall/edge_6104_lja_noec/uncompressing/alignments.txt\"\r\n",
      "00:01:51 3.8Gb  INFO: Reading and processing initial reads from [\"phasing-data/smoothed-reads-nomutsatall/edge_6104_smoothed_reads.fasta\"]\r\n",
      "00:02:43 5Gb  INFO: Uncompressing homopolymers in contigs\r\n",
      "00:02:44 5Gb  INFO: Total zero covered nucleotides 0\r\n",
      "00:02:44 5Gb  INFO: Calculating overlaps between adjacent uncompressed edges\r\n",
      "00:02:44 5Gb  INFO: Printing final gfa file to \"phasing-data/smoothed-reads-nomutsatall/edge_6104_lja_noec/mdbg.gfa\"\r\n",
      "00:02:44 5Gb  INFO: Printing final assembly to \"phasing-data/smoothed-reads-nomutsatall/edge_6104_lja_noec/assembly.fasta\"\r\n",
      "00:02:44 370Mb  INFO: Final homopolymer compressed and corrected reads can be found here: \"phasing-data/smoothed-reads-nomutsatall/edge_6104_lja_noec/k5001/corrected_reads.fasta\"\r\n",
      "00:02:44 370Mb  INFO: Final graph with homopolymer compressed edges can be found here: \"phasing-data/smoothed-reads-nomutsatall/edge_6104_lja_noec/mdbg/mdbg.hpc.gfa\"\r\n",
      "00:02:44 370Mb  INFO: Final graph can be found here: \"phasing-data/smoothed-reads-nomutsatall/edge_6104_lja_noec/mdbg.gfa\"\r\n",
      "00:02:44 370Mb  INFO: Final assembly can be found here: \"phasing-data/smoothed-reads-nomutsatall/edge_6104_lja_noec/assembly.fasta\"\r\n",
      "00:02:44 370Mb  INFO: LJA pipeline finished\r\n",
      "00:00:00 370Mb  INFO: Hello! You are running La Jolla Assembler (LJA), a tool for genome assembly from PacBio HiFi reads\r\n",
      "00:00:08 370Mb  INFO: 8730b3f45c8822e043b8710a1008f810faf73f63\r\n",
      "00:00:08 370Mb  INFO: LJA pipeline started\r\n",
      "00:00:08 370Mb  INFO: Performing initial correction with k = 5001\r\n",
      "00:00:08 0Mb  INFO: Reading reads\r\n",
      "00:00:08 0Mb  INFO: Extracting minimizers\r\n",
      "00:00:19 2.2Gb  INFO: Finished read processing\r\n",
      "00:00:19 2.2Gb  INFO: 3260868 hashs collected. Starting sorting.\r\n",
      "00:00:19 2.2Gb  INFO: Finished sorting. Total distinct minimizers: 6028\r\n",
      "00:00:19 2.2Gb  INFO: Starting construction of sparse de Bruijn graph\r\n",
      "00:00:19 2.2Gb  INFO: Vertex map constructed.\r\n",
      "00:00:19 2.2Gb  INFO: Filling edge sequences.\r\n",
      "00:00:29 2.3Gb  INFO: Finished sparse de Bruijn graph construction.\r\n",
      "00:00:29 2.3Gb  INFO:  Collecting tips \r\n",
      "00:00:29 2.3Gb  INFO: Added 6 artificial minimizers from tips.\r\n",
      "00:00:29 2.3Gb  INFO: Collected 12050 old edges.\r\n",
      "00:00:29 2.3Gb  INFO: New minimizers added to sparse graph.\r\n",
      "00:00:29 2.3Gb  INFO: Refilling graph with old edges.\r\n",
      "00:00:29 2.3Gb  INFO: Filling graph with new edges.\r\n",
      "00:00:29 2.3Gb  INFO: Finished fixing sparse de Bruijn graph.\r\n",
      "00:00:29 2.3Gb  INFO: Starting to extract disjointigs.\r\n",
      "00:00:30 2.3Gb  INFO: Finished extracting 3 disjointigs of total size 1513686\r\n",
      "00:00:30 0Mb  INFO: Loading disjointigs from file \"phasing-data/smoothed-reads-nomutsatall/edge_1671_lja_noec/k5001/disjointigs.fasta\"\r\n",
      "00:00:30 12Mb  INFO: Filling bloom filter with k+1-mers.\r\n",
      "00:00:30 12Mb  INFO: Filled 6936895 bits out of 47957856\r\n",
      "00:00:30 12Mb  INFO: Finished filling bloom filter. Selecting junctions.\r\n",
      "00:00:31 12Mb  INFO: Collected 609 junctions.\r\n",
      "00:00:31 12Mb  INFO: Starting DBG construction.\r\n",
      "00:00:31 12Mb  INFO: Vertices created.\r\n",
      "00:00:31 12Mb  INFO: Filled dbg edges. Adding hanging vertices \r\n",
      "00:00:31 12Mb  INFO: Added 0 hanging vertices\r\n",
      "00:00:31 12Mb  INFO: Merging unbranching paths\r\n",
      "00:00:31 12Mb  INFO: Ended merging edges. Resulting size 6\r\n",
      "00:00:31 12Mb  INFO: Cleaning edge coverages\r\n",
      "00:00:31 12Mb  INFO: Collecting alignments of sequences to the graph\r\n",
      "00:00:39 3.2Gb  INFO: Alignment collection finished. Total length of alignments is 233286\r\n",
      "00:00:40 3.2Gb  INFO: Printing reads to fasta file \"phasing-data/smoothed-reads-nomutsatall/edge_1671_lja_noec/k5001/corrected_reads.fasta\"\r\n",
      "00:01:00 370Mb  INFO: Performing repeat resolution by transforming de Bruijn graph into Multiplex de Bruijn graph\r\n",
      "00:01:00 0Mb  INFO: Loading graph from fasta\r\n",
      "00:01:00 8Mb  INFO: Finished loading graph\r\n",
      "00:01:01 126Mb  INFO: Looking for unique edges\r\n",
      "00:01:01 126Mb  INFO: Marked 4 long edges as unique\r\n",
      "00:01:01 126Mb  INFO: Marking extra edges as unique based on read paths\r\n",
      "00:01:01 126Mb  INFO: Marked 4 edges as unique\r\n",
      "00:01:01 126Mb  INFO: Splitting graph with unique edges\r\n",
      "00:01:01 126Mb  INFO: Processing 5 components\r\n",
      "00:01:01 126Mb  INFO: Finished unique edges search. Found 4 unique edges\r\n",
      "00:01:01 126Mb  INFO: Analysing repeats of multiplicity 2 and looking for additional unique edges\r\n",
      "00:01:01 126Mb  INFO: Finished processing of repeats of multiplicity 2. Found 0 erroneous edges.\r\n",
      "00:01:01 126Mb  INFO: Resolving repeats\r\n",
      "00:01:01 126Mb  INFO: Constructing paths\r\n",
      "00:01:03 173Mb  INFO: Building graph\r\n",
      "00:01:03 173Mb  INFO: Increasing k\r\n",
      "00:01:03 173Mb  INFO: Finished increasing k\r\n",
      "00:01:03 173Mb  INFO: Exporting remaining active transitions\r\n",
      "00:01:03 173Mb  INFO: Export to Dot\r\n",
      "00:01:03 173Mb  INFO: Export to GFA and compressed contigs\r\n",
      "00:01:03 185Mb  INFO: Finished repeat resolution\r\n",
      "00:01:03 370Mb  INFO: Performing polishing and homopolymer uncompression\r\n",
      "00:01:03 5Mb  INFO: Aligning reads back to assembly\r\n",
      "00:01:12 2.2Gb  INFO: Finished alignment.\r\n",
      "00:01:12 2.2Gb  INFO: Printing alignments to \"phasing-data/smoothed-reads-nomutsatall/edge_1671_lja_noec/uncompressing/alignments.txt\"\r\n",
      "00:01:12 2.2Gb  INFO: Reading and processing initial reads from [\"phasing-data/smoothed-reads-nomutsatall/edge_1671_smoothed_reads.fasta\"]\r\n",
      "00:01:41 3.4Gb  INFO: Uncompressing homopolymers in contigs\r\n",
      "00:01:42 3.4Gb  INFO: Total zero covered nucleotides 0\r\n",
      "00:01:42 3.4Gb  INFO: Calculating overlaps between adjacent uncompressed edges\r\n",
      "00:01:42 3.4Gb  INFO: Printing final gfa file to \"phasing-data/smoothed-reads-nomutsatall/edge_1671_lja_noec/mdbg.gfa\"\r\n",
      "00:01:42 3.4Gb  INFO: Printing final assembly to \"phasing-data/smoothed-reads-nomutsatall/edge_1671_lja_noec/assembly.fasta\"\r\n",
      "00:01:42 370Mb  INFO: Final homopolymer compressed and corrected reads can be found here: \"phasing-data/smoothed-reads-nomutsatall/edge_1671_lja_noec/k5001/corrected_reads.fasta\"\r\n",
      "00:01:42 370Mb  INFO: Final graph with homopolymer compressed edges can be found here: \"phasing-data/smoothed-reads-nomutsatall/edge_1671_lja_noec/mdbg/mdbg.hpc.gfa\"\r\n",
      "00:01:42 370Mb  INFO: Final graph can be found here: \"phasing-data/smoothed-reads-nomutsatall/edge_1671_lja_noec/mdbg.gfa\"\r\n",
      "00:01:42 370Mb  INFO: Final assembly can be found here: \"phasing-data/smoothed-reads-nomutsatall/edge_1671_lja_noec/assembly.fasta\"\r\n",
      "00:01:42 370Mb  INFO: LJA pipeline finished\r\n",
      "00:00:00 370Mb  INFO: Hello! You are running La Jolla Assembler (LJA), a tool for genome assembly from PacBio HiFi reads\r\n",
      "00:00:08 370Mb  INFO: 8730b3f45c8822e043b8710a1008f810faf73f63\r\n",
      "00:00:08 370Mb  INFO: LJA pipeline started\r\n",
      "00:00:08 370Mb  INFO: Performing initial correction with k = 5001\r\n",
      "00:00:08 0Mb  INFO: Reading reads\r\n",
      "00:00:08 0Mb  INFO: Extracting minimizers\r\n",
      "00:00:38 4.4Gb  INFO: Finished read processing\r\n",
      "00:00:38 4.4Gb  INFO: 9396054 hashs collected. Starting sorting.\r\n",
      "00:00:38 4.4Gb  INFO: Finished sorting. Total distinct minimizers: 8150\r\n",
      "00:00:38 4.4Gb  INFO: Starting construction of sparse de Bruijn graph\r\n",
      "00:00:38 4.4Gb  INFO: Vertex map constructed.\r\n",
      "00:00:38 4.4Gb  INFO: Filling edge sequences.\r\n",
      "00:01:08 6.7Gb  INFO: Finished sparse de Bruijn graph construction.\r\n",
      "00:01:08 6.7Gb  INFO:  Collecting tips \r\n",
      "00:01:08 6.7Gb  INFO: Added 3 artificial minimizers from tips.\r\n",
      "00:01:08 6.7Gb  INFO: Collected 16298 old edges.\r\n",
      "00:01:08 6.7Gb  INFO: New minimizers added to sparse graph.\r\n",
      "00:01:08 6.7Gb  INFO: Refilling graph with old edges.\r\n",
      "00:01:08 6.7Gb  INFO: Filling graph with new edges.\r\n",
      "00:01:08 6.7Gb  INFO: Finished fixing sparse de Bruijn graph.\r\n",
      "00:01:08 6.7Gb  INFO: Starting to extract disjointigs.\r\n",
      "00:01:08 6.7Gb  INFO: Finished extracting 3 disjointigs of total size 2058219\r\n",
      "00:01:08 0Mb  INFO: Loading disjointigs from file \"phasing-data/smoothed-reads-nomutsatall/edge_2358_lja_noec/k5001/disjointigs.fasta\"\r\n",
      "00:01:08 16Mb  INFO: Filling bloom filter with k+1-mers.\r\n",
      "00:01:09 16Mb  INFO: Filled 9458305 bits out of 65382912\r\n",
      "00:01:09 16Mb  INFO: Finished filling bloom filter. Selecting junctions.\r\n",
      "00:01:11 16Mb  INFO: Collected 779 junctions.\r\n",
      "00:01:11 16Mb  INFO: Starting DBG construction.\r\n",
      "00:01:11 16Mb  INFO: Vertices created.\r\n",
      "00:01:12 16Mb  INFO: Filled dbg edges. Adding hanging vertices \r\n",
      "00:01:12 16Mb  INFO: Added 0 hanging vertices\r\n",
      "00:01:12 16Mb  INFO: Merging unbranching paths\r\n",
      "00:01:12 16Mb  INFO: Ended merging edges. Resulting size 4\r\n",
      "00:01:12 16Mb  INFO: Cleaning edge coverages\r\n",
      "00:01:12 16Mb  INFO: Collecting alignments of sequences to the graph\r\n",
      "00:01:36 8Gb  INFO: Alignment collection finished. Total length of alignments is 653101\r\n",
      "00:01:38 8Gb  INFO: Printing reads to fasta file \"phasing-data/smoothed-reads-nomutsatall/edge_2358_lja_noec/k5001/corrected_reads.fasta\"\r\n",
      "00:02:32 370Mb  INFO: Performing repeat resolution by transforming de Bruijn graph into Multiplex de Bruijn graph\r\n",
      "00:02:32 0Mb  INFO: Loading graph from fasta\r\n",
      "00:02:32 12Mb  INFO: Finished loading graph\r\n",
      "00:02:34 246Mb  INFO: Looking for unique edges\r\n",
      "00:02:34 246Mb  INFO: Marked 4 long edges as unique\r\n",
      "00:02:34 246Mb  INFO: Marking extra edges as unique based on read paths\r\n",
      "00:02:34 246Mb  INFO: Marked 4 edges as unique\r\n",
      "00:02:34 246Mb  INFO: Splitting graph with unique edges\r\n",
      "00:02:34 246Mb  INFO: Processing 3 components\r\n",
      "00:02:34 246Mb  INFO: Finished unique edges search. Found 4 unique edges\r\n",
      "00:02:34 246Mb  INFO: Analysing repeats of multiplicity 2 and looking for additional unique edges\r\n",
      "00:02:34 246Mb  INFO: Finished processing of repeats of multiplicity 2. Found 0 erroneous edges.\r\n",
      "00:02:34 246Mb  INFO: Resolving repeats\r\n",
      "00:02:34 246Mb  INFO: Constructing paths\r\n",
      "00:02:39 472Mb  INFO: Building graph\r\n",
      "00:02:39 472Mb  INFO: Increasing k\r\n",
      "00:02:39 472Mb  INFO: Finished increasing k\r\n",
      "00:02:39 472Mb  INFO: Exporting remaining active transitions\r\n",
      "00:02:39 472Mb  INFO: Export to Dot\r\n",
      "00:02:39 472Mb  INFO: Export to GFA and compressed contigs\r\n",
      "00:02:40 490Mb  INFO: Finished repeat resolution\r\n",
      "00:02:40 370Mb  INFO: Performing polishing and homopolymer uncompression\r\n",
      "00:02:40 9Mb  INFO: Aligning reads back to assembly\r\n",
      "00:03:05 5.7Gb  INFO: Finished alignment.\r\n",
      "00:03:05 5.7Gb  INFO: Printing alignments to \"phasing-data/smoothed-reads-nomutsatall/edge_2358_lja_noec/uncompressing/alignments.txt\"\r\n",
      "00:03:05 5.7Gb  INFO: Reading and processing initial reads from [\"phasing-data/smoothed-reads-nomutsatall/edge_2358_smoothed_reads.fasta\"]\r\n",
      "00:04:35 5.8Gb  INFO: Uncompressing homopolymers in contigs\r\n",
      "00:04:35 5.8Gb  INFO: Total zero covered nucleotides 0\r\n",
      "00:04:35 5.8Gb  INFO: Calculating overlaps between adjacent uncompressed edges\r\n",
      "00:04:41 5.8Gb  INFO: Printing final gfa file to \"phasing-data/smoothed-reads-nomutsatall/edge_2358_lja_noec/mdbg.gfa\"\r\n",
      "00:04:41 5.8Gb  INFO: Printing final assembly to \"phasing-data/smoothed-reads-nomutsatall/edge_2358_lja_noec/assembly.fasta\"\r\n",
      "00:04:41 370Mb  INFO: Final homopolymer compressed and corrected reads can be found here: \"phasing-data/smoothed-reads-nomutsatall/edge_2358_lja_noec/k5001/corrected_reads.fasta\"\r\n",
      "00:04:41 370Mb  INFO: Final graph with homopolymer compressed edges can be found here: \"phasing-data/smoothed-reads-nomutsatall/edge_2358_lja_noec/mdbg/mdbg.hpc.gfa\"\r\n",
      "00:04:41 370Mb  INFO: Final graph can be found here: \"phasing-data/smoothed-reads-nomutsatall/edge_2358_lja_noec/mdbg.gfa\"\r\n",
      "00:04:41 370Mb  INFO: Final assembly can be found here: \"phasing-data/smoothed-reads-nomutsatall/edge_2358_lja_noec/assembly.fasta\"\r\n",
      "00:04:41 370Mb  INFO: LJA pipeline finished\r\n"
     ]
    }
   ],
   "source": [
    "!/home/mfedarko/software/LJA/bin/lja \\\n",
    "    --reads phasing-data/smoothed-reads/edge_6104_smoothed_reads.fasta \\\n",
    "    --noec \\\n",
    "    --output-dir phasing-data/smoothed-reads/edge_6104_lja_noec\n",
    "\n",
    "!/home/mfedarko/software/LJA/bin/lja \\\n",
    "    --reads phasing-data/smoothed-reads/edge_1671_smoothed_reads.fasta \\\n",
    "    --noec \\\n",
    "    --output-dir phasing-data/smoothed-reads/edge_1671_lja_noec\n",
    "\n",
    "!/home/mfedarko/software/LJA/bin/lja \\\n",
    "    --reads phasing-data/smoothed-reads/edge_2358_smoothed_reads.fasta \\\n",
    "    --noec \\\n",
    "    --output-dir phasing-data/smoothed-reads/edge_2358_lja_noec"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
