{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Perform read smoothing then assemble with LJA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-01-27T07:18:00.331386Z",
     "iopub.status.busy": "2022-01-27T07:18:00.330238Z",
     "iopub.status.idle": "2022-01-27T07:18:00.632113Z",
     "shell.execute_reply": "2022-01-27T07:18:00.631239Z"
    }
   },
   "outputs": [],
   "source": [
    "%run \"Header.ipynb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-01-27T07:18:00.636178Z",
     "iopub.status.busy": "2022-01-27T07:18:00.635383Z",
     "iopub.status.idle": "2022-01-27T07:18:01.143022Z",
     "shell.execute_reply": "2022-01-27T07:18:01.143672Z"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import time\n",
    "import pickle\n",
    "import pysam\n",
    "import skbio\n",
    "from collections import defaultdict\n",
    "from linked_mutations_utils import find_mutated_positions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Smooth reads\n",
    "\n",
    "Lots of this code is duplicated from the `Phasing-01-MakeGraph.ipynb` notebook in this folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-01-27T07:18:01.166636Z",
     "iopub.status.busy": "2022-01-27T07:18:01.147444Z",
     "iopub.status.idle": "2022-01-27T08:55:41.761668Z",
     "shell.execute_reply": "2022-01-27T08:55:41.760834Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Identifying mutated positions in genome CAMP...\n",
      "Found 284 mutated positions in CAMP.\n",
      "Going through these positions...\n",
      "\tOn aln 1,000 in seq CAMP. Time spent on CAMP so far: 1.73 sec.\n",
      "\tOn aln 2,000 in seq CAMP. Time spent on CAMP so far: 3.31 sec.\n",
      "\tOn aln 3,000 in seq CAMP. Time spent on CAMP so far: 5.09 sec.\n",
      "\tOn aln 4,000 in seq CAMP. Time spent on CAMP so far: 6.67 sec.\n",
      "\tOn aln 5,000 in seq CAMP. Time spent on CAMP so far: 8.66 sec.\n",
      "\tOn aln 6,000 in seq CAMP. Time spent on CAMP so far: 11.59 sec.\n",
      "\tOn aln 7,000 in seq CAMP. Time spent on CAMP so far: 14.43 sec.\n",
      "\tOn aln 8,000 in seq CAMP. Time spent on CAMP so far: 17.31 sec.\n",
      "\tOn aln 9,000 in seq CAMP. Time spent on CAMP so far: 20.22 sec.\n",
      "\tOn aln 10,000 in seq CAMP. Time spent on CAMP so far: 23.20 sec.\n",
      "\tOn aln 11,000 in seq CAMP. Time spent on CAMP so far: 26.11 sec.\n",
      "\tOn aln 12,000 in seq CAMP. Time spent on CAMP so far: 29.03 sec.\n",
      "\tOn aln 13,000 in seq CAMP. Time spent on CAMP so far: 31.90 sec.\n",
      "\tOn aln 14,000 in seq CAMP. Time spent on CAMP so far: 34.78 sec.\n",
      "\tOn aln 15,000 in seq CAMP. Time spent on CAMP so far: 37.70 sec.\n",
      "\tOn aln 16,000 in seq CAMP. Time spent on CAMP so far: 40.59 sec.\n",
      "\tOn aln 17,000 in seq CAMP. Time spent on CAMP so far: 43.49 sec.\n",
      "\tOn aln 18,000 in seq CAMP. Time spent on CAMP so far: 46.38 sec.\n",
      "\tOn aln 19,000 in seq CAMP. Time spent on CAMP so far: 49.33 sec.\n",
      "\tOn aln 20,000 in seq CAMP. Time spent on CAMP so far: 52.28 sec.\n",
      "\tOn aln 21,000 in seq CAMP. Time spent on CAMP so far: 55.19 sec.\n",
      "\tOn aln 22,000 in seq CAMP. Time spent on CAMP so far: 58.17 sec.\n",
      "\tOn aln 23,000 in seq CAMP. Time spent on CAMP so far: 61.04 sec.\n",
      "\tOn aln 24,000 in seq CAMP. Time spent on CAMP so far: 63.90 sec.\n",
      "\tOn aln 25,000 in seq CAMP. Time spent on CAMP so far: 66.77 sec.\n",
      "\tOn aln 26,000 in seq CAMP. Time spent on CAMP so far: 69.66 sec.\n",
      "\tOn aln 27,000 in seq CAMP. Time spent on CAMP so far: 72.51 sec.\n",
      "\tOn aln 28,000 in seq CAMP. Time spent on CAMP so far: 75.36 sec.\n",
      "\tOn aln 29,000 in seq CAMP. Time spent on CAMP so far: 78.27 sec.\n",
      "\tOn aln 30,000 in seq CAMP. Time spent on CAMP so far: 81.16 sec.\n",
      "\tOn aln 31,000 in seq CAMP. Time spent on CAMP so far: 84.03 sec.\n",
      "\tOn aln 32,000 in seq CAMP. Time spent on CAMP so far: 86.95 sec.\n",
      "\tOn aln 33,000 in seq CAMP. Time spent on CAMP so far: 89.78 sec.\n",
      "\tOn aln 34,000 in seq CAMP. Time spent on CAMP so far: 92.61 sec.\n",
      "\tOn aln 35,000 in seq CAMP. Time spent on CAMP so far: 95.52 sec.\n",
      "\tOn aln 36,000 in seq CAMP. Time spent on CAMP so far: 98.45 sec.\n",
      "\tOn aln 37,000 in seq CAMP. Time spent on CAMP so far: 101.40 sec.\n",
      "\tOn aln 38,000 in seq CAMP. Time spent on CAMP so far: 104.33 sec.\n",
      "\tOn aln 39,000 in seq CAMP. Time spent on CAMP so far: 107.28 sec.\n",
      "\tOn aln 40,000 in seq CAMP. Time spent on CAMP so far: 110.19 sec.\n",
      "\tOn aln 41,000 in seq CAMP. Time spent on CAMP so far: 113.09 sec.\n",
      "\tOn aln 42,000 in seq CAMP. Time spent on CAMP so far: 116.00 sec.\n",
      "\tOn aln 43,000 in seq CAMP. Time spent on CAMP so far: 118.92 sec.\n",
      "\tOn aln 44,000 in seq CAMP. Time spent on CAMP so far: 121.84 sec.\n",
      "\tOn aln 45,000 in seq CAMP. Time spent on CAMP so far: 124.70 sec.\n",
      "\tOn aln 46,000 in seq CAMP. Time spent on CAMP so far: 127.28 sec.\n",
      "\tOn aln 47,000 in seq CAMP. Time spent on CAMP so far: 129.62 sec.\n",
      "\tOn aln 48,000 in seq CAMP. Time spent on CAMP so far: 132.53 sec.\n",
      "\tOn aln 49,000 in seq CAMP. Time spent on CAMP so far: 135.48 sec.\n",
      "\tOn aln 50,000 in seq CAMP. Time spent on CAMP so far: 138.35 sec.\n",
      "\tOn aln 51,000 in seq CAMP. Time spent on CAMP so far: 141.22 sec.\n",
      "\tOn aln 52,000 in seq CAMP. Time spent on CAMP so far: 144.16 sec.\n",
      "\tOn aln 53,000 in seq CAMP. Time spent on CAMP so far: 147.15 sec.\n",
      "\tOn aln 54,000 in seq CAMP. Time spent on CAMP so far: 150.06 sec.\n",
      "\tOn aln 55,000 in seq CAMP. Time spent on CAMP so far: 152.96 sec.\n",
      "\tOn aln 56,000 in seq CAMP. Time spent on CAMP so far: 155.87 sec.\n",
      "\tOn aln 57,000 in seq CAMP. Time spent on CAMP so far: 158.75 sec.\n",
      "\tOn aln 58,000 in seq CAMP. Time spent on CAMP so far: 161.59 sec.\n",
      "\tOn aln 59,000 in seq CAMP. Time spent on CAMP so far: 164.46 sec.\n",
      "\tOn aln 60,000 in seq CAMP. Time spent on CAMP so far: 167.32 sec.\n",
      "\tOn aln 61,000 in seq CAMP. Time spent on CAMP so far: 170.21 sec.\n",
      "\tOn aln 62,000 in seq CAMP. Time spent on CAMP so far: 173.15 sec.\n",
      "\tOn aln 63,000 in seq CAMP. Time spent on CAMP so far: 176.06 sec.\n",
      "\tOn aln 64,000 in seq CAMP. Time spent on CAMP so far: 179.09 sec.\n",
      "\tOn aln 65,000 in seq CAMP. Time spent on CAMP so far: 182.18 sec.\n",
      "\tOn aln 66,000 in seq CAMP. Time spent on CAMP so far: 185.26 sec.\n",
      "\tOn aln 67,000 in seq CAMP. Time spent on CAMP so far: 188.33 sec.\n",
      "\tOn aln 68,000 in seq CAMP. Time spent on CAMP so far: 191.34 sec.\n",
      "\tOn aln 69,000 in seq CAMP. Time spent on CAMP so far: 194.31 sec.\n",
      "\tOn aln 70,000 in seq CAMP. Time spent on CAMP so far: 197.25 sec.\n",
      "\tOn aln 71,000 in seq CAMP. Time spent on CAMP so far: 200.18 sec.\n",
      "\tOn aln 72,000 in seq CAMP. Time spent on CAMP so far: 203.11 sec.\n",
      "\tOn aln 73,000 in seq CAMP. Time spent on CAMP so far: 206.09 sec.\n",
      "\tOn aln 74,000 in seq CAMP. Time spent on CAMP so far: 209.05 sec.\n",
      "\tOn aln 75,000 in seq CAMP. Time spent on CAMP so far: 211.99 sec.\n",
      "\tOn aln 76,000 in seq CAMP. Time spent on CAMP so far: 214.97 sec.\n",
      "\tOn aln 77,000 in seq CAMP. Time spent on CAMP so far: 217.94 sec.\n",
      "\tOn aln 78,000 in seq CAMP. Time spent on CAMP so far: 220.88 sec.\n",
      "\tOn aln 79,000 in seq CAMP. Time spent on CAMP so far: 223.82 sec.\n",
      "\tOn aln 80,000 in seq CAMP. Time spent on CAMP so far: 226.74 sec.\n",
      "\tOn aln 81,000 in seq CAMP. Time spent on CAMP so far: 229.65 sec.\n",
      "\tOn aln 82,000 in seq CAMP. Time spent on CAMP so far: 232.52 sec.\n",
      "\tOn aln 83,000 in seq CAMP. Time spent on CAMP so far: 235.43 sec.\n",
      "\tOn aln 84,000 in seq CAMP. Time spent on CAMP so far: 238.39 sec.\n",
      "\tOn aln 85,000 in seq CAMP. Time spent on CAMP so far: 241.36 sec.\n",
      "\tOn aln 86,000 in seq CAMP. Time spent on CAMP so far: 244.40 sec.\n",
      "\tOn aln 87,000 in seq CAMP. Time spent on CAMP so far: 247.48 sec.\n",
      "\tOn aln 88,000 in seq CAMP. Time spent on CAMP so far: 250.49 sec.\n",
      "\tOn aln 89,000 in seq CAMP. Time spent on CAMP so far: 253.50 sec.\n",
      "\tOn aln 90,000 in seq CAMP. Time spent on CAMP so far: 256.49 sec.\n",
      "\tOn aln 91,000 in seq CAMP. Time spent on CAMP so far: 259.41 sec.\n",
      "\tOn aln 92,000 in seq CAMP. Time spent on CAMP so far: 262.22 sec.\n",
      "\tOn aln 93,000 in seq CAMP. Time spent on CAMP so far: 265.13 sec.\n",
      "\tOn aln 94,000 in seq CAMP. Time spent on CAMP so far: 268.03 sec.\n",
      "\tOn aln 95,000 in seq CAMP. Time spent on CAMP so far: 270.98 sec.\n",
      "\tOn aln 96,000 in seq CAMP. Time spent on CAMP so far: 273.99 sec.\n",
      "\tOn aln 97,000 in seq CAMP. Time spent on CAMP so far: 276.98 sec.\n",
      "\tOn aln 98,000 in seq CAMP. Time spent on CAMP so far: 279.95 sec.\n",
      "\tOn aln 99,000 in seq CAMP. Time spent on CAMP so far: 282.97 sec.\n",
      "\tOn aln 100,000 in seq CAMP. Time spent on CAMP so far: 285.95 sec.\n",
      "\tOn aln 101,000 in seq CAMP. Time spent on CAMP so far: 288.95 sec.\n",
      "\tOn aln 102,000 in seq CAMP. Time spent on CAMP so far: 291.91 sec.\n",
      "\tOn aln 103,000 in seq CAMP. Time spent on CAMP so far: 294.91 sec.\n",
      "\tOn aln 104,000 in seq CAMP. Time spent on CAMP so far: 297.81 sec.\n",
      "\tOn aln 105,000 in seq CAMP. Time spent on CAMP so far: 300.74 sec.\n",
      "\tOn aln 106,000 in seq CAMP. Time spent on CAMP so far: 303.69 sec.\n",
      "\tOn aln 107,000 in seq CAMP. Time spent on CAMP so far: 306.61 sec.\n",
      "\tOn aln 108,000 in seq CAMP. Time spent on CAMP so far: 309.56 sec.\n",
      "\tOn aln 109,000 in seq CAMP. Time spent on CAMP so far: 312.52 sec.\n",
      "\tOn aln 110,000 in seq CAMP. Time spent on CAMP so far: 315.53 sec.\n",
      "\tOn aln 111,000 in seq CAMP. Time spent on CAMP so far: 318.57 sec.\n",
      "\tOn aln 112,000 in seq CAMP. Time spent on CAMP so far: 321.62 sec.\n",
      "\tOn aln 113,000 in seq CAMP. Time spent on CAMP so far: 324.65 sec.\n",
      "\tOn aln 114,000 in seq CAMP. Time spent on CAMP so far: 327.71 sec.\n",
      "\tOn aln 115,000 in seq CAMP. Time spent on CAMP so far: 330.74 sec.\n",
      "\tOn aln 116,000 in seq CAMP. Time spent on CAMP so far: 333.78 sec.\n",
      "\tOn aln 117,000 in seq CAMP. Time spent on CAMP so far: 336.90 sec.\n",
      "\tOn aln 118,000 in seq CAMP. Time spent on CAMP so far: 339.89 sec.\n",
      "\tOn aln 119,000 in seq CAMP. Time spent on CAMP so far: 342.87 sec.\n",
      "\tOn aln 120,000 in seq CAMP. Time spent on CAMP so far: 345.82 sec.\n",
      "\tOn aln 121,000 in seq CAMP. Time spent on CAMP so far: 348.83 sec.\n",
      "\tOn aln 122,000 in seq CAMP. Time spent on CAMP so far: 351.85 sec.\n",
      "\tOn aln 123,000 in seq CAMP. Time spent on CAMP so far: 354.87 sec.\n",
      "\tOn aln 124,000 in seq CAMP. Time spent on CAMP so far: 357.94 sec.\n",
      "\tOn aln 125,000 in seq CAMP. Time spent on CAMP so far: 360.95 sec.\n",
      "\tOn aln 126,000 in seq CAMP. Time spent on CAMP so far: 363.95 sec.\n",
      "\tOn aln 127,000 in seq CAMP. Time spent on CAMP so far: 366.91 sec.\n",
      "\tOn aln 128,000 in seq CAMP. Time spent on CAMP so far: 369.85 sec.\n",
      "\tOn aln 129,000 in seq CAMP. Time spent on CAMP so far: 372.82 sec.\n",
      "\tOn aln 130,000 in seq CAMP. Time spent on CAMP so far: 375.81 sec.\n",
      "\tOn aln 131,000 in seq CAMP. Time spent on CAMP so far: 378.82 sec.\n",
      "\tOn aln 132,000 in seq CAMP. Time spent on CAMP so far: 381.90 sec.\n",
      "\tOn aln 133,000 in seq CAMP. Time spent on CAMP so far: 384.94 sec.\n",
      "\tOn aln 134,000 in seq CAMP. Time spent on CAMP so far: 387.93 sec.\n",
      "\tOn aln 135,000 in seq CAMP. Time spent on CAMP so far: 390.96 sec.\n",
      "\tOn aln 136,000 in seq CAMP. Time spent on CAMP so far: 393.96 sec.\n",
      "\tOn aln 137,000 in seq CAMP. Time spent on CAMP so far: 396.98 sec.\n",
      "\tOn aln 138,000 in seq CAMP. Time spent on CAMP so far: 400.02 sec.\n",
      "\tOn aln 139,000 in seq CAMP. Time spent on CAMP so far: 403.09 sec.\n",
      "\tOn aln 140,000 in seq CAMP. Time spent on CAMP so far: 406.10 sec.\n",
      "\tOn aln 141,000 in seq CAMP. Time spent on CAMP so far: 409.17 sec.\n",
      "\tOn aln 142,000 in seq CAMP. Time spent on CAMP so far: 412.25 sec.\n",
      "\tOn aln 143,000 in seq CAMP. Time spent on CAMP so far: 415.34 sec.\n",
      "\tOn aln 144,000 in seq CAMP. Time spent on CAMP so far: 418.43 sec.\n",
      "\tOn aln 145,000 in seq CAMP. Time spent on CAMP so far: 421.45 sec.\n",
      "\tOn aln 146,000 in seq CAMP. Time spent on CAMP so far: 424.45 sec.\n",
      "\tOn aln 147,000 in seq CAMP. Time spent on CAMP so far: 427.47 sec.\n",
      "\tOn aln 148,000 in seq CAMP. Time spent on CAMP so far: 430.49 sec.\n",
      "\tOn aln 149,000 in seq CAMP. Time spent on CAMP so far: 433.45 sec.\n",
      "\tOn aln 150,000 in seq CAMP. Time spent on CAMP so far: 436.42 sec.\n",
      "\tOn aln 151,000 in seq CAMP. Time spent on CAMP so far: 439.42 sec.\n",
      "\tOn aln 152,000 in seq CAMP. Time spent on CAMP so far: 442.42 sec.\n",
      "\tOn aln 153,000 in seq CAMP. Time spent on CAMP so far: 445.42 sec.\n",
      "\tOn aln 154,000 in seq CAMP. Time spent on CAMP so far: 448.39 sec.\n",
      "\tOn aln 155,000 in seq CAMP. Time spent on CAMP so far: 451.41 sec.\n",
      "\tOn aln 156,000 in seq CAMP. Time spent on CAMP so far: 454.41 sec.\n",
      "\tOn aln 157,000 in seq CAMP. Time spent on CAMP so far: 457.52 sec.\n",
      "\tOn aln 158,000 in seq CAMP. Time spent on CAMP so far: 460.53 sec.\n",
      "\tOn aln 159,000 in seq CAMP. Time spent on CAMP so far: 463.62 sec.\n",
      "\tOn aln 160,000 in seq CAMP. Time spent on CAMP so far: 466.70 sec.\n",
      "\tOn aln 161,000 in seq CAMP. Time spent on CAMP so far: 469.75 sec.\n",
      "\tOn aln 162,000 in seq CAMP. Time spent on CAMP so far: 472.87 sec.\n",
      "\tOn aln 163,000 in seq CAMP. Time spent on CAMP so far: 475.79 sec.\n",
      "\tOn aln 164,000 in seq CAMP. Time spent on CAMP so far: 478.82 sec.\n",
      "\tOn aln 165,000 in seq CAMP. Time spent on CAMP so far: 481.88 sec.\n",
      "\tOn aln 166,000 in seq CAMP. Time spent on CAMP so far: 484.88 sec.\n",
      "\tOn aln 167,000 in seq CAMP. Time spent on CAMP so far: 487.82 sec.\n",
      "\tOn aln 168,000 in seq CAMP. Time spent on CAMP so far: 490.76 sec.\n",
      "\tOn aln 169,000 in seq CAMP. Time spent on CAMP so far: 493.70 sec.\n",
      "\tOn aln 170,000 in seq CAMP. Time spent on CAMP so far: 496.70 sec.\n",
      "\tOn aln 171,000 in seq CAMP. Time spent on CAMP so far: 499.72 sec.\n",
      "\tOn aln 172,000 in seq CAMP. Time spent on CAMP so far: 502.77 sec.\n",
      "\tOn aln 173,000 in seq CAMP. Time spent on CAMP so far: 505.83 sec.\n",
      "\tOn aln 174,000 in seq CAMP. Time spent on CAMP so far: 508.90 sec.\n",
      "\tOn aln 175,000 in seq CAMP. Time spent on CAMP so far: 511.88 sec.\n",
      "\tOn aln 176,000 in seq CAMP. Time spent on CAMP so far: 514.95 sec.\n",
      "\tOn aln 177,000 in seq CAMP. Time spent on CAMP so far: 517.95 sec.\n",
      "\tOn aln 178,000 in seq CAMP. Time spent on CAMP so far: 521.03 sec.\n",
      "\tOn aln 179,000 in seq CAMP. Time spent on CAMP so far: 523.99 sec.\n",
      "\tOn aln 180,000 in seq CAMP. Time spent on CAMP so far: 526.95 sec.\n",
      "\tOn aln 181,000 in seq CAMP. Time spent on CAMP so far: 530.02 sec.\n",
      "\tOn aln 182,000 in seq CAMP. Time spent on CAMP so far: 533.08 sec.\n",
      "\tOn aln 183,000 in seq CAMP. Time spent on CAMP so far: 536.23 sec.\n",
      "\tOn aln 184,000 in seq CAMP. Time spent on CAMP so far: 539.25 sec.\n",
      "\tOn aln 185,000 in seq CAMP. Time spent on CAMP so far: 542.34 sec.\n",
      "\tOn aln 186,000 in seq CAMP. Time spent on CAMP so far: 545.34 sec.\n",
      "\tOn aln 187,000 in seq CAMP. Time spent on CAMP so far: 548.32 sec.\n",
      "\tOn aln 188,000 in seq CAMP. Time spent on CAMP so far: 551.27 sec.\n",
      "\tOn aln 189,000 in seq CAMP. Time spent on CAMP so far: 554.29 sec.\n",
      "\tOn aln 190,000 in seq CAMP. Time spent on CAMP so far: 557.33 sec.\n",
      "\tOn aln 191,000 in seq CAMP. Time spent on CAMP so far: 560.36 sec.\n",
      "\tOn aln 192,000 in seq CAMP. Time spent on CAMP so far: 563.42 sec.\n",
      "\tOn aln 193,000 in seq CAMP. Time spent on CAMP so far: 566.55 sec.\n",
      "\tOn aln 194,000 in seq CAMP. Time spent on CAMP so far: 569.62 sec.\n",
      "\tOn aln 195,000 in seq CAMP. Time spent on CAMP so far: 572.68 sec.\n",
      "\tOn aln 196,000 in seq CAMP. Time spent on CAMP so far: 575.71 sec.\n",
      "\tOn aln 197,000 in seq CAMP. Time spent on CAMP so far: 578.81 sec.\n",
      "\tOn aln 198,000 in seq CAMP. Time spent on CAMP so far: 581.81 sec.\n",
      "\tOn aln 199,000 in seq CAMP. Time spent on CAMP so far: 584.80 sec.\n",
      "\tOn aln 200,000 in seq CAMP. Time spent on CAMP so far: 587.77 sec.\n",
      "\tOn aln 201,000 in seq CAMP. Time spent on CAMP so far: 590.78 sec.\n",
      "\tOn aln 202,000 in seq CAMP. Time spent on CAMP so far: 593.78 sec.\n",
      "\tOn aln 203,000 in seq CAMP. Time spent on CAMP so far: 596.70 sec.\n",
      "\tOn aln 204,000 in seq CAMP. Time spent on CAMP so far: 599.65 sec.\n",
      "\tOn aln 205,000 in seq CAMP. Time spent on CAMP so far: 602.59 sec.\n",
      "\tOn aln 206,000 in seq CAMP. Time spent on CAMP so far: 605.56 sec.\n",
      "\tOn aln 207,000 in seq CAMP. Time spent on CAMP so far: 608.54 sec.\n",
      "\tOn aln 208,000 in seq CAMP. Time spent on CAMP so far: 611.53 sec.\n",
      "\tOn aln 209,000 in seq CAMP. Time spent on CAMP so far: 614.67 sec.\n",
      "\tOn aln 210,000 in seq CAMP. Time spent on CAMP so far: 617.70 sec.\n",
      "\tOn aln 211,000 in seq CAMP. Time spent on CAMP so far: 620.79 sec.\n",
      "\tOn aln 212,000 in seq CAMP. Time spent on CAMP so far: 623.91 sec.\n",
      "\tOn aln 213,000 in seq CAMP. Time spent on CAMP so far: 627.03 sec.\n",
      "\tOn aln 214,000 in seq CAMP. Time spent on CAMP so far: 630.16 sec.\n",
      "\tOn aln 215,000 in seq CAMP. Time spent on CAMP so far: 633.17 sec.\n",
      "\tOn aln 216,000 in seq CAMP. Time spent on CAMP so far: 636.18 sec.\n",
      "\tOn aln 217,000 in seq CAMP. Time spent on CAMP so far: 639.19 sec.\n",
      "\tOn aln 218,000 in seq CAMP. Time spent on CAMP so far: 642.25 sec.\n",
      "\tOn aln 219,000 in seq CAMP. Time spent on CAMP so far: 645.32 sec.\n",
      "\tOn aln 220,000 in seq CAMP. Time spent on CAMP so far: 648.46 sec.\n",
      "\tOn aln 221,000 in seq CAMP. Time spent on CAMP so far: 651.59 sec.\n",
      "\tOn aln 222,000 in seq CAMP. Time spent on CAMP so far: 654.66 sec.\n",
      "\tOn aln 223,000 in seq CAMP. Time spent on CAMP so far: 657.75 sec.\n",
      "\tOn aln 224,000 in seq CAMP. Time spent on CAMP so far: 660.77 sec.\n",
      "\tOn aln 225,000 in seq CAMP. Time spent on CAMP so far: 663.76 sec.\n",
      "\tOn aln 226,000 in seq CAMP. Time spent on CAMP so far: 666.69 sec.\n",
      "\tOn aln 227,000 in seq CAMP. Time spent on CAMP so far: 669.73 sec.\n",
      "\tOn aln 228,000 in seq CAMP. Time spent on CAMP so far: 672.77 sec.\n",
      "\tOn aln 229,000 in seq CAMP. Time spent on CAMP so far: 675.81 sec.\n",
      "\tOn aln 230,000 in seq CAMP. Time spent on CAMP so far: 678.86 sec.\n",
      "\tOn aln 231,000 in seq CAMP. Time spent on CAMP so far: 681.87 sec.\n",
      "\tOn aln 232,000 in seq CAMP. Time spent on CAMP so far: 684.90 sec.\n",
      "\tOn aln 233,000 in seq CAMP. Time spent on CAMP so far: 688.01 sec.\n",
      "\tOn aln 234,000 in seq CAMP. Time spent on CAMP so far: 691.10 sec.\n",
      "\tOn aln 235,000 in seq CAMP. Time spent on CAMP so far: 694.23 sec.\n",
      "\tOn aln 236,000 in seq CAMP. Time spent on CAMP so far: 697.34 sec.\n",
      "\tOn aln 237,000 in seq CAMP. Time spent on CAMP so far: 700.43 sec.\n",
      "\tOn aln 238,000 in seq CAMP. Time spent on CAMP so far: 703.41 sec.\n",
      "\tOn aln 239,000 in seq CAMP. Time spent on CAMP so far: 706.37 sec.\n",
      "\tOn aln 240,000 in seq CAMP. Time spent on CAMP so far: 709.42 sec.\n",
      "\tOn aln 241,000 in seq CAMP. Time spent on CAMP so far: 712.52 sec.\n",
      "\tOn aln 242,000 in seq CAMP. Time spent on CAMP so far: 715.63 sec.\n",
      "\tOn aln 243,000 in seq CAMP. Time spent on CAMP so far: 718.67 sec.\n",
      "\tOn aln 244,000 in seq CAMP. Time spent on CAMP so far: 721.76 sec.\n",
      "\tOn aln 245,000 in seq CAMP. Time spent on CAMP so far: 724.82 sec.\n",
      "\tOn aln 246,000 in seq CAMP. Time spent on CAMP so far: 727.95 sec.\n",
      "\tOn aln 247,000 in seq CAMP. Time spent on CAMP so far: 731.02 sec.\n",
      "\tOn aln 248,000 in seq CAMP. Time spent on CAMP so far: 734.08 sec.\n",
      "\tOn aln 249,000 in seq CAMP. Time spent on CAMP so far: 737.12 sec.\n",
      "\tOn aln 250,000 in seq CAMP. Time spent on CAMP so far: 740.15 sec.\n",
      "\tOn aln 251,000 in seq CAMP. Time spent on CAMP so far: 743.15 sec.\n",
      "\tOn aln 252,000 in seq CAMP. Time spent on CAMP so far: 746.16 sec.\n",
      "\tOn aln 253,000 in seq CAMP. Time spent on CAMP so far: 749.17 sec.\n",
      "\tOn aln 254,000 in seq CAMP. Time spent on CAMP so far: 752.23 sec.\n",
      "\tOn aln 255,000 in seq CAMP. Time spent on CAMP so far: 755.33 sec.\n",
      "\tOn aln 256,000 in seq CAMP. Time spent on CAMP so far: 758.43 sec.\n",
      "\tOn aln 257,000 in seq CAMP. Time spent on CAMP so far: 761.43 sec.\n",
      "\tOn aln 258,000 in seq CAMP. Time spent on CAMP so far: 764.38 sec.\n",
      "\tOn aln 259,000 in seq CAMP. Time spent on CAMP so far: 767.34 sec.\n",
      "\tOn aln 260,000 in seq CAMP. Time spent on CAMP so far: 770.32 sec.\n",
      "\tOn aln 261,000 in seq CAMP. Time spent on CAMP so far: 773.33 sec.\n",
      "\tOn aln 262,000 in seq CAMP. Time spent on CAMP so far: 776.40 sec.\n",
      "\tOn aln 263,000 in seq CAMP. Time spent on CAMP so far: 779.49 sec.\n",
      "\tOn aln 264,000 in seq CAMP. Time spent on CAMP so far: 782.61 sec.\n",
      "\tOn aln 265,000 in seq CAMP. Time spent on CAMP so far: 785.69 sec.\n",
      "\tOn aln 266,000 in seq CAMP. Time spent on CAMP so far: 788.68 sec.\n",
      "\tOn aln 267,000 in seq CAMP. Time spent on CAMP so far: 791.76 sec.\n",
      "\tOn aln 268,000 in seq CAMP. Time spent on CAMP so far: 794.81 sec.\n",
      "\tOn aln 269,000 in seq CAMP. Time spent on CAMP so far: 797.92 sec.\n",
      "\tOn aln 270,000 in seq CAMP. Time spent on CAMP so far: 800.98 sec.\n",
      "\tOn aln 271,000 in seq CAMP. Time spent on CAMP so far: 804.00 sec.\n",
      "\tOn aln 272,000 in seq CAMP. Time spent on CAMP so far: 806.96 sec.\n",
      "\tOn aln 273,000 in seq CAMP. Time spent on CAMP so far: 810.03 sec.\n",
      "\tOn aln 274,000 in seq CAMP. Time spent on CAMP so far: 813.12 sec.\n",
      "\tOn aln 275,000 in seq CAMP. Time spent on CAMP so far: 816.26 sec.\n",
      "\tOn aln 276,000 in seq CAMP. Time spent on CAMP so far: 819.46 sec.\n",
      "\tOn aln 277,000 in seq CAMP. Time spent on CAMP so far: 822.60 sec.\n",
      "\tOn aln 278,000 in seq CAMP. Time spent on CAMP so far: 825.67 sec.\n",
      "\tOn aln 279,000 in seq CAMP. Time spent on CAMP so far: 828.68 sec.\n",
      "\tOn aln 280,000 in seq CAMP. Time spent on CAMP so far: 831.76 sec.\n",
      "\tOn aln 281,000 in seq CAMP. Time spent on CAMP so far: 834.90 sec.\n",
      "\tOn aln 282,000 in seq CAMP. Time spent on CAMP so far: 838.08 sec.\n",
      "\tOn aln 283,000 in seq CAMP. Time spent on CAMP so far: 841.25 sec.\n",
      "\tOn aln 284,000 in seq CAMP. Time spent on CAMP so far: 844.32 sec.\n",
      "\tOn aln 285,000 in seq CAMP. Time spent on CAMP so far: 847.37 sec.\n",
      "\tOn aln 286,000 in seq CAMP. Time spent on CAMP so far: 850.41 sec.\n",
      "\tOn aln 287,000 in seq CAMP. Time spent on CAMP so far: 853.46 sec.\n",
      "\tOn aln 288,000 in seq CAMP. Time spent on CAMP so far: 856.49 sec.\n",
      "\tOn aln 289,000 in seq CAMP. Time spent on CAMP so far: 859.58 sec.\n",
      "\tOn aln 290,000 in seq CAMP. Time spent on CAMP so far: 862.49 sec.\n",
      "\tOn aln 291,000 in seq CAMP. Time spent on CAMP so far: 865.43 sec.\n",
      "\tOn aln 292,000 in seq CAMP. Time spent on CAMP so far: 868.36 sec.\n",
      "\tOn aln 293,000 in seq CAMP. Time spent on CAMP so far: 871.35 sec.\n",
      "\tOn aln 294,000 in seq CAMP. Time spent on CAMP so far: 874.40 sec.\n",
      "\tOn aln 295,000 in seq CAMP. Time spent on CAMP so far: 877.45 sec.\n",
      "\tOn aln 296,000 in seq CAMP. Time spent on CAMP so far: 880.37 sec.\n",
      "\tOn aln 297,000 in seq CAMP. Time spent on CAMP so far: 883.30 sec.\n",
      "\tOn aln 298,000 in seq CAMP. Time spent on CAMP so far: 886.27 sec.\n",
      "\tOn aln 299,000 in seq CAMP. Time spent on CAMP so far: 889.24 sec.\n",
      "\tOn aln 300,000 in seq CAMP. Time spent on CAMP so far: 892.25 sec.\n",
      "\tOn aln 301,000 in seq CAMP. Time spent on CAMP so far: 895.35 sec.\n",
      "\tOn aln 302,000 in seq CAMP. Time spent on CAMP so far: 898.35 sec.\n",
      "\tOn aln 303,000 in seq CAMP. Time spent on CAMP so far: 901.42 sec.\n",
      "\tOn aln 304,000 in seq CAMP. Time spent on CAMP so far: 904.45 sec.\n",
      "\tOn aln 305,000 in seq CAMP. Time spent on CAMP so far: 907.56 sec.\n",
      "\tOn aln 306,000 in seq CAMP. Time spent on CAMP so far: 910.59 sec.\n",
      "\tOn aln 307,000 in seq CAMP. Time spent on CAMP so far: 913.61 sec.\n",
      "\tOn aln 308,000 in seq CAMP. Time spent on CAMP so far: 916.65 sec.\n",
      "\tOn aln 309,000 in seq CAMP. Time spent on CAMP so far: 919.62 sec.\n",
      "\tOn aln 310,000 in seq CAMP. Time spent on CAMP so far: 922.57 sec.\n",
      "\tOn aln 311,000 in seq CAMP. Time spent on CAMP so far: 925.63 sec.\n",
      "\tOn aln 312,000 in seq CAMP. Time spent on CAMP so far: 928.75 sec.\n",
      "\tOn aln 313,000 in seq CAMP. Time spent on CAMP so far: 931.86 sec.\n",
      "\tOn aln 314,000 in seq CAMP. Time spent on CAMP so far: 934.97 sec.\n",
      "\tOn aln 315,000 in seq CAMP. Time spent on CAMP so far: 938.12 sec.\n",
      "\tOn aln 316,000 in seq CAMP. Time spent on CAMP so far: 941.23 sec.\n",
      "\tOn aln 317,000 in seq CAMP. Time spent on CAMP so far: 944.34 sec.\n",
      "\tOn aln 318,000 in seq CAMP. Time spent on CAMP so far: 947.48 sec.\n",
      "\tOn aln 319,000 in seq CAMP. Time spent on CAMP so far: 950.60 sec.\n",
      "\tOn aln 320,000 in seq CAMP. Time spent on CAMP so far: 953.68 sec.\n",
      "\tOn aln 321,000 in seq CAMP. Time spent on CAMP so far: 956.66 sec.\n",
      "\tOn aln 322,000 in seq CAMP. Time spent on CAMP so far: 959.65 sec.\n",
      "\tOn aln 323,000 in seq CAMP. Time spent on CAMP so far: 962.63 sec.\n",
      "\tOn aln 324,000 in seq CAMP. Time spent on CAMP so far: 965.61 sec.\n",
      "\tOn aln 325,000 in seq CAMP. Time spent on CAMP so far: 968.61 sec.\n",
      "\tOn aln 326,000 in seq CAMP. Time spent on CAMP so far: 971.67 sec.\n",
      "\tOn aln 327,000 in seq CAMP. Time spent on CAMP so far: 974.82 sec.\n",
      "\tOn aln 328,000 in seq CAMP. Time spent on CAMP so far: 977.89 sec.\n",
      "\tOn aln 329,000 in seq CAMP. Time spent on CAMP so far: 981.05 sec.\n",
      "\tOn aln 330,000 in seq CAMP. Time spent on CAMP so far: 984.21 sec.\n",
      "\tOn aln 331,000 in seq CAMP. Time spent on CAMP so far: 987.26 sec.\n",
      "\tOn aln 332,000 in seq CAMP. Time spent on CAMP so far: 990.32 sec.\n",
      "\tOn aln 333,000 in seq CAMP. Time spent on CAMP so far: 993.43 sec.\n",
      "\tOn aln 334,000 in seq CAMP. Time spent on CAMP so far: 996.48 sec.\n",
      "\tOn aln 335,000 in seq CAMP. Time spent on CAMP so far: 999.59 sec.\n",
      "\tOn aln 336,000 in seq CAMP. Time spent on CAMP so far: 1,002.76 sec.\n",
      "\tOn aln 337,000 in seq CAMP. Time spent on CAMP so far: 1,005.87 sec.\n",
      "\tOn aln 338,000 in seq CAMP. Time spent on CAMP so far: 1,008.92 sec.\n",
      "\tOn aln 339,000 in seq CAMP. Time spent on CAMP so far: 1,012.00 sec.\n",
      "\tOn aln 340,000 in seq CAMP. Time spent on CAMP so far: 1,015.13 sec.\n",
      "\tOn aln 341,000 in seq CAMP. Time spent on CAMP so far: 1,018.21 sec.\n",
      "\tOn aln 342,000 in seq CAMP. Time spent on CAMP so far: 1,021.31 sec.\n",
      "\tOn aln 343,000 in seq CAMP. Time spent on CAMP so far: 1,024.38 sec.\n",
      "\tOn aln 344,000 in seq CAMP. Time spent on CAMP so far: 1,027.36 sec.\n",
      "\tOn aln 345,000 in seq CAMP. Time spent on CAMP so far: 1,030.36 sec.\n",
      "\tOn aln 346,000 in seq CAMP. Time spent on CAMP so far: 1,033.34 sec.\n",
      "\tOn aln 347,000 in seq CAMP. Time spent on CAMP so far: 1,036.29 sec.\n",
      "\tOn aln 348,000 in seq CAMP. Time spent on CAMP so far: 1,039.26 sec.\n",
      "\tOn aln 349,000 in seq CAMP. Time spent on CAMP so far: 1,042.26 sec.\n",
      "\tOn aln 350,000 in seq CAMP. Time spent on CAMP so far: 1,045.26 sec.\n",
      "\tOn aln 351,000 in seq CAMP. Time spent on CAMP so far: 1,048.32 sec.\n",
      "\tOn aln 352,000 in seq CAMP. Time spent on CAMP so far: 1,051.36 sec.\n",
      "\tOn aln 353,000 in seq CAMP. Time spent on CAMP so far: 1,054.39 sec.\n",
      "\tOn aln 354,000 in seq CAMP. Time spent on CAMP so far: 1,057.53 sec.\n",
      "\tOn aln 355,000 in seq CAMP. Time spent on CAMP so far: 1,060.72 sec.\n",
      "\tOn aln 356,000 in seq CAMP. Time spent on CAMP so far: 1,063.78 sec.\n",
      "\tOn aln 357,000 in seq CAMP. Time spent on CAMP so far: 1,066.98 sec.\n",
      "\tOn aln 358,000 in seq CAMP. Time spent on CAMP so far: 1,070.14 sec.\n",
      "\tOn aln 359,000 in seq CAMP. Time spent on CAMP so far: 1,073.30 sec.\n",
      "\tOn aln 360,000 in seq CAMP. Time spent on CAMP so far: 1,076.45 sec.\n",
      "\tOn aln 361,000 in seq CAMP. Time spent on CAMP so far: 1,079.55 sec.\n",
      "\tOn aln 362,000 in seq CAMP. Time spent on CAMP so far: 1,082.66 sec.\n",
      "\tOn aln 363,000 in seq CAMP. Time spent on CAMP so far: 1,085.76 sec.\n",
      "\tOn aln 364,000 in seq CAMP. Time spent on CAMP so far: 1,088.78 sec.\n",
      "\tOn aln 365,000 in seq CAMP. Time spent on CAMP so far: 1,091.91 sec.\n",
      "\tOn aln 366,000 in seq CAMP. Time spent on CAMP so far: 1,095.05 sec.\n",
      "\tOn aln 367,000 in seq CAMP. Time spent on CAMP so far: 1,098.22 sec.\n",
      "\tOn aln 368,000 in seq CAMP. Time spent on CAMP so far: 1,101.37 sec.\n",
      "\tOn aln 369,000 in seq CAMP. Time spent on CAMP so far: 1,104.45 sec.\n",
      "\tOn aln 370,000 in seq CAMP. Time spent on CAMP so far: 1,107.52 sec.\n",
      "\tOn aln 371,000 in seq CAMP. Time spent on CAMP so far: 1,110.62 sec.\n",
      "\tOn aln 372,000 in seq CAMP. Time spent on CAMP so far: 1,113.65 sec.\n",
      "\tOn aln 373,000 in seq CAMP. Time spent on CAMP so far: 1,116.79 sec.\n",
      "\tOn aln 374,000 in seq CAMP. Time spent on CAMP so far: 1,119.86 sec.\n",
      "\tOn aln 375,000 in seq CAMP. Time spent on CAMP so far: 1,122.94 sec.\n",
      "\tOn aln 376,000 in seq CAMP. Time spent on CAMP so far: 1,126.03 sec.\n",
      "\tOn aln 377,000 in seq CAMP. Time spent on CAMP so far: 1,129.26 sec.\n",
      "\tOn aln 378,000 in seq CAMP. Time spent on CAMP so far: 1,132.72 sec.\n",
      "\tOn aln 379,000 in seq CAMP. Time spent on CAMP so far: 1,136.33 sec.\n",
      "\tOn aln 380,000 in seq CAMP. Time spent on CAMP so far: 1,140.07 sec.\n",
      "\tOn aln 381,000 in seq CAMP. Time spent on CAMP so far: 1,143.71 sec.\n",
      "\tOn aln 382,000 in seq CAMP. Time spent on CAMP so far: 1,147.09 sec.\n",
      "\tOn aln 383,000 in seq CAMP. Time spent on CAMP so far: 1,150.25 sec.\n",
      "\tOn aln 384,000 in seq CAMP. Time spent on CAMP so far: 1,153.21 sec.\n",
      "\tOn aln 385,000 in seq CAMP. Time spent on CAMP so far: 1,156.15 sec.\n",
      "\tOn aln 386,000 in seq CAMP. Time spent on CAMP so far: 1,159.07 sec.\n",
      "\tOn aln 387,000 in seq CAMP. Time spent on CAMP so far: 1,162.02 sec.\n",
      "\tOn aln 388,000 in seq CAMP. Time spent on CAMP so far: 1,164.99 sec.\n",
      "\tOn aln 389,000 in seq CAMP. Time spent on CAMP so far: 1,168.05 sec.\n",
      "\tOn aln 390,000 in seq CAMP. Time spent on CAMP so far: 1,171.12 sec.\n",
      "\tOn aln 391,000 in seq CAMP. Time spent on CAMP so far: 1,174.23 sec.\n",
      "\tOn aln 392,000 in seq CAMP. Time spent on CAMP so far: 1,177.35 sec.\n",
      "\tOn aln 393,000 in seq CAMP. Time spent on CAMP so far: 1,180.33 sec.\n",
      "\tOn aln 394,000 in seq CAMP. Time spent on CAMP so far: 1,183.38 sec.\n",
      "\tOn aln 395,000 in seq CAMP. Time spent on CAMP so far: 1,186.43 sec.\n",
      "\tOn aln 396,000 in seq CAMP. Time spent on CAMP so far: 1,189.41 sec.\n",
      "\tOn aln 397,000 in seq CAMP. Time spent on CAMP so far: 1,192.42 sec.\n",
      "\tOn aln 398,000 in seq CAMP. Time spent on CAMP so far: 1,195.44 sec.\n",
      "\tOn aln 399,000 in seq CAMP. Time spent on CAMP so far: 1,198.39 sec.\n",
      "\tOn aln 400,000 in seq CAMP. Time spent on CAMP so far: 1,201.30 sec.\n",
      "\tOn aln 401,000 in seq CAMP. Time spent on CAMP so far: 1,204.25 sec.\n",
      "\tOn aln 402,000 in seq CAMP. Time spent on CAMP so far: 1,207.22 sec.\n",
      "\tOn aln 403,000 in seq CAMP. Time spent on CAMP so far: 1,210.23 sec.\n",
      "\tOn aln 404,000 in seq CAMP. Time spent on CAMP so far: 1,213.24 sec.\n",
      "\tOn aln 405,000 in seq CAMP. Time spent on CAMP so far: 1,216.20 sec.\n",
      "\tOn aln 406,000 in seq CAMP. Time spent on CAMP so far: 1,219.13 sec.\n",
      "\tOn aln 407,000 in seq CAMP. Time spent on CAMP so far: 1,222.44 sec.\n",
      "\tOn aln 408,000 in seq CAMP. Time spent on CAMP so far: 1,226.26 sec.\n",
      "\tOn aln 409,000 in seq CAMP. Time spent on CAMP so far: 1,229.70 sec.\n",
      "\tOn aln 410,000 in seq CAMP. Time spent on CAMP so far: 1,232.95 sec.\n",
      "\tOn aln 411,000 in seq CAMP. Time spent on CAMP so far: 1,235.92 sec.\n",
      "\tOn aln 412,000 in seq CAMP. Time spent on CAMP so far: 1,238.88 sec.\n",
      "\tOn aln 413,000 in seq CAMP. Time spent on CAMP so far: 1,241.84 sec.\n",
      "\tOn aln 414,000 in seq CAMP. Time spent on CAMP so far: 1,244.81 sec.\n",
      "\tOn aln 415,000 in seq CAMP. Time spent on CAMP so far: 1,247.80 sec.\n",
      "\tOn aln 416,000 in seq CAMP. Time spent on CAMP so far: 1,250.88 sec.\n",
      "\tOn aln 417,000 in seq CAMP. Time spent on CAMP so far: 1,253.92 sec.\n",
      "\tOn aln 418,000 in seq CAMP. Time spent on CAMP so far: 1,256.95 sec.\n",
      "\tOn aln 419,000 in seq CAMP. Time spent on CAMP so far: 1,260.03 sec.\n",
      "\tOn aln 420,000 in seq CAMP. Time spent on CAMP so far: 1,263.04 sec.\n",
      "\tOn aln 421,000 in seq CAMP. Time spent on CAMP so far: 1,266.08 sec.\n",
      "\tOn aln 422,000 in seq CAMP. Time spent on CAMP so far: 1,269.13 sec.\n",
      "\tOn aln 423,000 in seq CAMP. Time spent on CAMP so far: 1,272.13 sec.\n",
      "\tOn aln 424,000 in seq CAMP. Time spent on CAMP so far: 1,275.12 sec.\n",
      "\tOn aln 425,000 in seq CAMP. Time spent on CAMP so far: 1,278.11 sec.\n",
      "\tOn aln 426,000 in seq CAMP. Time spent on CAMP so far: 1,281.13 sec.\n",
      "\tOn aln 427,000 in seq CAMP. Time spent on CAMP so far: 1,284.13 sec.\n",
      "\tOn aln 428,000 in seq CAMP. Time spent on CAMP so far: 1,287.19 sec.\n",
      "\tOn aln 429,000 in seq CAMP. Time spent on CAMP so far: 1,290.21 sec.\n",
      "\tOn aln 430,000 in seq CAMP. Time spent on CAMP so far: 1,293.17 sec.\n",
      "\tOn aln 431,000 in seq CAMP. Time spent on CAMP so far: 1,296.18 sec.\n",
      "\tOn aln 432,000 in seq CAMP. Time spent on CAMP so far: 1,299.25 sec.\n",
      "\tOn aln 433,000 in seq CAMP. Time spent on CAMP so far: 1,302.33 sec.\n",
      "\tOn aln 434,000 in seq CAMP. Time spent on CAMP so far: 1,305.40 sec.\n",
      "\tOn aln 435,000 in seq CAMP. Time spent on CAMP so far: 1,308.42 sec.\n",
      "\tOn aln 436,000 in seq CAMP. Time spent on CAMP so far: 1,311.41 sec.\n",
      "\tOn aln 437,000 in seq CAMP. Time spent on CAMP so far: 1,314.41 sec.\n",
      "\tOn aln 438,000 in seq CAMP. Time spent on CAMP so far: 1,317.39 sec.\n",
      "\tOn aln 439,000 in seq CAMP. Time spent on CAMP so far: 1,320.45 sec.\n",
      "\tOn aln 440,000 in seq CAMP. Time spent on CAMP so far: 1,323.67 sec.\n",
      "\tOn aln 441,000 in seq CAMP. Time spent on CAMP so far: 1,327.17 sec.\n",
      "\tOn aln 442,000 in seq CAMP. Time spent on CAMP so far: 1,330.99 sec.\n",
      "\tOn aln 443,000 in seq CAMP. Time spent on CAMP so far: 1,335.02 sec.\n",
      "\tOn aln 444,000 in seq CAMP. Time spent on CAMP so far: 1,339.14 sec.\n",
      "\tOn aln 445,000 in seq CAMP. Time spent on CAMP so far: 1,343.35 sec.\n",
      "\tOn aln 446,000 in seq CAMP. Time spent on CAMP so far: 1,346.99 sec.\n",
      "\tOn aln 447,000 in seq CAMP. Time spent on CAMP so far: 1,350.08 sec.\n",
      "\tOn aln 448,000 in seq CAMP. Time spent on CAMP so far: 1,353.16 sec.\n",
      "\tOn aln 449,000 in seq CAMP. Time spent on CAMP so far: 1,356.30 sec.\n",
      "\tOn aln 450,000 in seq CAMP. Time spent on CAMP so far: 1,359.45 sec.\n",
      "\tOn aln 451,000 in seq CAMP. Time spent on CAMP so far: 1,362.54 sec.\n",
      "\tOn aln 452,000 in seq CAMP. Time spent on CAMP so far: 1,365.66 sec.\n",
      "\tOn aln 453,000 in seq CAMP. Time spent on CAMP so far: 1,368.69 sec.\n",
      "\tOn aln 454,000 in seq CAMP. Time spent on CAMP so far: 1,371.77 sec.\n",
      "\tOn aln 455,000 in seq CAMP. Time spent on CAMP so far: 1,374.89 sec.\n",
      "\tOn aln 456,000 in seq CAMP. Time spent on CAMP so far: 1,377.97 sec.\n",
      "\tOn aln 457,000 in seq CAMP. Time spent on CAMP so far: 1,381.01 sec.\n",
      "\tOn aln 458,000 in seq CAMP. Time spent on CAMP so far: 1,384.06 sec.\n",
      "\tOn aln 459,000 in seq CAMP. Time spent on CAMP so far: 1,387.19 sec.\n",
      "\tOn aln 460,000 in seq CAMP. Time spent on CAMP so far: 1,390.27 sec.\n",
      "\tOn aln 461,000 in seq CAMP. Time spent on CAMP so far: 1,393.32 sec.\n",
      "\tOn aln 462,000 in seq CAMP. Time spent on CAMP so far: 1,396.38 sec.\n",
      "\tOn aln 463,000 in seq CAMP. Time spent on CAMP so far: 1,399.38 sec.\n",
      "\tOn aln 464,000 in seq CAMP. Time spent on CAMP so far: 1,402.41 sec.\n",
      "\tOn aln 465,000 in seq CAMP. Time spent on CAMP so far: 1,405.46 sec.\n",
      "\tOn aln 466,000 in seq CAMP. Time spent on CAMP so far: 1,408.59 sec.\n",
      "\tOn aln 467,000 in seq CAMP. Time spent on CAMP so far: 1,411.68 sec.\n",
      "\tOn aln 468,000 in seq CAMP. Time spent on CAMP so far: 1,414.81 sec.\n",
      "\tOn aln 469,000 in seq CAMP. Time spent on CAMP so far: 1,417.96 sec.\n",
      "\tOn aln 470,000 in seq CAMP. Time spent on CAMP so far: 1,421.00 sec.\n",
      "\tOn aln 471,000 in seq CAMP. Time spent on CAMP so far: 1,424.09 sec.\n",
      "\tOn aln 472,000 in seq CAMP. Time spent on CAMP so far: 1,427.13 sec.\n",
      "\tOn aln 473,000 in seq CAMP. Time spent on CAMP so far: 1,430.18 sec.\n",
      "\tOn aln 474,000 in seq CAMP. Time spent on CAMP so far: 1,433.15 sec.\n",
      "\tOn aln 475,000 in seq CAMP. Time spent on CAMP so far: 1,435.93 sec.\n",
      "\tOn aln 476,000 in seq CAMP. Time spent on CAMP so far: 1,438.30 sec.\n",
      "Done with edge_6104! Took 1,438.86 sec.\n",
      "Identifying mutated positions in genome BACT1...\n",
      "Found 23,975 mutated positions in BACT1.\n",
      "Going through these positions...\n",
      "\tOn aln 1,000 in seq BACT1. Time spent on BACT1 so far: 2.54 sec.\n",
      "\tOn aln 2,000 in seq BACT1. Time spent on BACT1 so far: 6.41 sec.\n",
      "\tOn aln 3,000 in seq BACT1. Time spent on BACT1 so far: 10.80 sec.\n",
      "\tOn aln 4,000 in seq BACT1. Time spent on BACT1 so far: 14.43 sec.\n",
      "\tOn aln 5,000 in seq BACT1. Time spent on BACT1 so far: 18.45 sec.\n",
      "\tOn aln 6,000 in seq BACT1. Time spent on BACT1 so far: 22.65 sec.\n",
      "\tOn aln 7,000 in seq BACT1. Time spent on BACT1 so far: 27.23 sec.\n",
      "\tOn aln 8,000 in seq BACT1. Time spent on BACT1 so far: 31.95 sec.\n",
      "\tOn aln 9,000 in seq BACT1. Time spent on BACT1 so far: 36.63 sec.\n",
      "\tOn aln 10,000 in seq BACT1. Time spent on BACT1 so far: 42.26 sec.\n",
      "\tOn aln 11,000 in seq BACT1. Time spent on BACT1 so far: 47.98 sec.\n",
      "\tOn aln 12,000 in seq BACT1. Time spent on BACT1 so far: 53.42 sec.\n",
      "\tOn aln 13,000 in seq BACT1. Time spent on BACT1 so far: 59.38 sec.\n",
      "\tOn aln 14,000 in seq BACT1. Time spent on BACT1 so far: 64.47 sec.\n",
      "\tOn aln 15,000 in seq BACT1. Time spent on BACT1 so far: 69.86 sec.\n",
      "\tOn aln 16,000 in seq BACT1. Time spent on BACT1 so far: 75.35 sec.\n",
      "\tOn aln 17,000 in seq BACT1. Time spent on BACT1 so far: 80.12 sec.\n",
      "\tOn aln 18,000 in seq BACT1. Time spent on BACT1 so far: 85.23 sec.\n",
      "\tOn aln 19,000 in seq BACT1. Time spent on BACT1 so far: 90.40 sec.\n",
      "\tOn aln 20,000 in seq BACT1. Time spent on BACT1 so far: 95.96 sec.\n",
      "\tOn aln 21,000 in seq BACT1. Time spent on BACT1 so far: 101.50 sec.\n",
      "\tOn aln 22,000 in seq BACT1. Time spent on BACT1 so far: 108.33 sec.\n",
      "\tOn aln 23,000 in seq BACT1. Time spent on BACT1 so far: 115.75 sec.\n",
      "\tOn aln 24,000 in seq BACT1. Time spent on BACT1 so far: 121.15 sec.\n",
      "\tOn aln 25,000 in seq BACT1. Time spent on BACT1 so far: 127.27 sec.\n",
      "\tOn aln 26,000 in seq BACT1. Time spent on BACT1 so far: 133.63 sec.\n",
      "\tOn aln 27,000 in seq BACT1. Time spent on BACT1 so far: 141.13 sec.\n",
      "\tOn aln 28,000 in seq BACT1. Time spent on BACT1 so far: 147.67 sec.\n",
      "\tOn aln 29,000 in seq BACT1. Time spent on BACT1 so far: 152.68 sec.\n",
      "\tOn aln 30,000 in seq BACT1. Time spent on BACT1 so far: 158.17 sec.\n",
      "\tOn aln 31,000 in seq BACT1. Time spent on BACT1 so far: 164.61 sec.\n",
      "\tOn aln 32,000 in seq BACT1. Time spent on BACT1 so far: 170.85 sec.\n",
      "\tOn aln 33,000 in seq BACT1. Time spent on BACT1 so far: 176.34 sec.\n",
      "\tOn aln 34,000 in seq BACT1. Time spent on BACT1 so far: 181.84 sec.\n",
      "\tOn aln 35,000 in seq BACT1. Time spent on BACT1 so far: 188.08 sec.\n",
      "\tOn aln 36,000 in seq BACT1. Time spent on BACT1 so far: 193.06 sec.\n",
      "\tOn aln 37,000 in seq BACT1. Time spent on BACT1 so far: 197.68 sec.\n",
      "\tOn aln 38,000 in seq BACT1. Time spent on BACT1 so far: 202.73 sec.\n",
      "\tOn aln 39,000 in seq BACT1. Time spent on BACT1 so far: 208.70 sec.\n",
      "\tOn aln 40,000 in seq BACT1. Time spent on BACT1 so far: 214.13 sec.\n",
      "\tOn aln 41,000 in seq BACT1. Time spent on BACT1 so far: 219.41 sec.\n",
      "\tOn aln 42,000 in seq BACT1. Time spent on BACT1 so far: 224.68 sec.\n",
      "\tOn aln 43,000 in seq BACT1. Time spent on BACT1 so far: 229.37 sec.\n",
      "\tOn aln 44,000 in seq BACT1. Time spent on BACT1 so far: 235.91 sec.\n",
      "\tOn aln 45,000 in seq BACT1. Time spent on BACT1 so far: 245.16 sec.\n",
      "\tOn aln 46,000 in seq BACT1. Time spent on BACT1 so far: 252.58 sec.\n",
      "\tOn aln 47,000 in seq BACT1. Time spent on BACT1 so far: 257.81 sec.\n",
      "\tOn aln 48,000 in seq BACT1. Time spent on BACT1 so far: 262.30 sec.\n",
      "\tOn aln 49,000 in seq BACT1. Time spent on BACT1 so far: 267.31 sec.\n",
      "\tOn aln 50,000 in seq BACT1. Time spent on BACT1 so far: 273.05 sec.\n",
      "\tOn aln 51,000 in seq BACT1. Time spent on BACT1 so far: 278.96 sec.\n",
      "\tOn aln 52,000 in seq BACT1. Time spent on BACT1 so far: 284.45 sec.\n",
      "\tOn aln 53,000 in seq BACT1. Time spent on BACT1 so far: 290.18 sec.\n",
      "\tOn aln 54,000 in seq BACT1. Time spent on BACT1 so far: 295.54 sec.\n",
      "\tOn aln 55,000 in seq BACT1. Time spent on BACT1 so far: 301.04 sec.\n",
      "\tOn aln 56,000 in seq BACT1. Time spent on BACT1 so far: 307.55 sec.\n",
      "\tOn aln 57,000 in seq BACT1. Time spent on BACT1 so far: 314.85 sec.\n",
      "\tOn aln 58,000 in seq BACT1. Time spent on BACT1 so far: 322.22 sec.\n",
      "\tOn aln 59,000 in seq BACT1. Time spent on BACT1 so far: 328.86 sec.\n",
      "\tOn aln 60,000 in seq BACT1. Time spent on BACT1 so far: 334.54 sec.\n",
      "\tOn aln 61,000 in seq BACT1. Time spent on BACT1 so far: 340.17 sec.\n",
      "\tOn aln 62,000 in seq BACT1. Time spent on BACT1 so far: 346.26 sec.\n",
      "\tOn aln 63,000 in seq BACT1. Time spent on BACT1 so far: 353.50 sec.\n",
      "\tOn aln 64,000 in seq BACT1. Time spent on BACT1 so far: 361.07 sec.\n",
      "\tOn aln 65,000 in seq BACT1. Time spent on BACT1 so far: 368.18 sec.\n",
      "\tOn aln 66,000 in seq BACT1. Time spent on BACT1 so far: 375.24 sec.\n",
      "\tOn aln 67,000 in seq BACT1. Time spent on BACT1 so far: 380.62 sec.\n",
      "\tOn aln 68,000 in seq BACT1. Time spent on BACT1 so far: 385.59 sec.\n",
      "\tOn aln 69,000 in seq BACT1. Time spent on BACT1 so far: 390.60 sec.\n",
      "\tOn aln 70,000 in seq BACT1. Time spent on BACT1 so far: 396.13 sec.\n",
      "\tOn aln 71,000 in seq BACT1. Time spent on BACT1 so far: 402.10 sec.\n",
      "\tOn aln 72,000 in seq BACT1. Time spent on BACT1 so far: 408.43 sec.\n",
      "\tOn aln 73,000 in seq BACT1. Time spent on BACT1 so far: 415.57 sec.\n",
      "\tOn aln 74,000 in seq BACT1. Time spent on BACT1 so far: 422.70 sec.\n",
      "\tOn aln 75,000 in seq BACT1. Time spent on BACT1 so far: 431.21 sec.\n",
      "\tOn aln 76,000 in seq BACT1. Time spent on BACT1 so far: 441.37 sec.\n",
      "\tOn aln 77,000 in seq BACT1. Time spent on BACT1 so far: 446.98 sec.\n",
      "\tOn aln 78,000 in seq BACT1. Time spent on BACT1 so far: 452.38 sec.\n",
      "\tOn aln 79,000 in seq BACT1. Time spent on BACT1 so far: 457.88 sec.\n",
      "\tOn aln 80,000 in seq BACT1. Time spent on BACT1 so far: 463.05 sec.\n",
      "\tOn aln 81,000 in seq BACT1. Time spent on BACT1 so far: 468.50 sec.\n",
      "\tOn aln 82,000 in seq BACT1. Time spent on BACT1 so far: 474.13 sec.\n",
      "\tOn aln 83,000 in seq BACT1. Time spent on BACT1 so far: 480.01 sec.\n",
      "\tOn aln 84,000 in seq BACT1. Time spent on BACT1 so far: 486.13 sec.\n",
      "\tOn aln 85,000 in seq BACT1. Time spent on BACT1 so far: 493.26 sec.\n",
      "\tOn aln 86,000 in seq BACT1. Time spent on BACT1 so far: 499.79 sec.\n",
      "\tOn aln 87,000 in seq BACT1. Time spent on BACT1 so far: 505.86 sec.\n",
      "\tOn aln 88,000 in seq BACT1. Time spent on BACT1 so far: 511.76 sec.\n",
      "\tOn aln 89,000 in seq BACT1. Time spent on BACT1 so far: 517.47 sec.\n",
      "\tOn aln 90,000 in seq BACT1. Time spent on BACT1 so far: 524.33 sec.\n",
      "\tOn aln 91,000 in seq BACT1. Time spent on BACT1 so far: 532.78 sec.\n",
      "\tOn aln 92,000 in seq BACT1. Time spent on BACT1 so far: 539.43 sec.\n",
      "\tOn aln 93,000 in seq BACT1. Time spent on BACT1 so far: 544.59 sec.\n",
      "\tOn aln 94,000 in seq BACT1. Time spent on BACT1 so far: 550.52 sec.\n",
      "\tOn aln 95,000 in seq BACT1. Time spent on BACT1 so far: 557.22 sec.\n",
      "\tOn aln 96,000 in seq BACT1. Time spent on BACT1 so far: 564.42 sec.\n",
      "\tOn aln 97,000 in seq BACT1. Time spent on BACT1 so far: 571.02 sec.\n",
      "\tOn aln 98,000 in seq BACT1. Time spent on BACT1 so far: 577.16 sec.\n",
      "\tOn aln 99,000 in seq BACT1. Time spent on BACT1 so far: 584.78 sec.\n",
      "\tOn aln 100,000 in seq BACT1. Time spent on BACT1 so far: 593.22 sec.\n",
      "\tOn aln 101,000 in seq BACT1. Time spent on BACT1 so far: 600.23 sec.\n",
      "\tOn aln 102,000 in seq BACT1. Time spent on BACT1 so far: 606.98 sec.\n",
      "\tOn aln 103,000 in seq BACT1. Time spent on BACT1 so far: 613.42 sec.\n",
      "\tOn aln 104,000 in seq BACT1. Time spent on BACT1 so far: 620.07 sec.\n",
      "\tOn aln 105,000 in seq BACT1. Time spent on BACT1 so far: 627.00 sec.\n",
      "\tOn aln 106,000 in seq BACT1. Time spent on BACT1 so far: 634.06 sec.\n",
      "\tOn aln 107,000 in seq BACT1. Time spent on BACT1 so far: 641.85 sec.\n",
      "\tOn aln 108,000 in seq BACT1. Time spent on BACT1 so far: 650.24 sec.\n",
      "\tOn aln 109,000 in seq BACT1. Time spent on BACT1 so far: 659.81 sec.\n",
      "\tOn aln 110,000 in seq BACT1. Time spent on BACT1 so far: 667.22 sec.\n",
      "\tOn aln 111,000 in seq BACT1. Time spent on BACT1 so far: 673.70 sec.\n",
      "\tOn aln 112,000 in seq BACT1. Time spent on BACT1 so far: 680.28 sec.\n",
      "\tOn aln 113,000 in seq BACT1. Time spent on BACT1 so far: 688.17 sec.\n",
      "\tOn aln 114,000 in seq BACT1. Time spent on BACT1 so far: 696.42 sec.\n",
      "\tOn aln 115,000 in seq BACT1. Time spent on BACT1 so far: 702.85 sec.\n",
      "\tOn aln 116,000 in seq BACT1. Time spent on BACT1 so far: 709.03 sec.\n",
      "\tOn aln 117,000 in seq BACT1. Time spent on BACT1 so far: 715.22 sec.\n",
      "\tOn aln 118,000 in seq BACT1. Time spent on BACT1 so far: 722.02 sec.\n",
      "\tOn aln 119,000 in seq BACT1. Time spent on BACT1 so far: 730.86 sec.\n",
      "\tOn aln 120,000 in seq BACT1. Time spent on BACT1 so far: 738.79 sec.\n",
      "\tOn aln 121,000 in seq BACT1. Time spent on BACT1 so far: 745.28 sec.\n",
      "\tOn aln 122,000 in seq BACT1. Time spent on BACT1 so far: 755.56 sec.\n",
      "\tOn aln 123,000 in seq BACT1. Time spent on BACT1 so far: 763.58 sec.\n",
      "\tOn aln 124,000 in seq BACT1. Time spent on BACT1 so far: 771.13 sec.\n",
      "\tOn aln 125,000 in seq BACT1. Time spent on BACT1 so far: 779.67 sec.\n",
      "\tOn aln 126,000 in seq BACT1. Time spent on BACT1 so far: 786.48 sec.\n",
      "\tOn aln 127,000 in seq BACT1. Time spent on BACT1 so far: 795.12 sec.\n",
      "\tOn aln 128,000 in seq BACT1. Time spent on BACT1 so far: 808.44 sec.\n",
      "\tOn aln 129,000 in seq BACT1. Time spent on BACT1 so far: 817.15 sec.\n",
      "\tOn aln 130,000 in seq BACT1. Time spent on BACT1 so far: 824.62 sec.\n",
      "\tOn aln 131,000 in seq BACT1. Time spent on BACT1 so far: 831.63 sec.\n",
      "\tOn aln 132,000 in seq BACT1. Time spent on BACT1 so far: 838.12 sec.\n",
      "\tOn aln 133,000 in seq BACT1. Time spent on BACT1 so far: 845.15 sec.\n",
      "\tOn aln 134,000 in seq BACT1. Time spent on BACT1 so far: 852.46 sec.\n",
      "\tOn aln 135,000 in seq BACT1. Time spent on BACT1 so far: 860.49 sec.\n",
      "\tOn aln 136,000 in seq BACT1. Time spent on BACT1 so far: 867.27 sec.\n",
      "\tOn aln 137,000 in seq BACT1. Time spent on BACT1 so far: 873.51 sec.\n",
      "\tOn aln 138,000 in seq BACT1. Time spent on BACT1 so far: 879.79 sec.\n",
      "\tOn aln 139,000 in seq BACT1. Time spent on BACT1 so far: 886.18 sec.\n",
      "\tOn aln 140,000 in seq BACT1. Time spent on BACT1 so far: 892.90 sec.\n",
      "\tOn aln 141,000 in seq BACT1. Time spent on BACT1 so far: 899.90 sec.\n",
      "\tOn aln 142,000 in seq BACT1. Time spent on BACT1 so far: 906.96 sec.\n",
      "\tOn aln 143,000 in seq BACT1. Time spent on BACT1 so far: 913.92 sec.\n",
      "\tOn aln 144,000 in seq BACT1. Time spent on BACT1 so far: 920.60 sec.\n",
      "\tOn aln 145,000 in seq BACT1. Time spent on BACT1 so far: 925.94 sec.\n",
      "\tOn aln 146,000 in seq BACT1. Time spent on BACT1 so far: 931.42 sec.\n",
      "\tOn aln 147,000 in seq BACT1. Time spent on BACT1 so far: 937.18 sec.\n",
      "\tOn aln 148,000 in seq BACT1. Time spent on BACT1 so far: 943.74 sec.\n",
      "\tOn aln 149,000 in seq BACT1. Time spent on BACT1 so far: 951.64 sec.\n",
      "\tOn aln 150,000 in seq BACT1. Time spent on BACT1 so far: 960.38 sec.\n",
      "\tOn aln 151,000 in seq BACT1. Time spent on BACT1 so far: 967.97 sec.\n",
      "\tOn aln 152,000 in seq BACT1. Time spent on BACT1 so far: 975.06 sec.\n",
      "\tOn aln 153,000 in seq BACT1. Time spent on BACT1 so far: 983.59 sec.\n",
      "\tOn aln 154,000 in seq BACT1. Time spent on BACT1 so far: 991.95 sec.\n",
      "\tOn aln 155,000 in seq BACT1. Time spent on BACT1 so far: 1,000.88 sec.\n",
      "\tOn aln 156,000 in seq BACT1. Time spent on BACT1 so far: 1,010.59 sec.\n",
      "\tOn aln 157,000 in seq BACT1. Time spent on BACT1 so far: 1,019.79 sec.\n",
      "\tOn aln 158,000 in seq BACT1. Time spent on BACT1 so far: 1,031.22 sec.\n",
      "\tOn aln 159,000 in seq BACT1. Time spent on BACT1 so far: 1,043.74 sec.\n",
      "\tOn aln 160,000 in seq BACT1. Time spent on BACT1 so far: 1,051.48 sec.\n",
      "\tOn aln 161,000 in seq BACT1. Time spent on BACT1 so far: 1,058.97 sec.\n",
      "\tOn aln 162,000 in seq BACT1. Time spent on BACT1 so far: 1,066.53 sec.\n",
      "\tOn aln 163,000 in seq BACT1. Time spent on BACT1 so far: 1,074.89 sec.\n",
      "\tOn aln 164,000 in seq BACT1. Time spent on BACT1 so far: 1,083.06 sec.\n",
      "\tOn aln 165,000 in seq BACT1. Time spent on BACT1 so far: 1,092.02 sec.\n",
      "\tOn aln 166,000 in seq BACT1. Time spent on BACT1 so far: 1,100.94 sec.\n",
      "\tOn aln 167,000 in seq BACT1. Time spent on BACT1 so far: 1,112.54 sec.\n",
      "\tOn aln 168,000 in seq BACT1. Time spent on BACT1 so far: 1,126.28 sec.\n",
      "\tOn aln 169,000 in seq BACT1. Time spent on BACT1 so far: 1,135.96 sec.\n",
      "\tOn aln 170,000 in seq BACT1. Time spent on BACT1 so far: 1,143.34 sec.\n",
      "\tOn aln 171,000 in seq BACT1. Time spent on BACT1 so far: 1,151.30 sec.\n",
      "\tOn aln 172,000 in seq BACT1. Time spent on BACT1 so far: 1,159.16 sec.\n",
      "\tOn aln 173,000 in seq BACT1. Time spent on BACT1 so far: 1,167.16 sec.\n",
      "\tOn aln 174,000 in seq BACT1. Time spent on BACT1 so far: 1,176.32 sec.\n",
      "\tOn aln 175,000 in seq BACT1. Time spent on BACT1 so far: 1,186.93 sec.\n",
      "\tOn aln 176,000 in seq BACT1. Time spent on BACT1 so far: 1,195.13 sec.\n",
      "\tOn aln 177,000 in seq BACT1. Time spent on BACT1 so far: 1,203.14 sec.\n",
      "\tOn aln 178,000 in seq BACT1. Time spent on BACT1 so far: 1,211.48 sec.\n",
      "\tOn aln 179,000 in seq BACT1. Time spent on BACT1 so far: 1,219.15 sec.\n",
      "\tOn aln 180,000 in seq BACT1. Time spent on BACT1 so far: 1,227.02 sec.\n",
      "\tOn aln 181,000 in seq BACT1. Time spent on BACT1 so far: 1,235.11 sec.\n",
      "\tOn aln 182,000 in seq BACT1. Time spent on BACT1 so far: 1,243.36 sec.\n",
      "\tOn aln 183,000 in seq BACT1. Time spent on BACT1 so far: 1,251.40 sec.\n",
      "\tOn aln 184,000 in seq BACT1. Time spent on BACT1 so far: 1,259.67 sec.\n",
      "\tOn aln 185,000 in seq BACT1. Time spent on BACT1 so far: 1,268.90 sec.\n",
      "\tOn aln 186,000 in seq BACT1. Time spent on BACT1 so far: 1,276.84 sec.\n",
      "\tOn aln 187,000 in seq BACT1. Time spent on BACT1 so far: 1,284.53 sec.\n",
      "\tOn aln 188,000 in seq BACT1. Time spent on BACT1 so far: 1,292.62 sec.\n",
      "\tOn aln 189,000 in seq BACT1. Time spent on BACT1 so far: 1,300.99 sec.\n",
      "\tOn aln 190,000 in seq BACT1. Time spent on BACT1 so far: 1,311.00 sec.\n",
      "\tOn aln 191,000 in seq BACT1. Time spent on BACT1 so far: 1,322.45 sec.\n",
      "\tOn aln 192,000 in seq BACT1. Time spent on BACT1 so far: 1,332.74 sec.\n",
      "\tOn aln 193,000 in seq BACT1. Time spent on BACT1 so far: 1,340.85 sec.\n",
      "\tOn aln 194,000 in seq BACT1. Time spent on BACT1 so far: 1,348.31 sec.\n",
      "\tOn aln 195,000 in seq BACT1. Time spent on BACT1 so far: 1,355.14 sec.\n",
      "\tOn aln 196,000 in seq BACT1. Time spent on BACT1 so far: 1,361.31 sec.\n",
      "\tOn aln 197,000 in seq BACT1. Time spent on BACT1 so far: 1,369.37 sec.\n",
      "\tOn aln 198,000 in seq BACT1. Time spent on BACT1 so far: 1,384.03 sec.\n",
      "\tOn aln 199,000 in seq BACT1. Time spent on BACT1 so far: 1,395.68 sec.\n",
      "\tOn aln 200,000 in seq BACT1. Time spent on BACT1 so far: 1,403.60 sec.\n",
      "\tOn aln 201,000 in seq BACT1. Time spent on BACT1 so far: 1,412.08 sec.\n",
      "\tOn aln 202,000 in seq BACT1. Time spent on BACT1 so far: 1,420.32 sec.\n",
      "\tOn aln 203,000 in seq BACT1. Time spent on BACT1 so far: 1,428.33 sec.\n",
      "\tOn aln 204,000 in seq BACT1. Time spent on BACT1 so far: 1,436.31 sec.\n",
      "\tOn aln 205,000 in seq BACT1. Time spent on BACT1 so far: 1,443.87 sec.\n",
      "\tOn aln 206,000 in seq BACT1. Time spent on BACT1 so far: 1,451.49 sec.\n",
      "\tOn aln 207,000 in seq BACT1. Time spent on BACT1 so far: 1,459.49 sec.\n",
      "\tOn aln 208,000 in seq BACT1. Time spent on BACT1 so far: 1,467.59 sec.\n",
      "\tOn aln 209,000 in seq BACT1. Time spent on BACT1 so far: 1,476.41 sec.\n",
      "\tOn aln 210,000 in seq BACT1. Time spent on BACT1 so far: 1,486.36 sec.\n",
      "\tOn aln 211,000 in seq BACT1. Time spent on BACT1 so far: 1,495.43 sec.\n",
      "\tOn aln 212,000 in seq BACT1. Time spent on BACT1 so far: 1,504.82 sec.\n",
      "\tOn aln 213,000 in seq BACT1. Time spent on BACT1 so far: 1,514.37 sec.\n",
      "\tOn aln 214,000 in seq BACT1. Time spent on BACT1 so far: 1,525.40 sec.\n",
      "\tOn aln 215,000 in seq BACT1. Time spent on BACT1 so far: 1,535.99 sec.\n",
      "\tOn aln 216,000 in seq BACT1. Time spent on BACT1 so far: 1,545.73 sec.\n",
      "\tOn aln 217,000 in seq BACT1. Time spent on BACT1 so far: 1,557.00 sec.\n",
      "\tOn aln 218,000 in seq BACT1. Time spent on BACT1 so far: 1,565.76 sec.\n",
      "\tOn aln 219,000 in seq BACT1. Time spent on BACT1 so far: 1,573.13 sec.\n",
      "\tOn aln 220,000 in seq BACT1. Time spent on BACT1 so far: 1,580.29 sec.\n",
      "\tOn aln 221,000 in seq BACT1. Time spent on BACT1 so far: 1,588.68 sec.\n",
      "\tOn aln 222,000 in seq BACT1. Time spent on BACT1 so far: 1,598.79 sec.\n",
      "\tOn aln 223,000 in seq BACT1. Time spent on BACT1 so far: 1,608.11 sec.\n",
      "\tOn aln 224,000 in seq BACT1. Time spent on BACT1 so far: 1,616.65 sec.\n",
      "\tOn aln 225,000 in seq BACT1. Time spent on BACT1 so far: 1,625.85 sec.\n",
      "\tOn aln 226,000 in seq BACT1. Time spent on BACT1 so far: 1,634.58 sec.\n",
      "\tOn aln 227,000 in seq BACT1. Time spent on BACT1 so far: 1,642.60 sec.\n",
      "\tOn aln 228,000 in seq BACT1. Time spent on BACT1 so far: 1,651.13 sec.\n",
      "\tOn aln 229,000 in seq BACT1. Time spent on BACT1 so far: 1,660.07 sec.\n",
      "\tOn aln 230,000 in seq BACT1. Time spent on BACT1 so far: 1,668.56 sec.\n",
      "\tOn aln 231,000 in seq BACT1. Time spent on BACT1 so far: 1,676.82 sec.\n",
      "\tOn aln 232,000 in seq BACT1. Time spent on BACT1 so far: 1,685.04 sec.\n",
      "\tOn aln 233,000 in seq BACT1. Time spent on BACT1 so far: 1,693.30 sec.\n",
      "\tOn aln 234,000 in seq BACT1. Time spent on BACT1 so far: 1,701.33 sec.\n",
      "\tOn aln 235,000 in seq BACT1. Time spent on BACT1 so far: 1,709.37 sec.\n",
      "\tOn aln 236,000 in seq BACT1. Time spent on BACT1 so far: 1,717.73 sec.\n",
      "\tOn aln 237,000 in seq BACT1. Time spent on BACT1 so far: 1,726.96 sec.\n",
      "\tOn aln 238,000 in seq BACT1. Time spent on BACT1 so far: 1,736.36 sec.\n",
      "\tOn aln 239,000 in seq BACT1. Time spent on BACT1 so far: 1,746.66 sec.\n",
      "\tOn aln 240,000 in seq BACT1. Time spent on BACT1 so far: 1,756.05 sec.\n",
      "\tOn aln 241,000 in seq BACT1. Time spent on BACT1 so far: 1,767.59 sec.\n",
      "\tOn aln 242,000 in seq BACT1. Time spent on BACT1 so far: 1,780.97 sec.\n",
      "\tOn aln 243,000 in seq BACT1. Time spent on BACT1 so far: 1,790.07 sec.\n",
      "\tOn aln 244,000 in seq BACT1. Time spent on BACT1 so far: 1,798.36 sec.\n",
      "\tOn aln 245,000 in seq BACT1. Time spent on BACT1 so far: 1,806.77 sec.\n",
      "\tOn aln 246,000 in seq BACT1. Time spent on BACT1 so far: 1,815.44 sec.\n",
      "\tOn aln 247,000 in seq BACT1. Time spent on BACT1 so far: 1,827.87 sec.\n",
      "\tOn aln 248,000 in seq BACT1. Time spent on BACT1 so far: 1,838.71 sec.\n",
      "\tOn aln 249,000 in seq BACT1. Time spent on BACT1 so far: 1,848.35 sec.\n",
      "\tOn aln 250,000 in seq BACT1. Time spent on BACT1 so far: 1,857.62 sec.\n",
      "\tOn aln 251,000 in seq BACT1. Time spent on BACT1 so far: 1,866.70 sec.\n",
      "\tOn aln 252,000 in seq BACT1. Time spent on BACT1 so far: 1,875.59 sec.\n",
      "\tOn aln 253,000 in seq BACT1. Time spent on BACT1 so far: 1,884.13 sec.\n",
      "\tOn aln 254,000 in seq BACT1. Time spent on BACT1 so far: 1,892.59 sec.\n",
      "\tOn aln 255,000 in seq BACT1. Time spent on BACT1 so far: 1,901.31 sec.\n",
      "\tOn aln 256,000 in seq BACT1. Time spent on BACT1 so far: 1,912.02 sec.\n",
      "\tOn aln 257,000 in seq BACT1. Time spent on BACT1 so far: 1,926.12 sec.\n",
      "\tOn aln 258,000 in seq BACT1. Time spent on BACT1 so far: 1,937.29 sec.\n",
      "\tOn aln 259,000 in seq BACT1. Time spent on BACT1 so far: 1,946.43 sec.\n",
      "\tOn aln 260,000 in seq BACT1. Time spent on BACT1 so far: 1,955.86 sec.\n",
      "\tOn aln 261,000 in seq BACT1. Time spent on BACT1 so far: 1,965.14 sec.\n",
      "\tOn aln 262,000 in seq BACT1. Time spent on BACT1 so far: 1,972.99 sec.\n",
      "\tOn aln 263,000 in seq BACT1. Time spent on BACT1 so far: 1,978.93 sec.\n",
      "Done with edge_1671! Took 1,979.66 sec.\n",
      "Identifying mutated positions in genome BACT2...\n",
      "Found 1,631 mutated positions in BACT2.\n",
      "Going through these positions...\n",
      "\tOn aln 1,000 in seq BACT2. Time spent on BACT2 so far: 1.82 sec.\n",
      "\tOn aln 2,000 in seq BACT2. Time spent on BACT2 so far: 3.49 sec.\n",
      "\tOn aln 3,000 in seq BACT2. Time spent on BACT2 so far: 5.11 sec.\n",
      "\tOn aln 4,000 in seq BACT2. Time spent on BACT2 so far: 8.02 sec.\n",
      "\tOn aln 5,000 in seq BACT2. Time spent on BACT2 so far: 11.13 sec.\n",
      "\tOn aln 6,000 in seq BACT2. Time spent on BACT2 so far: 14.09 sec.\n",
      "\tOn aln 7,000 in seq BACT2. Time spent on BACT2 so far: 17.05 sec.\n",
      "\tOn aln 8,000 in seq BACT2. Time spent on BACT2 so far: 19.92 sec.\n",
      "\tOn aln 9,000 in seq BACT2. Time spent on BACT2 so far: 22.85 sec.\n",
      "\tOn aln 10,000 in seq BACT2. Time spent on BACT2 so far: 25.88 sec.\n",
      "\tOn aln 11,000 in seq BACT2. Time spent on BACT2 so far: 28.90 sec.\n",
      "\tOn aln 12,000 in seq BACT2. Time spent on BACT2 so far: 31.80 sec.\n",
      "\tOn aln 13,000 in seq BACT2. Time spent on BACT2 so far: 34.76 sec.\n",
      "\tOn aln 14,000 in seq BACT2. Time spent on BACT2 so far: 37.76 sec.\n",
      "\tOn aln 15,000 in seq BACT2. Time spent on BACT2 so far: 40.73 sec.\n",
      "\tOn aln 16,000 in seq BACT2. Time spent on BACT2 so far: 43.72 sec.\n",
      "\tOn aln 17,000 in seq BACT2. Time spent on BACT2 so far: 46.71 sec.\n",
      "\tOn aln 18,000 in seq BACT2. Time spent on BACT2 so far: 49.69 sec.\n",
      "\tOn aln 19,000 in seq BACT2. Time spent on BACT2 so far: 52.70 sec.\n",
      "\tOn aln 20,000 in seq BACT2. Time spent on BACT2 so far: 55.73 sec.\n",
      "\tOn aln 21,000 in seq BACT2. Time spent on BACT2 so far: 58.72 sec.\n",
      "\tOn aln 22,000 in seq BACT2. Time spent on BACT2 so far: 61.76 sec.\n",
      "\tOn aln 23,000 in seq BACT2. Time spent on BACT2 so far: 64.84 sec.\n",
      "\tOn aln 24,000 in seq BACT2. Time spent on BACT2 so far: 67.83 sec.\n",
      "\tOn aln 25,000 in seq BACT2. Time spent on BACT2 so far: 70.81 sec.\n",
      "\tOn aln 26,000 in seq BACT2. Time spent on BACT2 so far: 73.80 sec.\n",
      "\tOn aln 27,000 in seq BACT2. Time spent on BACT2 so far: 76.77 sec.\n",
      "\tOn aln 28,000 in seq BACT2. Time spent on BACT2 so far: 79.82 sec.\n",
      "\tOn aln 29,000 in seq BACT2. Time spent on BACT2 so far: 82.82 sec.\n",
      "\tOn aln 30,000 in seq BACT2. Time spent on BACT2 so far: 85.76 sec.\n",
      "\tOn aln 31,000 in seq BACT2. Time spent on BACT2 so far: 88.77 sec.\n",
      "\tOn aln 32,000 in seq BACT2. Time spent on BACT2 so far: 91.72 sec.\n",
      "\tOn aln 33,000 in seq BACT2. Time spent on BACT2 so far: 94.70 sec.\n",
      "\tOn aln 34,000 in seq BACT2. Time spent on BACT2 so far: 97.64 sec.\n",
      "\tOn aln 35,000 in seq BACT2. Time spent on BACT2 so far: 100.66 sec.\n",
      "\tOn aln 36,000 in seq BACT2. Time spent on BACT2 so far: 103.66 sec.\n",
      "\tOn aln 37,000 in seq BACT2. Time spent on BACT2 so far: 106.66 sec.\n",
      "\tOn aln 38,000 in seq BACT2. Time spent on BACT2 so far: 109.64 sec.\n",
      "\tOn aln 39,000 in seq BACT2. Time spent on BACT2 so far: 112.59 sec.\n",
      "\tOn aln 40,000 in seq BACT2. Time spent on BACT2 so far: 115.55 sec.\n",
      "\tOn aln 41,000 in seq BACT2. Time spent on BACT2 so far: 118.56 sec.\n",
      "\tOn aln 42,000 in seq BACT2. Time spent on BACT2 so far: 121.62 sec.\n",
      "\tOn aln 43,000 in seq BACT2. Time spent on BACT2 so far: 124.61 sec.\n",
      "\tOn aln 44,000 in seq BACT2. Time spent on BACT2 so far: 127.56 sec.\n",
      "\tOn aln 45,000 in seq BACT2. Time spent on BACT2 so far: 130.59 sec.\n",
      "\tOn aln 46,000 in seq BACT2. Time spent on BACT2 so far: 133.62 sec.\n",
      "\tOn aln 47,000 in seq BACT2. Time spent on BACT2 so far: 136.73 sec.\n",
      "\tOn aln 48,000 in seq BACT2. Time spent on BACT2 so far: 139.80 sec.\n",
      "\tOn aln 49,000 in seq BACT2. Time spent on BACT2 so far: 142.82 sec.\n",
      "\tOn aln 50,000 in seq BACT2. Time spent on BACT2 so far: 145.81 sec.\n",
      "\tOn aln 51,000 in seq BACT2. Time spent on BACT2 so far: 148.96 sec.\n",
      "\tOn aln 52,000 in seq BACT2. Time spent on BACT2 so far: 152.09 sec.\n",
      "\tOn aln 53,000 in seq BACT2. Time spent on BACT2 so far: 155.16 sec.\n",
      "\tOn aln 54,000 in seq BACT2. Time spent on BACT2 so far: 158.19 sec.\n",
      "\tOn aln 55,000 in seq BACT2. Time spent on BACT2 so far: 161.24 sec.\n",
      "\tOn aln 56,000 in seq BACT2. Time spent on BACT2 so far: 164.27 sec.\n",
      "\tOn aln 57,000 in seq BACT2. Time spent on BACT2 so far: 167.24 sec.\n",
      "\tOn aln 58,000 in seq BACT2. Time spent on BACT2 so far: 170.27 sec.\n",
      "\tOn aln 59,000 in seq BACT2. Time spent on BACT2 so far: 173.27 sec.\n",
      "\tOn aln 60,000 in seq BACT2. Time spent on BACT2 so far: 176.32 sec.\n",
      "\tOn aln 61,000 in seq BACT2. Time spent on BACT2 so far: 179.33 sec.\n",
      "\tOn aln 62,000 in seq BACT2. Time spent on BACT2 so far: 182.36 sec.\n",
      "\tOn aln 63,000 in seq BACT2. Time spent on BACT2 so far: 185.38 sec.\n",
      "\tOn aln 64,000 in seq BACT2. Time spent on BACT2 so far: 188.39 sec.\n",
      "\tOn aln 65,000 in seq BACT2. Time spent on BACT2 so far: 191.42 sec.\n",
      "\tOn aln 66,000 in seq BACT2. Time spent on BACT2 so far: 194.44 sec.\n",
      "\tOn aln 67,000 in seq BACT2. Time spent on BACT2 so far: 197.41 sec.\n",
      "\tOn aln 68,000 in seq BACT2. Time spent on BACT2 so far: 200.46 sec.\n",
      "\tOn aln 69,000 in seq BACT2. Time spent on BACT2 so far: 203.53 sec.\n",
      "\tOn aln 70,000 in seq BACT2. Time spent on BACT2 so far: 206.71 sec.\n",
      "\tOn aln 71,000 in seq BACT2. Time spent on BACT2 so far: 209.80 sec.\n",
      "\tOn aln 72,000 in seq BACT2. Time spent on BACT2 so far: 212.93 sec.\n",
      "\tOn aln 73,000 in seq BACT2. Time spent on BACT2 so far: 216.06 sec.\n",
      "\tOn aln 74,000 in seq BACT2. Time spent on BACT2 so far: 219.13 sec.\n",
      "\tOn aln 75,000 in seq BACT2. Time spent on BACT2 so far: 222.09 sec.\n",
      "\tOn aln 76,000 in seq BACT2. Time spent on BACT2 so far: 225.16 sec.\n",
      "\tOn aln 77,000 in seq BACT2. Time spent on BACT2 so far: 228.20 sec.\n",
      "\tOn aln 78,000 in seq BACT2. Time spent on BACT2 so far: 231.26 sec.\n",
      "\tOn aln 79,000 in seq BACT2. Time spent on BACT2 so far: 234.25 sec.\n",
      "\tOn aln 80,000 in seq BACT2. Time spent on BACT2 so far: 237.32 sec.\n",
      "\tOn aln 81,000 in seq BACT2. Time spent on BACT2 so far: 240.34 sec.\n",
      "\tOn aln 82,000 in seq BACT2. Time spent on BACT2 so far: 243.35 sec.\n",
      "\tOn aln 83,000 in seq BACT2. Time spent on BACT2 so far: 246.40 sec.\n",
      "\tOn aln 84,000 in seq BACT2. Time spent on BACT2 so far: 249.47 sec.\n",
      "\tOn aln 85,000 in seq BACT2. Time spent on BACT2 so far: 252.51 sec.\n",
      "\tOn aln 86,000 in seq BACT2. Time spent on BACT2 so far: 255.47 sec.\n",
      "\tOn aln 87,000 in seq BACT2. Time spent on BACT2 so far: 258.53 sec.\n",
      "\tOn aln 88,000 in seq BACT2. Time spent on BACT2 so far: 261.62 sec.\n",
      "\tOn aln 89,000 in seq BACT2. Time spent on BACT2 so far: 264.76 sec.\n",
      "\tOn aln 90,000 in seq BACT2. Time spent on BACT2 so far: 268.06 sec.\n",
      "\tOn aln 91,000 in seq BACT2. Time spent on BACT2 so far: 271.65 sec.\n",
      "\tOn aln 92,000 in seq BACT2. Time spent on BACT2 so far: 275.33 sec.\n",
      "\tOn aln 93,000 in seq BACT2. Time spent on BACT2 so far: 278.92 sec.\n",
      "\tOn aln 94,000 in seq BACT2. Time spent on BACT2 so far: 282.02 sec.\n",
      "\tOn aln 95,000 in seq BACT2. Time spent on BACT2 so far: 285.16 sec.\n",
      "\tOn aln 96,000 in seq BACT2. Time spent on BACT2 so far: 288.32 sec.\n",
      "\tOn aln 97,000 in seq BACT2. Time spent on BACT2 so far: 291.52 sec.\n",
      "\tOn aln 98,000 in seq BACT2. Time spent on BACT2 so far: 294.69 sec.\n",
      "\tOn aln 99,000 in seq BACT2. Time spent on BACT2 so far: 297.91 sec.\n",
      "\tOn aln 100,000 in seq BACT2. Time spent on BACT2 so far: 301.12 sec.\n",
      "\tOn aln 101,000 in seq BACT2. Time spent on BACT2 so far: 304.29 sec.\n",
      "\tOn aln 102,000 in seq BACT2. Time spent on BACT2 so far: 307.46 sec.\n",
      "\tOn aln 103,000 in seq BACT2. Time spent on BACT2 so far: 310.56 sec.\n",
      "\tOn aln 104,000 in seq BACT2. Time spent on BACT2 so far: 313.65 sec.\n",
      "\tOn aln 105,000 in seq BACT2. Time spent on BACT2 so far: 316.75 sec.\n",
      "\tOn aln 106,000 in seq BACT2. Time spent on BACT2 so far: 319.84 sec.\n",
      "\tOn aln 107,000 in seq BACT2. Time spent on BACT2 so far: 322.92 sec.\n",
      "\tOn aln 108,000 in seq BACT2. Time spent on BACT2 so far: 325.99 sec.\n",
      "\tOn aln 109,000 in seq BACT2. Time spent on BACT2 so far: 329.08 sec.\n",
      "\tOn aln 110,000 in seq BACT2. Time spent on BACT2 so far: 332.22 sec.\n",
      "\tOn aln 111,000 in seq BACT2. Time spent on BACT2 so far: 335.28 sec.\n",
      "\tOn aln 112,000 in seq BACT2. Time spent on BACT2 so far: 338.41 sec.\n",
      "\tOn aln 113,000 in seq BACT2. Time spent on BACT2 so far: 341.57 sec.\n",
      "\tOn aln 114,000 in seq BACT2. Time spent on BACT2 so far: 344.78 sec.\n",
      "\tOn aln 115,000 in seq BACT2. Time spent on BACT2 so far: 347.99 sec.\n",
      "\tOn aln 116,000 in seq BACT2. Time spent on BACT2 so far: 351.10 sec.\n",
      "\tOn aln 117,000 in seq BACT2. Time spent on BACT2 so far: 354.20 sec.\n",
      "\tOn aln 118,000 in seq BACT2. Time spent on BACT2 so far: 357.38 sec.\n",
      "\tOn aln 119,000 in seq BACT2. Time spent on BACT2 so far: 360.48 sec.\n",
      "\tOn aln 120,000 in seq BACT2. Time spent on BACT2 so far: 363.63 sec.\n",
      "\tOn aln 121,000 in seq BACT2. Time spent on BACT2 so far: 366.80 sec.\n",
      "\tOn aln 122,000 in seq BACT2. Time spent on BACT2 so far: 370.01 sec.\n",
      "\tOn aln 123,000 in seq BACT2. Time spent on BACT2 so far: 373.15 sec.\n",
      "\tOn aln 124,000 in seq BACT2. Time spent on BACT2 so far: 376.26 sec.\n",
      "\tOn aln 125,000 in seq BACT2. Time spent on BACT2 so far: 379.33 sec.\n",
      "\tOn aln 126,000 in seq BACT2. Time spent on BACT2 so far: 382.46 sec.\n",
      "\tOn aln 127,000 in seq BACT2. Time spent on BACT2 so far: 385.87 sec.\n",
      "\tOn aln 128,000 in seq BACT2. Time spent on BACT2 so far: 389.35 sec.\n",
      "\tOn aln 129,000 in seq BACT2. Time spent on BACT2 so far: 392.80 sec.\n",
      "\tOn aln 130,000 in seq BACT2. Time spent on BACT2 so far: 395.93 sec.\n",
      "\tOn aln 131,000 in seq BACT2. Time spent on BACT2 so far: 398.96 sec.\n",
      "\tOn aln 132,000 in seq BACT2. Time spent on BACT2 so far: 402.05 sec.\n",
      "\tOn aln 133,000 in seq BACT2. Time spent on BACT2 so far: 405.13 sec.\n",
      "\tOn aln 134,000 in seq BACT2. Time spent on BACT2 so far: 408.22 sec.\n",
      "\tOn aln 135,000 in seq BACT2. Time spent on BACT2 so far: 411.33 sec.\n",
      "\tOn aln 136,000 in seq BACT2. Time spent on BACT2 so far: 414.54 sec.\n",
      "\tOn aln 137,000 in seq BACT2. Time spent on BACT2 so far: 417.76 sec.\n",
      "\tOn aln 138,000 in seq BACT2. Time spent on BACT2 so far: 420.90 sec.\n",
      "\tOn aln 139,000 in seq BACT2. Time spent on BACT2 so far: 424.01 sec.\n",
      "\tOn aln 140,000 in seq BACT2. Time spent on BACT2 so far: 427.17 sec.\n",
      "\tOn aln 141,000 in seq BACT2. Time spent on BACT2 so far: 430.38 sec.\n",
      "\tOn aln 142,000 in seq BACT2. Time spent on BACT2 so far: 433.57 sec.\n",
      "\tOn aln 143,000 in seq BACT2. Time spent on BACT2 so far: 436.71 sec.\n",
      "\tOn aln 144,000 in seq BACT2. Time spent on BACT2 so far: 439.91 sec.\n",
      "\tOn aln 145,000 in seq BACT2. Time spent on BACT2 so far: 443.00 sec.\n",
      "\tOn aln 146,000 in seq BACT2. Time spent on BACT2 so far: 446.14 sec.\n",
      "\tOn aln 147,000 in seq BACT2. Time spent on BACT2 so far: 449.28 sec.\n",
      "\tOn aln 148,000 in seq BACT2. Time spent on BACT2 so far: 452.44 sec.\n",
      "\tOn aln 149,000 in seq BACT2. Time spent on BACT2 so far: 455.48 sec.\n",
      "\tOn aln 150,000 in seq BACT2. Time spent on BACT2 so far: 458.62 sec.\n",
      "\tOn aln 151,000 in seq BACT2. Time spent on BACT2 so far: 461.78 sec.\n",
      "\tOn aln 152,000 in seq BACT2. Time spent on BACT2 so far: 464.97 sec.\n",
      "\tOn aln 153,000 in seq BACT2. Time spent on BACT2 so far: 468.16 sec.\n",
      "\tOn aln 154,000 in seq BACT2. Time spent on BACT2 so far: 471.37 sec.\n",
      "\tOn aln 155,000 in seq BACT2. Time spent on BACT2 so far: 474.52 sec.\n",
      "\tOn aln 156,000 in seq BACT2. Time spent on BACT2 so far: 477.60 sec.\n",
      "\tOn aln 157,000 in seq BACT2. Time spent on BACT2 so far: 480.79 sec.\n",
      "\tOn aln 158,000 in seq BACT2. Time spent on BACT2 so far: 483.98 sec.\n",
      "\tOn aln 159,000 in seq BACT2. Time spent on BACT2 so far: 487.16 sec.\n",
      "\tOn aln 160,000 in seq BACT2. Time spent on BACT2 so far: 490.30 sec.\n",
      "\tOn aln 161,000 in seq BACT2. Time spent on BACT2 so far: 493.38 sec.\n",
      "\tOn aln 162,000 in seq BACT2. Time spent on BACT2 so far: 496.56 sec.\n",
      "\tOn aln 163,000 in seq BACT2. Time spent on BACT2 so far: 499.73 sec.\n",
      "\tOn aln 164,000 in seq BACT2. Time spent on BACT2 so far: 502.86 sec.\n",
      "\tOn aln 165,000 in seq BACT2. Time spent on BACT2 so far: 506.02 sec.\n",
      "\tOn aln 166,000 in seq BACT2. Time spent on BACT2 so far: 509.18 sec.\n",
      "\tOn aln 167,000 in seq BACT2. Time spent on BACT2 so far: 512.30 sec.\n",
      "\tOn aln 168,000 in seq BACT2. Time spent on BACT2 so far: 515.38 sec.\n",
      "\tOn aln 169,000 in seq BACT2. Time spent on BACT2 so far: 518.53 sec.\n",
      "\tOn aln 170,000 in seq BACT2. Time spent on BACT2 so far: 521.73 sec.\n",
      "\tOn aln 171,000 in seq BACT2. Time spent on BACT2 so far: 524.94 sec.\n",
      "\tOn aln 172,000 in seq BACT2. Time spent on BACT2 so far: 528.16 sec.\n",
      "\tOn aln 173,000 in seq BACT2. Time spent on BACT2 so far: 531.33 sec.\n",
      "\tOn aln 174,000 in seq BACT2. Time spent on BACT2 so far: 534.51 sec.\n",
      "\tOn aln 175,000 in seq BACT2. Time spent on BACT2 so far: 537.66 sec.\n",
      "\tOn aln 176,000 in seq BACT2. Time spent on BACT2 so far: 540.86 sec.\n",
      "\tOn aln 177,000 in seq BACT2. Time spent on BACT2 so far: 544.06 sec.\n",
      "\tOn aln 178,000 in seq BACT2. Time spent on BACT2 so far: 547.17 sec.\n",
      "\tOn aln 179,000 in seq BACT2. Time spent on BACT2 so far: 550.29 sec.\n",
      "\tOn aln 180,000 in seq BACT2. Time spent on BACT2 so far: 553.50 sec.\n",
      "\tOn aln 181,000 in seq BACT2. Time spent on BACT2 so far: 556.66 sec.\n",
      "\tOn aln 182,000 in seq BACT2. Time spent on BACT2 so far: 559.85 sec.\n",
      "\tOn aln 183,000 in seq BACT2. Time spent on BACT2 so far: 563.03 sec.\n",
      "\tOn aln 184,000 in seq BACT2. Time spent on BACT2 so far: 566.13 sec.\n",
      "\tOn aln 185,000 in seq BACT2. Time spent on BACT2 so far: 569.21 sec.\n",
      "\tOn aln 186,000 in seq BACT2. Time spent on BACT2 so far: 572.30 sec.\n",
      "\tOn aln 187,000 in seq BACT2. Time spent on BACT2 so far: 575.37 sec.\n",
      "\tOn aln 188,000 in seq BACT2. Time spent on BACT2 so far: 578.48 sec.\n",
      "\tOn aln 189,000 in seq BACT2. Time spent on BACT2 so far: 581.63 sec.\n",
      "\tOn aln 190,000 in seq BACT2. Time spent on BACT2 so far: 584.79 sec.\n",
      "\tOn aln 191,000 in seq BACT2. Time spent on BACT2 so far: 587.97 sec.\n",
      "\tOn aln 192,000 in seq BACT2. Time spent on BACT2 so far: 591.21 sec.\n",
      "\tOn aln 193,000 in seq BACT2. Time spent on BACT2 so far: 594.51 sec.\n",
      "\tOn aln 194,000 in seq BACT2. Time spent on BACT2 so far: 597.87 sec.\n",
      "\tOn aln 195,000 in seq BACT2. Time spent on BACT2 so far: 601.15 sec.\n",
      "\tOn aln 196,000 in seq BACT2. Time spent on BACT2 so far: 604.46 sec.\n",
      "\tOn aln 197,000 in seq BACT2. Time spent on BACT2 so far: 607.74 sec.\n",
      "\tOn aln 198,000 in seq BACT2. Time spent on BACT2 so far: 610.88 sec.\n",
      "\tOn aln 199,000 in seq BACT2. Time spent on BACT2 so far: 614.00 sec.\n",
      "\tOn aln 200,000 in seq BACT2. Time spent on BACT2 so far: 617.12 sec.\n",
      "\tOn aln 201,000 in seq BACT2. Time spent on BACT2 so far: 620.20 sec.\n",
      "\tOn aln 202,000 in seq BACT2. Time spent on BACT2 so far: 623.41 sec.\n",
      "\tOn aln 203,000 in seq BACT2. Time spent on BACT2 so far: 626.63 sec.\n",
      "\tOn aln 204,000 in seq BACT2. Time spent on BACT2 so far: 629.79 sec.\n",
      "\tOn aln 205,000 in seq BACT2. Time spent on BACT2 so far: 632.95 sec.\n",
      "\tOn aln 206,000 in seq BACT2. Time spent on BACT2 so far: 636.02 sec.\n",
      "\tOn aln 207,000 in seq BACT2. Time spent on BACT2 so far: 639.10 sec.\n",
      "\tOn aln 208,000 in seq BACT2. Time spent on BACT2 so far: 642.25 sec.\n",
      "\tOn aln 209,000 in seq BACT2. Time spent on BACT2 so far: 645.44 sec.\n",
      "\tOn aln 210,000 in seq BACT2. Time spent on BACT2 so far: 648.66 sec.\n",
      "\tOn aln 211,000 in seq BACT2. Time spent on BACT2 so far: 651.84 sec.\n",
      "\tOn aln 212,000 in seq BACT2. Time spent on BACT2 so far: 654.98 sec.\n",
      "\tOn aln 213,000 in seq BACT2. Time spent on BACT2 so far: 658.05 sec.\n",
      "\tOn aln 214,000 in seq BACT2. Time spent on BACT2 so far: 661.23 sec.\n",
      "\tOn aln 215,000 in seq BACT2. Time spent on BACT2 so far: 664.42 sec.\n",
      "\tOn aln 216,000 in seq BACT2. Time spent on BACT2 so far: 667.59 sec.\n",
      "\tOn aln 217,000 in seq BACT2. Time spent on BACT2 so far: 670.79 sec.\n",
      "\tOn aln 218,000 in seq BACT2. Time spent on BACT2 so far: 673.98 sec.\n",
      "\tOn aln 219,000 in seq BACT2. Time spent on BACT2 so far: 677.09 sec.\n",
      "\tOn aln 220,000 in seq BACT2. Time spent on BACT2 so far: 680.21 sec.\n",
      "\tOn aln 221,000 in seq BACT2. Time spent on BACT2 so far: 683.42 sec.\n",
      "\tOn aln 222,000 in seq BACT2. Time spent on BACT2 so far: 686.68 sec.\n",
      "\tOn aln 223,000 in seq BACT2. Time spent on BACT2 so far: 689.92 sec.\n",
      "\tOn aln 224,000 in seq BACT2. Time spent on BACT2 so far: 693.21 sec.\n",
      "\tOn aln 225,000 in seq BACT2. Time spent on BACT2 so far: 696.49 sec.\n",
      "\tOn aln 226,000 in seq BACT2. Time spent on BACT2 so far: 699.71 sec.\n",
      "\tOn aln 227,000 in seq BACT2. Time spent on BACT2 so far: 702.95 sec.\n",
      "\tOn aln 228,000 in seq BACT2. Time spent on BACT2 so far: 706.24 sec.\n",
      "\tOn aln 229,000 in seq BACT2. Time spent on BACT2 so far: 709.47 sec.\n",
      "\tOn aln 230,000 in seq BACT2. Time spent on BACT2 so far: 712.71 sec.\n",
      "\tOn aln 231,000 in seq BACT2. Time spent on BACT2 so far: 715.89 sec.\n",
      "\tOn aln 232,000 in seq BACT2. Time spent on BACT2 so far: 719.06 sec.\n",
      "\tOn aln 233,000 in seq BACT2. Time spent on BACT2 so far: 722.23 sec.\n",
      "\tOn aln 234,000 in seq BACT2. Time spent on BACT2 so far: 725.45 sec.\n",
      "\tOn aln 235,000 in seq BACT2. Time spent on BACT2 so far: 728.69 sec.\n",
      "\tOn aln 236,000 in seq BACT2. Time spent on BACT2 so far: 731.90 sec.\n",
      "\tOn aln 237,000 in seq BACT2. Time spent on BACT2 so far: 735.15 sec.\n",
      "\tOn aln 238,000 in seq BACT2. Time spent on BACT2 so far: 738.38 sec.\n",
      "\tOn aln 239,000 in seq BACT2. Time spent on BACT2 so far: 741.54 sec.\n",
      "\tOn aln 240,000 in seq BACT2. Time spent on BACT2 so far: 744.73 sec.\n",
      "\tOn aln 241,000 in seq BACT2. Time spent on BACT2 so far: 747.97 sec.\n",
      "\tOn aln 242,000 in seq BACT2. Time spent on BACT2 so far: 751.13 sec.\n",
      "\tOn aln 243,000 in seq BACT2. Time spent on BACT2 so far: 754.32 sec.\n",
      "\tOn aln 244,000 in seq BACT2. Time spent on BACT2 so far: 757.56 sec.\n",
      "\tOn aln 245,000 in seq BACT2. Time spent on BACT2 so far: 760.82 sec.\n",
      "\tOn aln 246,000 in seq BACT2. Time spent on BACT2 so far: 764.00 sec.\n",
      "\tOn aln 247,000 in seq BACT2. Time spent on BACT2 so far: 767.33 sec.\n",
      "\tOn aln 248,000 in seq BACT2. Time spent on BACT2 so far: 770.58 sec.\n",
      "\tOn aln 249,000 in seq BACT2. Time spent on BACT2 so far: 773.79 sec.\n",
      "\tOn aln 250,000 in seq BACT2. Time spent on BACT2 so far: 776.88 sec.\n",
      "\tOn aln 251,000 in seq BACT2. Time spent on BACT2 so far: 780.04 sec.\n",
      "\tOn aln 252,000 in seq BACT2. Time spent on BACT2 so far: 783.18 sec.\n",
      "\tOn aln 253,000 in seq BACT2. Time spent on BACT2 so far: 786.40 sec.\n",
      "\tOn aln 254,000 in seq BACT2. Time spent on BACT2 so far: 789.60 sec.\n",
      "\tOn aln 255,000 in seq BACT2. Time spent on BACT2 so far: 792.84 sec.\n",
      "\tOn aln 256,000 in seq BACT2. Time spent on BACT2 so far: 796.03 sec.\n",
      "\tOn aln 257,000 in seq BACT2. Time spent on BACT2 so far: 799.19 sec.\n",
      "\tOn aln 258,000 in seq BACT2. Time spent on BACT2 so far: 802.41 sec.\n",
      "\tOn aln 259,000 in seq BACT2. Time spent on BACT2 so far: 805.71 sec.\n",
      "\tOn aln 260,000 in seq BACT2. Time spent on BACT2 so far: 809.06 sec.\n",
      "\tOn aln 261,000 in seq BACT2. Time spent on BACT2 so far: 812.33 sec.\n",
      "\tOn aln 262,000 in seq BACT2. Time spent on BACT2 so far: 815.51 sec.\n",
      "\tOn aln 263,000 in seq BACT2. Time spent on BACT2 so far: 818.61 sec.\n",
      "\tOn aln 264,000 in seq BACT2. Time spent on BACT2 so far: 821.81 sec.\n",
      "\tOn aln 265,000 in seq BACT2. Time spent on BACT2 so far: 824.96 sec.\n",
      "\tOn aln 266,000 in seq BACT2. Time spent on BACT2 so far: 828.18 sec.\n",
      "\tOn aln 267,000 in seq BACT2. Time spent on BACT2 so far: 831.42 sec.\n",
      "\tOn aln 268,000 in seq BACT2. Time spent on BACT2 so far: 834.65 sec.\n",
      "\tOn aln 269,000 in seq BACT2. Time spent on BACT2 so far: 837.81 sec.\n",
      "\tOn aln 270,000 in seq BACT2. Time spent on BACT2 so far: 841.00 sec.\n",
      "\tOn aln 271,000 in seq BACT2. Time spent on BACT2 so far: 844.16 sec.\n",
      "\tOn aln 272,000 in seq BACT2. Time spent on BACT2 so far: 847.32 sec.\n",
      "\tOn aln 273,000 in seq BACT2. Time spent on BACT2 so far: 850.39 sec.\n",
      "\tOn aln 274,000 in seq BACT2. Time spent on BACT2 so far: 853.48 sec.\n",
      "\tOn aln 275,000 in seq BACT2. Time spent on BACT2 so far: 856.68 sec.\n",
      "\tOn aln 276,000 in seq BACT2. Time spent on BACT2 so far: 860.01 sec.\n",
      "\tOn aln 277,000 in seq BACT2. Time spent on BACT2 so far: 863.32 sec.\n",
      "\tOn aln 278,000 in seq BACT2. Time spent on BACT2 so far: 866.51 sec.\n",
      "\tOn aln 279,000 in seq BACT2. Time spent on BACT2 so far: 869.60 sec.\n",
      "\tOn aln 280,000 in seq BACT2. Time spent on BACT2 so far: 872.77 sec.\n",
      "\tOn aln 281,000 in seq BACT2. Time spent on BACT2 so far: 875.96 sec.\n",
      "\tOn aln 282,000 in seq BACT2. Time spent on BACT2 so far: 879.11 sec.\n",
      "\tOn aln 283,000 in seq BACT2. Time spent on BACT2 so far: 882.33 sec.\n",
      "\tOn aln 284,000 in seq BACT2. Time spent on BACT2 so far: 885.55 sec.\n",
      "\tOn aln 285,000 in seq BACT2. Time spent on BACT2 so far: 888.83 sec.\n",
      "\tOn aln 286,000 in seq BACT2. Time spent on BACT2 so far: 892.05 sec.\n",
      "\tOn aln 287,000 in seq BACT2. Time spent on BACT2 so far: 895.28 sec.\n",
      "\tOn aln 288,000 in seq BACT2. Time spent on BACT2 so far: 898.50 sec.\n",
      "\tOn aln 289,000 in seq BACT2. Time spent on BACT2 so far: 901.74 sec.\n",
      "\tOn aln 290,000 in seq BACT2. Time spent on BACT2 so far: 905.01 sec.\n",
      "\tOn aln 291,000 in seq BACT2. Time spent on BACT2 so far: 908.24 sec.\n",
      "\tOn aln 292,000 in seq BACT2. Time spent on BACT2 so far: 911.43 sec.\n",
      "\tOn aln 293,000 in seq BACT2. Time spent on BACT2 so far: 914.69 sec.\n",
      "\tOn aln 294,000 in seq BACT2. Time spent on BACT2 so far: 917.90 sec.\n",
      "\tOn aln 295,000 in seq BACT2. Time spent on BACT2 so far: 921.11 sec.\n",
      "\tOn aln 296,000 in seq BACT2. Time spent on BACT2 so far: 924.25 sec.\n",
      "\tOn aln 297,000 in seq BACT2. Time spent on BACT2 so far: 927.42 sec.\n",
      "\tOn aln 298,000 in seq BACT2. Time spent on BACT2 so far: 930.64 sec.\n",
      "\tOn aln 299,000 in seq BACT2. Time spent on BACT2 so far: 933.87 sec.\n",
      "\tOn aln 300,000 in seq BACT2. Time spent on BACT2 so far: 937.09 sec.\n",
      "\tOn aln 301,000 in seq BACT2. Time spent on BACT2 so far: 940.33 sec.\n",
      "\tOn aln 302,000 in seq BACT2. Time spent on BACT2 so far: 943.65 sec.\n",
      "\tOn aln 303,000 in seq BACT2. Time spent on BACT2 so far: 946.89 sec.\n",
      "\tOn aln 304,000 in seq BACT2. Time spent on BACT2 so far: 950.16 sec.\n",
      "\tOn aln 305,000 in seq BACT2. Time spent on BACT2 so far: 953.44 sec.\n",
      "\tOn aln 306,000 in seq BACT2. Time spent on BACT2 so far: 956.64 sec.\n",
      "\tOn aln 307,000 in seq BACT2. Time spent on BACT2 so far: 959.91 sec.\n",
      "\tOn aln 308,000 in seq BACT2. Time spent on BACT2 so far: 963.12 sec.\n",
      "\tOn aln 309,000 in seq BACT2. Time spent on BACT2 so far: 966.29 sec.\n",
      "\tOn aln 310,000 in seq BACT2. Time spent on BACT2 so far: 969.43 sec.\n",
      "\tOn aln 311,000 in seq BACT2. Time spent on BACT2 so far: 972.68 sec.\n",
      "\tOn aln 312,000 in seq BACT2. Time spent on BACT2 so far: 975.91 sec.\n",
      "\tOn aln 313,000 in seq BACT2. Time spent on BACT2 so far: 979.16 sec.\n",
      "\tOn aln 314,000 in seq BACT2. Time spent on BACT2 so far: 982.38 sec.\n",
      "\tOn aln 315,000 in seq BACT2. Time spent on BACT2 so far: 985.66 sec.\n",
      "\tOn aln 316,000 in seq BACT2. Time spent on BACT2 so far: 989.00 sec.\n",
      "\tOn aln 317,000 in seq BACT2. Time spent on BACT2 so far: 992.24 sec.\n",
      "\tOn aln 318,000 in seq BACT2. Time spent on BACT2 so far: 995.51 sec.\n",
      "\tOn aln 319,000 in seq BACT2. Time spent on BACT2 so far: 998.76 sec.\n",
      "\tOn aln 320,000 in seq BACT2. Time spent on BACT2 so far: 1,001.98 sec.\n",
      "\tOn aln 321,000 in seq BACT2. Time spent on BACT2 so far: 1,005.17 sec.\n",
      "\tOn aln 322,000 in seq BACT2. Time spent on BACT2 so far: 1,008.33 sec.\n",
      "\tOn aln 323,000 in seq BACT2. Time spent on BACT2 so far: 1,011.52 sec.\n",
      "\tOn aln 324,000 in seq BACT2. Time spent on BACT2 so far: 1,014.79 sec.\n",
      "\tOn aln 325,000 in seq BACT2. Time spent on BACT2 so far: 1,018.07 sec.\n",
      "\tOn aln 326,000 in seq BACT2. Time spent on BACT2 so far: 1,021.30 sec.\n",
      "\tOn aln 327,000 in seq BACT2. Time spent on BACT2 so far: 1,024.57 sec.\n",
      "\tOn aln 328,000 in seq BACT2. Time spent on BACT2 so far: 1,027.78 sec.\n",
      "\tOn aln 329,000 in seq BACT2. Time spent on BACT2 so far: 1,031.02 sec.\n",
      "\tOn aln 330,000 in seq BACT2. Time spent on BACT2 so far: 1,034.21 sec.\n",
      "\tOn aln 331,000 in seq BACT2. Time spent on BACT2 so far: 1,037.51 sec.\n",
      "\tOn aln 332,000 in seq BACT2. Time spent on BACT2 so far: 1,040.77 sec.\n",
      "\tOn aln 333,000 in seq BACT2. Time spent on BACT2 so far: 1,043.99 sec.\n",
      "\tOn aln 334,000 in seq BACT2. Time spent on BACT2 so far: 1,047.32 sec.\n",
      "\tOn aln 335,000 in seq BACT2. Time spent on BACT2 so far: 1,050.57 sec.\n",
      "\tOn aln 336,000 in seq BACT2. Time spent on BACT2 so far: 1,053.84 sec.\n",
      "\tOn aln 337,000 in seq BACT2. Time spent on BACT2 so far: 1,057.12 sec.\n",
      "\tOn aln 338,000 in seq BACT2. Time spent on BACT2 so far: 1,060.40 sec.\n",
      "\tOn aln 339,000 in seq BACT2. Time spent on BACT2 so far: 1,063.66 sec.\n",
      "\tOn aln 340,000 in seq BACT2. Time spent on BACT2 so far: 1,066.88 sec.\n",
      "\tOn aln 341,000 in seq BACT2. Time spent on BACT2 so far: 1,070.11 sec.\n",
      "\tOn aln 342,000 in seq BACT2. Time spent on BACT2 so far: 1,073.32 sec.\n",
      "\tOn aln 343,000 in seq BACT2. Time spent on BACT2 so far: 1,076.56 sec.\n",
      "\tOn aln 344,000 in seq BACT2. Time spent on BACT2 so far: 1,079.75 sec.\n",
      "\tOn aln 345,000 in seq BACT2. Time spent on BACT2 so far: 1,082.96 sec.\n",
      "\tOn aln 346,000 in seq BACT2. Time spent on BACT2 so far: 1,086.08 sec.\n",
      "\tOn aln 347,000 in seq BACT2. Time spent on BACT2 so far: 1,089.35 sec.\n",
      "\tOn aln 348,000 in seq BACT2. Time spent on BACT2 so far: 1,092.66 sec.\n",
      "\tOn aln 349,000 in seq BACT2. Time spent on BACT2 so far: 1,095.95 sec.\n",
      "\tOn aln 350,000 in seq BACT2. Time spent on BACT2 so far: 1,099.20 sec.\n",
      "\tOn aln 351,000 in seq BACT2. Time spent on BACT2 so far: 1,102.39 sec.\n",
      "\tOn aln 352,000 in seq BACT2. Time spent on BACT2 so far: 1,105.58 sec.\n",
      "\tOn aln 353,000 in seq BACT2. Time spent on BACT2 so far: 1,108.86 sec.\n",
      "\tOn aln 354,000 in seq BACT2. Time spent on BACT2 so far: 1,112.08 sec.\n",
      "\tOn aln 355,000 in seq BACT2. Time spent on BACT2 so far: 1,115.29 sec.\n",
      "\tOn aln 356,000 in seq BACT2. Time spent on BACT2 so far: 1,118.49 sec.\n",
      "\tOn aln 357,000 in seq BACT2. Time spent on BACT2 so far: 1,121.71 sec.\n",
      "\tOn aln 358,000 in seq BACT2. Time spent on BACT2 so far: 1,124.93 sec.\n",
      "\tOn aln 359,000 in seq BACT2. Time spent on BACT2 so far: 1,128.11 sec.\n",
      "\tOn aln 360,000 in seq BACT2. Time spent on BACT2 so far: 1,131.30 sec.\n",
      "\tOn aln 361,000 in seq BACT2. Time spent on BACT2 so far: 1,134.51 sec.\n",
      "\tOn aln 362,000 in seq BACT2. Time spent on BACT2 so far: 1,137.82 sec.\n",
      "\tOn aln 363,000 in seq BACT2. Time spent on BACT2 so far: 1,141.07 sec.\n",
      "\tOn aln 364,000 in seq BACT2. Time spent on BACT2 so far: 1,144.31 sec.\n",
      "\tOn aln 365,000 in seq BACT2. Time spent on BACT2 so far: 1,147.59 sec.\n",
      "\tOn aln 366,000 in seq BACT2. Time spent on BACT2 so far: 1,150.87 sec.\n",
      "\tOn aln 367,000 in seq BACT2. Time spent on BACT2 so far: 1,154.14 sec.\n",
      "\tOn aln 368,000 in seq BACT2. Time spent on BACT2 so far: 1,157.35 sec.\n",
      "\tOn aln 369,000 in seq BACT2. Time spent on BACT2 so far: 1,160.66 sec.\n",
      "\tOn aln 370,000 in seq BACT2. Time spent on BACT2 so far: 1,163.88 sec.\n",
      "\tOn aln 371,000 in seq BACT2. Time spent on BACT2 so far: 1,167.18 sec.\n",
      "\tOn aln 372,000 in seq BACT2. Time spent on BACT2 so far: 1,170.38 sec.\n",
      "\tOn aln 373,000 in seq BACT2. Time spent on BACT2 so far: 1,173.62 sec.\n",
      "\tOn aln 374,000 in seq BACT2. Time spent on BACT2 so far: 1,176.81 sec.\n",
      "\tOn aln 375,000 in seq BACT2. Time spent on BACT2 so far: 1,180.07 sec.\n",
      "\tOn aln 376,000 in seq BACT2. Time spent on BACT2 so far: 1,183.32 sec.\n",
      "\tOn aln 377,000 in seq BACT2. Time spent on BACT2 so far: 1,186.54 sec.\n",
      "\tOn aln 378,000 in seq BACT2. Time spent on BACT2 so far: 1,189.71 sec.\n",
      "\tOn aln 379,000 in seq BACT2. Time spent on BACT2 so far: 1,192.95 sec.\n",
      "\tOn aln 380,000 in seq BACT2. Time spent on BACT2 so far: 1,196.21 sec.\n",
      "\tOn aln 381,000 in seq BACT2. Time spent on BACT2 so far: 1,199.43 sec.\n",
      "\tOn aln 382,000 in seq BACT2. Time spent on BACT2 so far: 1,202.62 sec.\n",
      "\tOn aln 383,000 in seq BACT2. Time spent on BACT2 so far: 1,205.78 sec.\n",
      "\tOn aln 384,000 in seq BACT2. Time spent on BACT2 so far: 1,209.06 sec.\n",
      "\tOn aln 385,000 in seq BACT2. Time spent on BACT2 so far: 1,212.42 sec.\n",
      "\tOn aln 386,000 in seq BACT2. Time spent on BACT2 so far: 1,215.71 sec.\n",
      "\tOn aln 387,000 in seq BACT2. Time spent on BACT2 so far: 1,219.07 sec.\n",
      "\tOn aln 388,000 in seq BACT2. Time spent on BACT2 so far: 1,222.29 sec.\n",
      "\tOn aln 389,000 in seq BACT2. Time spent on BACT2 so far: 1,225.49 sec.\n",
      "\tOn aln 390,000 in seq BACT2. Time spent on BACT2 so far: 1,228.75 sec.\n",
      "\tOn aln 391,000 in seq BACT2. Time spent on BACT2 so far: 1,231.89 sec.\n",
      "\tOn aln 392,000 in seq BACT2. Time spent on BACT2 so far: 1,235.11 sec.\n",
      "\tOn aln 393,000 in seq BACT2. Time spent on BACT2 so far: 1,238.34 sec.\n",
      "\tOn aln 394,000 in seq BACT2. Time spent on BACT2 so far: 1,241.61 sec.\n",
      "\tOn aln 395,000 in seq BACT2. Time spent on BACT2 so far: 1,244.98 sec.\n",
      "\tOn aln 396,000 in seq BACT2. Time spent on BACT2 so far: 1,248.23 sec.\n",
      "\tOn aln 397,000 in seq BACT2. Time spent on BACT2 so far: 1,251.41 sec.\n",
      "\tOn aln 398,000 in seq BACT2. Time spent on BACT2 so far: 1,254.65 sec.\n",
      "\tOn aln 399,000 in seq BACT2. Time spent on BACT2 so far: 1,257.93 sec.\n",
      "\tOn aln 400,000 in seq BACT2. Time spent on BACT2 so far: 1,261.16 sec.\n",
      "\tOn aln 401,000 in seq BACT2. Time spent on BACT2 so far: 1,264.44 sec.\n",
      "\tOn aln 402,000 in seq BACT2. Time spent on BACT2 so far: 1,267.75 sec.\n",
      "\tOn aln 403,000 in seq BACT2. Time spent on BACT2 so far: 1,270.99 sec.\n",
      "\tOn aln 404,000 in seq BACT2. Time spent on BACT2 so far: 1,274.17 sec.\n",
      "\tOn aln 405,000 in seq BACT2. Time spent on BACT2 so far: 1,277.44 sec.\n",
      "\tOn aln 406,000 in seq BACT2. Time spent on BACT2 so far: 1,280.74 sec.\n",
      "\tOn aln 407,000 in seq BACT2. Time spent on BACT2 so far: 1,284.05 sec.\n",
      "\tOn aln 408,000 in seq BACT2. Time spent on BACT2 so far: 1,287.36 sec.\n",
      "\tOn aln 409,000 in seq BACT2. Time spent on BACT2 so far: 1,290.67 sec.\n",
      "\tOn aln 410,000 in seq BACT2. Time spent on BACT2 so far: 1,293.93 sec.\n",
      "\tOn aln 411,000 in seq BACT2. Time spent on BACT2 so far: 1,297.30 sec.\n",
      "\tOn aln 412,000 in seq BACT2. Time spent on BACT2 so far: 1,300.69 sec.\n",
      "\tOn aln 413,000 in seq BACT2. Time spent on BACT2 so far: 1,303.94 sec.\n",
      "\tOn aln 414,000 in seq BACT2. Time spent on BACT2 so far: 1,307.18 sec.\n",
      "\tOn aln 415,000 in seq BACT2. Time spent on BACT2 so far: 1,310.43 sec.\n",
      "\tOn aln 416,000 in seq BACT2. Time spent on BACT2 so far: 1,313.65 sec.\n",
      "\tOn aln 417,000 in seq BACT2. Time spent on BACT2 so far: 1,316.88 sec.\n",
      "\tOn aln 418,000 in seq BACT2. Time spent on BACT2 so far: 1,320.17 sec.\n",
      "\tOn aln 419,000 in seq BACT2. Time spent on BACT2 so far: 1,323.46 sec.\n",
      "\tOn aln 420,000 in seq BACT2. Time spent on BACT2 so far: 1,326.81 sec.\n",
      "\tOn aln 421,000 in seq BACT2. Time spent on BACT2 so far: 1,330.17 sec.\n",
      "\tOn aln 422,000 in seq BACT2. Time spent on BACT2 so far: 1,333.51 sec.\n",
      "\tOn aln 423,000 in seq BACT2. Time spent on BACT2 so far: 1,336.75 sec.\n",
      "\tOn aln 424,000 in seq BACT2. Time spent on BACT2 so far: 1,340.01 sec.\n",
      "\tOn aln 425,000 in seq BACT2. Time spent on BACT2 so far: 1,343.22 sec.\n",
      "\tOn aln 426,000 in seq BACT2. Time spent on BACT2 so far: 1,346.42 sec.\n",
      "\tOn aln 427,000 in seq BACT2. Time spent on BACT2 so far: 1,349.61 sec.\n",
      "\tOn aln 428,000 in seq BACT2. Time spent on BACT2 so far: 1,352.89 sec.\n",
      "\tOn aln 429,000 in seq BACT2. Time spent on BACT2 so far: 1,356.16 sec.\n",
      "\tOn aln 430,000 in seq BACT2. Time spent on BACT2 so far: 1,359.44 sec.\n",
      "\tOn aln 431,000 in seq BACT2. Time spent on BACT2 so far: 1,362.68 sec.\n",
      "\tOn aln 432,000 in seq BACT2. Time spent on BACT2 so far: 1,366.02 sec.\n",
      "\tOn aln 433,000 in seq BACT2. Time spent on BACT2 so far: 1,369.29 sec.\n",
      "\tOn aln 434,000 in seq BACT2. Time spent on BACT2 so far: 1,372.65 sec.\n",
      "\tOn aln 435,000 in seq BACT2. Time spent on BACT2 so far: 1,375.98 sec.\n",
      "\tOn aln 436,000 in seq BACT2. Time spent on BACT2 so far: 1,379.25 sec.\n",
      "\tOn aln 437,000 in seq BACT2. Time spent on BACT2 so far: 1,382.56 sec.\n",
      "\tOn aln 438,000 in seq BACT2. Time spent on BACT2 so far: 1,385.78 sec.\n",
      "\tOn aln 439,000 in seq BACT2. Time spent on BACT2 so far: 1,388.98 sec.\n",
      "\tOn aln 440,000 in seq BACT2. Time spent on BACT2 so far: 1,392.14 sec.\n",
      "\tOn aln 441,000 in seq BACT2. Time spent on BACT2 so far: 1,395.33 sec.\n",
      "\tOn aln 442,000 in seq BACT2. Time spent on BACT2 so far: 1,398.62 sec.\n",
      "\tOn aln 443,000 in seq BACT2. Time spent on BACT2 so far: 1,401.85 sec.\n",
      "\tOn aln 444,000 in seq BACT2. Time spent on BACT2 so far: 1,405.16 sec.\n",
      "\tOn aln 445,000 in seq BACT2. Time spent on BACT2 so far: 1,408.37 sec.\n",
      "\tOn aln 446,000 in seq BACT2. Time spent on BACT2 so far: 1,411.63 sec.\n",
      "\tOn aln 447,000 in seq BACT2. Time spent on BACT2 so far: 1,415.04 sec.\n",
      "\tOn aln 448,000 in seq BACT2. Time spent on BACT2 so far: 1,418.64 sec.\n",
      "\tOn aln 449,000 in seq BACT2. Time spent on BACT2 so far: 1,422.24 sec.\n",
      "\tOn aln 450,000 in seq BACT2. Time spent on BACT2 so far: 1,425.81 sec.\n",
      "\tOn aln 451,000 in seq BACT2. Time spent on BACT2 so far: 1,429.09 sec.\n",
      "\tOn aln 452,000 in seq BACT2. Time spent on BACT2 so far: 1,432.41 sec.\n",
      "\tOn aln 453,000 in seq BACT2. Time spent on BACT2 so far: 1,435.76 sec.\n",
      "\tOn aln 454,000 in seq BACT2. Time spent on BACT2 so far: 1,439.05 sec.\n",
      "\tOn aln 455,000 in seq BACT2. Time spent on BACT2 so far: 1,442.44 sec.\n",
      "\tOn aln 456,000 in seq BACT2. Time spent on BACT2 so far: 1,445.79 sec.\n",
      "\tOn aln 457,000 in seq BACT2. Time spent on BACT2 so far: 1,449.09 sec.\n",
      "\tOn aln 458,000 in seq BACT2. Time spent on BACT2 so far: 1,452.31 sec.\n",
      "\tOn aln 459,000 in seq BACT2. Time spent on BACT2 so far: 1,455.52 sec.\n",
      "\tOn aln 460,000 in seq BACT2. Time spent on BACT2 so far: 1,458.75 sec.\n",
      "\tOn aln 461,000 in seq BACT2. Time spent on BACT2 so far: 1,461.99 sec.\n",
      "\tOn aln 462,000 in seq BACT2. Time spent on BACT2 so far: 1,465.23 sec.\n",
      "\tOn aln 463,000 in seq BACT2. Time spent on BACT2 so far: 1,468.47 sec.\n",
      "\tOn aln 464,000 in seq BACT2. Time spent on BACT2 so far: 1,471.64 sec.\n",
      "\tOn aln 465,000 in seq BACT2. Time spent on BACT2 so far: 1,474.91 sec.\n",
      "\tOn aln 466,000 in seq BACT2. Time spent on BACT2 so far: 1,478.19 sec.\n",
      "\tOn aln 467,000 in seq BACT2. Time spent on BACT2 so far: 1,481.47 sec.\n",
      "\tOn aln 468,000 in seq BACT2. Time spent on BACT2 so far: 1,484.77 sec.\n",
      "\tOn aln 469,000 in seq BACT2. Time spent on BACT2 so far: 1,488.00 sec.\n",
      "\tOn aln 470,000 in seq BACT2. Time spent on BACT2 so far: 1,491.22 sec.\n",
      "\tOn aln 471,000 in seq BACT2. Time spent on BACT2 so far: 1,494.46 sec.\n",
      "\tOn aln 472,000 in seq BACT2. Time spent on BACT2 so far: 1,497.69 sec.\n",
      "\tOn aln 473,000 in seq BACT2. Time spent on BACT2 so far: 1,500.90 sec.\n",
      "\tOn aln 474,000 in seq BACT2. Time spent on BACT2 so far: 1,504.18 sec.\n",
      "\tOn aln 475,000 in seq BACT2. Time spent on BACT2 so far: 1,507.43 sec.\n",
      "\tOn aln 476,000 in seq BACT2. Time spent on BACT2 so far: 1,510.71 sec.\n",
      "\tOn aln 477,000 in seq BACT2. Time spent on BACT2 so far: 1,513.97 sec.\n",
      "\tOn aln 478,000 in seq BACT2. Time spent on BACT2 so far: 1,517.24 sec.\n",
      "\tOn aln 479,000 in seq BACT2. Time spent on BACT2 so far: 1,520.62 sec.\n",
      "\tOn aln 480,000 in seq BACT2. Time spent on BACT2 so far: 1,523.85 sec.\n",
      "\tOn aln 481,000 in seq BACT2. Time spent on BACT2 so far: 1,527.12 sec.\n",
      "\tOn aln 482,000 in seq BACT2. Time spent on BACT2 so far: 1,530.49 sec.\n",
      "\tOn aln 483,000 in seq BACT2. Time spent on BACT2 so far: 1,533.82 sec.\n",
      "\tOn aln 484,000 in seq BACT2. Time spent on BACT2 so far: 1,537.09 sec.\n",
      "\tOn aln 485,000 in seq BACT2. Time spent on BACT2 so far: 1,540.28 sec.\n",
      "\tOn aln 486,000 in seq BACT2. Time spent on BACT2 so far: 1,543.49 sec.\n",
      "\tOn aln 487,000 in seq BACT2. Time spent on BACT2 so far: 1,546.78 sec.\n",
      "\tOn aln 488,000 in seq BACT2. Time spent on BACT2 so far: 1,550.07 sec.\n",
      "\tOn aln 489,000 in seq BACT2. Time spent on BACT2 so far: 1,553.30 sec.\n",
      "\tOn aln 490,000 in seq BACT2. Time spent on BACT2 so far: 1,556.52 sec.\n",
      "\tOn aln 491,000 in seq BACT2. Time spent on BACT2 so far: 1,559.72 sec.\n",
      "\tOn aln 492,000 in seq BACT2. Time spent on BACT2 so far: 1,562.88 sec.\n",
      "\tOn aln 493,000 in seq BACT2. Time spent on BACT2 so far: 1,566.06 sec.\n",
      "\tOn aln 494,000 in seq BACT2. Time spent on BACT2 so far: 1,569.27 sec.\n",
      "\tOn aln 495,000 in seq BACT2. Time spent on BACT2 so far: 1,572.52 sec.\n",
      "\tOn aln 496,000 in seq BACT2. Time spent on BACT2 so far: 1,575.79 sec.\n",
      "\tOn aln 497,000 in seq BACT2. Time spent on BACT2 so far: 1,579.07 sec.\n",
      "\tOn aln 498,000 in seq BACT2. Time spent on BACT2 so far: 1,582.34 sec.\n",
      "\tOn aln 499,000 in seq BACT2. Time spent on BACT2 so far: 1,585.64 sec.\n",
      "\tOn aln 500,000 in seq BACT2. Time spent on BACT2 so far: 1,588.98 sec.\n",
      "\tOn aln 501,000 in seq BACT2. Time spent on BACT2 so far: 1,592.22 sec.\n",
      "\tOn aln 502,000 in seq BACT2. Time spent on BACT2 so far: 1,595.50 sec.\n",
      "\tOn aln 503,000 in seq BACT2. Time spent on BACT2 so far: 1,598.82 sec.\n",
      "\tOn aln 504,000 in seq BACT2. Time spent on BACT2 so far: 1,602.16 sec.\n",
      "\tOn aln 505,000 in seq BACT2. Time spent on BACT2 so far: 1,605.49 sec.\n",
      "\tOn aln 506,000 in seq BACT2. Time spent on BACT2 so far: 1,608.73 sec.\n",
      "\tOn aln 507,000 in seq BACT2. Time spent on BACT2 so far: 1,612.03 sec.\n",
      "\tOn aln 508,000 in seq BACT2. Time spent on BACT2 so far: 1,615.32 sec.\n",
      "\tOn aln 509,000 in seq BACT2. Time spent on BACT2 so far: 1,618.53 sec.\n",
      "\tOn aln 510,000 in seq BACT2. Time spent on BACT2 so far: 1,621.79 sec.\n",
      "\tOn aln 511,000 in seq BACT2. Time spent on BACT2 so far: 1,625.08 sec.\n",
      "\tOn aln 512,000 in seq BACT2. Time spent on BACT2 so far: 1,628.40 sec.\n",
      "\tOn aln 513,000 in seq BACT2. Time spent on BACT2 so far: 1,631.67 sec.\n",
      "\tOn aln 514,000 in seq BACT2. Time spent on BACT2 so far: 1,634.97 sec.\n",
      "\tOn aln 515,000 in seq BACT2. Time spent on BACT2 so far: 1,638.24 sec.\n",
      "\tOn aln 516,000 in seq BACT2. Time spent on BACT2 so far: 1,641.54 sec.\n",
      "\tOn aln 517,000 in seq BACT2. Time spent on BACT2 so far: 1,644.83 sec.\n",
      "\tOn aln 518,000 in seq BACT2. Time spent on BACT2 so far: 1,648.12 sec.\n",
      "\tOn aln 519,000 in seq BACT2. Time spent on BACT2 so far: 1,651.35 sec.\n",
      "\tOn aln 520,000 in seq BACT2. Time spent on BACT2 so far: 1,654.57 sec.\n",
      "\tOn aln 521,000 in seq BACT2. Time spent on BACT2 so far: 1,657.80 sec.\n",
      "\tOn aln 522,000 in seq BACT2. Time spent on BACT2 so far: 1,661.06 sec.\n",
      "\tOn aln 523,000 in seq BACT2. Time spent on BACT2 so far: 1,664.29 sec.\n",
      "\tOn aln 524,000 in seq BACT2. Time spent on BACT2 so far: 1,667.55 sec.\n",
      "\tOn aln 525,000 in seq BACT2. Time spent on BACT2 so far: 1,670.81 sec.\n",
      "\tOn aln 526,000 in seq BACT2. Time spent on BACT2 so far: 1,674.14 sec.\n",
      "\tOn aln 527,000 in seq BACT2. Time spent on BACT2 so far: 1,677.45 sec.\n",
      "\tOn aln 528,000 in seq BACT2. Time spent on BACT2 so far: 1,680.77 sec.\n",
      "\tOn aln 529,000 in seq BACT2. Time spent on BACT2 so far: 1,684.14 sec.\n",
      "\tOn aln 530,000 in seq BACT2. Time spent on BACT2 so far: 1,687.56 sec.\n",
      "\tOn aln 531,000 in seq BACT2. Time spent on BACT2 so far: 1,690.95 sec.\n",
      "\tOn aln 532,000 in seq BACT2. Time spent on BACT2 so far: 1,694.32 sec.\n",
      "\tOn aln 533,000 in seq BACT2. Time spent on BACT2 so far: 1,697.67 sec.\n",
      "\tOn aln 534,000 in seq BACT2. Time spent on BACT2 so far: 1,700.97 sec.\n",
      "\tOn aln 535,000 in seq BACT2. Time spent on BACT2 so far: 1,704.33 sec.\n",
      "\tOn aln 536,000 in seq BACT2. Time spent on BACT2 so far: 1,707.68 sec.\n",
      "\tOn aln 537,000 in seq BACT2. Time spent on BACT2 so far: 1,711.01 sec.\n",
      "\tOn aln 538,000 in seq BACT2. Time spent on BACT2 so far: 1,714.33 sec.\n",
      "\tOn aln 539,000 in seq BACT2. Time spent on BACT2 so far: 1,717.67 sec.\n",
      "\tOn aln 540,000 in seq BACT2. Time spent on BACT2 so far: 1,721.00 sec.\n",
      "\tOn aln 541,000 in seq BACT2. Time spent on BACT2 so far: 1,724.30 sec.\n",
      "\tOn aln 542,000 in seq BACT2. Time spent on BACT2 so far: 1,727.62 sec.\n",
      "\tOn aln 543,000 in seq BACT2. Time spent on BACT2 so far: 1,730.90 sec.\n",
      "\tOn aln 544,000 in seq BACT2. Time spent on BACT2 so far: 1,734.23 sec.\n",
      "\tOn aln 545,000 in seq BACT2. Time spent on BACT2 so far: 1,737.48 sec.\n",
      "\tOn aln 546,000 in seq BACT2. Time spent on BACT2 so far: 1,740.76 sec.\n",
      "\tOn aln 547,000 in seq BACT2. Time spent on BACT2 so far: 1,744.08 sec.\n",
      "\tOn aln 548,000 in seq BACT2. Time spent on BACT2 so far: 1,747.38 sec.\n",
      "\tOn aln 549,000 in seq BACT2. Time spent on BACT2 so far: 1,750.76 sec.\n",
      "\tOn aln 550,000 in seq BACT2. Time spent on BACT2 so far: 1,754.02 sec.\n",
      "\tOn aln 551,000 in seq BACT2. Time spent on BACT2 so far: 1,757.47 sec.\n",
      "\tOn aln 552,000 in seq BACT2. Time spent on BACT2 so far: 1,760.85 sec.\n",
      "\tOn aln 553,000 in seq BACT2. Time spent on BACT2 so far: 1,764.10 sec.\n",
      "\tOn aln 554,000 in seq BACT2. Time spent on BACT2 so far: 1,767.34 sec.\n",
      "\tOn aln 555,000 in seq BACT2. Time spent on BACT2 so far: 1,770.66 sec.\n",
      "\tOn aln 556,000 in seq BACT2. Time spent on BACT2 so far: 1,774.00 sec.\n",
      "\tOn aln 557,000 in seq BACT2. Time spent on BACT2 so far: 1,777.37 sec.\n",
      "\tOn aln 558,000 in seq BACT2. Time spent on BACT2 so far: 1,780.79 sec.\n",
      "\tOn aln 559,000 in seq BACT2. Time spent on BACT2 so far: 1,784.15 sec.\n",
      "\tOn aln 560,000 in seq BACT2. Time spent on BACT2 so far: 1,787.44 sec.\n",
      "\tOn aln 561,000 in seq BACT2. Time spent on BACT2 so far: 1,790.59 sec.\n",
      "\tOn aln 562,000 in seq BACT2. Time spent on BACT2 so far: 1,793.84 sec.\n",
      "\tOn aln 563,000 in seq BACT2. Time spent on BACT2 so far: 1,797.11 sec.\n",
      "\tOn aln 564,000 in seq BACT2. Time spent on BACT2 so far: 1,800.33 sec.\n",
      "\tOn aln 565,000 in seq BACT2. Time spent on BACT2 so far: 1,803.47 sec.\n",
      "\tOn aln 566,000 in seq BACT2. Time spent on BACT2 so far: 1,806.72 sec.\n",
      "\tOn aln 567,000 in seq BACT2. Time spent on BACT2 so far: 1,810.05 sec.\n",
      "\tOn aln 568,000 in seq BACT2. Time spent on BACT2 so far: 1,813.37 sec.\n",
      "\tOn aln 569,000 in seq BACT2. Time spent on BACT2 so far: 1,816.72 sec.\n",
      "\tOn aln 570,000 in seq BACT2. Time spent on BACT2 so far: 1,820.12 sec.\n",
      "\tOn aln 571,000 in seq BACT2. Time spent on BACT2 so far: 1,823.50 sec.\n",
      "\tOn aln 572,000 in seq BACT2. Time spent on BACT2 so far: 1,826.82 sec.\n",
      "\tOn aln 573,000 in seq BACT2. Time spent on BACT2 so far: 1,830.20 sec.\n",
      "\tOn aln 574,000 in seq BACT2. Time spent on BACT2 so far: 1,833.64 sec.\n",
      "\tOn aln 575,000 in seq BACT2. Time spent on BACT2 so far: 1,836.98 sec.\n",
      "\tOn aln 576,000 in seq BACT2. Time spent on BACT2 so far: 1,840.34 sec.\n",
      "\tOn aln 577,000 in seq BACT2. Time spent on BACT2 so far: 1,843.64 sec.\n",
      "\tOn aln 578,000 in seq BACT2. Time spent on BACT2 so far: 1,846.90 sec.\n",
      "\tOn aln 579,000 in seq BACT2. Time spent on BACT2 so far: 1,850.21 sec.\n",
      "\tOn aln 580,000 in seq BACT2. Time spent on BACT2 so far: 1,853.61 sec.\n",
      "\tOn aln 581,000 in seq BACT2. Time spent on BACT2 so far: 1,857.10 sec.\n",
      "\tOn aln 582,000 in seq BACT2. Time spent on BACT2 so far: 1,860.51 sec.\n",
      "\tOn aln 583,000 in seq BACT2. Time spent on BACT2 so far: 1,863.96 sec.\n",
      "\tOn aln 584,000 in seq BACT2. Time spent on BACT2 so far: 1,867.28 sec.\n",
      "\tOn aln 585,000 in seq BACT2. Time spent on BACT2 so far: 1,870.82 sec.\n",
      "\tOn aln 586,000 in seq BACT2. Time spent on BACT2 so far: 1,874.73 sec.\n",
      "\tOn aln 587,000 in seq BACT2. Time spent on BACT2 so far: 1,878.18 sec.\n",
      "\tOn aln 588,000 in seq BACT2. Time spent on BACT2 so far: 1,881.61 sec.\n",
      "\tOn aln 589,000 in seq BACT2. Time spent on BACT2 so far: 1,884.89 sec.\n",
      "\tOn aln 590,000 in seq BACT2. Time spent on BACT2 so far: 1,888.07 sec.\n",
      "\tOn aln 591,000 in seq BACT2. Time spent on BACT2 so far: 1,891.31 sec.\n",
      "\tOn aln 592,000 in seq BACT2. Time spent on BACT2 so far: 1,894.56 sec.\n",
      "\tOn aln 593,000 in seq BACT2. Time spent on BACT2 so far: 1,897.82 sec.\n",
      "\tOn aln 594,000 in seq BACT2. Time spent on BACT2 so far: 1,901.12 sec.\n",
      "\tOn aln 595,000 in seq BACT2. Time spent on BACT2 so far: 1,904.45 sec.\n",
      "\tOn aln 596,000 in seq BACT2. Time spent on BACT2 so far: 1,907.88 sec.\n",
      "\tOn aln 597,000 in seq BACT2. Time spent on BACT2 so far: 1,911.20 sec.\n",
      "\tOn aln 598,000 in seq BACT2. Time spent on BACT2 so far: 1,914.52 sec.\n",
      "\tOn aln 599,000 in seq BACT2. Time spent on BACT2 so far: 1,917.87 sec.\n",
      "\tOn aln 600,000 in seq BACT2. Time spent on BACT2 so far: 1,921.20 sec.\n",
      "\tOn aln 601,000 in seq BACT2. Time spent on BACT2 so far: 1,924.47 sec.\n",
      "\tOn aln 602,000 in seq BACT2. Time spent on BACT2 so far: 1,927.78 sec.\n",
      "\tOn aln 603,000 in seq BACT2. Time spent on BACT2 so far: 1,931.19 sec.\n",
      "\tOn aln 604,000 in seq BACT2. Time spent on BACT2 so far: 1,934.63 sec.\n",
      "\tOn aln 605,000 in seq BACT2. Time spent on BACT2 so far: 1,938.02 sec.\n",
      "\tOn aln 606,000 in seq BACT2. Time spent on BACT2 so far: 1,941.45 sec.\n",
      "\tOn aln 607,000 in seq BACT2. Time spent on BACT2 so far: 1,944.87 sec.\n",
      "\tOn aln 608,000 in seq BACT2. Time spent on BACT2 so far: 1,948.23 sec.\n",
      "\tOn aln 609,000 in seq BACT2. Time spent on BACT2 so far: 1,951.64 sec.\n",
      "\tOn aln 610,000 in seq BACT2. Time spent on BACT2 so far: 1,955.01 sec.\n",
      "\tOn aln 611,000 in seq BACT2. Time spent on BACT2 so far: 1,958.56 sec.\n",
      "\tOn aln 612,000 in seq BACT2. Time spent on BACT2 so far: 1,962.04 sec.\n",
      "\tOn aln 613,000 in seq BACT2. Time spent on BACT2 so far: 1,965.47 sec.\n",
      "\tOn aln 614,000 in seq BACT2. Time spent on BACT2 so far: 1,968.84 sec.\n",
      "\tOn aln 615,000 in seq BACT2. Time spent on BACT2 so far: 1,972.18 sec.\n",
      "\tOn aln 616,000 in seq BACT2. Time spent on BACT2 so far: 1,975.46 sec.\n",
      "\tOn aln 617,000 in seq BACT2. Time spent on BACT2 so far: 1,978.77 sec.\n",
      "\tOn aln 618,000 in seq BACT2. Time spent on BACT2 so far: 1,982.10 sec.\n",
      "\tOn aln 619,000 in seq BACT2. Time spent on BACT2 so far: 1,985.45 sec.\n",
      "\tOn aln 620,000 in seq BACT2. Time spent on BACT2 so far: 1,988.86 sec.\n",
      "\tOn aln 621,000 in seq BACT2. Time spent on BACT2 so far: 1,992.22 sec.\n",
      "\tOn aln 622,000 in seq BACT2. Time spent on BACT2 so far: 1,995.48 sec.\n",
      "\tOn aln 623,000 in seq BACT2. Time spent on BACT2 so far: 1,998.80 sec.\n",
      "\tOn aln 624,000 in seq BACT2. Time spent on BACT2 so far: 2,002.09 sec.\n",
      "\tOn aln 625,000 in seq BACT2. Time spent on BACT2 so far: 2,005.39 sec.\n",
      "\tOn aln 626,000 in seq BACT2. Time spent on BACT2 so far: 2,008.76 sec.\n",
      "\tOn aln 627,000 in seq BACT2. Time spent on BACT2 so far: 2,012.16 sec.\n",
      "\tOn aln 628,000 in seq BACT2. Time spent on BACT2 so far: 2,015.57 sec.\n",
      "\tOn aln 629,000 in seq BACT2. Time spent on BACT2 so far: 2,018.97 sec.\n",
      "\tOn aln 630,000 in seq BACT2. Time spent on BACT2 so far: 2,022.40 sec.\n",
      "\tOn aln 631,000 in seq BACT2. Time spent on BACT2 so far: 2,025.88 sec.\n",
      "\tOn aln 632,000 in seq BACT2. Time spent on BACT2 so far: 2,029.24 sec.\n",
      "\tOn aln 633,000 in seq BACT2. Time spent on BACT2 so far: 2,032.59 sec.\n",
      "\tOn aln 634,000 in seq BACT2. Time spent on BACT2 so far: 2,035.94 sec.\n",
      "\tOn aln 635,000 in seq BACT2. Time spent on BACT2 so far: 2,039.25 sec.\n",
      "\tOn aln 636,000 in seq BACT2. Time spent on BACT2 so far: 2,042.59 sec.\n",
      "\tOn aln 637,000 in seq BACT2. Time spent on BACT2 so far: 2,045.98 sec.\n",
      "\tOn aln 638,000 in seq BACT2. Time spent on BACT2 so far: 2,049.33 sec.\n",
      "\tOn aln 639,000 in seq BACT2. Time spent on BACT2 so far: 2,052.73 sec.\n",
      "\tOn aln 640,000 in seq BACT2. Time spent on BACT2 so far: 2,056.16 sec.\n",
      "\tOn aln 641,000 in seq BACT2. Time spent on BACT2 so far: 2,059.68 sec.\n",
      "\tOn aln 642,000 in seq BACT2. Time spent on BACT2 so far: 2,063.22 sec.\n",
      "\tOn aln 643,000 in seq BACT2. Time spent on BACT2 so far: 2,066.83 sec.\n",
      "\tOn aln 644,000 in seq BACT2. Time spent on BACT2 so far: 2,070.30 sec.\n",
      "\tOn aln 645,000 in seq BACT2. Time spent on BACT2 so far: 2,073.72 sec.\n",
      "\tOn aln 646,000 in seq BACT2. Time spent on BACT2 so far: 2,077.05 sec.\n",
      "\tOn aln 647,000 in seq BACT2. Time spent on BACT2 so far: 2,080.21 sec.\n",
      "\tOn aln 648,000 in seq BACT2. Time spent on BACT2 so far: 2,083.51 sec.\n",
      "\tOn aln 649,000 in seq BACT2. Time spent on BACT2 so far: 2,086.74 sec.\n",
      "\tOn aln 650,000 in seq BACT2. Time spent on BACT2 so far: 2,089.95 sec.\n",
      "\tOn aln 651,000 in seq BACT2. Time spent on BACT2 so far: 2,093.25 sec.\n",
      "\tOn aln 652,000 in seq BACT2. Time spent on BACT2 so far: 2,096.59 sec.\n",
      "\tOn aln 653,000 in seq BACT2. Time spent on BACT2 so far: 2,100.06 sec.\n",
      "\tOn aln 654,000 in seq BACT2. Time spent on BACT2 so far: 2,103.62 sec.\n",
      "\tOn aln 655,000 in seq BACT2. Time spent on BACT2 so far: 2,107.41 sec.\n",
      "\tOn aln 656,000 in seq BACT2. Time spent on BACT2 so far: 2,111.63 sec.\n",
      "\tOn aln 657,000 in seq BACT2. Time spent on BACT2 so far: 2,115.97 sec.\n",
      "\tOn aln 658,000 in seq BACT2. Time spent on BACT2 so far: 2,120.15 sec.\n",
      "\tOn aln 659,000 in seq BACT2. Time spent on BACT2 so far: 2,123.62 sec.\n",
      "\tOn aln 660,000 in seq BACT2. Time spent on BACT2 so far: 2,126.93 sec.\n",
      "\tOn aln 661,000 in seq BACT2. Time spent on BACT2 so far: 2,130.22 sec.\n",
      "\tOn aln 662,000 in seq BACT2. Time spent on BACT2 so far: 2,133.59 sec.\n",
      "\tOn aln 663,000 in seq BACT2. Time spent on BACT2 so far: 2,137.00 sec.\n",
      "\tOn aln 664,000 in seq BACT2. Time spent on BACT2 so far: 2,140.42 sec.\n",
      "\tOn aln 665,000 in seq BACT2. Time spent on BACT2 so far: 2,143.74 sec.\n",
      "\tOn aln 666,000 in seq BACT2. Time spent on BACT2 so far: 2,146.98 sec.\n",
      "\tOn aln 667,000 in seq BACT2. Time spent on BACT2 so far: 2,150.04 sec.\n",
      "\tOn aln 668,000 in seq BACT2. Time spent on BACT2 so far: 2,152.69 sec.\n",
      "\tOn aln 669,000 in seq BACT2. Time spent on BACT2 so far: 2,155.53 sec.\n",
      "\tOn aln 670,000 in seq BACT2. Time spent on BACT2 so far: 2,158.89 sec.\n",
      "\tOn aln 671,000 in seq BACT2. Time spent on BACT2 so far: 2,162.19 sec.\n",
      "\tOn aln 672,000 in seq BACT2. Time spent on BACT2 so far: 2,165.44 sec.\n",
      "\tOn aln 673,000 in seq BACT2. Time spent on BACT2 so far: 2,168.71 sec.\n",
      "\tOn aln 674,000 in seq BACT2. Time spent on BACT2 so far: 2,171.90 sec.\n",
      "\tOn aln 675,000 in seq BACT2. Time spent on BACT2 so far: 2,175.01 sec.\n",
      "\tOn aln 676,000 in seq BACT2. Time spent on BACT2 so far: 2,178.38 sec.\n",
      "\tOn aln 677,000 in seq BACT2. Time spent on BACT2 so far: 2,181.68 sec.\n",
      "\tOn aln 678,000 in seq BACT2. Time spent on BACT2 so far: 2,184.97 sec.\n",
      "\tOn aln 679,000 in seq BACT2. Time spent on BACT2 so far: 2,188.03 sec.\n",
      "\tOn aln 680,000 in seq BACT2. Time spent on BACT2 so far: 2,191.31 sec.\n",
      "\tOn aln 681,000 in seq BACT2. Time spent on BACT2 so far: 2,194.61 sec.\n",
      "\tOn aln 682,000 in seq BACT2. Time spent on BACT2 so far: 2,197.92 sec.\n",
      "\tOn aln 683,000 in seq BACT2. Time spent on BACT2 so far: 2,201.13 sec.\n",
      "\tOn aln 684,000 in seq BACT2. Time spent on BACT2 so far: 2,204.55 sec.\n",
      "\tOn aln 685,000 in seq BACT2. Time spent on BACT2 so far: 2,207.94 sec.\n",
      "\tOn aln 686,000 in seq BACT2. Time spent on BACT2 so far: 2,211.27 sec.\n",
      "\tOn aln 687,000 in seq BACT2. Time spent on BACT2 so far: 2,214.70 sec.\n",
      "\tOn aln 688,000 in seq BACT2. Time spent on BACT2 so far: 2,218.01 sec.\n",
      "\tOn aln 689,000 in seq BACT2. Time spent on BACT2 so far: 2,221.31 sec.\n",
      "\tOn aln 690,000 in seq BACT2. Time spent on BACT2 so far: 2,224.64 sec.\n",
      "\tOn aln 691,000 in seq BACT2. Time spent on BACT2 so far: 2,227.91 sec.\n",
      "\tOn aln 692,000 in seq BACT2. Time spent on BACT2 so far: 2,231.35 sec.\n",
      "\tOn aln 693,000 in seq BACT2. Time spent on BACT2 so far: 2,234.89 sec.\n",
      "\tOn aln 694,000 in seq BACT2. Time spent on BACT2 so far: 2,238.54 sec.\n",
      "\tOn aln 695,000 in seq BACT2. Time spent on BACT2 so far: 2,242.06 sec.\n",
      "\tOn aln 696,000 in seq BACT2. Time spent on BACT2 so far: 2,245.39 sec.\n",
      "\tOn aln 697,000 in seq BACT2. Time spent on BACT2 so far: 2,248.74 sec.\n",
      "\tOn aln 698,000 in seq BACT2. Time spent on BACT2 so far: 2,252.04 sec.\n",
      "\tOn aln 699,000 in seq BACT2. Time spent on BACT2 so far: 2,255.32 sec.\n",
      "\tOn aln 700,000 in seq BACT2. Time spent on BACT2 so far: 2,258.62 sec.\n",
      "\tOn aln 701,000 in seq BACT2. Time spent on BACT2 so far: 2,262.00 sec.\n",
      "\tOn aln 702,000 in seq BACT2. Time spent on BACT2 so far: 2,265.38 sec.\n",
      "\tOn aln 703,000 in seq BACT2. Time spent on BACT2 so far: 2,268.69 sec.\n",
      "\tOn aln 704,000 in seq BACT2. Time spent on BACT2 so far: 2,272.00 sec.\n",
      "\tOn aln 705,000 in seq BACT2. Time spent on BACT2 so far: 2,275.30 sec.\n",
      "\tOn aln 706,000 in seq BACT2. Time spent on BACT2 so far: 2,278.71 sec.\n",
      "\tOn aln 707,000 in seq BACT2. Time spent on BACT2 so far: 2,282.03 sec.\n",
      "\tOn aln 708,000 in seq BACT2. Time spent on BACT2 so far: 2,285.36 sec.\n",
      "\tOn aln 709,000 in seq BACT2. Time spent on BACT2 so far: 2,288.78 sec.\n",
      "\tOn aln 710,000 in seq BACT2. Time spent on BACT2 so far: 2,292.13 sec.\n",
      "\tOn aln 711,000 in seq BACT2. Time spent on BACT2 so far: 2,295.52 sec.\n",
      "\tOn aln 712,000 in seq BACT2. Time spent on BACT2 so far: 2,298.91 sec.\n",
      "\tOn aln 713,000 in seq BACT2. Time spent on BACT2 so far: 2,302.27 sec.\n",
      "\tOn aln 714,000 in seq BACT2. Time spent on BACT2 so far: 2,305.54 sec.\n",
      "\tOn aln 715,000 in seq BACT2. Time spent on BACT2 so far: 2,308.79 sec.\n",
      "\tOn aln 716,000 in seq BACT2. Time spent on BACT2 so far: 2,312.07 sec.\n",
      "\tOn aln 717,000 in seq BACT2. Time spent on BACT2 so far: 2,315.50 sec.\n",
      "\tOn aln 718,000 in seq BACT2. Time spent on BACT2 so far: 2,318.87 sec.\n",
      "\tOn aln 719,000 in seq BACT2. Time spent on BACT2 so far: 2,322.21 sec.\n",
      "\tOn aln 720,000 in seq BACT2. Time spent on BACT2 so far: 2,325.59 sec.\n",
      "\tOn aln 721,000 in seq BACT2. Time spent on BACT2 so far: 2,328.97 sec.\n",
      "\tOn aln 722,000 in seq BACT2. Time spent on BACT2 so far: 2,332.30 sec.\n",
      "\tOn aln 723,000 in seq BACT2. Time spent on BACT2 so far: 2,335.69 sec.\n",
      "\tOn aln 724,000 in seq BACT2. Time spent on BACT2 so far: 2,339.10 sec.\n",
      "\tOn aln 725,000 in seq BACT2. Time spent on BACT2 so far: 2,342.52 sec.\n",
      "\tOn aln 726,000 in seq BACT2. Time spent on BACT2 so far: 2,345.91 sec.\n",
      "\tOn aln 727,000 in seq BACT2. Time spent on BACT2 so far: 2,349.33 sec.\n",
      "\tOn aln 728,000 in seq BACT2. Time spent on BACT2 so far: 2,352.69 sec.\n",
      "\tOn aln 729,000 in seq BACT2. Time spent on BACT2 so far: 2,356.05 sec.\n",
      "\tOn aln 730,000 in seq BACT2. Time spent on BACT2 so far: 2,359.37 sec.\n",
      "\tOn aln 731,000 in seq BACT2. Time spent on BACT2 so far: 2,362.67 sec.\n",
      "\tOn aln 732,000 in seq BACT2. Time spent on BACT2 so far: 2,365.99 sec.\n",
      "\tOn aln 733,000 in seq BACT2. Time spent on BACT2 so far: 2,369.30 sec.\n",
      "\tOn aln 734,000 in seq BACT2. Time spent on BACT2 so far: 2,372.62 sec.\n",
      "\tOn aln 735,000 in seq BACT2. Time spent on BACT2 so far: 2,376.01 sec.\n",
      "\tOn aln 736,000 in seq BACT2. Time spent on BACT2 so far: 2,379.52 sec.\n",
      "\tOn aln 737,000 in seq BACT2. Time spent on BACT2 so far: 2,383.19 sec.\n",
      "\tOn aln 738,000 in seq BACT2. Time spent on BACT2 so far: 2,387.16 sec.\n",
      "\tOn aln 739,000 in seq BACT2. Time spent on BACT2 so far: 2,391.18 sec.\n",
      "\tOn aln 740,000 in seq BACT2. Time spent on BACT2 so far: 2,394.67 sec.\n",
      "\tOn aln 741,000 in seq BACT2. Time spent on BACT2 so far: 2,397.24 sec.\n",
      "\tOn aln 742,000 in seq BACT2. Time spent on BACT2 so far: 2,398.27 sec.\n",
      "Done with edge_2358! Took 2,398.73 sec.\n",
      "Time taken: 5,860.518506765366 sec.\n"
     ]
    }
   ],
   "source": [
    "bf = pysam.AlignmentFile(\"../main-workflow/output/fully-filtered-and-sorted-aln.bam\", \"rb\")\n",
    "output_dir = \"phasing-data/smoothed-reads/\"\n",
    "\n",
    "# verbose?\n",
    "no_indoor_voice = False\n",
    "\n",
    "def write_out_reads(filepath, readname2seq):\n",
    "    # Notably, this uses the \"a\" (append) method in order to add to the end of a file\n",
    "    with open(filepath, \"a\") as of:\n",
    "        for readname in readname2seq:\n",
    "            # Write out both the header and the sequence for each read\n",
    "            of.write(f\">{readname}\\n{str(readname2seq[readname])}\\n\")\n",
    "            \n",
    "ALN_UPDATE_FREQ = 1000\n",
    "ALN_BUFFER_FREQ = 1000\n",
    "            \n",
    "t1 = time.time()\n",
    "for seq in SEQS:\n",
    "    fasta = skbio.DNA.read(f\"../seqs/{seq}.fasta\")\n",
    "    \n",
    "    output_smoothed_reads_file = os.path.join(output_dir, f\"{seq}_smoothed_reads.fasta\")\n",
    "    \n",
    "    # Identify all (0-indexed, so compatible with skbio / pysam!)\n",
    "    # mutated positions in this genome up front to save time.\n",
    "    #\n",
    "    # Equivalently, we could also just take in an arbitrary VCF as input\n",
    "    # (e.g. one produced from another variant calling tool), although we'd\n",
    "    # need to be careful to only include SNVs and not indels/etc...\n",
    "    \n",
    "    print(f\"Identifying mutated positions in genome {seq2name[seq]}...\")\n",
    "    mutated_positions = find_mutated_positions(seq)\n",
    "    print(f\"Found {len(mutated_positions):,} mutated positions in {seq2name[seq]}.\")\n",
    "    \n",
    "    print(\"Going through these positions...\")\n",
    "    \n",
    "    # This should already be implicitly sorted, I think, but the code below relies on mutated_positions being\n",
    "    # in the exact same order as expected. So we may as well be paranoid.\n",
    "    mutated_positions = sorted(mutated_positions)\n",
    "    \n",
    "    # Instead of just writing out every smoothed alignment as soon as we generate it, we build up a \"buffer\"\n",
    "    # of these alignments and then write a bunch out at once. This way we limit slowdown due to constantly\n",
    "    # having to open/close files. I don't really have a good source for this as best practice, but I remembered\n",
    "    # to do it while writing this code, so somewhere in College Park the CS faculty at Maryland are smiling\n",
    "    #\n",
    "    # Also fyi this maps read name to smoothed alignment (well, at this point, just read) sequence. The read name\n",
    "    # is useful to preserve in fasta files so we have some idea of provenance (where smoothed reads came from)\n",
    "    smoothed_aln_buffer = {}\n",
    "    \n",
    "    # The first time we see an alignment of a read, it's 1; if we see a supp aln of this read, it's 2; etc.\n",
    "    # Lets us distinguish alignments with different names\n",
    "    readname2freq_so_far = defaultdict(int)\n",
    "    \n",
    "    # Go through all linear alignments of each read to this genome, focusing (for now) on just the primary\n",
    "    # alignments...\n",
    "    ts1 = time.time()\n",
    "    for ai, aln in enumerate(bf.fetch(seq), 1):\n",
    "        \n",
    "        if ai % ALN_UPDATE_FREQ == 0:\n",
    "            print(\n",
    "                f\"\\tOn aln {ai:,} in seq {seq2name[seq]}. \"\n",
    "                f\"Time spent on {seq2name[seq]} so far: {time.time() - ts1:,.2f} sec.\"\n",
    "            )\n",
    "            \n",
    "        if aln.is_secondary:\n",
    "            raise ValueError(\n",
    "                \"Not to get political or anything, but you should've already filtered secondary alns out\"\n",
    "            )\n",
    "            \n",
    "        # Note that supplementary alignments are ok, though! We implicitly handle these here.\n",
    "        #\n",
    "        # Different alignments of the same read will have different new_readnames, because we're gonna\n",
    "        # be treating them as distinct \"reads\". We should have already filtered reference-overlapping\n",
    "        # supp alns so this shouldn't be a problem\n",
    "        \n",
    "        readname = aln.query_name\n",
    "        readname2freq_so_far[readname] += 1\n",
    "        new_readname = f\"{readname}_{readname2freq_so_far[readname]}\"\n",
    "        \n",
    "        # should never happen\n",
    "        if new_readname in smoothed_aln_buffer:\n",
    "            raise ValueError(\"This exact read alignment has already been smoothed? Weird.\")\n",
    "            \n",
    "        # Figure out where on the MAG this alignment \"hits.\" These are 0-indexed positions from Pysam.\n",
    "        # (reference_end points to the position after the actual final position, since these are designed to\n",
    "        # be interoperable with Python's half-open intervals.)\n",
    "        #\n",
    "        # Of course, there likely will be indels within this range: we're purposefully ignoring those here.\n",
    "        ref_start = aln.reference_start\n",
    "        ref_end = aln.reference_end - 1\n",
    "        \n",
    "        # This should never happen (TM)\n",
    "        if ref_start >= ref_end:\n",
    "            # Du sollst jetzt mit Gott sprechen.\n",
    "            raise ValueError(\n",
    "                f\"Ref start {ref_start:,} >= ref end {ref_end:,} for read {new_readname}?\"\n",
    "            )\n",
    "        \n",
    "        # Smoothed sequence; we'll edit this so that if this read has (mis)matches to any called mutated\n",
    "        # positions, these positions are updated with the read's aligned nucleotides at these positions.\n",
    "        smoothed_aln_seq = fasta[ref_start: ref_end + 1]\n",
    "        \n",
    "        # just for debugging: track the exact edits made to smoothed_aln_seq\n",
    "        replacements_made = {}\n",
    "        \n",
    "        ap = aln.get_aligned_pairs(matches_only=True)\n",
    "        \n",
    "        # Iterating through the aligned pairs is expensive. Since read lengths are generally in the thousands\n",
    "        # to tens of thousands of bp (which is much less than the > 1 million bp length of any bacterial genome),\n",
    "        # we set things up so that we only iterate through the aligned pairs once. We maintain an integer, mpi,\n",
    "        # that is a poor man's \"pointer\" to an index in mutated_positions.\n",
    "        \n",
    "        mpi = 0\n",
    "        \n",
    "        # Go through this aln's aligned pairs. As we see each pair, compare the pair's reference position\n",
    "        # (refpos) to the mpi-th mutated position (herein referred to as \"mutpos\").\n",
    "        #\n",
    "        # If refpos >  mutpos, increment mpi until refpos <= mutpos (stopping as early as possible).\n",
    "        # If refpos == mutpos, we have a match! Update readname2mutpos2ismutated[mutpos] based on\n",
    "        #                      comparing the read to the reference at the aligned positions.\n",
    "        # If refpos <  mutpos, continue to the next pair.\n",
    "        \n",
    "        for pair in ap:\n",
    "            \n",
    "            refpos = pair[1]\n",
    "            mutpos = mutated_positions[mpi]\n",
    "            \n",
    "            no_mutations_to_right_of_here = False\n",
    "            \n",
    "            # Increment mpi until we get to the next mutated position at or after the reference pos for this\n",
    "            # aligned pair (or until we run out of mutated positions).\n",
    "            while refpos > mutpos:\n",
    "                mpi += 1\n",
    "                if mpi < len(mutated_positions):\n",
    "                    mutpos = mutated_positions[mpi]\n",
    "                else:\n",
    "                    no_mutations_to_right_of_here = True\n",
    "                    break\n",
    "            \n",
    "            # I expect this should happen only for reads aligned near the right end of the genome.\n",
    "            if no_mutations_to_right_of_here:\n",
    "                break\n",
    "            \n",
    "            # If the next mutation occurs after this aligned pair, continue on to a later pair.\n",
    "            if refpos < mutpos:\n",
    "                continue\n",
    "                \n",
    "            # If we've made it here, refpos == mutpos!\n",
    "            # (...unless I messed something up in how I designed this code.)\n",
    "            if refpos != mutpos:\n",
    "                raise ValueError(\"This should never happen!\")\n",
    "                \n",
    "            # Finally, get the nucleotide aligned to this mutated position from this read.\n",
    "            readpos = pair[0]\n",
    "            read_nt = aln.query_sequence[readpos]\n",
    "            \n",
    "            ref_nt = str(fasta[mutpos])\n",
    "            # We don't need to do anything if this read already matches the reference MAG at this position\n",
    "            if read_nt == ref_nt:\n",
    "                if no_indoor_voice:\n",
    "                    print(f\"Read {new_readname} matches ref at mutpos {mutpos + 1:,}: both {read_nt}\")\n",
    "            else:\n",
    "                # Record this specific \"allele\" for this read.\n",
    "                relative_pos_on_aln = mutpos - ref_start\n",
    "                smoothed_aln_seq = smoothed_aln_seq.replace([relative_pos_on_aln], read_nt)\n",
    "                replacements_made[relative_pos_on_aln] = read_nt\n",
    "                if no_indoor_voice:\n",
    "                    print(\n",
    "                        f\"Read {new_readname} mismatches ref at mutpos {mutpos + 1:,}: \"\n",
    "                        f\"ref = {ref_nt}, read = {read_nt}\"\n",
    "                    )\n",
    "                \n",
    "        if no_indoor_voice:\n",
    "            print(f\"Read {new_readname} required {len(replacements_made):,} replacements!\")\n",
    "        \n",
    "        # Now that we've finished processing all called mutations that this alignment spans, prepare it\n",
    "        # to be written out to a FASTA file. See comments above on smoothed_aln_buffer, and why we don't\n",
    "        # just write everything out as soon as it's ready.\n",
    "        #\n",
    "        # (Also, we've already guaranteed readname isn't already in smoothed_aln_buffer, so no need to worry\n",
    "        # about accidentally overwriting something from earlier.)\n",
    "        smoothed_aln_buffer[new_readname] = smoothed_aln_seq\n",
    "        \n",
    "        # Notably, we don't necessarily write out *exactly* ALN_BUFFER_FREQ reads at once -- skipping alignments\n",
    "        # due to them being supplementary, etc. (actually no need for an \"etc.\", that's literally the only\n",
    "        # possible reason as of writing, but whatever) doesn't stop ai from going up. Shouldn't make a difference\n",
    "        # unless we have a zillion supplementary alignments.\n",
    "        if ai % ALN_BUFFER_FREQ == 0:\n",
    "            write_out_reads(output_smoothed_reads_file, smoothed_aln_buffer)\n",
    "            # Clear the buffer\n",
    "            smoothed_aln_buffer = {}\n",
    "        \n",
    "    # We're probably going to have left over smoothed reads that we still haven't written out, unless things\n",
    "    # worked out so that on the final alignment we saw ai was exactly divisible by ALN_BUFFER_FREQ (and that's\n",
    "    # pretty unlikely unless you set the buffer freq to a low number). So make one last dump of the buffer.\n",
    "    if len(smoothed_aln_buffer) > 0:\n",
    "        write_out_reads(output_smoothed_reads_file, smoothed_aln_buffer)\n",
    "    \n",
    "    print(f\"Done with {seq}! Took {time.time() - ts1:,.2f} sec.\")\n",
    "        \n",
    "print(f\"Time taken: {time.time() - t1:,} sec.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.5. Stats about smoothed read lengths\n",
    "\n",
    "We could have just figured this out while computing the stuff above, but ... I didn't have the foresight to think of this earlier, and I don't want to rerun that stuff for another >1 hour, so we just loop through the FASTA files we just generated quickly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAMP: 476,356 smoothed reads\n",
      "\tmin / mean / median / max length = 200 / 11,264.7 / 11,154.0 / 35,099\n",
      "\tNum of reads with length ≥ 3,001: 472,064 / 476,356 (99.1%)\n",
      "BACT1: 263,145 smoothed reads\n",
      "\tmin / mean / median / max length = 202 / 11,597.6 / 11,515 / 39,723\n",
      "\tNum of reads with length ≥ 3,001: 260,087 / 263,145 (98.8%)\n",
      "BACT2: 742,926 smoothed reads\n",
      "\tmin / mean / median / max length = 200 / 11,318.0 / 11,259.0 / 40,346\n",
      "\tNum of reads with length ≥ 3,001: 732,240 / 742,926 (98.6%)\n"
     ]
    }
   ],
   "source": [
    "for seq in SEQS:\n",
    "    read_lengths = []\n",
    "    \n",
    "    # Parse a FASTA file -- I stole this code from myself in the Diversity Indices notebook\n",
    "    with open(f\"phasing-data/smoothed-reads/{seq}_smoothed_reads.fasta\", \"r\") as fastafile:\n",
    "\n",
    "        # Assumes that sequences are not split up over multiple lines (so a FASTA file with N sequences\n",
    "        # should have only 2N lines, maybe 2N + 1 if there's an extra empty newline at the bottom of the file)\n",
    "        for linenum, line in enumerate(fastafile):\n",
    "\n",
    "            if line.startswith(\">\"):\n",
    "                if linenum % 2 != 0:\n",
    "                    raise ValueError(\"something weird with > location in all_edges.fasta. Go yell at Marcus.\")\n",
    "            else:\n",
    "                if linenum % 2 != 1:\n",
    "                    raise ValueError(\"something weird with non > location in all_edges.fasta. Go yell at Marcus.\")\n",
    "\n",
    "                read_lengths.append(len(line.strip()))\n",
    "\n",
    "    num_reads = len(read_lengths)\n",
    "    minlen = min(read_lengths)\n",
    "    maxlen = max(read_lengths)\n",
    "    avglen = mean(read_lengths)\n",
    "    medlen = median(read_lengths)\n",
    "    \n",
    "    # We'll use a k-mer size of 1,001 for jumboDBG, and the default window size (w) is 2,000. Reads with length\n",
    "    # less than w + k = 3,001 will be ignored by jumboDB when constructing the graph, so we output stats about\n",
    "    # this to verify that we're not dropping a TON of reads (ideally we wouldn't drop any tho...)\n",
    "    threshold = 3001\n",
    "    geq_threshold = len([rl for rl in read_lengths if rl >= threshold])\n",
    "    pct = 100 * (geq_threshold / num_reads)\n",
    "    \n",
    "    print(f\"{seq2name[seq]}: {num_reads:,} smoothed reads\")\n",
    "    print(f\"\\tmin / mean / median / max length = {minlen:,} / {avglen:,.1f} / {medlen:,} / {maxlen:,}\")\n",
    "    print(f\"\\tNum of reads with length \\u2265 {threshold:,}: {geq_threshold:,} / {num_reads:,} ({pct:.1f}%)\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Run LJA on these smoothed reads\n",
    "\n",
    "More specifically, we just use the jumboDBG module of LJA, since error-correcting reads should (hopefully) not be required after the smoothing process we just did (remember that this was already based on the metaFlye-assembled MAGs).\n",
    "\n",
    "We use $k = 1{,}001$ and $w = 2{,}000$ (this is the default window size of jumboDBG as of writing), and also use `--coverage` (since this information could be useful when identifying low-coverage strains)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "00:00:00 162Mb  INFO: Hello! You are running jumboDBG, a tool for construction of de Bruijn graphs for arbitrarily large values of k\n",
      "00:00:00 162Mb  INFO: Note that jumboDBG does not perform any error correction and ignores all reads shorter than k + w = 3001\n",
      "00:00:00 0Mb  INFO: Reading reads\n",
      "00:00:00 0Mb  INFO: Extracting minimizers\n",
      "00:00:41 2.3Gb  INFO: Finished read processing\n",
      "00:00:41 2.3Gb  INFO: 4346161 hashs collected. Starting sorting.\n",
      "00:00:42 2.3Gb  INFO: Finished sorting. Total distinct minimizers: 1834\n",
      "00:00:42 2.3Gb  INFO: Starting construction of sparse de Bruijn graph\n",
      "00:00:42 2.3Gb  INFO: Vertex map constructed.\n",
      "00:00:42 2.3Gb  INFO: Filling edge sequences.\n",
      "00:01:20 3.5Gb  INFO: Finished sparse de Bruijn graph construction.\n",
      "00:01:20 3.5Gb  INFO:  Collecting tips \n",
      "00:01:20 3.5Gb  INFO: Added 53 artificial minimizers from tips.\n",
      "00:01:20 3.5Gb  INFO: Collected 4628 old edges.\n",
      "00:01:20 3.5Gb  INFO: New minimizers added to sparse graph.\n",
      "00:01:20 3.5Gb  INFO: Refilling graph with old edges.\n",
      "00:01:20 3.5Gb  INFO: Filling graph with new edges.\n",
      "00:01:20 3.5Gb  INFO: Finished fixing sparse de Bruijn graph.\n",
      "00:01:20 3.5Gb  INFO: Starting to extract disjointigs.\n",
      "00:01:20 3.5Gb  INFO: Finished extracting 854 disjointigs of total size 2642791\n",
      "00:01:20 162Mb  INFO: Loading disjointigs from file \"phasing-data/smoothed-reads/edge_6104_jumbodbg/disjointigs.fasta\"\n",
      "00:01:20 162Mb  INFO: Filling bloom filter with k+1-mers.\n",
      "00:01:21 162Mb  INFO: Filled 8242800 bits out of 57213984\n",
      "00:01:21 162Mb  INFO: Finished filling bloom filter. Selecting junctions.\n",
      "00:01:21 162Mb  INFO: Collected 1511 junctions.\n",
      "00:01:21 162Mb  INFO: Starting DBG construction.\n",
      "00:01:21 162Mb  INFO: Vertices created.\n",
      "00:01:21 162Mb  INFO: Filled dbg edges. Adding hanging vertices \n",
      "00:01:21 162Mb  INFO: Added 0 hanging vertices\n",
      "00:01:21 162Mb  INFO: Merging unbranching paths\n",
      "00:01:21 162Mb  INFO: Ended merging edges. Resulting size 812\n",
      "00:01:21 162Mb  INFO: Calculating edge coverage.\n",
      "00:01:21 162Mb  INFO: Starting to fill edge coverages\n",
      "00:01:37 4.3Gb  INFO: Edge coverage calculated.\n",
      "00:01:37 4.3Gb  INFO: Printing graph to fasta file \"phasing-data/smoothed-reads/edge_6104_jumbodbg/graph.fasta\"\n",
      "00:01:37 4.3Gb  INFO: Printing graph to gfa file \"phasing-data/smoothed-reads/edge_6104_jumbodbg/graph.gfa\"\n",
      "00:01:37 4.3Gb  INFO: Printing graph to dot file \"phasing-data/smoothed-reads/edge_6104_jumbodbg/graph.dot\"\n",
      "00:01:37 4.3Gb  INFO: DBG construction finished\n",
      "00:01:37 4.3Gb  INFO: Please cite our paper if you use jumboDBG in your research: https://www.biorxiv.org/content/10.1101/2020.12.10.420448\n",
      "00:00:00 162Mb  INFO: Hello! You are running jumboDBG, a tool for construction of de Bruijn graphs for arbitrarily large values of k\n",
      "00:00:00 162Mb  INFO: Note that jumboDBG does not perform any error correction and ignores all reads shorter than k + w = 3001\n",
      "00:00:00 0Mb  INFO: Reading reads\n",
      "00:00:00 0Mb  INFO: Extracting minimizers\n",
      "00:00:25 3.1Gb  INFO: Finished read processing\n",
      "00:00:25 3.1Gb  INFO: 2504810 hashs collected. Starting sorting.\n",
      "00:00:26 3.1Gb  INFO: Finished sorting. Total distinct minimizers: 20360\n",
      "00:00:26 3.1Gb  INFO: Starting construction of sparse de Bruijn graph\n",
      "00:00:26 3.1Gb  INFO: Vertex map constructed.\n",
      "00:00:26 3.1Gb  INFO: Filling edge sequences.\n",
      "00:00:52 3.1Gb  INFO: Finished sparse de Bruijn graph construction.\n",
      "00:00:52 3.1Gb  INFO:  Collecting tips \n",
      "00:00:52 3.1Gb  INFO: Added 3382 artificial minimizers from tips.\n",
      "00:00:52 3.1Gb  INFO: Collected 61608 old edges.\n",
      "00:00:52 3.1Gb  INFO: New minimizers added to sparse graph.\n",
      "00:00:52 3.1Gb  INFO: Refilling graph with old edges.\n",
      "00:00:53 3.1Gb  INFO: Filling graph with new edges.\n",
      "00:00:53 3.1Gb  INFO: Finished fixing sparse de Bruijn graph.\n",
      "00:00:53 3.1Gb  INFO: Starting to extract disjointigs.\n",
      "00:00:53 3.1Gb  INFO: Finished extracting 22337 disjointigs of total size 42522026\n",
      "00:00:54 162Mb  INFO: Loading disjointigs from file \"phasing-data/smoothed-reads/edge_1671_jumbodbg/disjointigs.fasta\"\n",
      "00:00:54 162Mb  INFO: Filling bloom filter with k+1-mers.\n",
      "00:01:00 162Mb  INFO: Filled 89659188 bits out of 645206048\n",
      "00:01:00 162Mb  INFO: Finished filling bloom filter. Selecting junctions.\n",
      "00:01:02 162Mb  INFO: Collected 33986 junctions.\n",
      "00:01:02 162Mb  INFO: Starting DBG construction.\n",
      "00:01:02 162Mb  INFO: Vertices created.\n",
      "00:01:02 162Mb  INFO: Filled dbg edges. Adding hanging vertices \n",
      "00:01:02 162Mb  INFO: Added 0 hanging vertices\n",
      "00:01:02 162Mb  INFO: Merging unbranching paths\n",
      "00:01:02 162Mb  INFO: Ended merging edges. Resulting size 28016\n",
      "00:01:02 162Mb  INFO: Calculating edge coverage.\n",
      "00:01:02 162Mb  INFO: Starting to fill edge coverages\n",
      "00:01:09 2.2Gb  INFO: Edge coverage calculated.\n",
      "00:01:09 2.2Gb  INFO: Printing graph to fasta file \"phasing-data/smoothed-reads/edge_1671_jumbodbg/graph.fasta\"\n",
      "00:01:11 2.2Gb  INFO: Printing graph to gfa file \"phasing-data/smoothed-reads/edge_1671_jumbodbg/graph.gfa\"\n",
      "00:01:12 2.2Gb  INFO: Printing graph to dot file \"phasing-data/smoothed-reads/edge_1671_jumbodbg/graph.dot\"\n",
      "00:01:12 2.2Gb  INFO: DBG construction finished\n",
      "00:01:12 2.2Gb  INFO: Please cite our paper if you use jumboDBG in your research: https://www.biorxiv.org/content/10.1101/2020.12.10.420448\n",
      "00:00:00 162Mb  INFO: Hello! You are running jumboDBG, a tool for construction of de Bruijn graphs for arbitrarily large values of k\n",
      "00:00:00 162Mb  INFO: Note that jumboDBG does not perform any error correction and ignores all reads shorter than k + w = 3001\n",
      "00:00:00 0Mb  INFO: Reading reads\n",
      "00:00:00 0Mb  INFO: Extracting minimizers\n",
      "00:01:07 4.5Gb  INFO: Finished read processing\n",
      "00:01:07 4.5Gb  INFO: 6836697 hashs collected. Starting sorting.\n",
      "00:01:07 4.5Gb  INFO: Finished sorting. Total distinct minimizers: 5810\n",
      "00:01:07 4.5Gb  INFO: Starting construction of sparse de Bruijn graph\n",
      "00:01:07 4.5Gb  INFO: Vertex map constructed.\n",
      "00:01:07 4.5Gb  INFO: Filling edge sequences.\n",
      "00:02:11 4.5Gb  INFO: Finished sparse de Bruijn graph construction.\n",
      "00:02:11 4.5Gb  INFO:  Collecting tips \n",
      "00:02:11 4.5Gb  INFO: Added 257 artificial minimizers from tips.\n",
      "00:02:11 4.5Gb  INFO: Collected 16826 old edges.\n",
      "00:02:11 4.5Gb  INFO: New minimizers added to sparse graph.\n",
      "00:02:11 4.5Gb  INFO: Refilling graph with old edges.\n",
      "00:02:12 4.5Gb  INFO: Filling graph with new edges.\n",
      "00:02:12 4.5Gb  INFO: Finished fixing sparse de Bruijn graph.\n",
      "00:02:12 4.5Gb  INFO: Starting to extract disjointigs.\n",
      "00:02:12 4.5Gb  INFO: Finished extracting 4612 disjointigs of total size 10109612\n",
      "00:02:12 162Mb  INFO: Loading disjointigs from file \"phasing-data/smoothed-reads/edge_2358_jumbodbg/disjointigs.fasta\"\n",
      "00:02:12 162Mb  INFO: Filling bloom filter with k+1-mers.\n",
      "00:02:13 162Mb  INFO: Filled 25080482 bits out of 175776000\n",
      "00:02:13 162Mb  INFO: Finished filling bloom filter. Selecting junctions.\n",
      "00:02:14 162Mb  INFO: Collected 6101 junctions.\n",
      "00:02:14 162Mb  INFO: Starting DBG construction.\n",
      "00:02:14 162Mb  INFO: Vertices created.\n",
      "00:02:14 162Mb  INFO: Filled dbg edges. Adding hanging vertices \n",
      "00:02:14 162Mb  INFO: Added 0 hanging vertices\n",
      "00:02:14 162Mb  INFO: Merging unbranching paths\n",
      "00:02:14 162Mb  INFO: Ended merging edges. Resulting size 4192\n",
      "00:02:14 162Mb  INFO: Calculating edge coverage.\n",
      "00:02:14 162Mb  INFO: Starting to fill edge coverages\n",
      "00:02:31 4Gb  INFO: Edge coverage calculated.\n",
      "00:02:32 4Gb  INFO: Printing graph to fasta file \"phasing-data/smoothed-reads/edge_2358_jumbodbg/graph.fasta\"\n",
      "00:02:32 4Gb  INFO: Printing graph to gfa file \"phasing-data/smoothed-reads/edge_2358_jumbodbg/graph.gfa\"\n",
      "00:02:32 4Gb  INFO: Printing graph to dot file \"phasing-data/smoothed-reads/edge_2358_jumbodbg/graph.dot\"\n",
      "00:02:32 4Gb  INFO: DBG construction finished\n",
      "00:02:32 4Gb  INFO: Please cite our paper if you use jumboDBG in your research: https://www.biorxiv.org/content/10.1101/2020.12.10.420448\n"
     ]
    }
   ],
   "source": [
    "!/home/mfedarko/software/LJA/bin/jumboDBG \\\n",
    "    --reads phasing-data/smoothed-reads/edge_6104_smoothed_reads.fasta \\\n",
    "    -k 1001 \\\n",
    "    --coverage \\\n",
    "    --output-dir phasing-data/smoothed-reads/edge_6104_jumbodbg\n",
    "\n",
    "!/home/mfedarko/software/LJA/bin/jumboDBG \\\n",
    "    --reads phasing-data/smoothed-reads/edge_1671_smoothed_reads.fasta \\\n",
    "    -k 1001 \\\n",
    "    --coverage \\\n",
    "    --output-dir phasing-data/smoothed-reads/edge_1671_jumbodbg\n",
    "\n",
    "!/home/mfedarko/software/LJA/bin/jumboDBG \\\n",
    "    --reads phasing-data/smoothed-reads/edge_2358_smoothed_reads.fasta \\\n",
    "    -k 1001 \\\n",
    "    --coverage \\\n",
    "    --output-dir phasing-data/smoothed-reads/edge_2358_jumbodbg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-01-27T08:55:41.781424Z",
     "iopub.status.busy": "2022-01-27T08:55:41.768139Z",
     "iopub.status.idle": "2022-01-27T09:29:48.238562Z",
     "shell.execute_reply": "2022-01-27T09:29:48.239237Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "00:00:00 351Mb  INFO: Hello! You are running La Jolla Assembler (LJA), a tool for genome assembly from PacBio HiFi reads\r\n",
      "00:00:00 351Mb  INFO: 0a4619b61b5f12d9bed1a84784942003687a8bda\r\n",
      "00:00:00 351Mb  INFO: LJA pipeline started\r\n",
      "00:00:00 351Mb  INFO: Performing initial correction with k = 501\r\n",
      "00:00:00 0Mb  INFO: Reading reads\r\n",
      "00:00:00 0Mb  INFO: Extracting minimizers\r\n",
      "00:00:33 3.3Gb  INFO: Finished read processing\r\n",
      "00:00:33 3.3Gb  INFO: 3056952 hashs collected. Starting sorting.\r\n",
      "00:00:33 3.3Gb  INFO: Finished sorting. Total distinct minimizers: 1092\r\n",
      "00:00:33 3.3Gb  INFO: Starting construction of sparse de Bruijn graph\r\n",
      "00:00:33 3.3Gb  INFO: Vertex map constructed.\r\n",
      "00:00:33 3.3Gb  INFO: Filling edge sequences.\r\n",
      "00:01:04 4.5Gb  INFO: Finished sparse de Bruijn graph construction.\r\n",
      "00:01:04 4.5Gb  INFO:  Collecting tips \r\n",
      "00:01:04 4.5Gb  INFO: Added 37 artificial minimizers from tips.\r\n",
      "00:01:04 4.5Gb  INFO: Collected 2716 old edges.\r\n",
      "00:01:04 4.5Gb  INFO: New minimizers added to sparse graph.\r\n",
      "00:01:04 4.5Gb  INFO: Refilling graph with old edges.\r\n",
      "00:01:04 4.5Gb  INFO: Filling graph with new edges.\r\n",
      "00:01:04 4.5Gb  INFO: Finished fixing sparse de Bruijn graph.\r\n",
      "00:01:04 4.5Gb  INFO: Starting to extract disjointigs.\r\n",
      "00:01:04 4.5Gb  INFO: Finished extracting 538 disjointigs of total size 1309348\r\n",
      "00:01:04 0Mb  INFO: Loading disjointigs from file \"phasing-data/smoothed-reads/edge_6104_lja/k501/disjointigs.fasta\"\r\n",
      "00:01:04 8Mb  INFO: Filling bloom filter with k+1-mers.\r\n",
      "00:01:04 8Mb  INFO: Filled 4805048 bits out of 33273920\r\n",
      "00:01:04 8Mb  INFO: Finished filling bloom filter. Selecting junctions.\r\n",
      "00:01:05 8Mb  INFO: Collected 972 junctions.\r\n",
      "00:01:05 8Mb  INFO: Starting DBG construction.\r\n",
      "00:01:05 8Mb  INFO: Vertices created.\r\n",
      "00:01:05 8Mb  INFO: Filled dbg edges. Adding hanging vertices \r\n",
      "00:01:05 8Mb  INFO: Added 0 hanging vertices\r\n",
      "00:01:05 8Mb  INFO: Merging unbranching paths\r\n",
      "00:01:05 8Mb  INFO: Ended merging edges. Resulting size 563\r\n",
      "00:01:05 8Mb  INFO: Cleaning edge coverages\r\n",
      "00:01:05 8Mb  INFO: Collecting alignments of sequences to the graph\r\n",
      "00:01:22 4.6Gb  INFO: Alignment collection finished. Total length of alignments is 2039543\r\n",
      "00:01:22 4.6Gb  INFO: Precorrecting reads\r\n",
      "00:01:22 4.6Gb  INFO: Applying corrections to reads\r\n",
      "00:01:22 4.6Gb  INFO: Applied correction to 54 reads\r\n",
      "00:01:22 4.6Gb  INFO: Corrected simple errors in 54 reads\r\n",
      "00:01:22 4.6Gb  INFO: Applying changes to the graph\r\n",
      "00:01:24 4.8Gb  INFO: Collecting and storing read suffixes\r\n",
      "00:01:24 4.8Gb  INFO: Correcting dinucleotide errors in reads\r\n",
      "00:01:25 4.8Gb  INFO: Applying corrections to reads\r\n",
      "00:01:25 4.8Gb  INFO: Applied correction to 3206 reads\r\n",
      "00:01:25 4.8Gb  INFO: Corrected 3206 dinucleotide sequences\r\n",
      "00:01:25 4.8Gb  INFO: Marking reliable edges\r\n",
      "00:01:25 4.8Gb  INFO: Marked 8 edges in 2 paths as reliable\r\n",
      "00:01:25 4.8Gb  INFO: Correcting low covered regions in reads with K = 800\r\n",
      "00:01:25 4.8Gb  INFO: Applying corrections to reads\r\n",
      "00:01:25 4.8Gb  INFO: Applied correction to 120 reads\r\n",
      "00:01:25 4.8Gb  INFO: Corrected low covered regions in 120 reads with K = 800\r\n",
      "00:01:25 4.8Gb  INFO: Applying changes to the graph\r\n",
      "00:01:27 4.8Gb  INFO: Marking reliable edges\r\n",
      "00:01:27 4.8Gb  INFO: Marked 6 edges in 2 paths as reliable\r\n",
      "00:01:27 4.8Gb  INFO: Correcting low covered regions in reads with K = 2000\r\n",
      "00:01:27 4.8Gb  INFO: Applying corrections to reads\r\n",
      "00:01:27 4.8Gb  INFO: Applied correction to 2 reads\r\n",
      "00:01:27 4.8Gb  INFO: Corrected low covered regions in 2 reads with K = 2000\r\n",
      "00:01:27 4.8Gb  INFO: Applying changes to the graph\r\n",
      "00:01:30 4.8Gb  INFO: Correcting dinucleotide errors in reads\r\n",
      "00:01:30 4.8Gb  INFO: Applying corrections to reads\r\n",
      "00:01:30 4.8Gb  INFO: Applied correction to 0 reads\r\n",
      "00:01:30 4.8Gb  INFO: Corrected 0 dinucleotide sequences\r\n",
      "00:01:30 4.8Gb  INFO: Marking reliable edges\r\n",
      "00:01:30 4.8Gb  INFO: Marked 6 edges in 2 paths as reliable\r\n",
      "00:01:30 4.8Gb  INFO: Correcting low covered regions in reads\r\n",
      "00:01:30 4.8Gb  INFO: Applying corrections to reads\r\n",
      "00:01:30 4.8Gb  INFO: Applied correction to 13 reads\r\n",
      "00:01:30 4.8Gb  INFO: Corrected low covered regions in 25 reads\r\n",
      "00:01:30 4.8Gb  INFO: Marking reliable edges\r\n",
      "00:01:30 4.8Gb  INFO: Marked 0 edges in 0 paths as reliable\r\n",
      "00:01:30 4.8Gb  INFO: Correcting low covered regions in reads with K = 3500\r\n",
      "00:01:31 4.8Gb  INFO: Applying corrections to reads\r\n",
      "00:01:31 4.8Gb  INFO: Applied correction to 2 reads\r\n",
      "00:01:31 4.8Gb  INFO: Corrected low covered regions in 2 reads with K = 3500\r\n",
      "00:01:31 4.8Gb  INFO: Applying changes to the graph\r\n",
      "00:01:33 4.8Gb  INFO: Printing reads to fasta file \"phasing-data/smoothed-reads/edge_6104_lja/k501/corrected.fasta\"\r\n",
      "00:02:09 351Mb  INFO: Initial correction results with k = 501 printed to \"phasing-data/smoothed-reads/edge_6104_lja/k501/corrected.fasta\"\r\n",
      "00:02:09 351Mb  INFO: Performing second phase of error correction using k = 5001\r\n",
      "00:02:09 0Mb  INFO: Reading reads\r\n",
      "00:02:09 0Mb  INFO: Extracting minimizers\r\n",
      "00:02:26 3.8Gb  INFO: Finished read processing\r\n",
      "00:02:26 3.8Gb  INFO: 5260183 hashs collected. Starting sorting.\r\n",
      "00:02:26 3.8Gb  INFO: Finished sorting. Total distinct minimizers: 6000\r\n",
      "00:02:26 3.8Gb  INFO: Starting construction of sparse de Bruijn graph\r\n",
      "00:02:26 3.8Gb  INFO: Vertex map constructed.\r\n",
      "00:02:26 3.8Gb  INFO: Filling edge sequences.\r\n",
      "00:02:42 4Gb  INFO: Finished sparse de Bruijn graph construction.\r\n",
      "00:02:42 4Gb  INFO:  Collecting tips \r\n",
      "00:02:42 4Gb  INFO: Added 71 artificial minimizers from tips.\r\n",
      "00:02:42 4Gb  INFO: Collected 12182 old edges.\r\n",
      "00:02:42 4Gb  INFO: New minimizers added to sparse graph.\r\n",
      "00:02:42 4Gb  INFO: Refilling graph with old edges.\r\n",
      "00:02:42 4Gb  INFO: Filling graph with new edges.\r\n",
      "00:02:42 4Gb  INFO: Finished fixing sparse de Bruijn graph.\r\n",
      "00:02:42 4Gb  INFO: Starting to extract disjointigs.\r\n",
      "00:02:42 4Gb  INFO: Finished extracting 372 disjointigs of total size 3368634\r\n",
      "00:02:42 0Mb  INFO: Loading disjointigs from file \"phasing-data/smoothed-reads/edge_6104_lja/k5001/disjointigs.fasta\"\r\n",
      "00:02:42 10Mb  INFO: Filling bloom filter with k+1-mers.\r\n",
      "00:02:43 10Mb  INFO: Filled 6981587 bits out of 48264384\r\n",
      "00:02:43 10Mb  INFO: Finished filling bloom filter. Selecting junctions.\r\n",
      "00:02:43 10Mb  INFO: Collected 878 junctions.\r\n",
      "00:02:43 10Mb  INFO: Starting DBG construction.\r\n",
      "00:02:43 10Mb  INFO: Vertices created.\r\n",
      "00:02:43 10Mb  INFO: Filled dbg edges. Adding hanging vertices \r\n",
      "00:02:43 10Mb  INFO: Added 0 hanging vertices\r\n",
      "00:02:43 10Mb  INFO: Merging unbranching paths\r\n",
      "00:02:43 10Mb  INFO: Ended merging edges. Resulting size 316\r\n",
      "00:02:43 10Mb  INFO: Cleaning edge coverages\r\n",
      "00:02:43 10Mb  INFO: Collecting alignments of sequences to the graph\r\n",
      "00:02:43 10Mb  INFO: Storing suffixes of read paths of length up to 10000000\r\n",
      "00:02:54 3.4Gb  INFO: Alignment collection finished. Total length of alignments is 731775\r\n",
      "00:02:54 3.4Gb  INFO: Correcting dinucleotide errors in reads\r\n",
      "00:02:54 3.4Gb  INFO: Applying corrections to reads\r\n",
      "00:02:54 3.4Gb  INFO: Applied correction to 0 reads\r\n",
      "00:02:54 3.4Gb  INFO: Corrected 0 dinucleotide sequences\r\n",
      "00:02:54 3.4Gb  INFO: Marking reliable edges\r\n",
      "00:02:54 3.4Gb  INFO: Marked 2 edges in 2 paths as reliable\r\n",
      "00:02:54 3.4Gb  INFO: Correcting low covered regions in reads\r\n",
      "00:02:55 3.5Gb  INFO: Applying corrections to reads\r\n",
      "00:02:55 3.5Gb  INFO: Applied correction to 72 reads\r\n",
      "00:02:55 3.5Gb  INFO: Corrected low covered regions in 72 reads\r\n",
      "00:02:55 3.5Gb  INFO: Collapsing bulges\r\n",
      "00:02:55 3.5Gb  INFO: Applying corrections to reads\r\n",
      "00:02:55 3.5Gb  INFO: Applied correction to 0 reads\r\n",
      "00:02:55 3.5Gb  INFO: Collapsed bulges in 54 reads\r\n",
      "00:02:55 3.5Gb  INFO: Applying changes to the graph\r\n",
      "00:02:57 3.5Gb  INFO: Running second round of error correction\r\n",
      "00:02:57 3.5Gb  INFO: Correcting dinucleotide errors in reads\r\n",
      "00:02:57 3.5Gb  INFO: Applying corrections to reads\r\n",
      "00:02:57 3.5Gb  INFO: Applied correction to 0 reads\r\n",
      "00:02:57 3.5Gb  INFO: Corrected 0 dinucleotide sequences\r\n",
      "00:02:57 3.5Gb  INFO: Correcting dinucleotide errors in reads\r\n",
      "00:02:57 3.5Gb  INFO: Applying corrections to reads\r\n",
      "00:02:57 3.5Gb  INFO: Applied correction to 0 reads\r\n",
      "00:02:57 3.5Gb  INFO: Corrected 0 dinucleotide sequences\r\n",
      "00:02:57 3.5Gb  INFO: Marking reliable edges\r\n",
      "00:02:57 3.5Gb  INFO: Marked 0 edges in 0 paths as reliable\r\n",
      "00:02:57 3.5Gb  INFO: Correcting low covered regions in reads\r\n",
      "00:02:57 3.6Gb  INFO: Applying corrections to reads\r\n",
      "00:02:57 3.6Gb  INFO: Applied correction to 0 reads\r\n",
      "00:02:57 3.6Gb  INFO: Corrected low covered regions in 0 reads\r\n",
      "00:02:57 3.6Gb  INFO: Correcting dinucleotide errors in reads\r\n",
      "00:02:57 3.6Gb  INFO: Applying corrections to reads\r\n",
      "00:02:57 3.6Gb  INFO: Applied correction to 0 reads\r\n",
      "00:02:57 3.6Gb  INFO: Corrected 0 dinucleotide sequences\r\n",
      "00:02:57 3.6Gb  INFO: Remarking reliable edges\r\n",
      "00:02:57 3.6Gb  INFO: Correcting tips using reliable edge marks\r\n",
      "00:02:58 3.6Gb  INFO: Applying corrections to reads\r\n",
      "00:02:58 3.6Gb  INFO: Applied correction to 2 reads\r\n",
      "00:02:58 3.6Gb  INFO: Collapsing bulges\r\n",
      "00:02:58 3.6Gb  INFO: Applying corrections to reads\r\n",
      "00:02:58 3.6Gb  INFO: Applied correction to 0 reads\r\n",
      "00:02:58 3.6Gb  INFO: Collapsed bulges in 60 reads\r\n",
      "00:02:58 3.6Gb  INFO: Applying changes to the graph\r\n",
      "00:03:00 3.7Gb  INFO: Started gap closing procedure\r\n",
      "00:03:00 3.7Gb  INFO: Found 0 potential overlaps. Aligning.\r\n",
      "00:03:00 3.7Gb  INFO: Collected 0 overlaps. Looking for unique overlaps\r\n",
      "00:03:00 3.7Gb  INFO: Collected 0 unique overlaps.\r\n",
      "00:03:00 3.7Gb  INFO: Could not correct 40 reads. They will be removed.\r\n",
      "00:03:00 3.7Gb  INFO: Uncorrected reads were removed.\r\n",
      "00:03:00 3.7Gb  INFO: Applying changes to the graph\r\n",
      "00:03:02 3.7Gb  INFO: Looking for unique edges\r\n",
      "00:03:02 3.7Gb  INFO: Marked 5 long edges as unique\r\n",
      "00:03:02 3.7Gb  INFO: Marking extra edges as unique based on read paths\r\n",
      "00:03:02 3.7Gb  INFO: Marked 5 edges as unique\r\n",
      "00:03:02 3.7Gb  INFO: Splitting graph with unique edges\r\n",
      "00:03:02 3.7Gb  INFO: Processing 5 components\r\n",
      "00:03:02 3.7Gb  INFO: Finished unique edges search. Found 5 unique edges\r\n",
      "00:03:02 3.7Gb  INFO: Analysing repeats of multiplicity 2 and looking for additional unique edges\r\n",
      "00:03:02 3.7Gb  INFO: Finished processing of repeats of multiplicity 2. Found 0 erroneous edges.\r\n",
      "00:03:02 3.7Gb  INFO: Correcting reads using unique edge extensions\r\n",
      "00:03:03 3.7Gb  INFO: Applying corrections to reads\r\n",
      "00:03:03 3.7Gb  INFO: Applied correction to 88 reads\r\n",
      "00:03:03 3.7Gb  INFO: Collecting bad edges\r\n",
      "00:03:03 3.7Gb  INFO: Removed 5 disconnected edges\r\n",
      "00:03:03 3.7Gb  INFO: Could not correct 59 reads. They will be removed.\r\n",
      "00:03:03 3.7Gb  INFO: Uncorrected reads were removed.\r\n",
      "00:03:03 3.7Gb  INFO: Looking for more unique edges\r\n",
      "00:03:03 3.7Gb  INFO: Finished unique edges search. Found 9 unique edges\r\n",
      "00:03:03 3.7Gb  INFO: Looking for more unique edges\r\n",
      "00:03:03 3.7Gb  INFO: Finished unique edges search. Found 20 unique edges\r\n",
      "00:03:03 3.7Gb  INFO: Correcting reads using unique edge extensions\r\n",
      "00:03:03 3.7Gb  INFO: Applying corrections to reads\r\n",
      "00:03:03 3.7Gb  INFO: Applied correction to 3082 reads\r\n",
      "00:03:03 3.7Gb  INFO: Collecting bad edges\r\n",
      "00:03:03 3.7Gb  INFO: Removed 11 disconnected edges\r\n",
      "00:03:03 3.7Gb  INFO: Could not correct 952 reads. They will be removed.\r\n",
      "00:03:03 3.7Gb  INFO: Uncorrected reads were removed.\r\n",
      "00:03:03 3.7Gb  INFO: Attempting to rescue small circular highly covered components\r\n",
      "00:03:03 3.7Gb  INFO: Could not correct 211 reads. They will be removed.\r\n",
      "00:03:03 3.7Gb  INFO: Uncorrected reads were removed.\r\n",
      "00:03:03 3.7Gb  INFO: Rescued 1 circular highly covered components\r\n",
      "00:03:03 3.7Gb  INFO: Applying changes to the graph\r\n",
      "00:03:06 3.7Gb  INFO: Started gap closing procedure\r\n",
      "00:03:06 3.7Gb  INFO: Found 0 potential overlaps. Aligning.\r\n",
      "00:03:06 3.7Gb  INFO: Collected 0 overlaps. Looking for unique overlaps\r\n",
      "00:03:06 3.7Gb  INFO: Collected 0 unique overlaps.\r\n",
      "00:03:08 3.7Gb  INFO: Printing reads to fasta file \"phasing-data/smoothed-reads/edge_6104_lja/k5001/corrected_reads.fasta\"\r\n",
      "00:03:40 351Mb  INFO: Second phase results with k = 5001 printed to \"phasing-data/smoothed-reads/edge_6104_lja/k5001/corrected_reads.fasta\"\r\n",
      "00:03:40 351Mb  INFO: Performing repeat resolution by transforming de Bruijn graph into Multiplex de Bruijn graph\r\n",
      "00:03:40 0Mb  INFO: Loading graph from fasta\r\n",
      "00:03:40 10Mb  INFO: Finished loading graph\r\n",
      "00:03:41 157Mb  INFO: Looking for unique edges\r\n",
      "00:03:41 157Mb  INFO: Marked 8 long edges as unique\r\n",
      "00:03:41 157Mb  INFO: Marking extra edges as unique based on read paths\r\n",
      "00:03:41 157Mb  INFO: Marked 8 edges as unique\r\n",
      "00:03:41 157Mb  INFO: Splitting graph with unique edges\r\n",
      "00:03:41 157Mb  INFO: Processing 6 components\r\n",
      "00:03:41 157Mb  INFO: Finished unique edges search. Found 8 unique edges\r\n",
      "00:03:41 157Mb  INFO: Analysing repeats of multiplicity 2 and looking for additional unique edges\r\n",
      "00:03:41 157Mb  INFO: Finished processing of repeats of multiplicity 2. Found 0 erroneous edges.\r\n",
      "00:03:41 157Mb  INFO: Resolving repeats\r\n",
      "00:03:41 157Mb  INFO: Constructing paths\r\n",
      "00:03:46 353Mb  INFO: Building graph\r\n",
      "00:03:46 353Mb  INFO: Increasing k\r\n",
      "00:03:47 353Mb  INFO: Finished increasing k\r\n",
      "00:03:47 353Mb  INFO: Exporting remaining active transitions\r\n",
      "00:03:47 353Mb  INFO: Export to Dot\r\n",
      "00:03:47 353Mb  INFO: Export to GFA and compressed contigs\r\n",
      "00:03:47 358Mb  INFO: Finished repeat resolution\r\n",
      "00:03:48 351Mb  INFO: Performing polishing and homopolymer uncompression\r\n",
      "00:03:48 8Mb  INFO: Aligning reads back to assembly\r\n",
      "00:04:10 4.6Gb  INFO: Finished alignment.\r\n",
      "00:04:10 4.6Gb  INFO: Printing alignments to \"phasing-data/smoothed-reads/edge_6104_lja/uncompressing/alignments.txt\"\r\n",
      "00:04:12 4.6Gb  INFO: Reading and processing initial reads from [\"phasing-data/smoothed-reads/edge_6104_smoothed_reads.fasta\"]\r\n",
      "00:08:49 4.6Gb  INFO: Uncompressing homopolymers in contigs\r\n",
      "00:08:50 4.6Gb  INFO: Total zero covered nucleotides 1\r\n",
      "00:08:50 4.6Gb  INFO: Calculating overlaps between adjacent uncompressed edges\r\n",
      "00:09:11 4.6Gb  INFO: Printing final gfa file to \"phasing-data/smoothed-reads/edge_6104_lja/mdbg.gfa\"\r\n",
      "00:09:12 4.6Gb  INFO: Printing final assembly to \"phasing-data/smoothed-reads/edge_6104_lja/assembly.fasta\"\r\n",
      "00:09:12 351Mb  INFO: Final homopolymer compressed and corrected reads can be found here: \"phasing-data/smoothed-reads/edge_6104_lja/k5001/corrected_reads.fasta\"\r\n",
      "00:09:12 351Mb  INFO: Final graph with homopolymer compressed edges can be found here: \"phasing-data/smoothed-reads/edge_6104_lja/mdbg/mdbg.hpc.gfa\"\r\n",
      "00:09:12 351Mb  INFO: Final graph can be found here: \"phasing-data/smoothed-reads/edge_6104_lja/mdbg.gfa\"\r\n",
      "00:09:12 351Mb  INFO: Final assembly can be found here: \"phasing-data/smoothed-reads/edge_6104_lja/assembly.fasta\"\r\n",
      "00:09:12 351Mb  INFO: LJA pipeline finished\r\n",
      "00:00:00 351Mb  INFO: Hello! You are running La Jolla Assembler (LJA), a tool for genome assembly from PacBio HiFi reads\r\n",
      "00:00:00 351Mb  INFO: 0a4619b61b5f12d9bed1a84784942003687a8bda\r\n",
      "00:00:00 351Mb  INFO: LJA pipeline started\r\n",
      "00:00:00 351Mb  INFO: Performing initial correction with k = 501\r\n",
      "00:00:00 0Mb  INFO: Reading reads\r\n",
      "00:00:00 0Mb  INFO: Extracting minimizers\r\n",
      "00:00:18 3.1Gb  INFO: Finished read processing\r\n",
      "00:00:18 3.1Gb  INFO: 1755662 hashs collected. Starting sorting.\r\n",
      "00:00:19 3.1Gb  INFO: Finished sorting. Total distinct minimizers: 9751\r\n",
      "00:00:19 3.1Gb  INFO: Starting construction of sparse de Bruijn graph\r\n",
      "00:00:19 3.1Gb  INFO: Vertex map constructed.\r\n",
      "00:00:19 3.1Gb  INFO: Filling edge sequences.\r\n",
      "00:00:37 3.2Gb  INFO: Finished sparse de Bruijn graph construction.\r\n",
      "00:00:37 3.2Gb  INFO:  Collecting tips \r\n",
      "00:00:37 3.2Gb  INFO: Added 2441 artificial minimizers from tips.\r\n",
      "00:00:37 3.2Gb  INFO: Collected 36732 old edges.\r\n",
      "00:00:37 3.2Gb  INFO: New minimizers added to sparse graph.\r\n",
      "00:00:37 3.2Gb  INFO: Refilling graph with old edges.\r\n",
      "00:00:37 3.2Gb  INFO: Filling graph with new edges.\r\n",
      "00:00:37 3.2Gb  INFO: Finished fixing sparse de Bruijn graph.\r\n",
      "00:00:37 3.2Gb  INFO: Starting to extract disjointigs.\r\n",
      "00:00:37 3.2Gb  INFO: Finished extracting 17122 disjointigs of total size 18397574\r\n",
      "00:00:38 0Mb  INFO: Loading disjointigs from file \"phasing-data/smoothed-reads/edge_1671_lja/k501/disjointigs.fasta\"\r\n",
      "00:00:38 48Mb  INFO: Filling bloom filter with k+1-mers.\r\n",
      "00:00:41 49Mb  INFO: Filled 41914843 bits out of 314222464\r\n",
      "00:00:41 49Mb  INFO: Finished filling bloom filter. Selecting junctions.\r\n",
      "00:00:42 53Mb  INFO: Collected 27157 junctions.\r\n",
      "00:00:42 53Mb  INFO: Starting DBG construction.\r\n",
      "00:00:42 53Mb  INFO: Vertices created.\r\n",
      "00:00:42 53Mb  INFO: Filled dbg edges. Adding hanging vertices \r\n",
      "00:00:42 53Mb  INFO: Added 0 hanging vertices\r\n",
      "00:00:42 53Mb  INFO: Merging unbranching paths\r\n",
      "00:00:42 53Mb  INFO: Ended merging edges. Resulting size 24853\r\n",
      "00:00:42 53Mb  INFO: Cleaning edge coverages\r\n",
      "00:00:42 53Mb  INFO: Collecting alignments of sequences to the graph\r\n",
      "00:00:48 3.2Gb  INFO: Alignment collection finished. Total length of alignments is 15485206\r\n",
      "00:00:48 3.2Gb  INFO: Precorrecting reads\r\n",
      "00:00:48 3.2Gb  INFO: Applying corrections to reads\r\n",
      "00:00:48 3.2Gb  INFO: Applied correction to 3850 reads\r\n",
      "00:00:48 3.2Gb  INFO: Corrected simple errors in 3850 reads\r\n",
      "00:00:48 3.2Gb  INFO: Applying changes to the graph\r\n",
      "00:00:52 3.2Gb  INFO: Collecting and storing read suffixes\r\n",
      "00:00:53 3.2Gb  INFO: Correcting dinucleotide errors in reads\r\n",
      "00:00:55 3.2Gb  INFO: Applying corrections to reads\r\n",
      "00:00:55 3.2Gb  INFO: Applied correction to 839 reads\r\n",
      "00:00:55 3.2Gb  INFO: Corrected 839 dinucleotide sequences\r\n",
      "00:00:55 3.2Gb  INFO: Marking reliable edges\r\n",
      "00:00:55 3.2Gb  INFO: Marked 46 edges in 30 paths as reliable\r\n",
      "00:00:55 3.2Gb  INFO: Correcting low covered regions in reads with K = 800\r\n",
      "00:00:55 3.2Gb  INFO: Applying corrections to reads\r\n",
      "00:00:55 3.2Gb  INFO: Applied correction to 3911 reads\r\n",
      "00:00:55 3.2Gb  INFO: Corrected low covered regions in 3911 reads with K = 800\r\n",
      "00:00:55 3.2Gb  INFO: Applying changes to the graph\r\n",
      "00:00:58 3.2Gb  INFO: Marking reliable edges\r\n",
      "00:00:58 3.2Gb  INFO: Marked 11 edges in 11 paths as reliable\r\n",
      "00:00:58 3.2Gb  INFO: Correcting low covered regions in reads with K = 2000\r\n",
      "00:00:59 3.2Gb  INFO: Applying corrections to reads\r\n",
      "00:00:59 3.2Gb  INFO: Applied correction to 128 reads\r\n",
      "00:00:59 3.2Gb  INFO: Corrected low covered regions in 128 reads with K = 2000\r\n",
      "00:00:59 3.2Gb  INFO: Applying changes to the graph\r\n",
      "00:01:01 3.2Gb  INFO: Correcting dinucleotide errors in reads\r\n",
      "00:01:02 3.2Gb  INFO: Applying corrections to reads\r\n",
      "00:01:02 3.2Gb  INFO: Applied correction to 1 reads\r\n",
      "00:01:02 3.2Gb  INFO: Corrected 1 dinucleotide sequences\r\n",
      "00:01:02 3.2Gb  INFO: Marking reliable edges\r\n",
      "00:01:02 3.2Gb  INFO: Marked 9 edges in 9 paths as reliable\r\n",
      "00:01:02 3.2Gb  INFO: Correcting low covered regions in reads\r\n",
      "00:01:03 3.2Gb  INFO: Applying corrections to reads\r\n",
      "00:01:03 3.2Gb  INFO: Applied correction to 488 reads\r\n",
      "00:01:03 3.2Gb  INFO: Corrected low covered regions in 888 reads\r\n",
      "00:01:03 3.2Gb  INFO: Marking reliable edges\r\n",
      "00:01:03 3.2Gb  INFO: Marked 6 edges in 6 paths as reliable\r\n",
      "00:01:03 3.2Gb  INFO: Correcting low covered regions in reads with K = 3500\r\n",
      "00:01:03 3.2Gb  INFO: Applying corrections to reads\r\n",
      "00:01:03 3.2Gb  INFO: Applied correction to 22 reads\r\n",
      "00:01:03 3.2Gb  INFO: Corrected low covered regions in 22 reads with K = 3500\r\n",
      "00:01:03 3.2Gb  INFO: Applying changes to the graph\r\n",
      "00:01:05 3.2Gb  INFO: Printing reads to fasta file \"phasing-data/smoothed-reads/edge_1671_lja/k501/corrected.fasta\"\r\n",
      "00:01:28 351Mb  INFO: Initial correction results with k = 501 printed to \"phasing-data/smoothed-reads/edge_1671_lja/k501/corrected.fasta\"\r\n",
      "00:01:28 351Mb  INFO: Performing second phase of error correction using k = 5001\r\n",
      "00:01:28 0Mb  INFO: Reading reads\r\n",
      "00:01:28 0Mb  INFO: Extracting minimizers\r\n",
      "00:01:38 2.2Gb  INFO: Finished read processing\r\n",
      "00:01:38 2.2Gb  INFO: 3259002 hashs collected. Starting sorting.\r\n",
      "00:01:38 2.3Gb  INFO: Finished sorting. Total distinct minimizers: 37160\r\n",
      "00:01:38 2.3Gb  INFO: Starting construction of sparse de Bruijn graph\r\n",
      "00:01:38 2.3Gb  INFO: Vertex map constructed.\r\n",
      "00:01:38 2.3Gb  INFO: Filling edge sequences.\r\n",
      "00:01:49 2.3Gb  INFO: Finished sparse de Bruijn graph construction.\r\n",
      "00:01:49 2.3Gb  INFO:  Collecting tips \r\n",
      "00:01:49 2.3Gb  INFO: Added 2018 artificial minimizers from tips.\r\n",
      "00:01:49 2.3Gb  INFO: Collected 74078 old edges.\r\n",
      "00:01:49 2.3Gb  INFO: New minimizers added to sparse graph.\r\n",
      "00:01:49 2.3Gb  INFO: Refilling graph with old edges.\r\n",
      "00:01:50 2.3Gb  INFO: Filling graph with new edges.\r\n",
      "00:01:50 2.3Gb  INFO: Finished fixing sparse de Bruijn graph.\r\n",
      "00:01:50 2.3Gb  INFO: Starting to extract disjointigs.\r\n",
      "00:01:50 2.3Gb  INFO: Finished extracting 3587 disjointigs of total size 27458858\r\n",
      "00:01:50 0Mb  INFO: Loading disjointigs from file \"phasing-data/smoothed-reads/edge_1671_lja/k5001/disjointigs.fasta\"\r\n",
      "00:01:51 48Mb  INFO: Filling bloom filter with k+1-mers.\r\n",
      "00:01:53 48Mb  INFO: Filled 44055818 bits out of 304648672\r\n",
      "00:01:53 48Mb  INFO: Finished filling bloom filter. Selecting junctions.\r\n",
      "00:01:54 49Mb  INFO: Collected 7353 junctions.\r\n",
      "00:01:54 49Mb  INFO: Starting DBG construction.\r\n",
      "00:01:54 49Mb  INFO: Vertices created.\r\n",
      "00:01:54 49Mb  INFO: Filled dbg edges. Adding hanging vertices \r\n",
      "00:01:54 49Mb  INFO: Added 0 hanging vertices\r\n",
      "00:01:54 49Mb  INFO: Merging unbranching paths\r\n",
      "00:01:54 49Mb  INFO: Ended merging edges. Resulting size 3765\r\n",
      "00:01:55 49Mb  INFO: Cleaning edge coverages\r\n",
      "00:01:55 49Mb  INFO: Collecting alignments of sequences to the graph\r\n",
      "00:01:55 49Mb  INFO: Storing suffixes of read paths of length up to 10000000\r\n",
      "00:02:00 1.2Gb  INFO: Alignment collection finished. Total length of alignments is 612718\r\n",
      "00:02:00 1.2Gb  INFO: Correcting dinucleotide errors in reads\r\n",
      "00:02:00 1.2Gb  INFO: Applying corrections to reads\r\n",
      "00:02:00 1.2Gb  INFO: Applied correction to 0 reads\r\n",
      "00:02:00 1.2Gb  INFO: Corrected 0 dinucleotide sequences\r\n",
      "00:02:00 1.2Gb  INFO: Marking reliable edges\r\n",
      "00:02:00 1.2Gb  INFO: Marked 36 edges in 26 paths as reliable\r\n",
      "00:02:00 1.2Gb  INFO: Correcting low covered regions in reads\r\n",
      "00:02:07 2.2Gb  INFO: Applying corrections to reads\r\n",
      "00:02:07 2.2Gb  INFO: Applied correction to 1728 reads\r\n",
      "00:02:07 2.2Gb  INFO: Corrected low covered regions in 1730 reads\r\n",
      "00:02:07 2.2Gb  INFO: Collapsing bulges\r\n",
      "00:02:07 2.2Gb  INFO: Applying corrections to reads\r\n",
      "00:02:07 2.2Gb  INFO: Applied correction to 0 reads\r\n",
      "00:02:07 2.2Gb  INFO: Collapsed bulges in 22 reads\r\n",
      "00:02:07 2.2Gb  INFO: Applying changes to the graph\r\n",
      "00:02:09 2.2Gb  INFO: Running second round of error correction\r\n",
      "00:02:09 2.2Gb  INFO: Correcting dinucleotide errors in reads\r\n",
      "00:02:09 2.2Gb  INFO: Applying corrections to reads\r\n",
      "00:02:09 2.2Gb  INFO: Applied correction to 0 reads\r\n",
      "00:02:09 2.2Gb  INFO: Corrected 0 dinucleotide sequences\r\n",
      "00:02:09 2.2Gb  INFO: Correcting dinucleotide errors in reads\r\n",
      "00:02:09 2.2Gb  INFO: Applying corrections to reads\r\n",
      "00:02:09 2.2Gb  INFO: Applied correction to 0 reads\r\n",
      "00:02:09 2.2Gb  INFO: Corrected 0 dinucleotide sequences\r\n",
      "00:02:09 2.2Gb  INFO: Marking reliable edges\r\n",
      "00:02:09 2.2Gb  INFO: Marked 0 edges in 0 paths as reliable\r\n",
      "00:02:09 2.2Gb  INFO: Correcting low covered regions in reads\r\n",
      "00:02:12 2.2Gb  INFO: Applying corrections to reads\r\n",
      "00:02:12 2.2Gb  INFO: Applied correction to 42 reads\r\n",
      "00:02:12 2.2Gb  INFO: Corrected low covered regions in 42 reads\r\n",
      "00:02:12 2.2Gb  INFO: Correcting dinucleotide errors in reads\r\n",
      "00:02:12 2.2Gb  INFO: Applying corrections to reads\r\n",
      "00:02:12 2.2Gb  INFO: Applied correction to 0 reads\r\n",
      "00:02:12 2.2Gb  INFO: Corrected 0 dinucleotide sequences\r\n",
      "00:02:12 2.2Gb  INFO: Remarking reliable edges\r\n",
      "00:02:12 2.2Gb  INFO: Correcting tips using reliable edge marks\r\n",
      "00:02:12 2.2Gb  INFO: Applying corrections to reads\r\n",
      "00:02:12 2.2Gb  INFO: Applied correction to 142 reads\r\n",
      "00:02:12 2.2Gb  INFO: Collapsing bulges\r\n",
      "00:02:12 2.2Gb  INFO: Applying corrections to reads\r\n",
      "00:02:12 2.2Gb  INFO: Applied correction to 0 reads\r\n",
      "00:02:12 2.2Gb  INFO: Collapsed bulges in 100 reads\r\n",
      "00:02:12 2.2Gb  INFO: Applying changes to the graph\r\n",
      "00:02:15 2.2Gb  INFO: Started gap closing procedure\r\n",
      "00:02:15 2.2Gb  INFO: Found 112 potential overlaps. Aligning.\r\n",
      "00:02:22 2.2Gb  INFO: Collected 21 overlaps. Looking for unique overlaps\r\n",
      "00:02:22 2.2Gb  INFO: Collected 6 unique overlaps.\r\n",
      "00:02:22 2.2Gb  INFO: Adding new connections to the graph\r\n",
      "00:02:27 2.2Gb  INFO: Correcting tips using reliable edge marks\r\n",
      "00:02:27 2.2Gb  INFO: Applying corrections to reads\r\n",
      "00:02:27 2.2Gb  INFO: Applied correction to 114 reads\r\n",
      "00:02:27 2.2Gb  INFO: Applying changes to the graph\r\n",
      "00:02:30 2.2Gb  INFO: Could not correct 621 reads. They will be removed.\r\n",
      "00:02:30 2.2Gb  INFO: Uncorrected reads were removed.\r\n",
      "00:02:30 2.2Gb  INFO: Applying changes to the graph\r\n",
      "00:02:33 2.2Gb  INFO: Looking for unique edges\r\n",
      "00:02:33 2.2Gb  INFO: Marked 135 long edges as unique\r\n",
      "00:02:33 2.2Gb  INFO: Marking extra edges as unique based on read paths\r\n",
      "00:02:33 2.2Gb  INFO: Marked 144 edges as unique\r\n",
      "00:02:33 2.2Gb  INFO: Splitting graph with unique edges\r\n",
      "00:02:33 2.2Gb  INFO: Processing 174 components\r\n",
      "00:02:33 2.2Gb  INFO: Finished unique edges search. Found 148 unique edges\r\n",
      "00:02:33 2.2Gb  INFO: Analysing repeats of multiplicity 2 and looking for additional unique edges\r\n",
      "00:02:33 2.2Gb  INFO: Finished processing of repeats of multiplicity 2. Found 0 erroneous edges.\r\n",
      "00:02:33 2.2Gb  INFO: Correcting reads using unique edge extensions\r\n",
      "00:02:33 2.2Gb  INFO: Applying corrections to reads\r\n",
      "00:02:33 2.2Gb  INFO: Applied correction to 73 reads\r\n",
      "00:02:33 2.2Gb  INFO: Collecting bad edges\r\n",
      "00:02:33 2.2Gb  INFO: Removed 12 disconnected edges\r\n",
      "00:02:33 2.2Gb  INFO: Could not correct 57 reads. They will be removed.\r\n",
      "00:02:33 2.2Gb  INFO: Uncorrected reads were removed.\r\n",
      "00:02:33 2.2Gb  INFO: Looking for more unique edges\r\n",
      "00:02:33 2.2Gb  INFO: Finished unique edges search. Found 110 unique edges\r\n",
      "00:02:33 2.2Gb  INFO: Looking for more unique edges\r\n",
      "00:02:33 2.2Gb  INFO: Finished unique edges search. Found 122 unique edges\r\n",
      "00:02:33 2.2Gb  INFO: Correcting reads using unique edge extensions\r\n",
      "00:02:33 2.2Gb  INFO: Applying corrections to reads\r\n",
      "00:02:33 2.2Gb  INFO: Applied correction to 24 reads\r\n",
      "00:02:33 2.2Gb  INFO: Collecting bad edges\r\n",
      "00:02:33 2.2Gb  INFO: Removed 16 disconnected edges\r\n",
      "00:02:33 2.2Gb  INFO: Could not correct 22 reads. They will be removed.\r\n",
      "00:02:33 2.2Gb  INFO: Uncorrected reads were removed.\r\n",
      "00:02:33 2.2Gb  INFO: Attempting to rescue small circular highly covered components\r\n",
      "00:02:33 2.2Gb  INFO: Could not correct 0 reads. They will be removed.\r\n",
      "00:02:33 2.2Gb  INFO: Uncorrected reads were removed.\r\n",
      "00:02:33 2.2Gb  INFO: Rescued 2 circular highly covered components\r\n",
      "00:02:33 2.2Gb  INFO: Applying changes to the graph\r\n",
      "00:02:36 2.2Gb  INFO: Started gap closing procedure\r\n",
      "00:02:37 2.2Gb  INFO: Found 138 potential overlaps. Aligning.\r\n",
      "00:02:46 2.2Gb  INFO: Collected 39 overlaps. Looking for unique overlaps\r\n",
      "00:02:46 2.2Gb  INFO: Collected 4 unique overlaps.\r\n",
      "00:02:46 2.2Gb  INFO: Adding new connections to the graph\r\n",
      "00:02:50 2.2Gb  INFO: Correcting tips using reliable edge marks\r\n",
      "00:02:50 2.2Gb  INFO: Applying corrections to reads\r\n",
      "00:02:50 2.2Gb  INFO: Applied correction to 16 reads\r\n",
      "00:02:50 2.2Gb  INFO: Applying changes to the graph\r\n",
      "00:02:54 2.2Gb  INFO: Printing reads to fasta file \"phasing-data/smoothed-reads/edge_1671_lja/k5001/corrected_reads.fasta\"\r\n",
      "00:03:14 351Mb  INFO: Second phase results with k = 5001 printed to \"phasing-data/smoothed-reads/edge_1671_lja/k5001/corrected_reads.fasta\"\r\n",
      "00:03:14 351Mb  INFO: Performing repeat resolution by transforming de Bruijn graph into Multiplex de Bruijn graph\r\n",
      "00:03:14 0Mb  INFO: Loading graph from fasta\r\n",
      "00:03:15 23Mb  INFO: Finished loading graph\r\n",
      "00:03:15 92Mb  INFO: Looking for unique edges\r\n",
      "00:03:15 92Mb  INFO: Marked 136 long edges as unique\r\n",
      "00:03:15 92Mb  INFO: Marking extra edges as unique based on read paths\r\n",
      "00:03:15 92Mb  INFO: Marked 147 edges as unique\r\n",
      "00:03:15 92Mb  INFO: Splitting graph with unique edges\r\n",
      "00:03:15 93Mb  INFO: Processing 168 components\r\n",
      "00:03:15 93Mb  INFO: Finished unique edges search. Found 175 unique edges\r\n",
      "00:03:15 93Mb  INFO: Analysing repeats of multiplicity 2 and looking for additional unique edges\r\n",
      "00:03:15 93Mb  INFO: Finished processing of repeats of multiplicity 2. Found 0 erroneous edges.\r\n",
      "00:03:15 93Mb  INFO: Resolving repeats\r\n",
      "00:03:15 93Mb  INFO: Constructing paths\r\n",
      "00:03:18 199Mb  INFO: Building graph\r\n",
      "00:03:18 200Mb  INFO: Increasing k\r\n",
      "00:03:18 200Mb  INFO: Finished increasing k\r\n",
      "00:03:18 200Mb  INFO: Exporting remaining active transitions\r\n",
      "00:03:18 200Mb  INFO: Export to Dot\r\n",
      "00:03:18 200Mb  INFO: Export to GFA and compressed contigs\r\n",
      "00:03:19 206Mb  INFO: Finished repeat resolution\r\n",
      "00:03:19 351Mb  INFO: Performing polishing and homopolymer uncompression\r\n",
      "00:03:19 11Mb  INFO: Aligning reads back to assembly\r\n",
      "00:03:31 2.4Gb  INFO: Finished alignment.\r\n",
      "00:03:31 2.4Gb  INFO: Printing alignments to \"phasing-data/smoothed-reads/edge_1671_lja/uncompressing/alignments.txt\"\r\n",
      "00:03:31 2.4Gb  INFO: Reading and processing initial reads from [\"phasing-data/smoothed-reads/edge_1671_smoothed_reads.fasta\"]\r\n",
      "00:05:14 2.8Gb  INFO: Uncompressing homopolymers in contigs\r\n",
      "00:05:15 2.8Gb  INFO: Total zero covered nucleotides 545\r\n",
      "00:05:15 2.8Gb  INFO: Calculating overlaps between adjacent uncompressed edges\r\n",
      "00:05:15 2.8Gb  INFO: Printing final gfa file to \"phasing-data/smoothed-reads/edge_1671_lja/mdbg.gfa\"\r\n",
      "00:05:15 2.8Gb  INFO: Printing final assembly to \"phasing-data/smoothed-reads/edge_1671_lja/assembly.fasta\"\r\n",
      "00:05:16 351Mb  INFO: Final homopolymer compressed and corrected reads can be found here: \"phasing-data/smoothed-reads/edge_1671_lja/k5001/corrected_reads.fasta\"\r\n",
      "00:05:16 351Mb  INFO: Final graph with homopolymer compressed edges can be found here: \"phasing-data/smoothed-reads/edge_1671_lja/mdbg/mdbg.hpc.gfa\"\r\n",
      "00:05:16 351Mb  INFO: Final graph can be found here: \"phasing-data/smoothed-reads/edge_1671_lja/mdbg.gfa\"\r\n",
      "00:05:16 351Mb  INFO: Final assembly can be found here: \"phasing-data/smoothed-reads/edge_1671_lja/assembly.fasta\"\r\n",
      "00:05:16 351Mb  INFO: LJA pipeline finished\r\n",
      "00:00:00 351Mb  INFO: Hello! You are running La Jolla Assembler (LJA), a tool for genome assembly from PacBio HiFi reads\r\n",
      "00:00:00 351Mb  INFO: 0a4619b61b5f12d9bed1a84784942003687a8bda\r\n",
      "00:00:00 351Mb  INFO: LJA pipeline started\r\n",
      "00:00:00 351Mb  INFO: Performing initial correction with k = 501\r\n",
      "00:00:00 0Mb  INFO: Reading reads\r\n",
      "00:00:00 0Mb  INFO: Extracting minimizers\r\n",
      "00:00:56 5.5Gb  INFO: Finished read processing\r\n",
      "00:00:56 5.5Gb  INFO: 5035894 hashs collected. Starting sorting.\r\n",
      "00:00:56 5.5Gb  INFO: Finished sorting. Total distinct minimizers: 2926\r\n",
      "00:00:56 5.5Gb  INFO: Starting construction of sparse de Bruijn graph\r\n",
      "00:00:56 5.5Gb  INFO: Vertex map constructed.\r\n",
      "00:00:56 5.5Gb  INFO: Filling edge sequences.\r\n",
      "00:01:46 5.5Gb  INFO: Finished sparse de Bruijn graph construction.\r\n",
      "00:01:46 5.5Gb  INFO:  Collecting tips \r\n",
      "00:01:46 5.5Gb  INFO: Added 188 artificial minimizers from tips.\r\n",
      "00:01:46 5.5Gb  INFO: Collected 8234 old edges.\r\n",
      "00:01:46 5.5Gb  INFO: New minimizers added to sparse graph.\r\n",
      "00:01:46 5.5Gb  INFO: Refilling graph with old edges.\r\n",
      "00:01:46 5.5Gb  INFO: Filling graph with new edges.\r\n",
      "00:01:46 5.5Gb  INFO: Finished fixing sparse de Bruijn graph.\r\n",
      "00:01:46 5.5Gb  INFO: Starting to extract disjointigs.\r\n",
      "00:01:46 5.5Gb  INFO: Finished extracting 2498 disjointigs of total size 3979146\r\n",
      "00:01:47 0Mb  INFO: Loading disjointigs from file \"phasing-data/smoothed-reads/edge_2358_lja/k501/disjointigs.fasta\"\r\n",
      "00:01:47 15Mb  INFO: Filling bloom filter with k+1-mers.\r\n",
      "00:01:47 16Mb  INFO: Filled 12410666 bits out of 87284736\r\n",
      "00:01:47 16Mb  INFO: Finished filling bloom filter. Selecting junctions.\r\n",
      "00:01:47 16Mb  INFO: Collected 3251 junctions.\r\n",
      "00:01:48 16Mb  INFO: Starting DBG construction.\r\n",
      "00:01:48 16Mb  INFO: Vertices created.\r\n",
      "00:01:48 16Mb  INFO: Filled dbg edges. Adding hanging vertices \r\n",
      "00:01:48 16Mb  INFO: Added 0 hanging vertices\r\n",
      "00:01:48 16Mb  INFO: Merging unbranching paths\r\n",
      "00:01:48 16Mb  INFO: Ended merging edges. Resulting size 2375\r\n",
      "00:01:48 16Mb  INFO: Cleaning edge coverages\r\n",
      "00:01:48 16Mb  INFO: Collecting alignments of sequences to the graph\r\n",
      "00:02:06 5.8Gb  INFO: Alignment collection finished. Total length of alignments is 6749817\r\n",
      "00:02:06 5.8Gb  INFO: Precorrecting reads\r\n",
      "00:02:07 5.8Gb  INFO: Applying corrections to reads\r\n",
      "00:02:07 5.8Gb  INFO: Applied correction to 373 reads\r\n",
      "00:02:07 5.8Gb  INFO: Corrected simple errors in 373 reads\r\n",
      "00:02:07 5.8Gb  INFO: Applying changes to the graph\r\n",
      "00:02:11 6Gb  INFO: Collecting and storing read suffixes\r\n",
      "00:02:11 6Gb  INFO: Correcting dinucleotide errors in reads\r\n",
      "00:02:12 6Gb  INFO: Applying corrections to reads\r\n",
      "00:02:12 6Gb  INFO: Applied correction to 852 reads\r\n",
      "00:02:12 6Gb  INFO: Corrected 852 dinucleotide sequences\r\n",
      "00:02:12 6Gb  INFO: Marking reliable edges\r\n",
      "00:02:12 6Gb  INFO: Marked 0 edges in 0 paths as reliable\r\n",
      "00:02:12 6Gb  INFO: Correcting low covered regions in reads with K = 800\r\n",
      "00:02:12 6Gb  INFO: Applying corrections to reads\r\n",
      "00:02:12 6Gb  INFO: Applied correction to 415 reads\r\n",
      "00:02:12 6Gb  INFO: Corrected low covered regions in 415 reads with K = 800\r\n",
      "00:02:12 6Gb  INFO: Applying changes to the graph\r\n",
      "00:02:17 6.1Gb  INFO: Marking reliable edges\r\n",
      "00:02:17 6.1Gb  INFO: Marked 0 edges in 0 paths as reliable\r\n",
      "00:02:17 6.1Gb  INFO: Correcting low covered regions in reads with K = 2000\r\n",
      "00:02:17 6.1Gb  INFO: Applying corrections to reads\r\n",
      "00:02:17 6.1Gb  INFO: Applied correction to 0 reads\r\n",
      "00:02:17 6.1Gb  INFO: Corrected low covered regions in 0 reads with K = 2000\r\n",
      "00:02:17 6.1Gb  INFO: Applying changes to the graph\r\n",
      "00:02:21 6.2Gb  INFO: Correcting dinucleotide errors in reads\r\n",
      "00:02:22 6.2Gb  INFO: Applying corrections to reads\r\n",
      "00:02:22 6.2Gb  INFO: Applied correction to 0 reads\r\n",
      "00:02:22 6.2Gb  INFO: Corrected 0 dinucleotide sequences\r\n",
      "00:02:22 6.2Gb  INFO: Marking reliable edges\r\n",
      "00:02:22 6.2Gb  INFO: Marked 0 edges in 0 paths as reliable\r\n",
      "00:02:22 6.2Gb  INFO: Correcting low covered regions in reads\r\n",
      "00:02:22 6.2Gb  INFO: Applying corrections to reads\r\n",
      "00:02:22 6.2Gb  INFO: Applied correction to 40 reads\r\n",
      "00:02:22 6.2Gb  INFO: Corrected low covered regions in 112 reads\r\n",
      "00:02:22 6.2Gb  INFO: Marking reliable edges\r\n",
      "00:02:22 6.2Gb  INFO: Marked 0 edges in 0 paths as reliable\r\n",
      "00:02:22 6.2Gb  INFO: Correcting low covered regions in reads with K = 3500\r\n",
      "00:02:22 6.2Gb  INFO: Applying corrections to reads\r\n",
      "00:02:22 6.2Gb  INFO: Applied correction to 0 reads\r\n",
      "00:02:22 6.2Gb  INFO: Corrected low covered regions in 0 reads with K = 3500\r\n",
      "00:02:22 6.2Gb  INFO: Applying changes to the graph\r\n",
      "00:02:27 6.2Gb  INFO: Printing reads to fasta file \"phasing-data/smoothed-reads/edge_2358_lja/k501/corrected.fasta\"\r\n",
      "00:03:27 351Mb  INFO: Initial correction results with k = 501 printed to \"phasing-data/smoothed-reads/edge_2358_lja/k501/corrected.fasta\"\r\n",
      "00:03:27 351Mb  INFO: Performing second phase of error correction using k = 5001\r\n",
      "00:03:27 0Mb  INFO: Reading reads\r\n",
      "00:03:27 0Mb  INFO: Extracting minimizers\r\n",
      "00:03:56 4.4Gb  INFO: Finished read processing\r\n",
      "00:03:56 4.4Gb  INFO: 9385689 hashs collected. Starting sorting.\r\n",
      "00:03:56 4.4Gb  INFO: Finished sorting. Total distinct minimizers: 19965\r\n",
      "00:03:56 4.4Gb  INFO: Starting construction of sparse de Bruijn graph\r\n",
      "00:03:56 4.4Gb  INFO: Vertex map constructed.\r\n",
      "00:03:56 4.4Gb  INFO: Filling edge sequences.\r\n",
      "00:04:25 5.7Gb  INFO: Finished sparse de Bruijn graph construction.\r\n",
      "00:04:25 5.7Gb  INFO:  Collecting tips \r\n",
      "00:04:25 5.7Gb  INFO: Added 314 artificial minimizers from tips.\r\n",
      "00:04:25 5.7Gb  INFO: Collected 40880 old edges.\r\n",
      "00:04:25 5.7Gb  INFO: New minimizers added to sparse graph.\r\n",
      "00:04:25 5.7Gb  INFO: Refilling graph with old edges.\r\n",
      "00:04:25 5.7Gb  INFO: Filling graph with new edges.\r\n",
      "00:04:25 5.7Gb  INFO: Finished fixing sparse de Bruijn graph.\r\n",
      "00:04:25 5.7Gb  INFO: Starting to extract disjointigs.\r\n",
      "00:04:25 5.7Gb  INFO: Finished extracting 1799 disjointigs of total size 14000010\r\n",
      "00:04:26 0Mb  INFO: Loading disjointigs from file \"phasing-data/smoothed-reads/edge_2358_lja/k5001/disjointigs.fasta\"\r\n",
      "00:04:26 26Mb  INFO: Filling bloom filter with k+1-mers.\r\n",
      "00:04:26 27Mb  INFO: Filled 23152980 bits out of 160102752\r\n",
      "00:04:26 27Mb  INFO: Finished filling bloom filter. Selecting junctions.\r\n",
      "00:04:27 27Mb  INFO: Collected 3422 junctions.\r\n",
      "00:04:27 27Mb  INFO: Starting DBG construction.\r\n",
      "00:04:27 27Mb  INFO: Vertices created.\r\n",
      "00:04:27 27Mb  INFO: Filled dbg edges. Adding hanging vertices \r\n",
      "00:04:27 27Mb  INFO: Added 0 hanging vertices\r\n",
      "00:04:27 27Mb  INFO: Merging unbranching paths\r\n",
      "00:04:27 27Mb  INFO: Ended merging edges. Resulting size 1453\r\n",
      "00:04:27 27Mb  INFO: Cleaning edge coverages\r\n",
      "00:04:27 27Mb  INFO: Collecting alignments of sequences to the graph\r\n",
      "00:04:27 27Mb  INFO: Storing suffixes of read paths of length up to 10000000\r\n",
      "00:04:42 6.3Gb  INFO: Alignment collection finished. Total length of alignments is 1806006\r\n",
      "00:04:42 6.3Gb  INFO: Correcting dinucleotide errors in reads\r\n",
      "00:04:43 6.3Gb  INFO: Applying corrections to reads\r\n",
      "00:04:43 6.3Gb  INFO: Applied correction to 0 reads\r\n",
      "00:04:43 6.3Gb  INFO: Corrected 0 dinucleotide sequences\r\n",
      "00:04:43 6.3Gb  INFO: Marking reliable edges\r\n",
      "00:04:43 6.3Gb  INFO: Marked 4 edges in 2 paths as reliable\r\n",
      "00:04:43 6.3Gb  INFO: Correcting low covered regions in reads\r\n",
      "00:04:45 6.7Gb  INFO: Applying corrections to reads\r\n",
      "00:04:45 6.7Gb  INFO: Applied correction to 380 reads\r\n",
      "00:04:45 6.7Gb  INFO: Corrected low covered regions in 380 reads\r\n",
      "00:04:45 6.7Gb  INFO: Collapsing bulges\r\n",
      "00:04:45 6.7Gb  INFO: Applying corrections to reads\r\n",
      "00:04:45 6.7Gb  INFO: Applied correction to 0 reads\r\n",
      "00:04:45 6.7Gb  INFO: Collapsed bulges in 72 reads\r\n",
      "00:04:45 6.7Gb  INFO: Applying changes to the graph\r\n",
      "00:04:49 6.7Gb  INFO: Running second round of error correction\r\n",
      "00:04:49 6.7Gb  INFO: Correcting dinucleotide errors in reads\r\n",
      "00:04:49 6.7Gb  INFO: Applying corrections to reads\r\n",
      "00:04:49 6.7Gb  INFO: Applied correction to 0 reads\r\n",
      "00:04:49 6.7Gb  INFO: Corrected 0 dinucleotide sequences\r\n",
      "00:04:49 6.7Gb  INFO: Correcting dinucleotide errors in reads\r\n",
      "00:04:49 6.7Gb  INFO: Applying corrections to reads\r\n",
      "00:04:49 6.7Gb  INFO: Applied correction to 0 reads\r\n",
      "00:04:49 6.7Gb  INFO: Corrected 0 dinucleotide sequences\r\n",
      "00:04:49 6.7Gb  INFO: Marking reliable edges\r\n",
      "00:04:49 6.7Gb  INFO: Marked 2 edges in 2 paths as reliable\r\n",
      "00:04:49 6.7Gb  INFO: Correcting low covered regions in reads\r\n",
      "00:04:50 6.7Gb  INFO: Applying corrections to reads\r\n",
      "00:04:50 6.7Gb  INFO: Applied correction to 22 reads\r\n",
      "00:04:50 6.7Gb  INFO: Corrected low covered regions in 22 reads\r\n",
      "00:04:50 6.7Gb  INFO: Correcting dinucleotide errors in reads\r\n",
      "00:04:50 6.7Gb  INFO: Applying corrections to reads\r\n",
      "00:04:50 6.7Gb  INFO: Applied correction to 0 reads\r\n",
      "00:04:50 6.7Gb  INFO: Corrected 0 dinucleotide sequences\r\n",
      "00:04:50 6.7Gb  INFO: Remarking reliable edges\r\n",
      "00:04:50 6.7Gb  INFO: Correcting tips using reliable edge marks\r\n",
      "00:04:50 6.7Gb  INFO: Applying corrections to reads\r\n",
      "00:04:50 6.7Gb  INFO: Applied correction to 3 reads\r\n",
      "00:04:50 6.7Gb  INFO: Collapsing bulges\r\n",
      "00:04:50 6.7Gb  INFO: Applying corrections to reads\r\n",
      "00:04:50 6.7Gb  INFO: Applied correction to 0 reads\r\n",
      "00:04:50 6.7Gb  INFO: Collapsed bulges in 82 reads\r\n",
      "00:04:50 6.7Gb  INFO: Applying changes to the graph\r\n",
      "00:04:55 6.7Gb  INFO: Started gap closing procedure\r\n",
      "00:04:55 6.7Gb  INFO: Found 2 potential overlaps. Aligning.\r\n",
      "00:04:55 6.7Gb  INFO: Collected 0 overlaps. Looking for unique overlaps\r\n",
      "00:04:55 6.7Gb  INFO: Collected 0 unique overlaps.\r\n",
      "00:04:55 6.7Gb  INFO: Could not correct 261 reads. They will be removed.\r\n",
      "00:04:55 6.7Gb  INFO: Uncorrected reads were removed.\r\n",
      "00:04:55 6.7Gb  INFO: Applying changes to the graph\r\n",
      "00:05:00 6.7Gb  INFO: Looking for unique edges\r\n",
      "00:05:00 6.7Gb  INFO: Marked 1 long edges as unique\r\n",
      "00:05:00 6.7Gb  INFO: Marking extra edges as unique based on read paths\r\n",
      "00:05:00 6.7Gb  INFO: Marked 1 edges as unique\r\n",
      "00:05:00 6.7Gb  INFO: Splitting graph with unique edges\r\n",
      "00:05:00 6.7Gb  INFO: Processing 3 components\r\n",
      "00:05:01 6.7Gb  INFO: Finished unique edges search. Found 1 unique edges\r\n",
      "00:05:01 6.7Gb  INFO: Analysing repeats of multiplicity 2 and looking for additional unique edges\r\n",
      "00:05:01 6.7Gb  INFO: Finished processing of repeats of multiplicity 2. Found 0 erroneous edges.\r\n",
      "00:05:01 6.7Gb  INFO: Correcting reads using unique edge extensions\r\n",
      "00:05:01 6.7Gb  INFO: Applying corrections to reads\r\n",
      "00:05:01 6.7Gb  INFO: Applied correction to 19 reads\r\n",
      "00:05:01 6.7Gb  INFO: Collecting bad edges\r\n",
      "00:05:01 6.7Gb  INFO: Removed 1 disconnected edges\r\n",
      "00:05:02 6.7Gb  INFO: Could not correct 21 reads. They will be removed.\r\n",
      "00:05:02 6.7Gb  INFO: Uncorrected reads were removed.\r\n",
      "00:05:02 6.7Gb  INFO: Looking for more unique edges\r\n",
      "00:05:02 6.7Gb  INFO: Finished unique edges search. Found 2 unique edges\r\n",
      "00:05:02 6.7Gb  INFO: Looking for more unique edges\r\n",
      "00:05:02 6.7Gb  INFO: Finished unique edges search. Found 3 unique edges\r\n",
      "00:05:02 6.7Gb  INFO: Correcting reads using unique edge extensions\r\n",
      "00:05:02 6.7Gb  INFO: Applying corrections to reads\r\n",
      "00:05:02 6.7Gb  INFO: Applied correction to 27 reads\r\n",
      "00:05:02 6.7Gb  INFO: Collecting bad edges\r\n",
      "00:05:02 6.7Gb  INFO: Removed 3 disconnected edges\r\n",
      "00:05:02 6.7Gb  INFO: Could not correct 12 reads. They will be removed.\r\n",
      "00:05:02 6.7Gb  INFO: Uncorrected reads were removed.\r\n",
      "00:05:02 6.7Gb  INFO: Attempting to rescue small circular highly covered components\r\n",
      "00:05:03 6.7Gb  INFO: Could not correct 0 reads. They will be removed.\r\n",
      "00:05:03 6.7Gb  INFO: Uncorrected reads were removed.\r\n",
      "00:05:03 6.7Gb  INFO: Rescued 0 circular highly covered components\r\n",
      "00:05:03 6.7Gb  INFO: Applying changes to the graph\r\n",
      "00:05:07 6.7Gb  INFO: Started gap closing procedure\r\n",
      "00:05:07 6.7Gb  INFO: Found 2 potential overlaps. Aligning.\r\n",
      "00:05:08 6.7Gb  INFO: Collected 0 overlaps. Looking for unique overlaps\r\n",
      "00:05:08 6.7Gb  INFO: Collected 0 unique overlaps.\r\n",
      "00:05:11 6.7Gb  INFO: Printing reads to fasta file \"phasing-data/smoothed-reads/edge_2358_lja/k5001/corrected_reads.fasta\"\r\n",
      "00:06:07 351Mb  INFO: Second phase results with k = 5001 printed to \"phasing-data/smoothed-reads/edge_2358_lja/k5001/corrected_reads.fasta\"\r\n",
      "00:06:07 351Mb  INFO: Performing repeat resolution by transforming de Bruijn graph into Multiplex de Bruijn graph\r\n",
      "00:06:07 0Mb  INFO: Loading graph from fasta\r\n",
      "00:06:08 30Mb  INFO: Finished loading graph\r\n",
      "00:06:09 250Mb  INFO: Looking for unique edges\r\n",
      "00:06:09 250Mb  INFO: Marked 1 long edges as unique\r\n",
      "00:06:09 250Mb  INFO: Marking extra edges as unique based on read paths\r\n",
      "00:06:09 250Mb  INFO: Marked 1 edges as unique\r\n",
      "00:06:09 250Mb  INFO: Splitting graph with unique edges\r\n",
      "00:06:09 250Mb  INFO: Processing 3 components\r\n",
      "00:06:10 250Mb  INFO: Finished unique edges search. Found 1 unique edges\r\n",
      "00:06:10 250Mb  INFO: Analysing repeats of multiplicity 2 and looking for additional unique edges\r\n",
      "00:06:10 250Mb  INFO: Finished processing of repeats of multiplicity 2. Found 0 erroneous edges.\r\n",
      "00:06:10 250Mb  INFO: Resolving repeats\r\n",
      "00:06:10 250Mb  INFO: Constructing paths\r\n",
      "00:06:22 0.6Gb  INFO: Building graph\r\n",
      "00:06:22 0.6Gb  INFO: Increasing k\r\n",
      "00:06:55 0.6Gb  INFO: Finished increasing k\r\n",
      "00:06:55 0.6Gb  INFO: Exporting remaining active transitions\r\n",
      "00:06:55 0.6Gb  INFO: Export to Dot\r\n",
      "00:06:55 0.6Gb  INFO: Export to GFA and compressed contigs\r\n",
      "00:06:56 0.7Gb  INFO: Finished repeat resolution\r\n",
      "00:06:57 351Mb  INFO: Performing polishing and homopolymer uncompression\r\n",
      "00:06:58 24Mb  INFO: Aligning reads back to assembly\r\n",
      "00:07:46 5.1Gb  INFO: Finished alignment.\r\n",
      "00:07:46 5.1Gb  INFO: Printing alignments to \"phasing-data/smoothed-reads/edge_2358_lja/uncompressing/alignments.txt\"\r\n",
      "00:07:50 5.1Gb  INFO: Reading and processing initial reads from [\"phasing-data/smoothed-reads/edge_2358_smoothed_reads.fasta\"]\r\n",
      "00:19:34 5.1Gb  INFO: Uncompressing homopolymers in contigs\r\n",
      "00:19:36 5.1Gb  INFO: Total zero covered nucleotides 0\r\n",
      "00:19:36 5.1Gb  INFO: Calculating overlaps between adjacent uncompressed edges\r\n",
      "00:19:37 5.1Gb  INFO: Printing final gfa file to \"phasing-data/smoothed-reads/edge_2358_lja/mdbg.gfa\"\r\n",
      "00:19:37 5.1Gb  INFO: Printing final assembly to \"phasing-data/smoothed-reads/edge_2358_lja/assembly.fasta\"\r\n",
      "00:19:37 351Mb  INFO: Final homopolymer compressed and corrected reads can be found here: \"phasing-data/smoothed-reads/edge_2358_lja/k5001/corrected_reads.fasta\"\r\n",
      "00:19:37 351Mb  INFO: Final graph with homopolymer compressed edges can be found here: \"phasing-data/smoothed-reads/edge_2358_lja/mdbg/mdbg.hpc.gfa\"\r\n",
      "00:19:37 351Mb  INFO: Final graph can be found here: \"phasing-data/smoothed-reads/edge_2358_lja/mdbg.gfa\"\r\n",
      "00:19:37 351Mb  INFO: Final assembly can be found here: \"phasing-data/smoothed-reads/edge_2358_lja/assembly.fasta\"\r\n",
      "00:19:37 351Mb  INFO: LJA pipeline finished\r\n"
     ]
    }
   ],
   "source": [
    "# Old approach: use all of LJA:\n",
    "# 1. jumboDBG     [make the de Bruijn graph]\n",
    "# 2. mowerDBG     [error correct reads]\n",
    "# 3. multiplexDBG [create multiplex de Bruijn graph]\n",
    "# !/home/mfedarko/software/LJA/bin/lja \\\n",
    "#     --reads phasing-data/smoothed-reads/edge_6104_smoothed_reads.fasta \\\n",
    "#     --output-dir phasing-data/smoothed-reads/edge_6104_lja\n",
    "\n",
    "# !/home/mfedarko/software/LJA/bin/lja \\\n",
    "#     --reads phasing-data/smoothed-reads/edge_1671_smoothed_reads.fasta \\\n",
    "#     --output-dir phasing-data/smoothed-reads/edge_1671_lja\n",
    "\n",
    "# !/home/mfedarko/software/LJA/bin/lja \\\n",
    "#     --reads phasing-data/smoothed-reads/edge_2358_smoothed_reads.fasta \\\n",
    "#     --output-dir phasing-data/smoothed-reads/edge_2358_lja"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
