{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# $R_S$ vs. $R_N$ and $R_{NNS}$ vs. $R_{NS}$  barplots for each genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-11T23:10:33.663085Z",
     "iopub.status.busy": "2021-06-11T23:10:33.662193Z",
     "iopub.status.idle": "2021-06-11T23:10:33.952730Z",
     "shell.execute_reply": "2021-06-11T23:10:33.951742Z"
    }
   },
   "outputs": [],
   "source": [
    "%run \"Header.ipynb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-11T23:10:33.957180Z",
     "iopub.status.busy": "2021-06-11T23:10:33.956355Z",
     "iopub.status.idle": "2021-06-11T23:10:36.326771Z",
     "shell.execute_reply": "2021-06-11T23:10:36.326016Z"
    }
   },
   "outputs": [],
   "source": [
    "import json\n",
    "import pickle\n",
    "import skbio\n",
    "import pileup\n",
    "from statistics import median\n",
    "from math import log\n",
    "from collections import defaultdict\n",
    "from matplotlib import pyplot\n",
    "from parse_sco import parse_sco"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-11T23:10:36.332320Z",
     "iopub.status.busy": "2021-06-11T23:10:36.331528Z",
     "iopub.status.idle": "2021-06-11T23:10:38.527108Z",
     "shell.execute_reply": "2021-06-11T23:10:38.526318Z"
    }
   },
   "outputs": [],
   "source": [
    "with open(\"matrix-jsons/seq2gene2codon2alignedcodons.json\", \"r\") as jf:\n",
    "    seq2gene2codon2alignedcodons = json.load(jf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'266': {'TAA': 4797, 'AAA': 2, 'TTA': 1, 'TCA': 1, 'TAT': 1, 'TAC': 1},\n",
       " '269': {'GTA': 4779, 'TGA': 1, 'GTT': 1, 'CTA': 1},\n",
       " '272': {'TTG': 4728, 'TGG': 1, 'TTT': 7, 'GTG': 1},\n",
       " '275': {'AAT': 4744, 'ATT': 1, 'AAA': 8, 'CAT': 1, 'AAC': 1},\n",
       " '278': {'GCA': 4786, 'TCA': 1, 'GCC': 1},\n",
       " '281': {'AAG': 4776, 'AGG': 1, 'ATG': 2, 'CAG': 1, 'ACG': 1},\n",
       " '284': {'GTA': 4811, 'ACA': 1},\n",
       " '287': {'CTT': 4804, 'CGT': 1, 'CAT': 1, 'CTG': 1, 'TCT': 1},\n",
       " '290': {'ATG': 4813, 'TTG': 1, 'AAG': 5},\n",
       " '293': {'GCT': 4815, 'GCA': 1, 'GCG': 1, 'GAT': 1},\n",
       " '296': {'TAT': 4786, 'TTT': 1, 'AGT': 1},\n",
       " '299': {'CTT': 4822, 'CTC': 1, 'CAT': 1},\n",
       " '302': {'ATT': 4810, 'CTT': 1},\n",
       " '305': {'GTG': 4819, 'TTG': 1, 'GAG': 1, 'GGA': 1},\n",
       " '308': {'TTT': 4809, 'TTG': 2, 'GTT': 1},\n",
       " '311': {'ATG': 4805, 'ATT': 3, 'CTG': 2, 'TAG': 1},\n",
       " '314': {'ATA': 4827},\n",
       " '317': {'ATT': 4804, 'AAT': 2, 'CTT': 1},\n",
       " '320': {'ACA': 4821, 'ACT': 1, 'ACC': 1, 'CCA': 1},\n",
       " '323': {'CCT': 4737, 'CAT': 1},\n",
       " '326': {'GTC': 4813, 'GTT': 2, 'GTA': 1},\n",
       " '329': {'GAA': 4819},\n",
       " '332': {'AAC': 4804, 'ATC': 1, 'TAC': 1},\n",
       " '335': {'TAT': 4754, 'TTA': 2, 'TAA': 1},\n",
       " '338': {'ATT': 4789,\n",
       "  'ATG': 1,\n",
       "  'AGT': 1,\n",
       "  'ATA': 1,\n",
       "  'TTT': 1,\n",
       "  'AAT': 1,\n",
       "  'CTT': 1},\n",
       " '341': {'AGA': 4829, 'GAA': 2, 'AGG': 1, 'CGA': 1},\n",
       " '344': {'TTT': 4799, 'TTG': 1, 'TTA': 2, 'GTT': 2, 'CTT': 3, 'ATT': 1},\n",
       " '347': {'TTC': 4806, 'ATT': 2, 'TTT': 1},\n",
       " '350': {'AAA': 4815, 'AAT': 1},\n",
       " '353': {'GGA': 4797, 'CGA': 1, 'GGG': 1, 'TGA': 1, 'GGC': 1},\n",
       " '356': {'AGT': 4688, 'CGT': 1, 'AAT': 1, 'AAG': 1, 'GTT': 1},\n",
       " '359': {'AAA': 4839, 'CAA': 1},\n",
       " '362': {'ACA': 4815, 'CAC': 4, 'ACC': 1, 'CCA': 1, 'GAC': 1},\n",
       " '365': {'TGC': 4820, 'TGG': 1},\n",
       " '368': {'CGC': 4819},\n",
       " '371': {'AAT': 4643, 'CAT': 2, 'AAA': 1, 'AAG': 2, 'ACT': 1, 'ATT': 2},\n",
       " '374': {'GAA': 4818, 'AGA': 4, 'GAC': 2, 'AAA': 3, 'GCA': 1, 'TAA': 1},\n",
       " '377': {'AAA': 4818, 'TAA': 3, 'AAG': 2},\n",
       " '380': {'TTT': 4815, 'TTA': 3, 'ATT': 2},\n",
       " '383': {'CAA': 4817, 'AAA': 1, 'ACA': 1, 'CAT': 2},\n",
       " '386': {'AAA': 4809, 'AAC': 3, 'ACA': 1},\n",
       " '389': {'ATG': 4827, 'ATT': 1},\n",
       " '392': {'TTT': 4783, 'GTT': 1, 'TTA': 1},\n",
       " '395': {'TAT': 4822, 'TTA': 2, 'GAT': 1},\n",
       " '398': {'GGG': 4806, 'GTG': 1, 'TGG': 1, 'GGC': 1},\n",
       " '401': {'TTA': 4766, 'GTA': 2, 'TTT': 6, 'TAG': 1},\n",
       " '404': {'AGT': 4822, 'GTT': 3, 'ATT': 1, 'GGT': 1},\n",
       " '407': {'TTT': 4732, 'TTA': 4, 'GTT': 1},\n",
       " '410': {'CAT': 4831, 'CAA': 1},\n",
       " '413': {'TAT': 4785, 'TAG': 1, 'AAT': 1},\n",
       " '416': {'ATT': 4828, 'AAT': 1, 'AGT': 1, 'ATA': 1},\n",
       " '419': {'GTG': 4832},\n",
       " '422': {'TTA': 4741},\n",
       " '425': {'ATT': 4830, 'ATC': 1, 'AAT': 2, 'ACT': 1},\n",
       " '428': {'GTG': 4824, 'TTG': 2, 'GTT': 1},\n",
       " '431': {'GCT': 4832},\n",
       " '434': {'TGT': 4801, 'GTG': 1, 'GTT': 1},\n",
       " '437': {'CTT': 4832, 'ATT': 1},\n",
       " '440': {'TTA': 4822, 'GTA': 2},\n",
       " '443': {'AAA': 4826, 'AAT': 1},\n",
       " '446': {'ATG': 4836, 'AGG': 2},\n",
       " '449': {'CAT': 4834},\n",
       " '452': {'TTT': 4804, 'GTT': 1, 'TTA': 2, 'TGT': 1, 'ATT': 1},\n",
       " '455': {'TAT': 4824,\n",
       "  'ATA': 1,\n",
       "  'TTT': 1,\n",
       "  'TAA': 1,\n",
       "  'AAT': 1,\n",
       "  'ATT': 1,\n",
       "  'TTA': 1},\n",
       " '458': {'GGA': 4822, 'GTA': 1, 'GGT': 2, 'AGG': 1},\n",
       " '461': {'CAA': 4829, 'CAG': 2, 'CTA': 1, 'CAT': 1},\n",
       " '464': {'AAG': 4829, 'ACG': 1},\n",
       " '467': {'ATG': 4816, 'CTG': 2, 'AAG': 2},\n",
       " '470': {'CTA': 4835, 'ATA': 1},\n",
       " '473': {'GAA': 4826},\n",
       " '476': {'CCT': 4816, 'CCG': 1},\n",
       " '479': {'AAA': 4803, 'ACA': 1, 'AAT': 1},\n",
       " '482': {'GCT': 4830, 'GAT': 1},\n",
       " '485': {'GGC': 4820},\n",
       " '488': {'ATT': 4826, 'TTT': 1},\n",
       " '491': {'TTA': 4819, 'TAA': 1, 'ATA': 1, 'TTT': 1, 'CTA': 1},\n",
       " '494': {'ATC': 4833, 'CTC': 1, 'ATT': 1},\n",
       " '497': {'TTT': 4816, 'TGT': 1, 'GTT': 2, 'ATT': 3},\n",
       " '500': {'TTA': 4821, 'GTA': 1, 'TTT': 1},\n",
       " '503': {'GGA': 4825, 'TGA': 1},\n",
       " '506': {'TCA': 4814},\n",
       " '509': {'GCT': 4830, 'GCA': 1},\n",
       " '512': {'ATA': 4835, 'AGA': 2},\n",
       " '515': {'ACG': 4836, 'AAG': 2, 'TCG': 1},\n",
       " '518': {'ATT': 4834},\n",
       " '521': {'ATT': 4832, 'AGT': 1},\n",
       " '524': {'ATG': 4836},\n",
       " '527': {'GCG': 4830, 'GCA': 1},\n",
       " '530': {'ATT': 4832, 'AGT': 1, 'TTT': 1},\n",
       " '533': {'TGG': 4833, 'TGA': 1},\n",
       " '536': {'GGC': 4798, 'TGC': 4, 'GTC': 1, 'GCT': 1},\n",
       " '539': {'ATT': 4821, 'AAT': 3, 'AGT': 1, 'ATG': 2, 'TAT': 1},\n",
       " '542': {'GGA': 4808, 'GAA': 1, 'TGA': 1},\n",
       " '545': {'CAT': 4823, 'CAA': 2, 'CAG': 1},\n",
       " '548': {'TTT': 4820},\n",
       " '551': {'CTA': 4837, 'TTA': 1, 'ATA': 1},\n",
       " '554': {'AAA': 4819, 'ACA': 2, 'CAA': 1, 'AAT': 1, 'AAC': 1},\n",
       " '557': {'TCC': 4834, 'ATC': 1, 'GCC': 1},\n",
       " '560': {'GAA': 4830, 'GAC': 1, 'GAT': 2},\n",
       " '563': {'CAA': 4831, 'AAA': 1},\n",
       " '566': {'ATT': 4719, 'TTC': 2},\n",
       " '569': {'AAA': 4829, 'AAC': 1, 'TAA': 2, 'ACA': 2, 'ATA': 1},\n",
       " '572': {'GTT': 4826, 'ATT': 1, 'GGT': 2, 'GTA': 1},\n",
       " '575': {'GTA': 4837},\n",
       " '578': {'GAT': 4837},\n",
       " '581': {'TGC': 4737, 'GCT': 1, 'GGC': 2},\n",
       " '584': {'TTT': 4837, 'TTG': 1, 'ATT': 1},\n",
       " '587': {'GGC': 4816, 'GTC': 1, 'CGG': 1, 'GCC': 1, 'GCT': 1},\n",
       " '590': {'AAC': 4804, 'AAA': 3},\n",
       " '593': {'AAT': 4788},\n",
       " '596': {'GCA': 4830, 'GAA': 1},\n",
       " '599': {'ATC': 4828, 'TAC': 1},\n",
       " '602': {'CAT': 4819, 'CAG': 2, 'CCT': 1},\n",
       " '605': {'GAA': 4822, 'AAA': 1, 'GAC': 3, 'TAA': 1},\n",
       " '608': {'GCT': 4819, 'GCG': 1, 'TCT': 2, 'ACT': 1},\n",
       " '611': {'CAC': 4837, 'CCG': 1},\n",
       " '614': {'TAT': 4842},\n",
       " '617': {'GTC': 4834},\n",
       " '620': {'TTT': 4829},\n",
       " '623': {'CTA': 4834, 'ATA': 2, 'CGA': 1},\n",
       " '626': {'AGA': 4800, 'TGA': 1},\n",
       " '629': {'CCT': 4831},\n",
       " '632': {'CTG': 4849, 'CGG': 1},\n",
       " '635': {'TAT': 4776, 'AAT': 1, 'TAA': 1},\n",
       " '638': {'TTT': 4844, 'TGT': 1, 'GTT': 2, 'TAT': 1, 'ATT': 1},\n",
       " '641': {'CTA': 4841, 'CCT': 1},\n",
       " '644': {'ATG': 4836, 'TTG': 1},\n",
       " '647': {'GCG': 4830, 'GGG': 1},\n",
       " '650': {'ATG': 4845, 'ATT': 1, 'AGG': 1},\n",
       " '653': {'TGG': 4843, 'GGG': 1},\n",
       " '656': {'TCG': 4851, 'TAG': 1},\n",
       " '659': {'ATA': 4841, 'TAA': 2},\n",
       " '662': {'TTT': 4810, 'TGT': 1, 'TTA': 2},\n",
       " '665': {'GCT': 4837, 'GCA': 3, 'CTT': 1, 'GCG': 2, 'GCC': 1},\n",
       " '668': {'GCT': 4839, 'GCG': 1, 'GAT': 1},\n",
       " '671': {'TAT': 4844, 'TAG': 1, 'AAT': 1},\n",
       " '674': {'CAC': 4832},\n",
       " '677': {'TTG': 4765, 'GTG': 2, 'ATG': 1},\n",
       " '680': {'TGG': 4828, 'ATG': 2, 'AGG': 1, 'TGT': 1},\n",
       " '683': {'AAA': 4818, 'AAG': 1, 'AAT': 1, 'TAA': 1},\n",
       " '686': {'TTA': 4784, 'GTT': 2, 'TTC': 1, 'TTT': 2, 'ATT': 1, 'TCA': 1},\n",
       " '689': {'TGG': 4845, 'AGG': 1, 'GGG': 1, 'TGT': 2, 'TTG': 1},\n",
       " '692': {'TTA': 4801, 'TTT': 1, 'TAT': 1, 'GTA': 2},\n",
       " '695': {'TAT': 4844, 'GAT': 1, 'CAT': 1, 'TCC': 1, 'TTA': 1},\n",
       " '698': {'ACC': 4764, 'AAC': 3, 'ACT': 1},\n",
       " '701': {'GAA': 4845, 'AAA': 3, 'TAA': 1, 'TGA': 2},\n",
       " '704': {'TTT': 4834, 'TTG': 3, 'ATT': 2},\n",
       " '707': {'GAA': 4831, 'GGA': 1},\n",
       " '710': {'ATG': 4852}}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq2gene2codon2alignedcodons[\"edge_6104\"][\"1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Percentages (values of $p$) for naive mutation calling\n",
    "\n",
    "### Values we use for plotting\n",
    "Currently the same as in the codon mutation notebook. I guess these should be a shared variable, ideally? But that might be too much work for its own good."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-11T23:10:38.533094Z",
     "iopub.status.busy": "2021-06-11T23:10:38.532287Z",
     "iopub.status.idle": "2021-06-11T23:10:38.534094Z",
     "shell.execute_reply": "2021-06-11T23:10:38.534727Z"
    }
   },
   "outputs": [],
   "source": [
    "percentages = [2, 1, 0.5, 0.25, 0.15]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Full values, used for plotting a FDR curve elsewhere\n",
    "\n",
    "Copied from the target/decoy notebook. may be nice to generalize to shared code too i guess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First two percentages: [4.99, 4.98]\n",
      "Last two percentages: [0.16, 0.15]\n",
      "Number of percentages: 485\n"
     ]
    }
   ],
   "source": [
    "# Percentages go from 4.99%, 4.98%, ..., 0.16%, 0.15%\n",
    "full_percentages = [p / 100 for p in range(15, 500, 1)][::-1]\n",
    "print(f\"First two percentages: {full_percentages[:2]}\")\n",
    "print(f\"Last two percentages: {full_percentages[-2:]}\")\n",
    "print(f\"Number of percentages: {len(full_percentages):,}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sanity check -- we can just compute this data for all of the stuff in full_percentages\n",
    "# and then for the stuff we wanna plot in this notebook just focus on the values in percentages\n",
    "for p in percentages:\n",
    "    assert p in full_percentages, f\"{p}% not in full_percentages\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define the number of \"potential\" synonymous / nonsynonymous and non-nonse / nonsense single-nucleotide mutations for each codon\n",
    "\n",
    "From the paper:\n",
    "\n",
    "> Of the 64 unique DNA codons, each codon i has Si synonymous and Ni nonsynonymous\n",
    "single-nucleotide mutations into other codons. For all i, Si + Ni = 9. For example, in the\n",
    "standard genetic code the codon GGG encoding Glycine has Si = 3 synonymous singlenucleotide mutations (GGT, GGC, GGA) and Ni = 6 nonsynonymous single-nucleotide\n",
    "mutations (AGG, CGG, TGG, GAG, GCG, GTG)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sum of Si = 138\n",
      "sum of Ni = 438\n",
      "sum of NNSi = 526\n",
      "sum of NSi = 23\n"
     ]
    }
   ],
   "source": [
    "# This list contains all 64 DNA codons (AAA, AAC, ..., TTT).\n",
    "codons = []\n",
    "for i in \"ACGT\":\n",
    "    for j in \"ACGT\":\n",
    "        for k in \"ACGT\":\n",
    "            codons.append(i + j + k)\n",
    "\n",
    "# Maps the 64 DNA codons in \"codons\" above to integers: the number of potential single-nucleotide...\n",
    "# synonymous mutations\n",
    "c2si = {}\n",
    "# nonsynonymous mutations\n",
    "c2ni = {}\n",
    "# non-nonsense mutations\n",
    "c2nnsi = {}\n",
    "# nonsense mutations\n",
    "c2nsi = {}\n",
    "\n",
    "# c2si[c] + c2ni[c] should equal 9 for all 64 DNA codons c, and\n",
    "# c2nnsi[c] + c2nsi[c] should also equal 9 for all 61 sense DNA codons c (these two dicts are undefined\n",
    "# for stop DNA codons, because you can't really have a \"nonsense\" mutation occur at an extant stop codon --\n",
    "# that'd be a \"nonstop\" mutation. See wikipedia: https://en.wikipedia.org/wiki/Nonsense_mutation)\n",
    "            \n",
    "# These bools just impact how much we print out during this cell. mostly useful for debugging.\n",
    "syn_verbose = False\n",
    "nonsense_verbose = False\n",
    "codon_verbose = False\n",
    "\n",
    "for c in codons:\n",
    "    aa = str(skbio.DNA.translate(skbio.DNA(c)))\n",
    "    \n",
    "    si = 0\n",
    "    ni = 0\n",
    "    nnsi = 0\n",
    "    nsi = 0\n",
    "    \n",
    "    num_alt_codons_considered = 0\n",
    "    \n",
    "    for pos in (0, 1, 2):\n",
    "        posnt = c[pos]\n",
    "        for altnt in sorted(set(\"ACGT\") - set(posnt)):\n",
    "            # it should be possible to do this without checking pos and just\n",
    "            # using a single fancy slice operation, but this is more foolproof\n",
    "            # imo and i am nothing if not a fool a lot of the time\n",
    "            if pos == 0:\n",
    "                alt_codon = altnt + c[1:]\n",
    "            elif pos == 1:\n",
    "                alt_codon = c[0] + altnt + c[2]\n",
    "            else:\n",
    "                alt_codon = c[:2] + altnt\n",
    "\n",
    "            aa2 = str(skbio.DNA.translate(skbio.DNA(alt_codon)))\n",
    "\n",
    "            prefix = f\"{c} ({aa}) -> {alt_codon} ({aa2}) is \"\n",
    "            if aa2 == aa:\n",
    "                si += 1\n",
    "                if syn_verbose:\n",
    "                    print(prefix + \"syn\")\n",
    "            else:\n",
    "                ni += 1\n",
    "                if syn_verbose:\n",
    "                    print(prefix + \"nonsyn\")\n",
    "            num_alt_codons_considered += 1\n",
    "            \n",
    "            # Record NNSi and NSi info for mutations *from* the 61 sense codons\n",
    "            # (we don't consider codons that are already stop codons)\n",
    "            if aa != \"*\":\n",
    "                if aa2 == \"*\":\n",
    "                    # Nonsense mutation!\n",
    "                    nsi += 1\n",
    "                    if nonsense_verbose:\n",
    "                        print(prefix + \"nonsense\")\n",
    "                else:\n",
    "                    # Non-nonsense mutation!\n",
    "                    nnsi += 1\n",
    "                    if nonsense_verbose:\n",
    "                        print(prefix + \"non-nonsense\")\n",
    "                        \n",
    "    # Sanity checks\n",
    "    if num_alt_codons_considered != 9:\n",
    "        raise ValueError(\"each codon should only have 9 alt codons???\")\n",
    "        \n",
    "    if si + ni != 9:\n",
    "        raise ValueError(f\"For codon {c}, Si + Ni = {si} + {ni} = {si + ni}. should sum to 9???\")\n",
    "        \n",
    "    if aa != \"*\":\n",
    "        if nnsi + nsi != 9:\n",
    "            raise ValueError(f\"For codon {c}, NNSi + NSi = {nnsi} + {nsi} = {nnsi + nsi}. should sum to 9???\")\n",
    "    else:\n",
    "        if nnsi != 0 or nsi != 0:\n",
    "            raise ValueError(f\"For codon {c}, NNSi and NSi should both be 0, i.e. undefined.\")\n",
    "\n",
    "    # Print info about this codon\n",
    "    if codon_verbose:\n",
    "        print(f\"{c} ({aa}) has:\\n\\t{si} synonymous and {ni} nonsynonymous s.n. muts\")\n",
    "        if aa != \"*\":\n",
    "            print(f\"\\t{nnsi} non-nonsense and {nsi} nonsense s.n. muts\")\n",
    "        else:\n",
    "            print(f\"\\tThis is a stop codon, so we're not counting (non-)nonsense s.n. muts from this codon\")\n",
    "            \n",
    "    # Update mappings we will need for later\n",
    "    c2si[c] = si\n",
    "    c2ni[c] = ni\n",
    "    if aa != \"*\":\n",
    "        c2nnsi[c] = nnsi\n",
    "        c2nsi[c] = nsi\n",
    "        \n",
    "print(f\"sum of Si = {sum(c2si.values())}\")\n",
    "print(f\"sum of Ni = {sum(c2ni.values())}\")\n",
    "print(f\"sum of NNSi = {sum(c2nnsi.values())}\")\n",
    "print(f\"sum of NSi = {sum(c2nsi.values())}\")\n",
    "\n",
    "# Just to back up some of the intro to this section in the paper...\n",
    "assert c2si[\"GGG\"] == 3 and c2ni[\"GGG\"] == 6"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute numbers of possible and observed single-nucleotide mutations of the types we care about\n",
    "\n",
    "This takes a while to run. With just a few percentage values it takes a few minutes, but using all of `full_percentages` (i.e. everything from 50% to 0.01% using increments of 0.01%) it takes around an hour, I think.\n",
    "\n",
    "... update, not sure how limiting the top to 5% helps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "On gene 1 / 1,297 in seq CAMP.\n",
      "On gene 2 / 1,297 in seq CAMP.\n",
      "On gene 3 / 1,297 in seq CAMP.\n",
      "On gene 4 / 1,297 in seq CAMP.\n",
      "On gene 5 / 1,297 in seq CAMP.\n",
      "On gene 6 / 1,297 in seq CAMP.\n",
      "On gene 7 / 1,297 in seq CAMP.\n",
      "On gene 8 / 1,297 in seq CAMP.\n",
      "On gene 9 / 1,297 in seq CAMP.\n",
      "On gene 10 / 1,297 in seq CAMP.\n",
      "On gene 11 / 1,297 in seq CAMP.\n",
      "On gene 12 / 1,297 in seq CAMP.\n",
      "On gene 13 / 1,297 in seq CAMP.\n",
      "On gene 14 / 1,297 in seq CAMP.\n",
      "On gene 15 / 1,297 in seq CAMP.\n",
      "On gene 16 / 1,297 in seq CAMP.\n",
      "On gene 17 / 1,297 in seq CAMP.\n",
      "On gene 18 / 1,297 in seq CAMP.\n",
      "On gene 19 / 1,297 in seq CAMP.\n",
      "On gene 20 / 1,297 in seq CAMP.\n",
      "On gene 21 / 1,297 in seq CAMP.\n",
      "On gene 22 / 1,297 in seq CAMP.\n",
      "On gene 23 / 1,297 in seq CAMP.\n",
      "On gene 24 / 1,297 in seq CAMP.\n",
      "On gene 25 / 1,297 in seq CAMP.\n",
      "On gene 26 / 1,297 in seq CAMP.\n",
      "On gene 27 / 1,297 in seq CAMP.\n",
      "On gene 28 / 1,297 in seq CAMP.\n",
      "On gene 29 / 1,297 in seq CAMP.\n",
      "On gene 30 / 1,297 in seq CAMP.\n",
      "On gene 31 / 1,297 in seq CAMP.\n",
      "On gene 32 / 1,297 in seq CAMP.\n",
      "On gene 33 / 1,297 in seq CAMP.\n",
      "On gene 34 / 1,297 in seq CAMP.\n",
      "On gene 35 / 1,297 in seq CAMP.\n",
      "On gene 36 / 1,297 in seq CAMP.\n",
      "On gene 37 / 1,297 in seq CAMP.\n",
      "On gene 38 / 1,297 in seq CAMP.\n",
      "On gene 39 / 1,297 in seq CAMP.\n",
      "On gene 40 / 1,297 in seq CAMP.\n",
      "On gene 41 / 1,297 in seq CAMP.\n",
      "On gene 42 / 1,297 in seq CAMP.\n",
      "On gene 43 / 1,297 in seq CAMP.\n",
      "On gene 44 / 1,297 in seq CAMP.\n",
      "On gene 45 / 1,297 in seq CAMP.\n",
      "On gene 46 / 1,297 in seq CAMP.\n",
      "On gene 47 / 1,297 in seq CAMP.\n",
      "On gene 48 / 1,297 in seq CAMP.\n",
      "On gene 49 / 1,297 in seq CAMP.\n",
      "On gene 50 / 1,297 in seq CAMP.\n",
      "On gene 51 / 1,297 in seq CAMP.\n",
      "On gene 52 / 1,297 in seq CAMP.\n",
      "On gene 53 / 1,297 in seq CAMP.\n",
      "On gene 54 / 1,297 in seq CAMP.\n",
      "On gene 55 / 1,297 in seq CAMP.\n",
      "On gene 56 / 1,297 in seq CAMP.\n",
      "On gene 57 / 1,297 in seq CAMP.\n",
      "On gene 58 / 1,297 in seq CAMP.\n",
      "On gene 59 / 1,297 in seq CAMP.\n",
      "On gene 60 / 1,297 in seq CAMP.\n",
      "On gene 61 / 1,297 in seq CAMP.\n",
      "On gene 62 / 1,297 in seq CAMP.\n",
      "On gene 63 / 1,297 in seq CAMP.\n",
      "On gene 64 / 1,297 in seq CAMP.\n",
      "On gene 65 / 1,297 in seq CAMP.\n",
      "On gene 66 / 1,297 in seq CAMP.\n",
      "On gene 67 / 1,297 in seq CAMP.\n",
      "On gene 68 / 1,297 in seq CAMP.\n",
      "On gene 69 / 1,297 in seq CAMP.\n",
      "On gene 70 / 1,297 in seq CAMP.\n",
      "On gene 71 / 1,297 in seq CAMP.\n",
      "On gene 72 / 1,297 in seq CAMP.\n",
      "On gene 73 / 1,297 in seq CAMP.\n",
      "On gene 74 / 1,297 in seq CAMP.\n",
      "On gene 75 / 1,297 in seq CAMP.\n",
      "On gene 76 / 1,297 in seq CAMP.\n",
      "On gene 77 / 1,297 in seq CAMP.\n",
      "On gene 78 / 1,297 in seq CAMP.\n",
      "On gene 79 / 1,297 in seq CAMP.\n",
      "On gene 80 / 1,297 in seq CAMP.\n",
      "On gene 81 / 1,297 in seq CAMP.\n",
      "On gene 82 / 1,297 in seq CAMP.\n",
      "On gene 83 / 1,297 in seq CAMP.\n",
      "On gene 84 / 1,297 in seq CAMP.\n",
      "On gene 85 / 1,297 in seq CAMP.\n",
      "On gene 86 / 1,297 in seq CAMP.\n",
      "On gene 87 / 1,297 in seq CAMP.\n",
      "On gene 88 / 1,297 in seq CAMP.\n",
      "On gene 89 / 1,297 in seq CAMP.\n",
      "On gene 90 / 1,297 in seq CAMP.\n",
      "On gene 91 / 1,297 in seq CAMP.\n",
      "On gene 92 / 1,297 in seq CAMP.\n",
      "On gene 93 / 1,297 in seq CAMP.\n",
      "On gene 94 / 1,297 in seq CAMP.\n",
      "On gene 95 / 1,297 in seq CAMP.\n",
      "On gene 96 / 1,297 in seq CAMP.\n",
      "On gene 97 / 1,297 in seq CAMP.\n",
      "On gene 98 / 1,297 in seq CAMP.\n",
      "On gene 99 / 1,297 in seq CAMP.\n",
      "On gene 100 / 1,297 in seq CAMP.\n",
      "On gene 101 / 1,297 in seq CAMP.\n",
      "On gene 102 / 1,297 in seq CAMP.\n",
      "On gene 103 / 1,297 in seq CAMP.\n",
      "On gene 104 / 1,297 in seq CAMP.\n",
      "On gene 105 / 1,297 in seq CAMP.\n",
      "On gene 106 / 1,297 in seq CAMP.\n",
      "On gene 107 / 1,297 in seq CAMP.\n",
      "On gene 108 / 1,297 in seq CAMP.\n",
      "On gene 109 / 1,297 in seq CAMP.\n",
      "On gene 110 / 1,297 in seq CAMP.\n",
      "On gene 111 / 1,297 in seq CAMP.\n",
      "On gene 112 / 1,297 in seq CAMP.\n",
      "On gene 113 / 1,297 in seq CAMP.\n",
      "On gene 114 / 1,297 in seq CAMP.\n",
      "On gene 115 / 1,297 in seq CAMP.\n",
      "On gene 116 / 1,297 in seq CAMP.\n",
      "On gene 117 / 1,297 in seq CAMP.\n",
      "On gene 118 / 1,297 in seq CAMP.\n",
      "On gene 119 / 1,297 in seq CAMP.\n",
      "On gene 120 / 1,297 in seq CAMP.\n",
      "On gene 121 / 1,297 in seq CAMP.\n",
      "On gene 122 / 1,297 in seq CAMP.\n",
      "On gene 123 / 1,297 in seq CAMP.\n",
      "On gene 124 / 1,297 in seq CAMP.\n",
      "On gene 125 / 1,297 in seq CAMP.\n",
      "On gene 126 / 1,297 in seq CAMP.\n",
      "On gene 127 / 1,297 in seq CAMP.\n",
      "On gene 128 / 1,297 in seq CAMP.\n",
      "On gene 129 / 1,297 in seq CAMP.\n",
      "On gene 130 / 1,297 in seq CAMP.\n",
      "On gene 131 / 1,297 in seq CAMP.\n",
      "On gene 132 / 1,297 in seq CAMP.\n",
      "On gene 133 / 1,297 in seq CAMP.\n",
      "On gene 134 / 1,297 in seq CAMP.\n",
      "On gene 135 / 1,297 in seq CAMP.\n",
      "On gene 136 / 1,297 in seq CAMP.\n",
      "On gene 137 / 1,297 in seq CAMP.\n",
      "On gene 138 / 1,297 in seq CAMP.\n",
      "On gene 139 / 1,297 in seq CAMP.\n",
      "On gene 140 / 1,297 in seq CAMP.\n",
      "On gene 141 / 1,297 in seq CAMP.\n",
      "On gene 142 / 1,297 in seq CAMP.\n",
      "On gene 143 / 1,297 in seq CAMP.\n",
      "On gene 144 / 1,297 in seq CAMP.\n",
      "On gene 145 / 1,297 in seq CAMP.\n",
      "On gene 146 / 1,297 in seq CAMP.\n",
      "On gene 147 / 1,297 in seq CAMP.\n",
      "On gene 148 / 1,297 in seq CAMP.\n",
      "On gene 149 / 1,297 in seq CAMP.\n",
      "On gene 150 / 1,297 in seq CAMP.\n",
      "On gene 151 / 1,297 in seq CAMP.\n",
      "On gene 152 / 1,297 in seq CAMP.\n",
      "On gene 153 / 1,297 in seq CAMP.\n",
      "On gene 154 / 1,297 in seq CAMP.\n",
      "On gene 155 / 1,297 in seq CAMP.\n",
      "On gene 156 / 1,297 in seq CAMP.\n",
      "On gene 157 / 1,297 in seq CAMP.\n",
      "On gene 158 / 1,297 in seq CAMP.\n",
      "On gene 159 / 1,297 in seq CAMP.\n",
      "On gene 160 / 1,297 in seq CAMP.\n",
      "On gene 161 / 1,297 in seq CAMP.\n",
      "On gene 162 / 1,297 in seq CAMP.\n",
      "On gene 163 / 1,297 in seq CAMP.\n",
      "On gene 164 / 1,297 in seq CAMP.\n",
      "On gene 165 / 1,297 in seq CAMP.\n",
      "On gene 166 / 1,297 in seq CAMP.\n",
      "On gene 167 / 1,297 in seq CAMP.\n",
      "On gene 168 / 1,297 in seq CAMP.\n",
      "On gene 169 / 1,297 in seq CAMP.\n",
      "On gene 170 / 1,297 in seq CAMP.\n",
      "On gene 171 / 1,297 in seq CAMP.\n",
      "On gene 172 / 1,297 in seq CAMP.\n",
      "On gene 173 / 1,297 in seq CAMP.\n",
      "On gene 174 / 1,297 in seq CAMP.\n",
      "On gene 175 / 1,297 in seq CAMP.\n",
      "On gene 176 / 1,297 in seq CAMP.\n",
      "On gene 177 / 1,297 in seq CAMP.\n",
      "On gene 178 / 1,297 in seq CAMP.\n",
      "On gene 179 / 1,297 in seq CAMP.\n",
      "On gene 180 / 1,297 in seq CAMP.\n",
      "On gene 181 / 1,297 in seq CAMP.\n",
      "On gene 182 / 1,297 in seq CAMP.\n",
      "On gene 183 / 1,297 in seq CAMP.\n",
      "On gene 184 / 1,297 in seq CAMP.\n",
      "On gene 185 / 1,297 in seq CAMP.\n",
      "On gene 186 / 1,297 in seq CAMP.\n",
      "On gene 187 / 1,297 in seq CAMP.\n",
      "On gene 188 / 1,297 in seq CAMP.\n",
      "On gene 189 / 1,297 in seq CAMP.\n",
      "On gene 190 / 1,297 in seq CAMP.\n",
      "On gene 191 / 1,297 in seq CAMP.\n",
      "On gene 192 / 1,297 in seq CAMP.\n",
      "On gene 193 / 1,297 in seq CAMP.\n",
      "On gene 194 / 1,297 in seq CAMP.\n",
      "On gene 195 / 1,297 in seq CAMP.\n",
      "On gene 196 / 1,297 in seq CAMP.\n",
      "On gene 197 / 1,297 in seq CAMP.\n",
      "On gene 198 / 1,297 in seq CAMP.\n",
      "On gene 199 / 1,297 in seq CAMP.\n",
      "On gene 200 / 1,297 in seq CAMP.\n",
      "On gene 201 / 1,297 in seq CAMP.\n",
      "On gene 202 / 1,297 in seq CAMP.\n",
      "On gene 203 / 1,297 in seq CAMP.\n",
      "On gene 204 / 1,297 in seq CAMP.\n",
      "On gene 205 / 1,297 in seq CAMP.\n",
      "On gene 206 / 1,297 in seq CAMP.\n",
      "On gene 207 / 1,297 in seq CAMP.\n",
      "On gene 208 / 1,297 in seq CAMP.\n",
      "On gene 209 / 1,297 in seq CAMP.\n",
      "On gene 210 / 1,297 in seq CAMP.\n",
      "On gene 211 / 1,297 in seq CAMP.\n",
      "On gene 212 / 1,297 in seq CAMP.\n",
      "On gene 213 / 1,297 in seq CAMP.\n",
      "On gene 214 / 1,297 in seq CAMP.\n",
      "On gene 215 / 1,297 in seq CAMP.\n",
      "On gene 216 / 1,297 in seq CAMP.\n",
      "On gene 217 / 1,297 in seq CAMP.\n",
      "On gene 218 / 1,297 in seq CAMP.\n",
      "On gene 219 / 1,297 in seq CAMP.\n",
      "On gene 220 / 1,297 in seq CAMP.\n",
      "On gene 221 / 1,297 in seq CAMP.\n",
      "On gene 222 / 1,297 in seq CAMP.\n",
      "On gene 223 / 1,297 in seq CAMP.\n",
      "On gene 224 / 1,297 in seq CAMP.\n",
      "On gene 225 / 1,297 in seq CAMP.\n",
      "On gene 226 / 1,297 in seq CAMP.\n",
      "On gene 227 / 1,297 in seq CAMP.\n",
      "On gene 228 / 1,297 in seq CAMP.\n",
      "On gene 229 / 1,297 in seq CAMP.\n",
      "On gene 230 / 1,297 in seq CAMP.\n",
      "On gene 231 / 1,297 in seq CAMP.\n",
      "On gene 232 / 1,297 in seq CAMP.\n",
      "On gene 233 / 1,297 in seq CAMP.\n",
      "On gene 234 / 1,297 in seq CAMP.\n",
      "On gene 235 / 1,297 in seq CAMP.\n",
      "On gene 236 / 1,297 in seq CAMP.\n",
      "On gene 237 / 1,297 in seq CAMP.\n",
      "On gene 238 / 1,297 in seq CAMP.\n",
      "On gene 239 / 1,297 in seq CAMP.\n",
      "On gene 240 / 1,297 in seq CAMP.\n",
      "On gene 241 / 1,297 in seq CAMP.\n",
      "On gene 242 / 1,297 in seq CAMP.\n",
      "On gene 243 / 1,297 in seq CAMP.\n",
      "On gene 244 / 1,297 in seq CAMP.\n",
      "On gene 245 / 1,297 in seq CAMP.\n",
      "On gene 246 / 1,297 in seq CAMP.\n",
      "On gene 247 / 1,297 in seq CAMP.\n",
      "On gene 248 / 1,297 in seq CAMP.\n",
      "On gene 249 / 1,297 in seq CAMP.\n",
      "On gene 250 / 1,297 in seq CAMP.\n",
      "On gene 251 / 1,297 in seq CAMP.\n",
      "On gene 252 / 1,297 in seq CAMP.\n",
      "On gene 253 / 1,297 in seq CAMP.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "On gene 254 / 1,297 in seq CAMP.\n",
      "On gene 255 / 1,297 in seq CAMP.\n",
      "On gene 256 / 1,297 in seq CAMP.\n",
      "On gene 257 / 1,297 in seq CAMP.\n",
      "On gene 258 / 1,297 in seq CAMP.\n",
      "On gene 259 / 1,297 in seq CAMP.\n",
      "On gene 260 / 1,297 in seq CAMP.\n",
      "On gene 261 / 1,297 in seq CAMP.\n",
      "On gene 262 / 1,297 in seq CAMP.\n",
      "On gene 263 / 1,297 in seq CAMP.\n",
      "On gene 264 / 1,297 in seq CAMP.\n",
      "On gene 265 / 1,297 in seq CAMP.\n",
      "On gene 266 / 1,297 in seq CAMP.\n",
      "On gene 267 / 1,297 in seq CAMP.\n",
      "On gene 268 / 1,297 in seq CAMP.\n",
      "On gene 269 / 1,297 in seq CAMP.\n",
      "On gene 270 / 1,297 in seq CAMP.\n",
      "On gene 271 / 1,297 in seq CAMP.\n",
      "On gene 272 / 1,297 in seq CAMP.\n",
      "On gene 273 / 1,297 in seq CAMP.\n",
      "On gene 274 / 1,297 in seq CAMP.\n",
      "On gene 275 / 1,297 in seq CAMP.\n",
      "On gene 276 / 1,297 in seq CAMP.\n",
      "On gene 277 / 1,297 in seq CAMP.\n",
      "On gene 278 / 1,297 in seq CAMP.\n",
      "On gene 279 / 1,297 in seq CAMP.\n",
      "On gene 280 / 1,297 in seq CAMP.\n",
      "On gene 281 / 1,297 in seq CAMP.\n",
      "On gene 282 / 1,297 in seq CAMP.\n",
      "On gene 283 / 1,297 in seq CAMP.\n",
      "On gene 284 / 1,297 in seq CAMP.\n",
      "On gene 285 / 1,297 in seq CAMP.\n",
      "On gene 286 / 1,297 in seq CAMP.\n",
      "On gene 287 / 1,297 in seq CAMP.\n",
      "On gene 288 / 1,297 in seq CAMP.\n",
      "On gene 289 / 1,297 in seq CAMP.\n",
      "On gene 290 / 1,297 in seq CAMP.\n",
      "On gene 291 / 1,297 in seq CAMP.\n",
      "On gene 292 / 1,297 in seq CAMP.\n",
      "On gene 293 / 1,297 in seq CAMP.\n",
      "On gene 294 / 1,297 in seq CAMP.\n",
      "On gene 295 / 1,297 in seq CAMP.\n",
      "On gene 296 / 1,297 in seq CAMP.\n",
      "On gene 297 / 1,297 in seq CAMP.\n",
      "On gene 298 / 1,297 in seq CAMP.\n",
      "On gene 299 / 1,297 in seq CAMP.\n",
      "On gene 300 / 1,297 in seq CAMP.\n",
      "On gene 301 / 1,297 in seq CAMP.\n",
      "On gene 302 / 1,297 in seq CAMP.\n",
      "On gene 303 / 1,297 in seq CAMP.\n",
      "On gene 304 / 1,297 in seq CAMP.\n",
      "On gene 305 / 1,297 in seq CAMP.\n",
      "On gene 306 / 1,297 in seq CAMP.\n",
      "On gene 307 / 1,297 in seq CAMP.\n",
      "On gene 308 / 1,297 in seq CAMP.\n",
      "On gene 309 / 1,297 in seq CAMP.\n",
      "On gene 310 / 1,297 in seq CAMP.\n",
      "On gene 311 / 1,297 in seq CAMP.\n",
      "On gene 312 / 1,297 in seq CAMP.\n",
      "On gene 313 / 1,297 in seq CAMP.\n",
      "On gene 314 / 1,297 in seq CAMP.\n",
      "On gene 315 / 1,297 in seq CAMP.\n",
      "On gene 316 / 1,297 in seq CAMP.\n",
      "On gene 317 / 1,297 in seq CAMP.\n",
      "On gene 318 / 1,297 in seq CAMP.\n",
      "On gene 319 / 1,297 in seq CAMP.\n",
      "On gene 320 / 1,297 in seq CAMP.\n",
      "On gene 321 / 1,297 in seq CAMP.\n",
      "On gene 322 / 1,297 in seq CAMP.\n",
      "On gene 323 / 1,297 in seq CAMP.\n",
      "On gene 324 / 1,297 in seq CAMP.\n",
      "On gene 325 / 1,297 in seq CAMP.\n",
      "On gene 326 / 1,297 in seq CAMP.\n",
      "On gene 327 / 1,297 in seq CAMP.\n",
      "On gene 328 / 1,297 in seq CAMP.\n",
      "On gene 329 / 1,297 in seq CAMP.\n",
      "On gene 330 / 1,297 in seq CAMP.\n",
      "On gene 331 / 1,297 in seq CAMP.\n",
      "On gene 332 / 1,297 in seq CAMP.\n",
      "On gene 333 / 1,297 in seq CAMP.\n",
      "On gene 334 / 1,297 in seq CAMP.\n",
      "On gene 335 / 1,297 in seq CAMP.\n",
      "On gene 336 / 1,297 in seq CAMP.\n",
      "On gene 337 / 1,297 in seq CAMP.\n",
      "On gene 338 / 1,297 in seq CAMP.\n",
      "On gene 339 / 1,297 in seq CAMP.\n",
      "On gene 340 / 1,297 in seq CAMP.\n",
      "On gene 341 / 1,297 in seq CAMP.\n",
      "On gene 342 / 1,297 in seq CAMP.\n",
      "On gene 343 / 1,297 in seq CAMP.\n",
      "On gene 344 / 1,297 in seq CAMP.\n",
      "On gene 345 / 1,297 in seq CAMP.\n",
      "On gene 346 / 1,297 in seq CAMP.\n",
      "On gene 347 / 1,297 in seq CAMP.\n",
      "On gene 348 / 1,297 in seq CAMP.\n",
      "On gene 349 / 1,297 in seq CAMP.\n",
      "On gene 350 / 1,297 in seq CAMP.\n",
      "On gene 351 / 1,297 in seq CAMP.\n",
      "On gene 352 / 1,297 in seq CAMP.\n",
      "On gene 353 / 1,297 in seq CAMP.\n",
      "On gene 354 / 1,297 in seq CAMP.\n",
      "On gene 355 / 1,297 in seq CAMP.\n",
      "On gene 356 / 1,297 in seq CAMP.\n",
      "On gene 357 / 1,297 in seq CAMP.\n",
      "On gene 358 / 1,297 in seq CAMP.\n",
      "On gene 359 / 1,297 in seq CAMP.\n",
      "On gene 360 / 1,297 in seq CAMP.\n",
      "On gene 361 / 1,297 in seq CAMP.\n",
      "On gene 362 / 1,297 in seq CAMP.\n",
      "On gene 363 / 1,297 in seq CAMP.\n",
      "On gene 364 / 1,297 in seq CAMP.\n",
      "On gene 365 / 1,297 in seq CAMP.\n",
      "On gene 366 / 1,297 in seq CAMP.\n",
      "On gene 367 / 1,297 in seq CAMP.\n",
      "On gene 368 / 1,297 in seq CAMP.\n",
      "On gene 369 / 1,297 in seq CAMP.\n",
      "On gene 370 / 1,297 in seq CAMP.\n",
      "On gene 371 / 1,297 in seq CAMP.\n",
      "On gene 372 / 1,297 in seq CAMP.\n",
      "On gene 373 / 1,297 in seq CAMP.\n",
      "On gene 374 / 1,297 in seq CAMP.\n",
      "On gene 375 / 1,297 in seq CAMP.\n",
      "On gene 376 / 1,297 in seq CAMP.\n",
      "On gene 377 / 1,297 in seq CAMP.\n",
      "On gene 378 / 1,297 in seq CAMP.\n",
      "On gene 379 / 1,297 in seq CAMP.\n",
      "On gene 380 / 1,297 in seq CAMP.\n",
      "On gene 381 / 1,297 in seq CAMP.\n",
      "On gene 382 / 1,297 in seq CAMP.\n",
      "On gene 383 / 1,297 in seq CAMP.\n",
      "On gene 384 / 1,297 in seq CAMP.\n",
      "On gene 385 / 1,297 in seq CAMP.\n",
      "On gene 386 / 1,297 in seq CAMP.\n",
      "On gene 387 / 1,297 in seq CAMP.\n",
      "On gene 388 / 1,297 in seq CAMP.\n",
      "On gene 389 / 1,297 in seq CAMP.\n",
      "On gene 390 / 1,297 in seq CAMP.\n",
      "On gene 391 / 1,297 in seq CAMP.\n",
      "On gene 392 / 1,297 in seq CAMP.\n",
      "On gene 393 / 1,297 in seq CAMP.\n",
      "On gene 394 / 1,297 in seq CAMP.\n",
      "On gene 395 / 1,297 in seq CAMP.\n",
      "On gene 396 / 1,297 in seq CAMP.\n",
      "On gene 397 / 1,297 in seq CAMP.\n",
      "On gene 398 / 1,297 in seq CAMP.\n",
      "On gene 399 / 1,297 in seq CAMP.\n",
      "On gene 400 / 1,297 in seq CAMP.\n",
      "On gene 401 / 1,297 in seq CAMP.\n",
      "On gene 402 / 1,297 in seq CAMP.\n",
      "On gene 403 / 1,297 in seq CAMP.\n",
      "On gene 404 / 1,297 in seq CAMP.\n",
      "On gene 405 / 1,297 in seq CAMP.\n",
      "On gene 406 / 1,297 in seq CAMP.\n",
      "On gene 407 / 1,297 in seq CAMP.\n",
      "On gene 408 / 1,297 in seq CAMP.\n",
      "On gene 409 / 1,297 in seq CAMP.\n",
      "On gene 410 / 1,297 in seq CAMP.\n",
      "On gene 411 / 1,297 in seq CAMP.\n",
      "On gene 412 / 1,297 in seq CAMP.\n",
      "On gene 413 / 1,297 in seq CAMP.\n",
      "On gene 414 / 1,297 in seq CAMP.\n",
      "On gene 415 / 1,297 in seq CAMP.\n",
      "On gene 416 / 1,297 in seq CAMP.\n",
      "On gene 417 / 1,297 in seq CAMP.\n",
      "On gene 418 / 1,297 in seq CAMP.\n",
      "On gene 419 / 1,297 in seq CAMP.\n",
      "On gene 420 / 1,297 in seq CAMP.\n",
      "On gene 421 / 1,297 in seq CAMP.\n",
      "On gene 422 / 1,297 in seq CAMP.\n",
      "On gene 423 / 1,297 in seq CAMP.\n",
      "On gene 424 / 1,297 in seq CAMP.\n",
      "On gene 425 / 1,297 in seq CAMP.\n",
      "On gene 426 / 1,297 in seq CAMP.\n",
      "On gene 427 / 1,297 in seq CAMP.\n",
      "On gene 428 / 1,297 in seq CAMP.\n",
      "On gene 429 / 1,297 in seq CAMP.\n",
      "On gene 430 / 1,297 in seq CAMP.\n",
      "On gene 431 / 1,297 in seq CAMP.\n",
      "On gene 432 / 1,297 in seq CAMP.\n",
      "On gene 433 / 1,297 in seq CAMP.\n",
      "On gene 434 / 1,297 in seq CAMP.\n",
      "On gene 435 / 1,297 in seq CAMP.\n",
      "On gene 436 / 1,297 in seq CAMP.\n",
      "On gene 437 / 1,297 in seq CAMP.\n",
      "On gene 438 / 1,297 in seq CAMP.\n",
      "On gene 439 / 1,297 in seq CAMP.\n",
      "On gene 440 / 1,297 in seq CAMP.\n",
      "On gene 441 / 1,297 in seq CAMP.\n",
      "On gene 442 / 1,297 in seq CAMP.\n",
      "On gene 443 / 1,297 in seq CAMP.\n",
      "On gene 444 / 1,297 in seq CAMP.\n",
      "On gene 445 / 1,297 in seq CAMP.\n",
      "On gene 446 / 1,297 in seq CAMP.\n",
      "On gene 447 / 1,297 in seq CAMP.\n",
      "On gene 448 / 1,297 in seq CAMP.\n",
      "On gene 449 / 1,297 in seq CAMP.\n",
      "On gene 450 / 1,297 in seq CAMP.\n",
      "On gene 451 / 1,297 in seq CAMP.\n",
      "On gene 452 / 1,297 in seq CAMP.\n",
      "On gene 453 / 1,297 in seq CAMP.\n",
      "On gene 454 / 1,297 in seq CAMP.\n",
      "On gene 455 / 1,297 in seq CAMP.\n",
      "On gene 456 / 1,297 in seq CAMP.\n",
      "On gene 457 / 1,297 in seq CAMP.\n",
      "On gene 458 / 1,297 in seq CAMP.\n",
      "On gene 459 / 1,297 in seq CAMP.\n",
      "On gene 460 / 1,297 in seq CAMP.\n",
      "On gene 461 / 1,297 in seq CAMP.\n",
      "On gene 462 / 1,297 in seq CAMP.\n",
      "On gene 463 / 1,297 in seq CAMP.\n",
      "On gene 464 / 1,297 in seq CAMP.\n",
      "On gene 465 / 1,297 in seq CAMP.\n",
      "On gene 466 / 1,297 in seq CAMP.\n",
      "On gene 467 / 1,297 in seq CAMP.\n",
      "On gene 468 / 1,297 in seq CAMP.\n",
      "On gene 469 / 1,297 in seq CAMP.\n",
      "On gene 470 / 1,297 in seq CAMP.\n",
      "On gene 471 / 1,297 in seq CAMP.\n",
      "On gene 472 / 1,297 in seq CAMP.\n",
      "On gene 473 / 1,297 in seq CAMP.\n",
      "On gene 474 / 1,297 in seq CAMP.\n",
      "On gene 475 / 1,297 in seq CAMP.\n",
      "On gene 476 / 1,297 in seq CAMP.\n",
      "On gene 477 / 1,297 in seq CAMP.\n",
      "On gene 478 / 1,297 in seq CAMP.\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-15-1406368d4f0c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     84\u001b[0m                         \u001b[0;32mif\u001b[0m \u001b[0mp\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     85\u001b[0m                             is_mutation = pileup.naively_call_mutation_directly(\n\u001b[0;32m---> 86\u001b[0;31m                                 \u001b[0mmax_freq_alt_codon_freq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnum_aligned_codons\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0monly_call_if_rare\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     87\u001b[0m                             )\n\u001b[1;32m     88\u001b[0m                         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# Compute \"N\", the total length of all nucleotides in all codons in all predicted genes in each genome\n",
    "# This allows for positions to be counted multiple times in the case of gene overlaps. Not ideal, I guess,\n",
    "# but throwing out overlapping sections might bias this against the observation of stop codons.\n",
    "seq2total_codon_nt_len = {}\n",
    "\n",
    "# This should be less than N for each genome. It only includes the lengths of sense (i.e. not stop) codons;\n",
    "# we use this in place of N when doing the (non)nonsense stuff because \n",
    "seq2sense_codon_nt_len = {}\n",
    "\n",
    "# Just needed for misc text stuff later\n",
    "seq2num_genes = {}\n",
    "\n",
    "# Possible numbers of single-nucleotide (non)synonymous and (non)nonsense mutations throughout the genomes\n",
    "seq2poss_si = defaultdict(int)\n",
    "seq2poss_ni = defaultdict(int)\n",
    "seq2poss_nnsi = defaultdict(int)\n",
    "seq2poss_nsi = defaultdict(int)\n",
    "\n",
    "# For each percentage p in percentages, the observed numbers of the same types of mutations\n",
    "p2seq2obs_si = {p: defaultdict(int) for p in full_percentages}\n",
    "p2seq2obs_ni = {p: defaultdict(int) for p in full_percentages}\n",
    "p2seq2obs_nnsi = {p: defaultdict(int) for p in full_percentages}\n",
    "p2seq2obs_nsi = {p: defaultdict(int) for p in full_percentages}\n",
    "\n",
    "for seq in SEQS:\n",
    "    fasta = skbio.DNA.read(f\"../seqs/{seq}.fasta\")\n",
    "    genes_df = parse_sco(f\"../seqs/genes/{seq}.sco\")\n",
    "    \n",
    "    seq2num_genes[seq] = len(genes_df.index)\n",
    "    \n",
    "    # We can compute N easily since we already know gene lengths\n",
    "    seq2total_codon_nt_len[seq] = genes_df[\"Length\"].sum()\n",
    "    \n",
    "    # ... but we might as well sanity-check it, since we'll be iterating through all genes anyway\n",
    "    total_codon_nt_len = 0\n",
    "    sense_codon_nt_len = 0\n",
    "    \n",
    "    for gene_data in genes_df.itertuples():\n",
    "        print(f\"On gene {gene_data.Index:,} / {len(genes_df.index):,} in seq {seq2name[seq]}.\")\n",
    "        for cpleft in range(gene_data.LeftEnd, gene_data.RightEnd + 1, 3):\n",
    "            \n",
    "            # Get the exact DNA sequence of this codon\n",
    "            codon_dna = fasta[cpleft - 1: cpleft + 2]\n",
    "            if gene_data.Strand == \"-\":\n",
    "                codon_dna = codon_dna.reverse_complement()\n",
    "            codon_seq = str(codon_dna)\n",
    "            \n",
    "            # Increment N and N' values, and update numbers of possible single-nucleotide types of mutations\n",
    "            if str(codon_dna.translate()) != \"*\":\n",
    "                sense_codon_nt_len += 3\n",
    "                seq2poss_nnsi[seq] += c2nnsi[codon_seq]\n",
    "                seq2poss_nsi[seq] += c2nsi[codon_seq]\n",
    "                \n",
    "            total_codon_nt_len += 3\n",
    "            seq2poss_si[seq] += c2si[codon_seq]\n",
    "            seq2poss_ni[seq] += c2ni[codon_seq]\n",
    "            \n",
    "            # Use codon alignment data (already computed and stored in seq2gene2codon2alignedcodons\n",
    "            # thanks to the MutationMatrices notebook) to naively call p-mutations at each codon, and thus\n",
    "            # count \"observed\" types of single-nucleotide mutations\n",
    "            aligned_codons = seq2gene2codon2alignedcodons[seq][str(gene_data.Index)][str(cpleft)]\n",
    "            \n",
    "            # NOTE this code is copied over from the MutationMatrices notebook -- should eventually generalize\n",
    "            # to a separate function somewhere\n",
    "            if len(aligned_codons) > 0:\n",
    "                max_ac_freq = max(aligned_codons.values())\n",
    "                if codon_seq in aligned_codons and aligned_codons[codon_seq] == max_ac_freq:\n",
    "                    num_aligned_codons = sum(aligned_codons.values())\n",
    "                    # Only call a mutation using the max-freq alt codon, not the sum of all alternate codon freqs.\n",
    "                    # The max(d, key=d.get) trick is from https://stackoverflow.com/a/280156 (I've used it a lot\n",
    "                    # here :) Notably, this breaks ties arbitrarily.\n",
    "                    alt_codons = {c: aligned_codons[c] for c in aligned_codons if c != codon_seq}\n",
    "                    if len(alt_codons) > 0:\n",
    "                        max_freq_alt_codon = max(alt_codons, key=aligned_codons.get)\n",
    "                        max_freq_alt_codon_freq = alt_codons[max_freq_alt_codon]\n",
    "                    else:\n",
    "                        # The only codon aligned to this position is the reference codon, so\n",
    "                        # there's no mutation here.\n",
    "                        max_freq_alt_codon_freq = 0\n",
    "                        \n",
    "                    # So far, so good -- this codon has multiple codons aligned to it, including the reference\n",
    "                    # codon. So it could have a p-mutation.\n",
    "                    for p in full_percentages:\n",
    "                        if p != 0:\n",
    "                            is_mutation = pileup.naively_call_mutation_directly(\n",
    "                                max_freq_alt_codon_freq, num_aligned_codons, p, only_call_if_rare=True\n",
    "                            )\n",
    "                        else:\n",
    "                            raise ValueError(\"unsupported, go yell at marcus\")\n",
    "                            \n",
    "                        if is_mutation:\n",
    "                            # OK, so this is a p-mutation (or p = 0% and we're treating this like a mutation).\n",
    "                            \n",
    "                            # If the max freq alt codon is only a single nucleotide different from the\n",
    "                            # reference codon, update our counts. Otherwise, ignore this codon. (Note that\n",
    "                            # we're still breaking ties re: \"max freq alt codon\" arbitrarily; we could change\n",
    "                            # this a bit if desired, but shouldn't be a big deal.)\n",
    "\n",
    "                            num_diffs = 0\n",
    "                            for pos in (0, 1, 2):\n",
    "                                if max_freq_alt_codon[pos] != codon_seq[pos]:\n",
    "                                    num_diffs += 1\n",
    "\n",
    "                            if num_diffs > 1:\n",
    "                                continue\n",
    "\n",
    "                            elif num_diffs == 1:\n",
    "                                # Nice, the max freq alt codon is a single-nucleotide mutation (compared to\n",
    "                                # the reference codon)!\n",
    "                                aa = str(codon_dna.translate())\n",
    "                                aa2 = str(skbio.DNA(max_freq_alt_codon).translate())\n",
    "\n",
    "                                # Record Syn/NonSyn stuff\n",
    "                                if aa == aa2:\n",
    "                                    p2seq2obs_si[p][seq] += 1\n",
    "                                else:\n",
    "                                    p2seq2obs_ni[p][seq] += 1\n",
    "\n",
    "                                # Record Non-nonsense/Nonsense stuff.\n",
    "                                # We could probs merge this into the second case of the if statement above\n",
    "                                # but imo the small time savings isn't worth the added complexity of weird code\n",
    "                                if aa != \"*\":\n",
    "                                    if aa2 == \"*\":\n",
    "                                        p2seq2obs_nsi[p][seq] += 1\n",
    "                                    else:\n",
    "                                        p2seq2obs_nnsi[p][seq] += 1\n",
    "                            else:\n",
    "                                # This should never happen, since we filtered out codon_seq from the alt_codons\n",
    "                                # dict already.\n",
    "                                raise ValueError(f\"Seems like {max_freq_alt_codon} is identical to {codon_seq}???\")\n",
    "    \n",
    "    # Sanity checks\n",
    "    if total_codon_nt_len != seq2total_codon_nt_len[seq]:\n",
    "        raise ValueError(f\"Total length of all codons in seq {seq2name[seq]} doesn't seem right?\")\n",
    "        \n",
    "    # If this check fails, it means Prodigal predicted genes where there were stop codons within the middle\n",
    "    # of a gene sequence. That's not cool! I guess that could happen in the case of alternate genetic codes,\n",
    "    # but that shouldn't happen within the context of just this analysis.\n",
    "    gl3 = len(genes_df.index) * 3\n",
    "    exp_sense_codon_len = total_codon_nt_len - gl3\n",
    "    if sense_codon_nt_len != exp_sense_codon_len:\n",
    "        print(f\"Total length of all codons in seq {seq2name[seq]} = {total_codon_nt_len:,}\")\n",
    "        print(f\"Total length of all sense codons in seq {seq2name[seq]} = {sense_codon_nt_len:,}\")\n",
    "        print(f\"Diff btwn these lengths should be 3 * # genes = 3 * {len(genes_df.index):,} = {gl3:,}\")\n",
    "        print(f\"Actual diff is {total_codon_nt_len - sense_codon_nt_len:,}???\")\n",
    "        raise ValueError(f\"Length of all sense codons in seq {seq2name[seq]} doesn't seem right?\")\n",
    "    seq2sense_codon_nt_len[seq] = sense_codon_nt_len"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Output numbers of observed and possible mutations\n",
    "\n",
    "Slow to compute since we use a ton of values of $p$, but since these are just numbers of observed mutations -- not coordinates -- doesn't take up a ton of space thankfully."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"misc-output/p2seq2obs_si.pickle\", \"wb\") as pf:\n",
    "    pickle.dump(p2seq2obs_si, pf)\n",
    "    \n",
    "with open(\"misc-output/p2seq2obs_ni.pickle\", \"wb\") as pf:\n",
    "    pickle.dump(p2seq2obs_ni, pf)\n",
    "    \n",
    "with open(\"misc-output/p2seq2obs_nnsi.pickle\", \"wb\") as pf:\n",
    "    pickle.dump(p2seq2obs_nnsi, pf)\n",
    "    \n",
    "with open(\"misc-output/p2seq2obs_nsi.pickle\", \"wb\") as pf:\n",
    "    pickle.dump(p2seq2obs_nsi, pf)\n",
    "\n",
    "with open(\"misc-output/seq2poss_si.pickle\", \"wb\") as pf:\n",
    "    pickle.dump(seq2poss_si, pf)\n",
    "    \n",
    "with open(\"misc-output/seq2poss_ni.pickle\", \"wb\") as pf:\n",
    "    pickle.dump(seq2poss_ni, pf)\n",
    "    \n",
    "with open(\"misc-output/seq2poss_nnsi.pickle\", \"wb\") as pf:\n",
    "    pickle.dump(seq2poss_nnsi, pf)\n",
    "\n",
    "with open(\"misc-output/seq2poss_nsi.pickle\", \"wb\") as pf:\n",
    "    pickle.dump(seq2poss_nsi, pf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load numbers of observed and possible mutations\n",
    "\n",
    "So that we could skip the slow step above if desired"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_pf(basename):\n",
    "    with open(f\"misc-output/{basename}.pickle\", \"rb\") as pf:\n",
    "        return pickle.load(pf)\n",
    "\n",
    "p2seq2obs_si = load_pf(\"p2seq2obs_si\")\n",
    "p2seq2obs_ni = load_pf(\"p2seq2obs_ni\")\n",
    "p2seq2obs_nnsi = load_pf(\"p2seq2obs_nnsi\")\n",
    "p2seq2obs_nsi = load_pf(\"p2seq2obs_nsi\")\n",
    "\n",
    "seq2poss_si = load_pf(\"seq2poss_si\")\n",
    "seq2poss_ni = load_pf(\"seq2poss_ni\")\n",
    "seq2poss_nnsi = load_pf(\"seq2poss_nnsi\")\n",
    "seq2poss_nsi = load_pf(\"seq2poss_nsi\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute rates based on the above data that are easy to plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rates for synonymous and nonsynonymous s.n. mutations\n",
    "p2seq2synvals = {p: {} for p in percentages}\n",
    "# Rates for non-nonsense and nonsense s.n. mutations\n",
    "p2seq2nonsensevals = {p: {} for p in percentages}\n",
    "\n",
    "def get_rate(p, seq, p2seq2obs, seq2poss):\n",
    "    # This should never happen (TM)\n",
    "    if seq not in seq2poss:\n",
    "        raise ValueError(\"No possible mutations of some type defined for this seq?\")\n",
    "        \n",
    "    if seq not in p2seq2obs[p]:\n",
    "        rate = 0\n",
    "    else:\n",
    "        rate = p2seq2obs[p][seq] / seq2poss[seq]\n",
    "\n",
    "    return rate\n",
    "\n",
    "for p in percentages:\n",
    "    for seq in SEQS:\n",
    "        p2seq2synvals[p][seq] = {}\n",
    "        p2seq2synvals[p][seq][\"syn\"] = get_rate(p, seq, p2seq2obs_si, seq2poss_si)\n",
    "        p2seq2synvals[p][seq][\"nonsyn\"] = get_rate(p, seq, p2seq2obs_ni, seq2poss_ni)\n",
    "\n",
    "        p2seq2nonsensevals[p][seq] = {}\n",
    "        p2seq2nonsensevals[p][seq][\"nonnonsense\"] = get_rate(p, seq, p2seq2obs_nnsi, seq2poss_nnsi)\n",
    "        p2seq2nonsensevals[p][seq][\"nonsense\"] = get_rate(p, seq, p2seq2obs_nsi, seq2poss_nsi)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_suffix(p):\n",
    "    # Like in the CP 1/2/3 notebook.\n",
    "    if p == 0:\n",
    "        return \"alt(codon) > 0\"\n",
    "    else:\n",
    "         return f\"na\\u00efve, $p$ = {p}%\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-11T23:21:58.712882Z",
     "iopub.status.busy": "2021-06-11T23:21:58.698229Z",
     "iopub.status.idle": "2021-06-11T23:22:01.680859Z",
     "shell.execute_reply": "2021-06-11T23:22:01.681538Z"
    }
   },
   "outputs": [],
   "source": [
    "fig, axes = pyplot.subplots(len(percentages), 3, gridspec_kw={\"hspace\": 0.5, \"wspace\": 0.35})\n",
    "\n",
    "if len(percentages) % 2 == 0:\n",
    "    # Really, the precise middle is between two rows, right? anyway as of writing we use 7 percentages\n",
    "    # so this never happens\n",
    "    middle_row = len(percentages) / 2\n",
    "else:\n",
    "    middle_row = percentages.index(median(percentages))\n",
    "\n",
    "for pi, p in enumerate(percentages):\n",
    "    for si, seq in enumerate(SEQS):\n",
    "        rs = p2seq2synvals[p][seq][\"syn\"]\n",
    "        rn = p2seq2synvals[p][seq][\"nonsyn\"]\n",
    "        axes[pi, si].bar(\n",
    "            [\"Synonymous ($R_S$)\", \"Nonsynonymous ($R_N$)\"],\n",
    "            [rs,rn],\n",
    "            color=[\"#2233cc\", \"#cc3322\"]\n",
    "        )\n",
    "        axes[pi, si].ticklabel_format(axis=\"y\", style=\"plain\")\n",
    "        \n",
    "        titlecolor=\"#000000\"\n",
    "        titleweight=\"normal\"\n",
    "        if rs <= rn:\n",
    "            titlecolor = \"#cc3322\"\n",
    "            titleweight = \"semibold\"\n",
    "            \n",
    "        axes[pi, si].set_title(f\"{seq2name[seq]}: {get_suffix(p)}\", color=titlecolor, fontweight=titleweight)\n",
    "        \n",
    "        if pi == middle_row and si == 0:\n",
    "            axes[pi, si].set_ylabel(\"Ratio of observed to possible mutations\", labelpad=15, fontsize=15)\n",
    "\n",
    "axes[len(percentages) - 1, 1].set_xlabel(\"Type of mutation\", labelpad=15, fontsize=15)\n",
    "fig.suptitle(\"Single-nucleotide synonymous vs. nonsynonymous codon rare mutation rates\", y=0.935, x=0.51, fontsize=20)\n",
    "fig.set_size_inches(15, 12.5)\n",
    "fig.savefig(\"figs/syn-nonsyn-snmut-barplots.png\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-11T23:22:01.883530Z",
     "iopub.status.busy": "2021-06-11T23:22:01.694805Z",
     "iopub.status.idle": "2021-06-11T23:22:05.860654Z",
     "shell.execute_reply": "2021-06-11T23:22:05.861322Z"
    }
   },
   "outputs": [],
   "source": [
    "fig2, axes2 = pyplot.subplots(len(percentages), 3, gridspec_kw={\"hspace\": 0.5, \"wspace\": 0.35})\n",
    "\n",
    "for pi, p in enumerate(percentages):\n",
    "    for si, seq in enumerate(SEQS):\n",
    "        rnns = p2seq2nonsensevals[p][seq][\"nonnonsense\"]\n",
    "        rns = p2seq2nonsensevals[p][seq][\"nonsense\"]\n",
    "        axes2[pi, si].bar(\n",
    "            [\"Non-nonsense ($R_{NNS}$)\", \"Nonsense ($R_{NS}$)\"],\n",
    "            [rnns, rns],\n",
    "            color=[\"#cc9922\", \"#9922cc\"]\n",
    "        )\n",
    "        \n",
    "        # Turn off scientific notation on the y-axis' ticks -- for some very small numbers\n",
    "        # for CAMP's mutation rates, matplotlib will default to scaling the y-axis tick values by 1e-5.\n",
    "        # See https://stackoverflow.com/a/28373421: the axis=\"y\" argument is needed because by default\n",
    "        # this'll try to apply this to *both* axes, and the categorical x-axis causes an error.\n",
    "        axes2[pi, si].ticklabel_format(axis=\"y\", style=\"plain\")\n",
    "        \n",
    "        titlecolor = \"#000000\"\n",
    "        titleweight = \"normal\"\n",
    "        if rnns <= rns:\n",
    "            titlecolor = \"#9922cc\"\n",
    "            titleweight = \"semibold\"\n",
    "            \n",
    "        axes2[pi, si].set_title(f\"{seq2name[seq]}: {get_suffix(p)}\", color=titlecolor, fontweight=titleweight)\n",
    "        \n",
    "        if pi == middle_row and si == 0:\n",
    "            axes2[pi, si].set_ylabel(\"Ratio of observed to possible mutations\", labelpad=15, fontsize=15)\n",
    "        \n",
    "axes2[len(percentages) - 1, 1].set_xlabel(\"Type of mutation\", labelpad=15, fontsize=15)\n",
    "fig2.suptitle(\"Single-nucleotide non-nonsense vs. nonsense codon rare mutation rates\", y=0.935, x=0.51, fontsize=20)\n",
    "fig2.set_size_inches(15, 12.5)\n",
    "fig2.savefig(\"figs/nonnonsense-nonsense-snmut-barplots.png\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Misc. text info about BACT1, to help with explanation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For the $R_S$ and $R_N$ section\n",
    "\n",
    "#### Example of rate computation with BACT1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "outputtext = fr\"\"\"For the {seq2name['edge_1671']} genome\n",
    "(where $N$ = {seq2total_codon_nt_len['edge_1671']:,}),\n",
    "there are $J$ = {seq2poss_si['edge_1671']:,} possible synonymous and\n",
    "$K$ = {seq2poss_ni['edge_1671']:,} possible nonsynonymous single-nucleotide mutations.\n",
    "%\n",
    "Using the mutation frequency threshold $p = 0.5\\%$,\n",
    "the {seq2name['edge_1671']} genome contains\n",
    "$X$ = {p2seq2obs_si[0.5]['edge_1671']:,} synonymous and\n",
    "$Y$ = {p2seq2obs_ni[0.5]['edge_1671']:,} nonsynonymous rare single-nucleotide mutations.\n",
    "%\n",
    "Thus, the ratio of observed to possible synonymous single-nucleotide mutations\n",
    "($R_S = X / J = \\frac{{{p2seq2obs_si[0.5]['edge_1671']:,}}}{{{seq2poss_si['edge_1671']:,}}}\n",
    "\\approx {p2seq2synvals[0.5]['edge_1671']['syn']:.4f}$)\n",
    "is much larger than the ratio of observed to possible nonsynonymous single-nucleotide mutations\n",
    "($R_N = Y / K = \\frac{{{p2seq2obs_ni[0.5]['edge_1671']:,}}}{{{seq2poss_ni['edge_1671']:,}}}\n",
    "\\approx {p2seq2synvals[0.5]['edge_1671']['nonsyn']:.4f}$)\n",
    "for {seq2name['edge_1671']}.\n",
    "\"\"\"\n",
    "with open(\"misc-text/bact1-rs-rn-info.tex\", \"w\") as of:\n",
    "    of.write(f\"{outputtext}\\endinput\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(outputtext)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For the $R_{NNS}$ and $R_{NS}$ section"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Some unused text I started writing before realizing that it isn't super important probably, but oh\n",
    "# well the extra assertion is nice to keep in\n",
    "\n",
    "n = seq2total_codon_nt_len['edge_1671']\n",
    "s = seq2num_genes['edge_1671']\n",
    "np = seq2sense_codon_nt_len['edge_1671']\n",
    "assert np == (n - (3 * s))\n",
    "\n",
    "print(fr\"\"\"\n",
    "For all genomes, $N' = N - 3S$, where $N$ is the total number of\n",
    "nucleotides located in all codons (sense and stop), and $S$ is the total number of stop codons\n",
    "located in all predicted genes throughout the the genome (this number should be equal to the\n",
    "total number of predicted genes, since each gene should have only a single associated stop codon).\n",
    "The {seq2name['edge_1671']} genome has $N = {n:,}$,\n",
    "$3S = {3 * s:,}$, and $N' = {np:,}$,\n",
    "so $N' = N - S$ holds.\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: we use {{ }} (doubling up on curly braces) in order to escape them --\n",
    "# https://stackoverflow.com/a/5466478\n",
    "# We could probs also just split this up into multiple strings but that's woooork\n",
    "\n",
    "outputtext = fr\"\"\"To illustrate this, we again consider\n",
    "the {seq2name['edge_1671']} genome at the threshold $p = 0.5\\%$.\n",
    "%\n",
    "In this genome, there are $N'$ = {seq2sense_codon_nt_len['edge_1671']:,} nucleotides\n",
    "located throughout all sense codons (we ignore stop codons in the reference genome).\n",
    "%\n",
    "{seq2name['edge_1671']} has $J$ = {seq2poss_nnsi['edge_1671']:,} possible non-nonsense and\n",
    "$K$ = {seq2poss_nsi['edge_1671']:,} possible nonsense single-nucleotide mutations.\n",
    "%\n",
    "Using $p = 0.5\\%$, {seq2name['edge_1671']} contains\n",
    "$X$ = {p2seq2obs_nnsi[0.5]['edge_1671']:,} non-nonsense and\n",
    "$Y$ = {p2seq2obs_nsi[0.5]['edge_1671']:,} nonsense rare single-nucleotide mutations.\n",
    "%\n",
    "At $p = 0.5\\%$, then, the ratio of observed to possible non-nonsense single-nucleotide mutations\n",
    "($R_{{NNS}} = X / J = \\frac{{{p2seq2obs_nnsi[0.5]['edge_1671']:,}}}{{{seq2poss_nnsi['edge_1671']:,}}}\n",
    "\\approx {p2seq2nonsensevals[0.5]['edge_1671']['nonnonsense']:.4f}$)\n",
    "is much larger than the ratio of observed to possible nonsynonymous single-nucleotide mutations\n",
    "($R_{{NS}} = Y / K = \\frac{{{p2seq2obs_nsi[0.5]['edge_1671']:,}}}{{{seq2poss_nsi['edge_1671']:,}}}\n",
    "\\approx {p2seq2nonsensevals[0.5]['edge_1671']['nonsense']:.6f}$)\n",
    "for {seq2name['edge_1671']}.\n",
    "\"\"\"\n",
    "with open(\"misc-text/bact1-rnns-rns-info.tex\", \"w\") as of:\n",
    "    of.write(f\"{outputtext}\\endinput\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(outputtext)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
